{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 451
        },
        "collapsed": true,
        "id": "in1LgqjMGezX",
        "outputId": "1f503bec-ed27-4036-e041-ba97455ba2ad"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]'</li><li>'Get:2 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]'</li><li>'Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]'</li><li>'Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease'</li><li>'Get:5 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,017 kB]'</li><li>'Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]'</li><li>'Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease'</li><li>'Hit:8 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease'</li><li>'Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,742 kB]'</li><li>'Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]'</li><li>'Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,246 kB]'</li><li>'Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,553 kB]'</li><li>'Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,984 kB]'</li><li>'Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,295 kB]'</li><li>'Fetched 21.2 MB in 4s (6,024 kB/s)'</li><li>'Reading package lists...'</li></ol>\n"
            ],
            "text/markdown": "1. 'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]'\n2. 'Get:2 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]'\n3. 'Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]'\n4. 'Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease'\n5. 'Get:5 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,017 kB]'\n6. 'Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]'\n7. 'Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease'\n8. 'Hit:8 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease'\n9. 'Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,742 kB]'\n10. 'Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]'\n11. 'Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,246 kB]'\n12. 'Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,553 kB]'\n13. 'Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,984 kB]'\n14. 'Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,295 kB]'\n15. 'Fetched 21.2 MB in 4s (6,024 kB/s)'\n16. 'Reading package lists...'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item 'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease {[}3,632 B{]}'\n\\item 'Get:2 https://r2u.stat.illinois.edu/ubuntu jammy InRelease {[}6,555 B{]}'\n\\item 'Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease {[}129 kB{]}'\n\\item 'Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease'\n\\item 'Get:5 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages {[}9,017 kB{]}'\n\\item 'Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease {[}128 kB{]}'\n\\item 'Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease'\n\\item 'Hit:8 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease'\n\\item 'Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages {[}2,742 kB{]}'\n\\item 'Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease {[}127 kB{]}'\n\\item 'Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages {[}1,246 kB{]}'\n\\item 'Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages {[}1,553 kB{]}'\n\\item 'Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages {[}2,984 kB{]}'\n\\item 'Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages {[}3,295 kB{]}'\n\\item 'Fetched 21.2 MB in 4s (6,024 kB/s)'\n\\item 'Reading package lists...'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] \"Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\"      \n",
              " [2] \"Get:2 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\"                      \n",
              " [3] \"Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\"                 \n",
              " [4] \"Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease\"                                    \n",
              " [5] \"Get:5 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,017 kB]\"             \n",
              " [6] \"Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\"                   \n",
              " [7] \"Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\"              \n",
              " [8] \"Hit:8 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\"               \n",
              " [9] \"Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,742 kB]\"           \n",
              "[10] \"Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\"                \n",
              "[11] \"Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,246 kB]\"\n",
              "[12] \"Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,553 kB]\"  \n",
              "[13] \"Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,984 kB]\"    \n",
              "[14] \"Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,295 kB]\"      \n",
              "[15] \"Fetched 21.2 MB in 4s (6,024 kB/s)\"                                                        \n",
              "[16] \"Reading package lists...\"                                                                  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'The following NEW packages will be installed:'</li><li><span style=white-space:pre-wrap>'  jags'</span></li><li>'0 upgraded, 1 newly installed, 0 to remove and 9 not upgraded.'</li><li>'Need to get 1,644 kB of archives.'</li><li>'After this operation, 3,492 kB of additional disk space will be used.'</li><li>'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ jags 4.3.2-1.2204.0 [1,644 kB]'</li><li>'Fetched 1,644 kB in 0s (8,429 kB/s)'</li><li>'Selecting previously unselected package jags.\\r'</li><li>'(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 121954 files and directories currently installed.)\\r'</li><li>'Preparing to unpack .../jags_4.3.2-1.2204.0_amd64.deb ...\\r'</li><li>'Unpacking jags (4.3.2-1.2204.0) ...\\r'</li><li>'Setting up jags (4.3.2-1.2204.0) ...\\r'</li><li>'Processing triggers for man-db (2.10.2-1) ...\\r'</li><li>'Processing triggers for libc-bin (2.35-0ubuntu3.10) ...\\r'</li></ol>\n"
            ],
            "text/markdown": "1. 'Reading package lists...'\n2. 'Building dependency tree...'\n3. 'Reading state information...'\n4. 'The following NEW packages will be installed:'\n5. <span style=white-space:pre-wrap>'  jags'</span>\n6. '0 upgraded, 1 newly installed, 0 to remove and 9 not upgraded.'\n7. 'Need to get 1,644 kB of archives.'\n8. 'After this operation, 3,492 kB of additional disk space will be used.'\n9. 'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ jags 4.3.2-1.2204.0 [1,644 kB]'\n10. 'Fetched 1,644 kB in 0s (8,429 kB/s)'\n11. 'Selecting previously unselected package jags.\\r'\n12. '(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 121954 files and directories currently installed.)\\r'\n13. 'Preparing to unpack .../jags_4.3.2-1.2204.0_amd64.deb ...\\r'\n14. 'Unpacking jags (4.3.2-1.2204.0) ...\\r'\n15. 'Setting up jags (4.3.2-1.2204.0) ...\\r'\n16. 'Processing triggers for man-db (2.10.2-1) ...\\r'\n17. 'Processing triggers for libc-bin (2.35-0ubuntu3.10) ...\\r'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item 'Reading package lists...'\n\\item 'Building dependency tree...'\n\\item 'Reading state information...'\n\\item 'The following NEW packages will be installed:'\n\\item '  jags'\n\\item '0 upgraded, 1 newly installed, 0 to remove and 9 not upgraded.'\n\\item 'Need to get 1,644 kB of archives.'\n\\item 'After this operation, 3,492 kB of additional disk space will be used.'\n\\item 'Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ jags 4.3.2-1.2204.0 {[}1,644 kB{]}'\n\\item 'Fetched 1,644 kB in 0s (8,429 kB/s)'\n\\item 'Selecting previously unselected package jags.\\textbackslash{}r'\n\\item '(Reading database ... \\textbackslash{}r(Reading database ... 5\\%\\textbackslash{}r(Reading database ... 10\\%\\textbackslash{}r(Reading database ... 15\\%\\textbackslash{}r(Reading database ... 20\\%\\textbackslash{}r(Reading database ... 25\\%\\textbackslash{}r(Reading database ... 30\\%\\textbackslash{}r(Reading database ... 35\\%\\textbackslash{}r(Reading database ... 40\\%\\textbackslash{}r(Reading database ... 45\\%\\textbackslash{}r(Reading database ... 50\\%\\textbackslash{}r(Reading database ... 55\\%\\textbackslash{}r(Reading database ... 60\\%\\textbackslash{}r(Reading database ... 65\\%\\textbackslash{}r(Reading database ... 70\\%\\textbackslash{}r(Reading database ... 75\\%\\textbackslash{}r(Reading database ... 80\\%\\textbackslash{}r(Reading database ... 85\\%\\textbackslash{}r(Reading database ... 90\\%\\textbackslash{}r(Reading database ... 95\\%\\textbackslash{}r(Reading database ... 100\\%\\textbackslash{}r(Reading database ... 121954 files and directories currently installed.)\\textbackslash{}r'\n\\item 'Preparing to unpack .../jags\\_4.3.2-1.2204.0\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking jags (4.3.2-1.2204.0) ...\\textbackslash{}r'\n\\item 'Setting up jags (4.3.2-1.2204.0) ...\\textbackslash{}r'\n\\item 'Processing triggers for man-db (2.10.2-1) ...\\textbackslash{}r'\n\\item 'Processing triggers for libc-bin (2.35-0ubuntu3.10) ...\\textbackslash{}r'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] \"Reading package lists...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              " [2] \"Building dependency tree...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              " [3] \"Reading state information...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
              " [4] \"The following NEW packages will be installed:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              " [5] \"  jags\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
              " [6] \"0 upgraded, 1 newly installed, 0 to remove and 9 not upgraded.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
              " [7] \"Need to get 1,644 kB of archives.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
              " [8] \"After this operation, 3,492 kB of additional disk space will be used.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
              " [9] \"Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ jags 4.3.2-1.2204.0 [1,644 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[10] \"Fetched 1,644 kB in 0s (8,429 kB/s)\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              "[11] \"Selecting previously unselected package jags.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[12] \"(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 121954 files and directories currently installed.)\\r\"\n",
              "[13] \"Preparing to unpack .../jags_4.3.2-1.2204.0_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              "[14] \"Unpacking jags (4.3.2-1.2204.0) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
              "[15] \"Setting up jags (4.3.2-1.2204.0) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
              "[16] \"Processing triggers for man-db (2.10.2-1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[17] \"Processing triggers for libc-bin (2.35-0ubuntu3.10) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     "
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "system(\"apt-get update\", intern = TRUE)\n",
        "system(\"apt-get install -y jags\", intern = TRUE)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "34HU_aGHGkYm",
        "outputId": "f7bd504f-af15-4a5d-de85-51277b75d93e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.21 (BiocManager 1.30.26), R 4.5.0 (2025-04-11)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'infercnv'\n",
            "\n",
            "also installing the dependencies ‘dotCall64’, ‘tensor’, ‘BH’, ‘sitmo’, ‘formatR’, ‘statmod’, ‘TH.data’, ‘sandwich’, ‘sp’, ‘spam’, ‘zoo’, ‘lazyeval’, ‘crosstalk’, ‘RcppTOML’, ‘here’, ‘spatstat.data’, ‘spatstat.univar’, ‘spatstat.random’, ‘spatstat.utils’, ‘spatstat.sparse’, ‘goftest’, ‘abind’, ‘deldir’, ‘polyclip’, ‘FNN’, ‘dqrng’, ‘XVector’, ‘UCSC.utils’, ‘GenomeInfoDbData’, ‘SparseArray’, ‘gtools’, ‘lambda.r’, ‘futile.options’, ‘RcppParallel’, ‘RcppArmadillo’, ‘limma’, ‘locfit’, ‘libcoin’, ‘matrixStats’, ‘modeltools’, ‘mvtnorm’, ‘multcomp’, ‘bitops’, ‘plyr’, ‘globals’, ‘listenv’, ‘parallelly’, ‘iterators’, ‘SeuratObject’, ‘cowplot’, ‘fastDummies’, ‘future.apply’, ‘ggrepel’, ‘ggridges’, ‘ica’, ‘irlba’, ‘leidenbase’, ‘lmtest’, ‘patchwork’, ‘pbapply’, ‘plotly’, ‘png’, ‘progressr’, ‘RcppAnnoy’, ‘RcppHNSW’, ‘reticulate’, ‘ROCR’, ‘RSpectra’, ‘Rtsne’, ‘scattermore’, ‘sctransform’, ‘spatstat.explore’, ‘spatstat.geom’, ‘uwot’, ‘RcppEigen’, ‘RcppProgress’, ‘MatrixGenerics’, ‘GenomicRanges’, ‘Biobase’, ‘S4Vectors’, ‘IRanges’, ‘GenomeInfoDb’, ‘S4Arrays’, ‘DelayedArray’, ‘findpython’, ‘gplots’, ‘futile.logger’, ‘ape’, ‘phyclust’, ‘fastcluster’, ‘parallelDist’, ‘HiddenMarkov’, ‘edgeR’, ‘coin’, ‘caTools’, ‘RANN’, ‘igraph’, ‘reshape2’, ‘rjags’, ‘fitdistrplus’, ‘future’, ‘foreach’, ‘doParallel’, ‘Seurat’, ‘BiocGenerics’, ‘SummarizedExperiment’, ‘SingleCellExperiment’, ‘coda’, ‘gridExtra’, ‘argparse’\n",
            "\n",
            "\n",
            "Old packages: 'curl', 'tibble', 'utf8'\n",
            "\n"
          ]
        }
      ],
      "source": [
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "     install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"infercnv\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OtCBbpFWF04k"
      },
      "outputs": [],
      "source": [
        "library(infercnv)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UfR0bcQDLI2r",
        "outputId": "d7419eb2-6f4d-4d39-d0c2-d349fe79183a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 06:47:45] Parsing matrix: /content/1_and_2.matrix\n",
            "INFO [2025-06-10 06:48:50] Parsing gene order file: gene_pos_file.txt\n",
            "INFO [2025-06-10 06:48:50] Parsing cell annotations file: /content/1_vs_2.annotations.tsv\n",
            "INFO [2025-06-10 06:48:50] ::order_reduce:Start.\n",
            "INFO [2025-06-10 06:48:51] .order_reduce(): expr and order match.\n",
            "INFO [2025-06-10 06:48:52] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 36601,5631 Total=52464582 Min=0 Max=5216.\n",
            "INFO [2025-06-10 06:48:53] num genes removed taking into account provided gene ordering list: 1280 = 3.49717220840961% removed.\n",
            "INFO [2025-06-10 06:48:53] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
            "WARN [2025-06-10 06:48:54] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-06-10 06:48:58] validating infercnv_obj\n"
          ]
        }
      ],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/1_and_2.matrix\",\n",
        "                                    annotations_file=\"/content/1_vs_2.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5iw3iuv5LIkU",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a7cbe074-39c5-4c18-8fd1-9d9cddbcd3dd",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 06:50:04] ::process_data:Start\n",
            "INFO [2025-06-10 06:50:04] Creating output path 1_vs_2\n",
            "INFO [2025-06-10 06:50:04] Checking for saved results.\n",
            "INFO [2025-06-10 06:50:04] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-06-10 06:50:30] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-06-10 06:50:30] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-06-10 06:50:30] Removing 19366 genes from matrix as below mean expr threshold: 0.01\n",
            "INFO [2025-06-10 06:50:31] validating infercnv_obj\n",
            "INFO [2025-06-10 06:50:31] There are 15955 genes and 5630 cells remaining in the expr matrix.\n",
            "INFO [2025-06-10 06:50:35] Removed 1 genes having fewer than 3 min cells per gene = 0.00626763 % genes removed here\n",
            "INFO [2025-06-10 06:50:36] validating infercnv_obj\n",
            "INFO [2025-06-10 06:50:55] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-06-10 06:50:55] normalizing counts matrix by depth\n",
            "INFO [2025-06-10 06:50:58] Computed total sum normalization factor as median libsize: 6972.500000\n",
            "INFO [2025-06-10 06:51:19] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-06-10 06:51:19] transforming log2xplus1()\n",
            "INFO [2025-06-10 06:51:43] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-06-10 06:51:43] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 06:51:43] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 06:51:52] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 06:52:47] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-06-10 06:52:47] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-06-10 06:53:35] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-06-10 06:53:35] smooth_by_chromosome: chr: chr1\n",
            "INFO [2025-06-10 06:53:47] smooth_by_chromosome: chr: chr2\n",
            "INFO [2025-06-10 06:53:57] smooth_by_chromosome: chr: chr3\n",
            "INFO [2025-06-10 06:54:08] smooth_by_chromosome: chr: chr4\n",
            "INFO [2025-06-10 06:54:17] smooth_by_chromosome: chr: chr5\n",
            "INFO [2025-06-10 06:54:26] smooth_by_chromosome: chr: chr6\n",
            "INFO [2025-06-10 06:54:35] smooth_by_chromosome: chr: chr7\n",
            "INFO [2025-06-10 06:54:45] smooth_by_chromosome: chr: chr8\n",
            "INFO [2025-06-10 06:54:54] smooth_by_chromosome: chr: chr9\n",
            "INFO [2025-06-10 06:55:03] smooth_by_chromosome: chr: chr10\n",
            "INFO [2025-06-10 06:55:11] smooth_by_chromosome: chr: chr11\n",
            "INFO [2025-06-10 06:55:22] smooth_by_chromosome: chr: chr12\n",
            "INFO [2025-06-10 06:55:32] smooth_by_chromosome: chr: chr13\n",
            "INFO [2025-06-10 06:55:39] smooth_by_chromosome: chr: chr14\n",
            "INFO [2025-06-10 06:55:48] smooth_by_chromosome: chr: chr15\n",
            "INFO [2025-06-10 06:55:57] smooth_by_chromosome: chr: chr16\n",
            "INFO [2025-06-10 06:56:06] smooth_by_chromosome: chr: chr17\n",
            "INFO [2025-06-10 06:56:16] smooth_by_chromosome: chr: chr18\n",
            "INFO [2025-06-10 06:56:24] smooth_by_chromosome: chr: chr19\n",
            "INFO [2025-06-10 06:56:34] smooth_by_chromosome: chr: chr20\n",
            "INFO [2025-06-10 06:56:43] smooth_by_chromosome: chr: chr21\n",
            "INFO [2025-06-10 06:56:50] smooth_by_chromosome: chr: chr22\n",
            "INFO [2025-06-10 06:56:58] smooth_by_chromosome: chr: GL000194.1\n",
            "INFO [2025-06-10 06:56:58] smooth_by_chromosome: chr: GL000218.1\n",
            "INFO [2025-06-10 06:56:58] smooth_by_chromosome: chr: GL000219.1\n",
            "INFO [2025-06-10 06:56:58] smooth_by_chromosome: chr: KI270721.1\n",
            "INFO [2025-06-10 06:57:46] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-06-10 06:57:46] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-06-10 06:58:49] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-06-10 06:58:49] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 06:58:49] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 06:58:58] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 06:59:55] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-06-10 06:59:55] invert_log2(), computing 2^x\n",
            "INFO [2025-06-10 07:01:09] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-06-10 07:01:09] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-06-10 07:01:14] define_signif_tumor_subclusters(), tumor: CD4 T-cells\n",
            "INFO [2025-06-10 07:01:14] Less cells in group CD4 T-cells than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:14] define_signif_tumor_subclusters(), tumor: CD8 T-cells\n",
            "INFO [2025-06-10 07:01:14] Setting auto leiden resolution for CD8 T-cells to 0.531821\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  DIRAS1, SLC39A3, SGTA, THOP1, ZNF554, ZNF555, ZNF57, AC119403.1, ZNF77, TLE2 \n",
            "\t   AC005944.1, TLE5, GNA11, SLC29A2, MRPL11, AP001107.9, GNA15, B4GAT1, PELI3, BRMS1 \n",
            "\t   AC005264.1, CACTIN, HMG20B, AP002748.4, AC005786.3, C19orf71, RIN1, S1PR4, MFSD12, FZR1 \n",
            "Negative:  EIF2S1, ATP6V1D, AL135978.2, TSG101, SPTY2D1OS, UEVLD, SPTY2D1, LDHAL6A, LDHC, ZDHHC13 \n",
            "\t   CSRP3-AS1, LDHA, GTF2H1, HPS5, KCNC1, NCR3LG1, E2F8, NUCB2, TPH1, SAAL1 \n",
            "\t   PIK3C2A, SERGEF, NAV2, RPS13, PLEKHA7, C11orf58, HTATIP2, PRMT3, ANO5, AC103794.1 \n",
            "PC_ 2 \n",
            "Positive:  ANKS1B, APAF1, UHRF1BP1L, IKBIP, ACTR6, SLC25A3, TMPO, DEPDC4, TMPO-AS1, AC007513.1 \n",
            "\t   LINC02453, CDK17, SCYL2, CFAP54, AC007656.2, AC008149.2, NEDD1, ELK3, GAS2L3, LTA4H \n",
            "\t   HAL, CCDC38, UTP20, SNRPF, USP44, METAP2, VEZT, ARL1, FGD6, NR2C1 \n",
            "Negative:  LMTK2, TECPR1, BRI3, AC079781.5, BAIAP2L1, TRRAP, SDHAF3, AC004893.3, SMURF1, SEM1 \n",
            "\t   ARPC1A, SLC25A13, ARPC1B, PDAP1, DYNC1I1, BUD31, PTCD1, CPSF4, PDK4, ATP5MF \n",
            "\t   AC002451.1, PON2, AC002429.2, PPP1R9A, PEG10, SGCE, CASD1, BET1, GNG11, VPS50 \n",
            "PC_ 3 \n",
            "Positive:  ADPGK, ADPGK-AS1, BBS4, NEO1, LINC02259, ARIH1, CLK3, ARID3B, NPTN-IT1, NPTN \n",
            "\t   LOXL1-AS1, AC100835.2, UBL7-AS1, UBL7, STOML1, EDC3, SEMA7A, TMEM202-AS1, PML, CSK \n",
            "\t   ULK3, HEXA, SCAMP2, CELF6, AC009690.2, PARP6, MPI, PKM, FAM219B, SENP8 \n",
            "Negative:  ST8SIA5, AC090241.2, PIAS2, LOXHD1, RNF165, C18orf25, HAUS1, KATNAL2, ATP5F1A, PSTPIP2 \n",
            "\t   AC012254.1, EPG5, HDHD2, SIGLEC15, SLC14A1, IER3IP1, AC023421.2, SLC14A2, AC090376.1, SMAD2 \n",
            "\t   SETBP1, CTIF, LINC01478, SMAD7, AC110014.1, LINC00907, AC016866.1, PIK3C3, DYM, AC016866.2 \n",
            "PC_ 4 \n",
            "Positive:  PLXNC1, VEZT, AC123567.2, AC073655.1, AC012464.1, FGD6, METAP2, AC073655.2, CEP83, USP44 \n",
            "\t   NR2C1, AC012085.2, NDUFA12, TMCC3, SNRPF, CCDC38, CRADD, HAL, LTA4H, ELK3 \n",
            "\t   SOCS2, AC008149.2, CDK17, SOCS2-AS1, AC007513.1, CFAP54, NEDD1, AC007656.2, LINC02453, TMPO-AS1 \n",
            "Negative:  LINC02666, MGMT, EBF3, C10orf143, LINC01163, GLRX3, MKI67, TCERG1L, PTPRE, PPP2R2D \n",
            "\t   INSYN2A, DOCK1, ADAM12, BNIP3, FANK1, AL162274.2, DHX32, BCCIP, JAKMIP3, UROS \n",
            "\t   MMP21, EDRF1-AS1, EDRF1, EDRF1-DT, STK32C, CTBP2, LRRC27, AL731577.1, PWWP2B, AL451069.2 \n",
            "PC_ 5 \n",
            "Positive:  CNIH1, GMFB, CGRRF1, SAMD4A, GCH1, WDHD1, SOCS4, MAPK1IP1L, LGALS3, DLGAP5 \n",
            "\t   AL158801.2, FBXO34, AL158801.3, ATG14, KTN1-AS1, KTN1, PELI2, AL355073.1, LINC02284, TMEM260 \n",
            "\t   AL161757.5, EXOC5, AP5M1, NAA30, SLC35F4, ARMH4, AC091271.1, CLTC, PTRH2, ACTR10 \n",
            "Negative:  ABCA3, CCNF, AC009065.4, TEDC2, RNPS1, ECI1, TBC1D24, PGP, BRICD5, E4F1 \n",
            "\t   MLST8, ATP6V0C, TRAF7, SNHG19, OVCH1, OVCH1-AS1, ERGIC2, TMTC1, IPO8, AC009318.3 \n",
            "\t   AC009318.1, FAR2, CAPRIN2, AC012150.1, AC010198.2, AC022364.1, CCDC91, TSPAN11, KLHL42, MRPS35 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:16] define_signif_tumor_subclusters(), tumor: GMP\n",
            "INFO [2025-06-10 07:01:16] Less cells in group GMP than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:16] define_signif_tumor_subclusters(), tumor: HSC\n",
            "INFO [2025-06-10 07:01:16] Less cells in group HSC than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:16] define_signif_tumor_subclusters(), tumor: komórki B pamięci\n",
            "INFO [2025-06-10 07:01:16] Setting auto leiden resolution for komórki B pamięci to 0.514215\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  OSBPL1A, CABYR, AC023983.1, AC090772.1, IMPACT, AC090772.2, ZNF521, TTC39C, LAMA3, AC105114.2 \n",
            "\t   NPC1, SS18, RMC1, TAF4B, RIOK3, TMEM241, KCTD1, CABLES1, AQP4-AS1, RBBP8 \n",
            "\t   AC090912.1, PCAT18, MIB1, AC106037.3, ABHD3, SNRPD1, CDH2, ESCO1, GREB1L, DSG2 \n",
            "Negative:  BEST1, RAB3IL1, FTH1, FADS3, INCENP, FADS1, ASRGL1, FADS2, AHNAK, FEN1 \n",
            "\t   AP001363.1, TMEM258, EEF1G, SDHAF2, CPSF7, TUT1, TMEM216, TMEM138, MTA2, CYB561A3 \n",
            "\t   TKFC, EML3, PGPEP1, LSM4, LRRC25, JUND, B3GAT3, SSBP4, IQCN, AC008397.1 \n",
            "PC_ 2 \n",
            "Positive:  TRANK1, STAC, ARPP21, EPM2AIP1, MLH1, LRRFIP2, GOLGA4, C3orf35, COMMD6, ITGA9 \n",
            "\t   UCHL3, ITGA9-AS1, LMO7-AS1, CTDSPL, LMO7, VILL, HSD17B12, AC087521.2, ALKBH3, ACCS \n",
            "\t   TTC17, EXT2, AC087276.2, CD82, API5, AL136441.1, TSPAN18, LRRC4C, PLCD1, TP53I11 \n",
            "Negative:  HEXD-IT1, HEXD, OGFOD3, SECTM1, CD7, AC132872.1, LINC01970, AC132872.3, CSNK1D, SLC16A3 \n",
            "\t   CCDC57, FASN, DUS1L, GPS1, RFNG, DCXR, RAC3, LRRC45, CENPX, ASPSCR1 \n",
            "\t   PYCR1, MAFG, SIRT7, PCYT2, ANAPC11, ALYREF, AC145207.5, ARHGDIA, AC145207.2, P4HB \n",
            "PC_ 3 \n",
            "Positive:  PRMT9, ARHGAP10, NR3C2, AC069272.1, AC002460.2, LINC02355, DCLK2, LRBA, RPS3A, SH3D19 \n",
            "\t   AC095055.1, FAM160A1, GATB, AC092611.1, AC097375.1, FBXW7, AC023424.3, MIR4453HG, TMEM154, ARFIP1 \n",
            "\t   FHDC1, TRIM2, MND1, TMEM131L, AC106865.2, TLR2, RNF175, AC020703.1, PLRG1, AC097467.3 \n",
            "Negative:  TP53RK, SLC13A3, ZNF334, ELMO2, AL133227.1, SLC35C2, CD40, NCOA5, SLC12A5, ZNF335 \n",
            "\t   PCIF1, AL162458.1, B4GALT5, PLTP, SLC9A8, CTSA, SPATA2, RNF114, ZSWIM1, ZSWIM3 \n",
            "\t   SNAI1, ACOT8, UBE2V1, SNX21, TMEM189, TNNC2, CEBPB-AS1, CEBPB, UBE2C, SMIM25 \n",
            "PC_ 4 \n",
            "Positive:  LUZP1, AL031428.1, HNRNPR, KDM1A, EPHB2, ZBTB40, CDC42-IT1, ZNF436, ZNF436-AS1, TCEA3 \n",
            "\t   E2F2, ID3, NCMAP, MDS2, NCMAP-DT, RCAN3, RPL11, ELOA-AS1, RCAN3AS, STPG1 \n",
            "\t   IFNLR1, NIPAL3, HMGCL, GALE, LYPLA2, PITHD1, FUCA1, ELOA, SRSF10, CNR2 \n",
            "Negative:  GAS2, AC006299.1, FANCF, SVIP, ANO5, LINC02718, PRMT3, HTATIP2, LUZP2, NAV2 \n",
            "\t   CSRP3-AS1, E2F8, ZDHHC13, SPTY2D1, SPTY2D1OS, CCDC34, UEVLD, TSG101, LGR4, LDHAL6A \n",
            "\t   LIN7C, LDHC, BDNF-AS, LDHA, KIF18A, GTF2H1, HPS5, METTL15, SAAL1, TPH1 \n",
            "PC_ 5 \n",
            "Positive:  ZDHHC24, AP002748.6, CTSF, BBS1, DPP3, CCS, AP002748.4, RBM14, PELI3, MRPL11 \n",
            "\t   SLC29A2, RBM4, AP001107.9, B4GAT1, RBM4B, BRMS1, SPTBN2, RIN1, YIF1A, C11orf80 \n",
            "\t   AP001107.1, RCE1, RAB1B, PC, KLC2, PACS1, LRFN4, SF3B2, KDM2A, CATSPER1 \n",
            "Negative:  MXD4, ZFYVE28, HAUS3, RNF4, C4orf48, NELFA, FAM193A, NSD2, TNIP2, LETM1 \n",
            "\t   SH3BP2, TMEM129, ADD1, MFSD10, TACC3, NOP14-AS1, AC016773.1, SLBP, FAM53A, AC147067.1 \n",
            "\t   CRIPAK, UVSSA, MAEA, CTBP1-DT, CTBP1, CTBP1-AS, SPON2, AC092535.4, FGFRL1, IDUA \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:17] define_signif_tumor_subclusters(), tumor: Komórki blastyczne (preB/CLP)\n",
            "INFO [2025-06-10 07:01:17] Setting auto leiden resolution for Komórki blastyczne (preB/CLP) to 0.531821\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  GPSM1, CARD9, SNAPC4, QSOX2, ENTR1, PMPCA, TMEM250, INPP5E, NACC2, SEC16A \n",
            "\t   NOTCH1, UBAC1, AL590226.1, EGFL7, AL355574.1, AGPAT2, DIPK1B, CAMSAP1, SNHG7, LCN6 \n",
            "\t   LCN8, TMEM141, CCDC183, MRPS2, RABL6, PHPT1, EDF1, TRAF2, PPP1R26, FBXW5 \n",
            "Negative:  PSMC6, AL133453.1, ERO1A, GPR137C, TXNDC16, PTGER2, PTGDR, NID2, RTRAF, GNG2 \n",
            "\t   FRMD6-AS1, FRMD6, TMX1, TRIM9, PYGL, AL606834.2, NIN, ARF6, AL627171.1, LINC01588 \n",
            "\t   AL627171.2, SAV1, VCPKMT, SOS2, NEMF, L2HGDH, ATL1, DMAC2L, AL118556.1, CDKL1 \n",
            "PC_ 2 \n",
            "Positive:  PARD3B, INO80D, AC007383.2, ICOS, CD28, NDUFS1, RAPH1, EEF1B2, ZDBF2, ADAM23 \n",
            "\t   DYTN, FASTKD2, AC008269.1, KLF7, KLF7-IT1, MYOSLID, AC009226.1, CREB1, METTL21A, CCNYL1 \n",
            "\t   FZD5, PLEKHM3, CRYGD, IDH1, IDH1-AS1, PIKFYVE, PTH2R, RPE, KANSL1L, AC007038.2 \n",
            "Negative:  JMJD6, MXRA7, METTL23, AC005837.1, SRSF2, ST6GALNAC1, ST6GALNAC2, MFSD11, MGAT5B, AC015802.6 \n",
            "\t   SNHG16, AC016168.2, AC015802.4, PRCD, SNHG20, CYGB, RHBDF2, SEC14L1, UBE2O, SPHK1 \n",
            "\t   PRPSAP1, QRICH2, SEPTIN9, UBALD2, TNRC6C, AC111182.1, AC021683.2, AC111170.1, TNRC6C-AS1, RNF157 \n",
            "PC_ 3 \n",
            "Positive:  ZBTB4, POLR2A, CHRNB1, TNFSF12, TMEM102, TNFSF13, NLGN2, SENP3, TMEM256, PLSCR3 \n",
            "\t   ACAP1, NEURL4, GPS2, EIF5A, YBX2, SLC2A4, AL354809.1, LHFPL6, NHLRC3, COG6 \n",
            "\t   FOXO1, LINC00598, PROSER1, UFM1, ELP5, MRPS31, LINC00571, SUPT20H, SLC25A15, CTDNEP1 \n",
            "Negative:  TOPORS, SMIM27, DDX58, NDUFB6, APTX, ACO1, DNAJA1, SMU1, LINGO2, B4GALT1 \n",
            "\t   C9orf72, SPINK4, BAG1, MOB3B, CHMP5, NFX1, AQP3, TEK, NOL6, PTENP1-AS \n",
            "\t   UBE2R2-AS1, LRRC19, UBE2R2, UBAP2, IFT74, DCAF12, UBAP1, PLAA, KIF24, NUDT2 \n",
            "PC_ 4 \n",
            "Positive:  PJA2, MAN2A1, TMEM232, GRB2, AC011933.4, SLC25A19, AC022211.2, SLC25A46, AC011933.2, MIF4GD \n",
            "\t   MRPS7, TMEM94, WDR36, GGA3, TSEN54, NUP85, LLGL2, CAMK4, SUMO2, MYO15B \n",
            "\t   RECQL5, AC010275.1, JPT1, SMIM5, STARD4, NT5C, SMIM6, STARD4-AS1, SAP30BP, ARMC7 \n",
            "Negative:  NDUFAF6, STK32B, CYTL1, STX18-AS1, EVC2, STX18, EVC, NSG1, ZBTB49, CRMP1 \n",
            "\t   LYAR, WFS1, TMEM128, LRPAP1, MAN2B2, RGS12, MRFAP1, MSANTD1, FAM184A, MAN1A1 \n",
            "\t   AL137009.1, HTT-AS, ASF1A, AL078600.1, MCM9, HTT, LINC02482, TBC1D32, GRK4, Z99496.2 \n",
            "PC_ 5 \n",
            "Positive:  DCAF15, RFX1, IL27RA, SAMD1, PRKACA, ASF1B, AC022098.1, ADGRL1, AC011509.2, LINC01841 \n",
            "\t   ADGRE5, AC008569.1, DDX39A, PKN1, GIPC1, DNAJB1, TMEM173, DNAJC18, UBE2D2, SPATA24 \n",
            "\t   MZB1, TECR, CXXC5, SLC23A1, AC135457.1, NDUFB7, AC008667.1, PAIP2, SNHG4, MATR3 \n",
            "Negative:  AC009318.1, FAR2, AC012150.1, AC022364.1, CCDC91, KLHL42, MRPS35, PPFIBP1, ARNTL2-AS1, ARNTL2 \n",
            "\t   STK38L, AC092747.1, AC092747.4, MED21, TM7SF3, FGFR1OP2, INTS13, AC023051.1, ITPR2, AC024145.1 \n",
            "\t   AC022509.2, SSPN, RASSF8, AC087239.1, KRAS, AC022382.2, CPNE9, MTMR14, BRPF1, LHFPL4 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:18] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 1 (preB)\n",
            "INFO [2025-06-10 07:01:18] Less cells in group Komórki blastyczne 1 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:18] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 2 (preB)\n",
            "INFO [2025-06-10 07:01:18] Less cells in group Komórki blastyczne 2 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:18] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 3 (preB)\n",
            "INFO [2025-06-10 07:01:18] Setting auto leiden resolution for Komórki blastyczne 3 (preB) to 0.130032\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AL627171.2, AL627171.1, ARF6, LINC01588, NEMF, VCPKMT, KLHDC2, KLHDC1, SOS2, POLE2 \n",
            "\t   L2HGDH, DNAAF2, TRIM9, TMX1, PYGL, DMAC2L, FRMD6, AL606834.2, CDKL1, FRMD6-AS1 \n",
            "\t   NIN, MAP4K5, SAV1, AL118556.1, ATL1, GNG2, RTRAF, NID2, PTGDR, PTGER2 \n",
            "Negative:  SLITRK5, AL162493.1, MIR17HG, AL590807.1, GPC5, NDFIP2, SPRY2, RBM26-AS1, GPC5-AS2, RBM26 \n",
            "\t   AL356737.2, OBI1, GPC5-AS1, POU4F1, TGDS, MPPED2, DNAJC24, IMMP1L, AL035078.1, AC131571.1 \n",
            "\t   SLAIN1, ELP4, RCN1, ARL14EP, GPR180, AL078612.1, LINC02755, MYCBP2, WT1, METTL15 \n",
            "PC_ 2 \n",
            "Positive:  CHKB-DT, ARSA, CHKB, SHANK3, RABL2B, CPT1B, U62317.4, ODF3B, TYMP, SCO2 \n",
            "\t   NCAPH2, LMF2, ADM2, SBF1, PPP6R2, DENND6B, PLXNB2, MAPK12, HDAC10, TUBGCP6 \n",
            "\t   SELENOO, AL022328.4, THOP1, ZNF554, SGTA, ZNF555, ZNF57, AL022328.3, SLC39A3, AC119403.1 \n",
            "Negative:  RFX7, NEDD4, DNAAF4, C15orf65, AC069368.2, CCPG1, ANKDD1A, PIGB, AC103691.1, PIGBOS1 \n",
            "\t   RAPGEF2, C4orf45, FNIP2, FSTL5, AC018926.2, NAF1, PPID, SPG21, TMA16, ETFDH \n",
            "\t   C4orf46, MARCH1, RXFP1, APELA, TMEM144, RAB27A, TRIM61, FAM198B-AS1, GASK1B, TMEM192 \n",
            "PC_ 3 \n",
            "Positive:  NCMAP, NCMAP-DT, RCAN3, RCAN3AS, NIPAL3, STPG1, IFNLR1, SRSF10, PNRC2, CNR2 \n",
            "\t   FUCA1, HMGCL, GALE, LYPLA2, PITHD1, ELOA, ELOA-AS1, RPL11, MDS2, ID3 \n",
            "\t   E2F2, TCEA3, ZNF436-AS1, ZNF436, HNRNPR, LUZP1, AL031428.1, KDM1A, FANCE, EPHB2 \n",
            "Negative:  KIAA0930, UPK3A, FAM118A, NUP50, SMC1B, RIBC2, NUP50-DT, AC113410.4, Z95331.1, AFF4 \n",
            "\t   HSPA4, ZCCHC10, FSTL4, ATXN10, LEAP2, LINC00899, PRR34-AS1, UQCRQ, MIRLET7BHG, C5orf15 \n",
            "\t   GTSE1, TTC38, TRMU, CDPF1, CELSR1, PPARA, GRAMD4, SHROOM1, CERK, VDAC1 \n",
            "PC_ 4 \n",
            "Positive:  AK7, GSKIP, ATG2B, AL163051.1, AL139020.1, PAPOLA, TCL1A, TCL1B, TCL6, GLRX5 \n",
            "\t   SNHG10, DEGS2, SYNE3, AL133523.1, CLMN, YY1, AL157871.6, DICER1-AS1, SLC25A29, AL157871.4 \n",
            "\t   DICER1, WARS, WDR25, SERPINA1, MEG3, AL132709.8, IFI27L2, LINC02320, PPP2R5C, IFI27 \n",
            "Negative:  ADIPOR2, FBXL14, WNT5B, AC004672.1, ERC1, AC004803.1, RAD52, WNK1, AC021054.1, NINJ2 \n",
            "\t   CCDC77, KDM5A, AC007406.5, AC026369.3, IQSEC3, FBXO48, PLEK, APLF, ARHGAP25, CNRIP1 \n",
            "\t   AC017083.1, GFPT1, PPP3R1, PNO1, WDR92, C1D, ETAA1, PNPLA1, PXT1, BRPF3 \n",
            "PC_ 5 \n",
            "Positive:  VDAC1, AC008608.2, TCF7, SKP1, C5orf15, PPP2CA, FSTL4, CDKL3, HSPA4, UBE2B \n",
            "\t   AC113410.4, AC109454.2, ZCCHC10, AFF4, CDKN2AIPNL, LEAP2, UQCRQ, AC005355.1, SHROOM1, SEPTIN8 \n",
            "\t   JADE2, KIF3A, AC004039.1, SAR1B, TH2LCRR, RAD50, SEC24A, IRF1, CAMLG, AC116366.1 \n",
            "Negative:  ARID4A, AL139021.2, PSMA3-AS1, PSMA3, TOMM20L, ACTR10, ARMH4, AL139021.1, SLC35F4, NAA30 \n",
            "\t   TIMM9, AP5M1, KIAA0586, EXOC5, LINC01500, AL161757.5, DAAM1, TMEM260, GPR135, LINC02284 \n",
            "\t   L3HYPDH, AL355073.1, JKAMP, PELI2, CCDC175, KTN1, KTN1-AS1, RTN1, ATG14, AL157911.1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:20] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 4 (preB)\n",
            "INFO [2025-06-10 07:01:20] Setting auto leiden resolution for Komórki blastyczne 4 (preB) to 0.0650662\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  VAV2, SARDH, BRD3OS, SLC2A6, BRD3, ADAMTS13, WDR5, REXO4, RXRA, STKLD1 \n",
            "\t   COL5A1, SURF4, FCN1, PPP1R26, SURF2, SURF1, MRPS2, RPL7A, CAMSAP1, AL355574.1 \n",
            "\t   MED22, UBAC1, NACC2, SURF6, TMEM250, ABO, QSOX2, GPSM1, GBGT1, CARD9 \n",
            "Negative:  CHRAC1, AGO2, AC107375.1, PTK2, DENND3, AC040970.1, SLC45A4, ATRNL1, AC016042.1, TRUB1 \n",
            "\t   AL137025.1, FAM160B1, ENO4, ABLIM1, SHTN1, VWA2, SLC18A2, AC011676.3, CCDC186, AL391988.1 \n",
            "\t   PDZD8, AL592546.3, AC005871.2, EMX2OS, PTP4A3, NHLRC2, RAB11FIP2, AC022395.1, DCLRE1A, CASC2 \n",
            "PC_ 2 \n",
            "Positive:  TMEM131L, MND1, AC106865.2, TRIM2, TLR2, FHDC1, RNF175, ARFIP1, AC020703.1, TMEM154 \n",
            "\t   PLRG1, MIR4453HG, AC023424.3, AC097467.3, FBXW7, AC097375.1, AC092611.1, GATB, FAM160A1, AC095055.1 \n",
            "\t   SH3D19, DNTTIP2, GCLM, RPS3A, ARHGAP29, ABCD3, F3, CNN3, LRBA, ALG14 \n",
            "Negative:  MAP4K2, AP001462.1, MEN1, SF1, CDC42BPG, PYGM, EHD1, RASGRP2, NRXN2, ATG2A \n",
            "\t   PPP2R5B, ARL2, SNX15, SAC3D1, AP005273.2, NAALADL1, CDCA5, ZFPL1, RPS6KA4, VPS51 \n",
            "\t   TM7SF2, CCDC88B, ZNHIT2, FAU, SYVN1, AP003774.2, MRPL49, AP003068.2, PRDX5, CAPN1 \n",
            "PC_ 3 \n",
            "Positive:  AP001505.1, ADARB1, FAM207A, POFUT2, LINC00205, ITGB2-AS1, ITGB2, COL18A1, PTTG1IP, SLC19A1 \n",
            "\t   SUMO3, LINC01694, UBE2G2, PCBP3, TRPM2, AJ011931.2, AP001062.1, COL6A2, CFAP410, PFKL \n",
            "\t   SPATC1L, AP001059.2, LSS, AP001059.3, MCM3AP-AS1, ICOSLG, MCM3AP, AP001057.1, LINC01678, GATD3A \n",
            "Negative:  FAT4, ANKRD50, AC104664.1, LINC01091, INTU, SPATA5, HSPA4L, SPRY1, NUDT6, SNHG8 \n",
            "\t   LARP7, NDST3, METTL14, KIAA1109, AC093765.2, CAMK2D, ZGRF1, BBS7, SEC24D, AC073475.1 \n",
            "\t   PLK4, ALPK1, CCNA2, EXOSC9, ANXA5, MAD2L1, SYNPO2, PDE5A, C4orf3, USP53 \n",
            "PC_ 4 \n",
            "Positive:  FOXRED1, SRPRA, TIRAP, FAM118B, RPUSD4, AP001318.2, PUS3, HYLS1, DCPS, CHEK1 \n",
            "\t   STT3A, GSEC, EI24, FEZ1, ST3GAL4, TMEM218, KIRREL3, SLC37A2, ETS1, CCDC15 \n",
            "\t   AP001122.1, ROBO4, FLI1, SENCR, ROBO3, ARHGAP32, NFRKB, MSANTD2, PRDM10, APLP2 \n",
            "Negative:  RPRD1B, TTI1, TGM2, BPI, AL391095.1, CTNNBL1, SNHG17, BLCAP, SNHG11, SRC \n",
            "\t   RALGAPB, MANBAL, ACTR5, RPN2, MROH8, PPP1R16B, FAM83D, RBL1, DHX35, SAMHD1 \n",
            "\t   MAFB, TOP1, TLDC2, PLCG1-AS1, PLCG1, SOGA1, ZHX3, DSN1, CHD6, PTPRT \n",
            "PC_ 5 \n",
            "Positive:  GPSM1, CARD9, QSOX2, SNAPC4, ENTR1, TMEM250, PMPCA, NACC2, INPP5E, UBAC1 \n",
            "\t   SEC16A, AL355574.1, CAMSAP1, MRPS2, NOTCH1, PPP1R26, AL590226.1, FCN1, COL5A1, RXRA \n",
            "\t   EGFL7, WDR5, AGPAT2, BRD3, BRD3OS, DIPK1B, VAV2, SNHG7, LCN6, SARDH \n",
            "Negative:  TENM4, PRCP, DDIAS, RAB30, NARS2, RAB30-DT, LINC02728, AP001767.3, AP003086.1, PCF11 \n",
            "\t   GAB2, AP000873.2, USP35, ANKRD42, KCTD21, CCDC90B, KCTD21-AS1, AP000446.1, ALG8, DLG2 \n",
            "\t   NDUFC2, INTS4, AP001825.1, AP002812.2, TMEM126B, AAMDC, RSF1, TMEM126A, AQP11, CREBZF \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:22] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 5 (preB)\n",
            "INFO [2025-06-10 07:01:22] Setting auto leiden resolution for Komórki blastyczne 5 (preB) to 0.482523\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  NIBAN2, LRSAM1, RPL12, STXBP1, PTRH1, ZNF79, TOR2A, SH2D3C, SLC2A8, GARNL3 \n",
            "\t   RALGPS1, ZBTB34, MAP2K2, ZBTB7A, ZBTB43, PIAS4, EEF2, DAPK3, MATK, MVB12B \n",
            "\t   MRPL54, APBA3, PIP5K1C, CACTIN, HMG20B, AC005786.3, C19orf71, PBX3, MFSD12, FZR1 \n",
            "Negative:  AC018752.1, COMMD10, AC034236.2, STARD4-AS1, STARD4, AC010275.1, NREP, AP3S1, CAMK4, TMED7 \n",
            "\t   EPB41L4A, AC010226.1, ATG12, TICAM2, WDR36, SRP19, FEM1C, SLC25A46, EPB41L4A-AS1, APC \n",
            "\t   REEP5, TMEM232, DCP2, CCDC112, MCC, YTHDC2, PGGT1B, TRIM36, MAN2A1, ZNF608 \n",
            "PC_ 2 \n",
            "Positive:  EEF1D, NAPRT, TIGD5, MROH6, GSDMD, PYCR3, AC067930.8, TSTA3, ZC3H3, ZNF623 \n",
            "\t   RHPN1, ZNF707, TOP1MT, ZNF696, MINCR, AC105219.1, GLI4, ANO2, VWF, CD9 \n",
            "\t   ZFP41, AC006206.2, LY6H, SCRIB, LY6E, GALNT8, PLEKHG6, PUF60, NDUFA9, THEM6 \n",
            "Negative:  AC006299.1, LINC02718, SVIP, LUZP2, GAS2, CCDC34, LGR4, FANCF, LIN7C, BDNF-AS \n",
            "\t   ANO5, KIF18A, METTL15, LINC02755, PRMT3, ARL14EP, MPPED2, DNAJC24, IMMP1L, ELP4 \n",
            "\t   AC131571.1, AL035078.1, RCN1, AL078612.1, WT1, WT1-AS, EIF3M, CCDC73, PRRG4, QSER1 \n",
            "PC_ 3 \n",
            "Positive:  CBR3, CBR3-AS1, CBR1, DOP1B, SETD4, AP000692.2, AP000331.1, MORC3, RUNX1, AP000692.1 \n",
            "\t   RCAN1, KCNE1, AP000317.1, SLC5A3, CHAF1B, MRPS6, HLCS, LINC00649, ATP5PO, AP000704.1 \n",
            "\t   ITSN1, PIGP, CRYZL1, TTC3, DONSON, TTC3-AS1, SON, VPS26C, GART, AP001437.2 \n",
            "Negative:  CXXC5, CHI3L2, C1orf162, POT1, POT1-AS1, ZNF800, DENND2D, OVGP1, TMIGD3, AL390195.1 \n",
            "\t   AC006333.2, GCC1, RAP1A, WDR77, ATP5PB, WASL, ARF5, INKA2, UBE2D2, LMOD2 \n",
            "\t   DDX20, FSCN3, CTTNBP2NL, NDUFA5, SND1, TMEM173, WNT2B, AASS, SND1-IT1, ST7L \n",
            "PC_ 4 \n",
            "Positive:  FAM153CP, N4BP3, RMND5B, AC106795.2, NHP2, AC106795.3, HNRNPAB, PHYKPL, FAM153A, COL23A1 \n",
            "\t   CLK4, ZNF354A, AC139795.2, ZNF354B, ZFP2, B4GALT7, ZNF879, TMED9, ZNF354C, ADAMTS2 \n",
            "\t   FAM193B, RUFY1, DDX41, AC136604.2, DOK3, HNRNPH1, CANX, MAML1, PDLIM7, LTC4S \n",
            "Negative:  UGGT2, MBNL2, AL161430.1, AL442067.3, RAP2A, IPO5, FARP1, STK24, DOCK9, UBAC2-AS1 \n",
            "\t   UBAC2, GPR18, AL161891.1, LINC02343, LINC00457, RFC3, NBEA, AL139383.1, AL138999.2, SOHLH2 \n",
            "\t   STARD13-AS, STARD13, GPR183, CCDC169, KL, SPART, SPART-AS1, TM9SF2, CCNA1, RFXAP \n",
            "PC_ 5 \n",
            "Positive:  HAL, CCDC38, LTA4H, ELK3, CDK17, AC008149.2, SNRPF, AC007513.1, USP44, CFAP54 \n",
            "\t   NEDD1, METAP2, VEZT, AC007656.2, FGD6, LINC02453, NR2C1, TMPO-AS1, NDUFA12, TMPO \n",
            "\t   TMCC3, CEP83, SLC25A3, IKBIP, APAF1, ZNF736, ANKS1B, ZNF727, LINC02848, CHCHD2 \n",
            "Negative:  SMG5, UBQLN4, LAMTOR2, PAQR6, MEX3A, BGLAP, SSR2, LMNA, PMF1, SLC25A44 \n",
            "\t   AL355388.3, SEMA4A, ARHGEF2, KHDC4, RIT1, SYT11, GON4L, DAP3, YY1AP1, AL353807.5 \n",
            "\t   MSTO1, ASH1L-AS1, ASH1L, RUSC1, TGIF2, DLGAP4-AS1, DLGAP4, AAR2, EPB41L1, NORAD \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:23] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 6 (preB)\n",
            "INFO [2025-06-10 07:01:23] Less cells in group Komórki blastyczne 6 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:23] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 7 (preB)\n",
            "INFO [2025-06-10 07:01:23] Less cells in group Komórki blastyczne 7 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:23] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 8 (preB)\n",
            "INFO [2025-06-10 07:01:23] Setting auto leiden resolution for Komórki blastyczne 8 (preB) to 0.497824\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  SLC38A10, NDUFAF8, LINC00482, AC027601.6, TEPSIN, CEP131, AC110285.2, AATK, BAHCC1, BAIAP2 \n",
            "\t   ACTG1, BAIAP2-DT, FAAP100, CHMP6, RPTOR, ENDOV, NPLOC4, RNF213-AS1, PDE6G, RNF213 \n",
            "\t   SLC26A11, OXLD1, SGSH, AC087741.1, CCDC137, EIF4A3, GAA, FDXR, ARL16, CCDC40 \n",
            "Negative:  AC016723.1, STK39, AC073050.1, CERS6, XIRP2, CERS6-AS1, SCN7A, NOSTRIN, SCN9A, SPC25 \n",
            "\t   SCN1A-AS1, DHRS9, TTC21B-AS1, TTC21B, AC009495.1, AC009495.3, GALNT3, SCN3A, AC019197.1, COBLL1 \n",
            "\t   GRB14, GCA, IFIH1, DPP4, SLC4A10, PSMD14, LINC01806, ASTN2-AS1, TRIM32, ASTN2 \n",
            "PC_ 2 \n",
            "Positive:  MNT, SGSM2, TSR1, METTL16, SRR, PAFAH1B1, AC090617.4, SMG6, DPH1, RPA1 \n",
            "\t   CLUH, RAP1GAP2, OR3A2, TRPV1, SHPK, CTNS, TAX1BP3, EMC6, P2RX5, ITGAE \n",
            "\t   HASPIN, NCBP3, CAMKK1, P2RX1, ATP2A3, ZZEF1, CYB5D2, ANKFY1, UBE2G1, SPNS3 \n",
            "Negative:  ARFGAP3, A4GALT, PACSIN2, ATP5MGL, TTLL1, CYB5R3, Z93241.1, BIK, POLDIP3, SERHL2 \n",
            "\t   MCAT, RRP7A, NFAM1, TSPO, TCF20, NDUFA6, NDUFA6-DT, SMDT1, PHETA2, TTLL12 \n",
            "\t   NAGA, EFCAB6, WBP2NL, SAMM50, PARVB, CENPM, AL033543.1, AL031595.3, PARVG, AL031595.2 \n",
            "PC_ 3 \n",
            "Positive:  RPS17, AC105339.3, GOLGA6L9, AC105339.6, SAXO2, EFL1, SNHG21, MEX3B, AC104041.1, WHAMM \n",
            "\t   TMC3-AS1, HOMER2, AC103858.3, AC022558.2, STARD5, RAMAC, AC103858.1, C15orf40, BTBD1, IL16 \n",
            "\t   AC024270.5, CFAP161, TM6SF1, HDGFL3, AC103876.1, TLNRD1, SH3GL3, GOLGA6L4, MESD, AC048382.5 \n",
            "Negative:  AC009226.1, CREB1, MYOSLID, METTL21A, KLF7-IT1, CCNYL1, KLF7, FZD5, PLEKHM3, CRYGD \n",
            "\t   IDH1, AC008269.1, FASTKD2, IDH1-AS1, RSRC1, AC106707.1, PTX3, AC092944.1, VEPH1, MLF1 \n",
            "\t   DYTN, PIKFYVE, CCNL1, GFM1, ADAM23, LXN, LEKR1, RARRES1, PTH2R, ZDBF2 \n",
            "PC_ 4 \n",
            "Positive:  RAD23B, AL807761.4, KLF4, ZNF462, ELP1, LINC01505, ABITRAM, TMEM38B, CTNNAL1, FKTN \n",
            "\t   TMEM245, PALM2-AKAP2, FSD1L, TXN, LPAR1, SLC44A1, ECPAS, ZNF483, AL591506.1, PTGR1 \n",
            "\t   CT70, ABCA1, DNAJC25, NIPSNAP3B, NIPSNAP3A, SMC2, GNG10, GRIN3A, RNF20, TMEM246 \n",
            "Negative:  CIB2, IDH3A, TBC1D2B, ACSBG1, HMG20A, PEAK1, AC090181.2, TSPAN3, PSTPIP1, RCN2 \n",
            "\t   SCAPER, AC027243.3, ETFA, TMEM266, NRG4, EFL1, FBXO22, MEX3B, SAXO2, AC104041.1 \n",
            "\t   GOLGA6L9, TMC3-AS1, AC105339.3, RPS17, AC103858.3, UBE2Q2, STARD5, AC105339.6, AC103858.1, SNHG21 \n",
            "PC_ 5 \n",
            "Positive:  SNX8, EIF3B, NUDT1, MRM2, ZFAND2A, MAD1L1, AC091729.3, PSMG3-AS1, PSMG3, CHST12 \n",
            "\t   GPER1, MICALL2, TMEM184A, INTS1, MAFK, GPR146, LFNG, AC073957.1, C7orf50, BRAT1 \n",
            "\t   COX19, IQCE, ADAP1, AC073957.3, TTYH3, GET4, SUN1, DNAAF5, AC147651.4, AMZ1 \n",
            "Negative:  DDX52, MRPL45, GON7, TMEM251, MOAP1, PSMC1, NRDE2, KCNK13, ITPK1, EFCAB11 \n",
            "\t   TDP1, FOXN3-AS2, GOLGA5, FOXN3-AS1, AL356805.1, AL512791.2, AL357093.2, LGMN, CPSF2, CALM1 \n",
            "\t   SLC24A4, ATXN3, RIN3, NDUFB1, TRIP11, FOXN3, TTC7B, TC2N, DGLUCY, GPR68 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:24] define_signif_tumor_subclusters(), tumor: Komórki blastyczne proliferujące\n",
            "INFO [2025-06-10 07:01:24] Setting auto leiden resolution for Komórki blastyczne proliferujące to 0.298471\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  POLR3D, BMP1, REEP4, NUDT18, FAM160B2, DMTN, NPM2, XPO7, DOK2, AC018541.1 \n",
            "\t   ATP6V1B2, LPL, INTS10, AC090541.1, CSGALNACT1, SH2D4A, AC009884.1, AC100800.1, PSD3, NAT1 \n",
            "\t   AC124242.1, ASAH1, PCM1, MTUS1, MTMR7, KIF13B, WASHC5, VPS37A, SQLE, AC108449.1 \n",
            "Negative:  SS18L1, KLHL26, SNAPC4, ENTR1, CARD9, PSMA7, GPSM1, PMPCA, INPP5E, QSOX2 \n",
            "\t   TMEM250, NACC2, SEC16A, UBAC1, LSM14B, NOTCH1, AL355574.1, AL590226.1, REX1BD, CAMSAP1 \n",
            "\t   EGFL7, AGPAT2, TAF4, DIPK1B, CRLF1, SNHG7, CDH4, LCN6, PACS1, KLC2 \n",
            "PC_ 2 \n",
            "Positive:  SMARCE1, KRT10, CCR7, TOP2A, IGFBP4, TMEM99, KRT13, RARA-AS1, EIF1, AC080112.4 \n",
            "\t   JUP, RARA, NT5C3B, KIFC2, CDC6, KLHL11, ACLY, TTC25, CNP, DNAJC7 \n",
            "\t   NKIRAS2, ZNF385C, DHX58, KAT2A, RAB5C, GHDC, STAT5B, AC099811.1, STAT5A, STAT3 \n",
            "Negative:  ECM2, AL157827.2, IPPK, BICD2, ZNF484, FGD3, SUSD3, CARD19, NINJ1, FAM120AOS \n",
            "\t   FAM120A, PHF2, POU4F1, RBM26, OBI1, SLAIN1, RBM26-AS1, MYCBP2, NDFIP2, PTPDC1 \n",
            "\t   LINC01806, SPRY2, MYCBP2-AS1, AL590807.1, AL158152.1, AL162493.1, FBXL3, SLITRK5, PSMD14, MIR17HG \n",
            "PC_ 3 \n",
            "Positive:  RNF213-AS1, RNF213, ENDOV, RPTOR, SLC26A11, CHMP6, SGSH, AC087741.1, EIF4A3, GAA \n",
            "\t   CCDC40, TBC1D16, CBX4, CBX8, CBX2, RBFOX3, ENGASE, CANT1, ACTG1, FAAP100 \n",
            "\t   NPLOC4, PDE6G, OXLD1, CCDC137, ARL16, TK1, HGS, SYNGR2, AC139530.1, C17orf99 \n",
            "Negative:  DNAAF2, POLE2, KLHDC1, KLHDC2, AL627171.1, AL627171.2, NEMF, ARF6, LINC01588, VCPKMT \n",
            "\t   SOS2, L2HGDH, DMAC2L, CDKL1, MAP4K5, AL118556.1, ATL1, SAV1, NIN, AL606834.2 \n",
            "\t   PYGL, TRIM9, TMX1, FRMD6, FRMD6-AS1, GNG2, RTRAF, PIK3R6, MYH10, NID2 \n",
            "PC_ 4 \n",
            "Positive:  SEPHS1, BEND7, PRPF18, PHYH, AL157392.3, MCM10, FRMD4A, FAM107B, OPTN, CDNF \n",
            "\t   CCDC3, HSPA14, AC069544.1, CAMK1D, SUV39H2, AL512770.1, CDC123, DCLRE1C, NUDT5, OLAH \n",
            "\t   SEC61A2, ACBD7, RPP38, NMT2, FAM171A1, MINDY3, PTER, AL353576.1, RSU1, CUBN \n",
            "Negative:  TMEM140, CYREN, AGBL3, AC083870.1, CALD1, BPGM, AKR1B1, SLC35B4, LRGUK, EXOC4 \n",
            "\t   AC009365.4, CHCHD3, PLXNA4, PODXL, AC008264.2, MKLN1-AS, MKLN1, LINC-PINT, AC058791.1, AC016831.5 \n",
            "\t   AC016831.1, LINC00513, AC016831.7, AUXG01000058.1, COPG2, MEST, CEP41, AC087071.2, TMEM209, KLHDC10 \n",
            "PC_ 5 \n",
            "Positive:  RPA2, THEMIS2, PPP1R8, STX12, FAM76A, IFI6, FGR, AHDC1, FO393419.2, WASF2 \n",
            "\t   MAP3K6, SYTL1, TMEM222, WDTC1, SLC9A1, NUDC, GPATCH3, GPN2, ZDHHC18, PIGV \n",
            "\t   ARID1A, RPS6KA1, HMGN2, AL513365.2, EEF1AKMT2, FAM53B-AS1, DHDDS, FAM53B, LHPP, OAT \n",
            "Negative:  HS3ST1, AC006230.1, AC005699.1, RAB28, BOD1L1, LINC00504, CPEB2-DT, CPEB2, AC098829.1, C1QTNF7 \n",
            "\t   CC2D2A, FBXL5, FAM200B, BST1, CD38, FGFBP2, PROM1, TAPT1, AC108063.1, TAPT1-AS1 \n",
            "\t   AC006427.2, AC097515.1, LDB2, QDPR, LAP3, AC006160.1, PIK3R6, MED28, MYH10, NDEL1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: komórki dendrytyczne mieloidalne\n",
            "INFO [2025-06-10 07:01:25] Less cells in group komórki dendrytyczne mieloidalne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: komórki NK T\n",
            "INFO [2025-06-10 07:01:25] Less cells in group komórki NK T than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: monocyty\n",
            "INFO [2025-06-10 07:01:25] Less cells in group monocyty than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: MPP\n",
            "INFO [2025-06-10 07:01:25] Less cells in group MPP than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: prekursory erytroidalne i megakariocytarne\n",
            "INFO [2025-06-10 07:01:25] Less cells in group prekursory erytroidalne i megakariocytarne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: wczesne komórki erytroidalne\n",
            "INFO [2025-06-10 07:01:25] Less cells in group wczesne komórki erytroidalne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:01:25] define_signif_tumor_subclusters(), tumor: control\n",
            "INFO [2025-06-10 07:01:25] Setting auto leiden resolution for control to 0.0057241\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  MBNL1, GMPS, MBNL1-AS1, AC108718.1, P2RY12, P2RY13, IGSF10, AC026347.1, SUCNR1, SLC33A1 \n",
            "\t   AADACL2-AS1, GPR87, P2RY1, P2RY14, C3orf33, RAP2B, PLCH1, ARHGEF26, DHX36, MME-AS1 \n",
            "\t   MME, ATP2A2, ARPC3, ANAPC7, IFT81, GPN3, C12orf76, ANKRD13A, FAM216A, GIT2 \n",
            "Negative:  CARMIL2, ACD, PARD6A, ENKD1, C16orf86, GFOD2, RANBP10, CENPT, THAP11, NUTF2 \n",
            "\t   EDC4, PSKH1, CTRL, PSMB10, LCAT, SLC12A4, DPEP2, DUS2, DDX28, NFATC3 \n",
            "\t   AC020978.3, ASPHD1, CDIPT, KCTD13, MVP, TAOK2, PAGR1, TMEM219, HIRIP3, AC009133.1 \n",
            "PC_ 2 \n",
            "Positive:  MARK1, MARC1, RAB3GAP2, HLX, HLX-AS1, IARS2, DUSP10, TAF1A, BPNT1, MIA3 \n",
            "\t   EPRS, AIDA, LYPLAL1, BROX, LYPLAL1-DT, FAM177B, C1orf143, AL392172.2, TGFB2, AL392172.1 \n",
            "\t   RRP15, DISP1, GPATCH2, TLR5, USH2A, SUSD4, KCTD3, CAPN2, CENPF, TP53BP2 \n",
            "Negative:  NPIPB2, SNX29, AC010333.2, AC007216.3, CPPED1, GSPT1, AC007216.4, AC109597.1, NOMO1, RRN3 \n",
            "\t   NPIPA1, NTAN1, BFAR, RSL1D1, PDXDC1, MPV17L, PARN, MRTFB, ERCC4, BMERB1 \n",
            "\t   ZC3H7A, MARF1, NDE1, AC026401.3, AC007613.1, MYH11, TXNDC11, AF001548.2, SNN, LITAF \n",
            "PC_ 3 \n",
            "Positive:  TRMT44, AC108519.1, AC105916.1, ACOX3, SH3TC1, SLC2A9, ABLIM2, AC005674.2, AFAP1, WDR1 \n",
            "\t   SORCS2, ZNF518B, GRPEL1, AC005599.1, CLNK, TADA2B, HS3ST1, AC097382.2, AC006230.1, AC005699.1 \n",
            "\t   AC097382.3, RAB28, TBC1D14, BOD1L1, KIAA0232, LINC00504, CPEB2-DT, BLOC1S4, CPEB2, MRFAP1L1 \n",
            "Negative:  KIFC2, CYHR1, TONSL, RGPD1, CD8B, PLGLB1, AC092651.3, LINC01943, CD8A, RMND5A \n",
            "\t   CYTOR, VPS28, RNF103, PLGLB2, AC015971.1, RGPD2, KRCC1, AC104134.1, CHMP3, EIF2AK3 \n",
            "\t   MAL, MRPS5, ZNF514, EIF2AK3-DT, RPIA, IGKC, ZNF2, AC092835.1, FAHD2A, LINC00342 \n",
            "PC_ 4 \n",
            "Positive:  PYHIN1, IFI16, AIM2, MNDA, ACKR1, FCER1A, DUSP23, FCRL6, CFAP45, SPTA1 \n",
            "\t   TAGLN2, LINC01133, PIGM, IGSF8, CD1C, AL121987.2, PEA15, DCAF8, CD1D, PEX19 \n",
            "\t   COPA, KIRREL1, NCSTN, VANGL2, SLAMF6, FCRL1, AL138930.1, CD84, SLAMF1, FCRL3 \n",
            "Negative:  IGLC2, IGLC1, IGLC3, RSPH14, IGLV3-21, BCR, PRAME, ZNF280B, IGLL1, AC245060.5 \n",
            "\t   AC245060.6, AP000345.2, VPREB1, RGL4, IGLV10-54, ZNF70, TOP3B, VPREB3, AC245452.1, CHCHD10 \n",
            "\t   PPM1F, MMP11, MAPK1, YPEL1, SMARCB1, PPIL2, DERL3, AP000553.2, SLC2A11, SDF2L1 \n",
            "PC_ 5 \n",
            "Positive:  CASP4, AC008574.1, AC113418.1, HSD17B4, TNFAIP8, PRR16, CASP1, DMXL1, AC008568.1, CARD16 \n",
            "\t   AC008629.1, SRFBP1, MSANTD4, DTWD2, KBTBD3, AASDHPPT, LINC02208, LOX, AP000766.1, LINC02147 \n",
            "\t   CWF19L2, SNX2, SEMA6A-AS1, ALKBH8, SLC35F2, SNX24, SEMA6A, RAB39A, AC018752.1, CEP120 \n",
            "Negative:  EGFL7, AGPAT2, DIPK1B, SNHG7, LCN6, LCN8, TMEM141, CCDC183, RABL6, PHPT1 \n",
            "\t   EDF1, PCYT2, SIRT7, MAFG, PYCR1, ASPSCR1, TRAF2, CENPX, LRRC45, FBXW5 \n",
            "\t   RAC3, DCXR, PTGDS, RFNG, GPS1, LCNL1, DUS1L, PAXX, FASN, CLIC3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:03:14] ::plot_cnv:Start\n",
            "INFO [2025-06-10 07:03:14] ::plot_cnv:Current data dimensions (r,c)=15954,5630 Total=89896366.4199627 Min=0.815313367210792 Max=8.67281305982181.\n",
            "INFO [2025-06-10 07:03:14] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 07:03:17] plot_cnv(): auto thresholding at: (0.896486 , 1.105191)\n",
            "INFO [2025-06-10 07:03:32] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 07:03:32] Observation data size: Cells= 724 Genes= 15954\n",
            "INFO [2025-06-10 07:03:32] clustering observations via method: ward.D\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(1) is 12\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  12,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(2) is 25\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  25,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(3) is 5\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  5,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(4) is 15\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  15,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(5) is 26\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  26,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(6) is 25\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  25,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(7) is 9\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  9,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(8) is 6\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  6,15954\n",
            "INFO [2025-06-10 07:03:32] Number of cells in group(9) is 59\n",
            "INFO [2025-06-10 07:03:32] group size being clustered:  59,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(10) is 57\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  57,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(11) is 13\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  13,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(12) is 88\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  88,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(13) is 84\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  84,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(14) is 78\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  78,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(15) is 35\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  35,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(16) is 2\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  2,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(17) is 1\n",
            "INFO [2025-06-10 07:03:33] Skipping group: 17, since less than 2 entries\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(18) is 1\n",
            "INFO [2025-06-10 07:03:33] Skipping group: 18, since less than 2 entries\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(19) is 28\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  28,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(20) is 10\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  10,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(21) is 4\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  4,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(22) is 27\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  27,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(23) is 27\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  27,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(24) is 15\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  15,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(25) is 7\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  7,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(26) is 4\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  4,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(27) is 8\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  8,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(28) is 20\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  20,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(29) is 12\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  12,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(30) is 16\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  16,15954\n",
            "INFO [2025-06-10 07:03:33] Number of cells in group(31) is 5\n",
            "INFO [2025-06-10 07:03:33] group size being clustered:  5,15954\n",
            "INFO [2025-06-10 07:03:33] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 07:03:33] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 07:03:33] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:03:33] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:03:36] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:03:36] Quantiles of plotted data range: 0.896486380417822,0.972924999033669,0.998376818305305,1.02632746113481,1.10519132085884\n",
            "INFO [2025-06-10 07:03:37] plot_cnv_references:Start\n",
            "INFO [2025-06-10 07:03:37] Reference data size: Cells= 4906 Genes= 15954\n",
            "INFO [2025-06-10 07:03:54] plot_cnv_references:Number reference groups= 46\n",
            "INFO [2025-06-10 07:03:55] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 07:04:18] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:04:18] Quantiles of plotted data range: 0.896486380417822,0.97586068281717,0.998364280379089,1.02284270676345,1.10519132085884\n",
            "INFO [2025-06-10 07:05:35] ::plot_cnv:Start\n",
            "INFO [2025-06-10 07:05:35] ::plot_cnv:Current data dimensions (r,c)=15954,5630 Total=89896366.4199627 Min=0.815313367210792 Max=8.67281305982181.\n",
            "INFO [2025-06-10 07:05:35] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 07:05:38] plot_cnv(): auto thresholding at: (0.896486 , 1.105191)\n",
            "INFO [2025-06-10 07:05:41] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 07:05:41] Observation data size: Cells= 724 Genes= 15954\n",
            "INFO [2025-06-10 07:05:41] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 07:05:41] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 07:05:41] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:05:41] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:05:43] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:05:43] Quantiles of plotted data range: 0.896486380417822,0.972924999033669,0.998376818305305,1.02632746113481,1.10519132085884\n",
            "INFO [2025-06-10 07:05:45] plot_cnv_references:Start\n",
            "INFO [2025-06-10 07:05:45] Reference data size: Cells= 4906 Genes= 15954\n",
            "INFO [2025-06-10 07:25:55] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 07:25:56] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 07:26:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:26:19] Quantiles of plotted data range: 0.896486380417822,0.97586068281717,0.998364280379089,1.02284270676345,1.10519132085884\n",
            "INFO [2025-06-10 07:26:25] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-06-10 07:26:25] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-06-10 07:26:25] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 07:26:28] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.945409717800508 - 1.05595546686792\n",
            "INFO [2025-06-10 07:27:08] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-06-10 07:27:08] ::plot_cnv:Start\n",
            "INFO [2025-06-10 07:27:08] ::plot_cnv:Current data dimensions (r,c)=15954,5630 Total=90074961.4250345 Min=0.815313367210792 Max=8.67281305982181.\n",
            "INFO [2025-06-10 07:27:09] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 07:27:12] plot_cnv(): auto thresholding at: (0.894809 , 1.105191)\n",
            "INFO [2025-06-10 07:27:14] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 07:27:14] Observation data size: Cells= 724 Genes= 15954\n",
            "INFO [2025-06-10 07:27:14] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 07:27:14] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 07:27:14] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:27:14] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 07:27:17] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:27:17] Quantiles of plotted data range: 0.894808679141156,1.00068259233421,1.00068259233421,1.00068259233421,1.10519132085884\n",
            "INFO [2025-06-10 07:27:18] plot_cnv_references:Start\n",
            "INFO [2025-06-10 07:27:18] Reference data size: Cells= 4906 Genes= 15954\n",
            "INFO [2025-06-10 07:47:23] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 07:47:24] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 07:47:47] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 07:47:47] Quantiles of plotted data range: 0.894808679141156,1.00068259233421,1.00068259233421,1.00068259233421,1.10519132085884\n"
          ]
        }
      ],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"1_vs_2\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PWuQ-vl_SdWA",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5067c4cb-9222-437e-e107-6077f77240ac"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:47:52] Parsing matrix: /content/1_and_3.matrix\n",
            "INFO [2025-06-10 07:48:40] Parsing gene order file: gene_pos_file.txt\n",
            "INFO [2025-06-10 07:48:40] Parsing cell annotations file: /content/1_vs_3.annotations.tsv\n",
            "INFO [2025-06-10 07:48:40] ::order_reduce:Start.\n",
            "INFO [2025-06-10 07:48:40] .order_reduce(): expr and order match.\n",
            "INFO [2025-06-10 07:48:41] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 36601,5593 Total=51838276 Min=0 Max=15474.\n",
            "INFO [2025-06-10 07:48:41] num genes removed taking into account provided gene ordering list: 1280 = 3.49717220840961% removed.\n",
            "INFO [2025-06-10 07:48:42] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
            "WARN [2025-06-10 07:48:43] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-06-10 07:48:47] validating infercnv_obj\n"
          ]
        }
      ],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/1_and_3.matrix\",\n",
        "                                    annotations_file=\"/content/1_vs_3.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "XFyRkLD0ShM6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2e9dec5d-3e53-44bd-e585-358846659e2e",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:48:47] ::process_data:Start\n",
            "INFO [2025-06-10 07:48:47] Creating output path 1_vs_3\n",
            "INFO [2025-06-10 07:48:47] Checking for saved results.\n",
            "INFO [2025-06-10 07:48:47] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-06-10 07:49:12] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-06-10 07:49:12] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-06-10 07:49:13] Removing 19151 genes from matrix as below mean expr threshold: 0.01\n",
            "INFO [2025-06-10 07:49:13] validating infercnv_obj\n",
            "INFO [2025-06-10 07:49:13] There are 16170 genes and 5593 cells remaining in the expr matrix.\n",
            "INFO [2025-06-10 07:49:17] Removed 3 genes having fewer than 3 min cells per gene = 0.0185529 % genes removed here\n",
            "INFO [2025-06-10 07:49:17] validating infercnv_obj\n",
            "INFO [2025-06-10 07:49:36] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-06-10 07:49:36] normalizing counts matrix by depth\n",
            "INFO [2025-06-10 07:49:39] Computed total sum normalization factor as median libsize: 6983.000000\n",
            "INFO [2025-06-10 07:50:00] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-06-10 07:50:00] transforming log2xplus1()\n",
            "INFO [2025-06-10 07:50:22] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-06-10 07:50:22] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 07:50:22] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 07:50:32] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 07:51:27] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-06-10 07:51:27] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-06-10 07:52:15] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-06-10 07:52:15] smooth_by_chromosome: chr: chr1\n",
            "INFO [2025-06-10 07:52:26] smooth_by_chromosome: chr: chr2\n",
            "INFO [2025-06-10 07:52:36] smooth_by_chromosome: chr: chr3\n",
            "INFO [2025-06-10 07:52:46] smooth_by_chromosome: chr: chr4\n",
            "INFO [2025-06-10 07:52:55] smooth_by_chromosome: chr: chr5\n",
            "INFO [2025-06-10 07:53:04] smooth_by_chromosome: chr: chr6\n",
            "INFO [2025-06-10 07:53:14] smooth_by_chromosome: chr: chr7\n",
            "INFO [2025-06-10 07:53:23] smooth_by_chromosome: chr: chr8\n",
            "INFO [2025-06-10 07:53:32] smooth_by_chromosome: chr: chr9\n",
            "INFO [2025-06-10 07:53:41] smooth_by_chromosome: chr: chr10\n",
            "INFO [2025-06-10 07:53:50] smooth_by_chromosome: chr: chr11\n",
            "INFO [2025-06-10 07:53:59] smooth_by_chromosome: chr: chr12\n",
            "INFO [2025-06-10 07:54:09] smooth_by_chromosome: chr: chr13\n",
            "INFO [2025-06-10 07:54:17] smooth_by_chromosome: chr: chr14\n",
            "INFO [2025-06-10 07:54:25] smooth_by_chromosome: chr: chr15\n",
            "INFO [2025-06-10 07:54:34] smooth_by_chromosome: chr: chr16\n",
            "INFO [2025-06-10 07:54:44] smooth_by_chromosome: chr: chr17\n",
            "INFO [2025-06-10 07:54:54] smooth_by_chromosome: chr: chr18\n",
            "INFO [2025-06-10 07:55:01] smooth_by_chromosome: chr: chr19\n",
            "INFO [2025-06-10 07:55:12] smooth_by_chromosome: chr: chr20\n",
            "INFO [2025-06-10 07:55:20] smooth_by_chromosome: chr: chr21\n",
            "INFO [2025-06-10 07:55:28] smooth_by_chromosome: chr: chr22\n",
            "INFO [2025-06-10 07:55:35] smooth_by_chromosome: chr: GL000194.1\n",
            "INFO [2025-06-10 07:55:35] smooth_by_chromosome: chr: GL000218.1\n",
            "INFO [2025-06-10 07:55:35] smooth_by_chromosome: chr: GL000219.1\n",
            "INFO [2025-06-10 07:55:35] smooth_by_chromosome: chr: KI270711.1\n",
            "INFO [2025-06-10 07:55:35] smooth_by_chromosome: chr: KI270721.1\n",
            "INFO [2025-06-10 07:56:25] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-06-10 07:56:25] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-06-10 07:57:28] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-06-10 07:57:28] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 07:57:28] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 07:57:37] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 07:58:34] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-06-10 07:58:34] invert_log2(), computing 2^x\n",
            "INFO [2025-06-10 07:59:47] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-06-10 07:59:47] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-06-10 07:59:54] define_signif_tumor_subclusters(), tumor: CD4 T-cells\n",
            "INFO [2025-06-10 07:59:54] Setting auto leiden resolution for CD4 T-cells to 0.497824\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  SGK3, MCMDC2, SNHG6, PPP1R42, COPS5, CSPP1, ARFGEF1, AC021321.1, CPA6, PREX2 \n",
            "\t   AC011853.2, SLCO5A1, NCOA2, TRAM1, LACTB2-AS1, LACTB2, XKR9, MSC-AS1, TERF1, AC022893.1 \n",
            "\t   RPL7, RDH10, STAU2, RPSA, AC027018.1, SLC25A38, AC099332.2, AC104850.2, AC099332.1, CX3CR1 \n",
            "Negative:  CMC2, CENPN, CDYL2, DYNLRB2, AC092718.2, AC092718.4, AC105411.1, ATMIN, C16orf46, MAFTRR \n",
            "\t   GCSH, GAN, LINC01229, CMIP, MAF, PLCG2, AC027279.1, SDR42E1, AC046158.1, AC092142.1 \n",
            "\t   AC079414.3, MPHOSPH6, WWOX, CDH13, VAT1L, NUDT7, HSBP1, AC009139.2, MLYCD, SYCE1L \n",
            "PC_ 2 \n",
            "Positive:  PRRC2C, MYOCOS, VAMP4, FMO4, EEF1AKNMT, DNM3, MROH9, PIGC, SUCO, BX284613.2 \n",
            "\t   AL031599.1, TNFSF4, AL645568.1, PRDX6, KLHL20, RC3H1, CENPL, ZBTB37, DARS2, RC3H1-IT1 \n",
            "\t   GAS5, RABGAP1L-DT, RABGAP1L, Z99127.1, Z99127.4, CACYBP, MRPS14, KIAA0040, LINC02803, COP1 \n",
            "Negative:  EPB41, GMEB1, AL360012.1, YTHDF2, SRSF4, LINC01715, TAF12, MECR, SNHG12, MATN1 \n",
            "\t   TRNAU1AP, MATN1-AS1, SNHG3, AL137857.1, RCC1, PHACTR4, LAPTM5, MED18, PUM1, SESN2 \n",
            "\t   NKAIN1, SNRNP40, AL353622.1, ZCCHC17, ATP5IF1, DNAJC8, AL451070.1, PTAFR, EYA3, SERINC2 \n",
            "PC_ 3 \n",
            "Positive:  CFL1, AP5B1, SNX32, RNASEH2C, KAT5, RELA-DT, RELA, MUS81, SIPA1, CTSW \n",
            "\t   PCNX3, FIBP, CCDC85B, MAP3K11, FOSL1, C11orf68, EHBP1L1, DRAP1, SART1, FAM89B \n",
            "\t   EIF1AD, ZNRD2, BANF1, CATSPER1, SSSCA1-AS1, SF3B2, LTBP3, SCYL1, MALAT1, NEAT1 \n",
            "Negative:  NUP50-DT, NUP50, ARHGAP8, KIAA0930, UPK3A, FAM118A, SMC1B, RIBC2, Z95331.1, ATXN10 \n",
            "\t   LINC00899, PRR34-AS1, MIRLET7BHG, AL121672.2, PPARA, CDPF1, TTC38, GTSE1, TRMU, CELSR1 \n",
            "\t   GRAMD4, CERK, AL118516.1, TBC1D22A, TAFA5, AC207130.1, C22orf34, Z97192.1, Z97192.3, Z97192.4 \n",
            "PC_ 4 \n",
            "Positive:  PTPRT, CHD6, SRSF6, ZHX3, L3MBTL1, PLCG1, PLCG1-AS1, TOP1, MAFB, DHX35 \n",
            "\t   SGK2, FAM83D, IFT52, MYBL2, PPP1R16B, TOX2, ACTR5, OSER1, RALGAPB, OSER1-DT \n",
            "\t   GDAP1L1, FITM2, SNHG11, TTPAL, SERINC3, PKIG, SNHG17, ADA, YWHAB, PABPC1L \n",
            "Negative:  CDC14A, RTCA, RTCA-AS1, AL445928.2, DBT, LRRC39, TRMT13, SASS6, AC093019.2, MFSD14A \n",
            "\t   SLC35A3, AC118553.1, AGL, FRRS1, DPYD-AS1, DPYD, PTBP2, RWDD3, AC092802.1, TLCD4 \n",
            "\t   EGLN3, SPTSSA, ALG14, EAPP, CNN3, SNX6, F3, CFL2, ABCD3, BAZ1A \n",
            "PC_ 5 \n",
            "Positive:  AC096992.2, NCK1-DT, SLC35G2, STAG1, PCCB, MSL2, PPP2R3A, AC092969.1, EPHB1, CEP63 \n",
            "\t   ANAPC13, RYK, RAB6B, SRPRB, TOPBP1, CDV3, NPHP3-AS1, TUBE1, NPHP3, FAM229B \n",
            "\t   UBA5, ACKR4, AL360015.1, ACAD11, DNAJC13, ACPP, CPNE4, MRPL3, NUDT16, AC010210.1 \n",
            "Negative:  TMEM256, ATF5, NLGN2, NUP62, CHRNB1, IL4I1, ZBTB4, POLR2A, TNFSF12, TNFSF13 \n",
            "\t   SENP3, EIF4A1, CD68, AC016876.1, MPDU1, FXR2, SAT2, ATP1B2, TP53, WRAP53 \n",
            "\t   KDM6B, NAA38, CYB5D1, CHD3, RNF227, KCNAB3, TRAPPC1, CNTROB, ALOX15B, ALOX12B \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:59:55] define_signif_tumor_subclusters(), tumor: CD8 T-cells\n",
            "INFO [2025-06-10 07:59:55] Setting auto leiden resolution for CD8 T-cells to 0.177114\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AP001628.1, WDR4, PDE9A, SLC37A1, RSPH1, UBASH3A, NDUFV3, ABCG1, PKNOX1, U2AF1 \n",
            "\t   HSF2BP, UMODL1, RRP1B, PDXK, ZNF295-AS1, CSTB, ZBTB21, RRP1, AATBC, C2CD2 \n",
            "\t   AGPAT3, PRDM15, TRAPPC10, AP001610.2, LINC01678, MX1, ICOSLG, MX2, AP001059.3, AP001059.2 \n",
            "Negative:  LEF1-AS1, LEF1, RPL34-AS1, RPL34, HADH, CYP2U1, AC096564.1, OSTC, SGMS2, PAPSS1 \n",
            "\t   COL25A1, AC092445.1, SEC24B-AS1, AIMP1, TBCK, SEC24B, AC109361.1, MCUB, GSTCD, CASP6 \n",
            "\t   PLA2G12A, INTS12, GAR1, ARHGEF38, ELOVL6, PPA2, AC004704.1, FAM241A, TET2-AS1, AP1AR \n",
            "PC_ 2 \n",
            "Positive:  OBSCN-AS1, IBA57, GUK1, MRPL55, ELOVL6, GAR1, AC004704.1, PLA2G12A, FAM241A, CASP6 \n",
            "\t   AP1AR, MCUB, TIFA, C1orf35, SEC24B, ALPK1, ZGRF1, SEC24B-AS1, COL25A1, LARP7 \n",
            "\t   OSTC, ARF1, RPL34, ANK2, RPL34-AS1, LEF1-AS1, CAMK2D, LEF1, WNT9A, HADH \n",
            "Negative:  SSH1, CORO1C, USP30, AC007569.1, SELPLG, USP30-AS1, ISCU, AC008119.1, SART3, ALKBH2 \n",
            "\t   FICD, CMKLR1, UNG, AC007622.2, PRDM4, ACACB, PWP1, MYO1H, BTBD11, KCTD10 \n",
            "\t   UBE3B, AC078929.1, MMAB, MVK, CRY1, FAM222A, GLTP, MTERF2, TCHP, GIT2 \n",
            "PC_ 3 \n",
            "Positive:  SPSB2, LRRC23, TPI1, ENO2, ATN1, USP5, C12orf57, CDCA3, PTPN6, CD4 \n",
            "\t   LAG3, PTMS, PHB2, MLF2, COPS7A, EMG1, ZNF384, AC125494.1, LPCAT3, AC004540.2 \n",
            "\t   SNX10, ING4, CBX3, AC004540.1, HNRNPA2B1, C1S, SKAP2, ACRBP, NFE2L3, HOTAIRM1 \n",
            "Negative:  SEC23IP, MCMBP, PLPP4, INPP5F, BAG3, AC023282.1, TIAL1, WDR11-AS1, WDR11, RGS10 \n",
            "\t   ATE1, NSMCE4A, AL731566.2, GRK5, PLEKHA1, FAM24B, PRDX3, C10orf88, SFXN4, PSTK \n",
            "\t   DENND10, IKZF5, EIF3A, ACADSB, BUB3, AL157388.1, LINC02641, AL139407.1, CHST15, CACUL1 \n",
            "PC_ 4 \n",
            "Positive:  HMGCR, ANKRD31, CERT1, POLK, ANKDD1B, POC5, IL18RAP, AC007278.1, IL18R1, MFSD9 \n",
            "\t   IL1R1, TMEM182, MRPS9, AC007271.1, AC104655.1, IL1R2, TGFBRAP1, LINC01127, AC012360.1, MAP4K4 \n",
            "\t   AC012360.3, RFX8, RNF149, C2orf49, CNOT11, FHL2, TBC1D8, AC108058.1, AC018638.7, FAM71F2 \n",
            "Negative:  RRM2, HPCAL1, KLF11, ODC1, AC104794.3, NOL10, GRHL1, AC007314.1, AC010969.2, ATP6V1C2 \n",
            "\t   OAT, LHPP, FAM53B, CHST15, TAF1B, FAM53B-AS1, EEF1AKMT2, LINC02641, PDIA6, ABRAXAS2 \n",
            "\t   BUB3, AL731577.2, YWHAQ, ZRANB1, SLC66A3, CTBP2, ACADSB, EDRF1-DT, EDRF1, ROCK2 \n",
            "PC_ 5 \n",
            "Positive:  SPSB2, LRRC23, TPI1, ENO2, USP5, ATN1, CDCA3, C12orf57, CD4, PTPN6 \n",
            "\t   LAG3, PHB2, PTMS, MLF2, COPS7A, EMG1, ZNF384, AC125494.1, ING4, LPCAT3 \n",
            "\t   ACRBP, LPAR5, AC006064.2, CHD4, C1S, NOP2, IFFO1, C1R, AC006064.4, C1RL \n",
            "Negative:  HOXA6, HOXA5, HOXA7, HOXA-AS3, HOXA9, HOXA-AS2, HOXA3, HOXA10-AS, HOTAIRM1, HOXA10 \n",
            "\t   SKAP2, HIBADH, AC004540.1, AC007130.1, AC004540.2, TAX1BP1, SNX10, JAZF1, CBX3, HNRNPA2B1 \n",
            "\t   CREB5, NFE2L3, AC005162.3, C7orf31, CPVL, CYCS, CHN2, OSBPL3, AC003093.1, AC004593.1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:59:56] define_signif_tumor_subclusters(), tumor: HSC\n",
            "INFO [2025-06-10 07:59:56] Setting auto leiden resolution for HSC to 0.57128\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AP005131.7, AP005131.1, AP005131.6, AP005131.5, LDLRAD4-AS1, LDLRAD4, SREK1IP1, CWC27, SETD9, AC008937.2 \n",
            "\t   IPO11, MAP3K1, MIER3, DIMT1, LINC01948, AC008937.3, ERCC8, NDUFAF2, KIF2A, ANKRD55 \n",
            "\t   SMIM15, SMIM15-AS1, AC016644.1, C5orf64, ZSWIM6, AC008892.1, GPBP1, PLK2, ELOVL7, GAPT \n",
            "Negative:  ZBTB4, CHRNB1, POLR2A, TNFSF12, TNFSF13, NLGN2, SENP3, EIF4A1, CD68, TMEM256 \n",
            "\t   AC016876.1, PLSCR3, MPDU1, FXR2, TNK1, SAT2, ATP1B2, ACAP1, TP53, NEURL4 \n",
            "\t   WRAP53, KDM6B, GPS2, NAA38, EIF5A, CYB5D1, YBX2, CHD3, SLC2A4, RNF227 \n",
            "PC_ 2 \n",
            "Positive:  MAP3K5, AL024508.1, AL024508.2, MAP7, AL121933.2, PEX7, MTFR2, BCLAF1, AL138828.1, IFNGR1 \n",
            "\t   AL360178.1, PDE7B, AL356234.2, XPNPEP1, ADD3-AS1, LINC01435, ADD3, SORCS1, MXI1, LINC00271 \n",
            "\t   SPON1, ITPRIP, SMNDC1, GSTO2, AL355512.1, GSTO1, LINC02539, CFAP43, DUSP5, AL049552.1 \n",
            "Negative:  AL034550.2, C20orf203, COMMD7, NOL4L, ASXL1, DNMT3B, KIF3B, MAPRE1, POFUT1, AL035071.2 \n",
            "\t   PLAGL2, AL035071.1, EFCAB8, TM9SF4, CDK5RAP1, SNTA1, HCK, CBFA2T2, NECAB3, PDRG1 \n",
            "\t   E2F1, PXMP4, TTLL9, ZNF341, TPX2, CHMP4B, RALY-AS1, AL117381.1, RALY, AL031668.1 \n",
            "PC_ 3 \n",
            "Positive:  COX19, C7orf50, AC073957.1, GPR146, AC073957.3, GET4, ADAP1, GPER1, SUN1, DNAAF5 \n",
            "\t   AC147651.4, ZFAND2A, PRKAR1B, FAM20C, AC091729.3, AC093627.4, MICALL2, AC093627.5, AC215522.2, INTS1 \n",
            "\t   MAFK, TMEM184A, PSMG3, PSMG3-AS1, MAD1L1, MRM2, NUDT1, SNX8, EIF3B, CHST12 \n",
            "Negative:  TET2-AS1, TET2, PPA2, AC004069.1, ARHGEF38, GSTCD, INTS12, AC109361.1, AC096577.1, TBCK \n",
            "\t   AC004053.1, AIMP1, CXXC4-AS1, AC092445.1, LINC02428, PAPSS1, CENPE, BDH2, SGMS2, SLC9B2 \n",
            "\t   AC096564.1, SLC9B1, CYP2U1, CISD2, HADH, UBE2D3-AS1, LEF1, AC018797.3, LEF1-AS1, UBE2D3 \n",
            "PC_ 4 \n",
            "Positive:  ITPR1, EGOT, AC018816.1, BHLHE40-AS1, BHLHE40, ARL8B, AC026202.2, EDEM1, AC069277.1, GRM7-AS3 \n",
            "\t   GRM7, LMCD1-AS1, CAV3, RAD18, SRGAP3, THUMPD3-AS1, THUMPD3, SETD5, HARBI1, AC018410.1 \n",
            "\t   ATG13, NR1H3, ACP2, MADD, ARHGAP1, DDB2, ZNF408, ARFGAP2, F2, AC090589.2 \n",
            "Negative:  MAGI3, LRIG2, PHTF1, LRIG2-DT, RSBN1, AL390242.1, SLC16A1-AS1, SLC16A1, AL137856.1, RHOC \n",
            "\t   AL603832.1, PTPN22, MOV10, AP4B1-AS1, CAPZA1, AP4B1, ST7L, DCLRE1B, WNT2B, HIPK1-AS1 \n",
            "\t   CTTNBP2NL, HIPK1, KCND3, TRIM33, BCAS2, DDX20, DENND2C, NPR3, MED10, NRAS \n",
            "PC_ 5 \n",
            "Positive:  OVOL2, MGME1, PET117, KAT14, ZNF133, AL049646.1, DZANK1, POLR3F, RBBP9, SEC23B \n",
            "\t   SMIM26, DTD1, AL121900.1, LINC00653, SLC24A3, GVQW2, AL590617.2, RIN2, TMEM67, RBM12B \n",
            "\t   PDP1, NHSL1, FAM92A, RAD54B, LINC00535, FSBP, TRIQK, NAA20, VIRMA, HEBP2 \n",
            "Negative:  EDNRB-AS1, SLAIN1, POU4F1, RBM26, RBM26-AS1, OBI1, MYCBP2, MYCBP2-AS1, NDFIP2, FBXL3 \n",
            "\t   CLN5, SPRY2, KCTD12, AL590807.1, AL136441.1, LMO7, AL162493.1, LMO7-AS1, UCHL3, LINC00351 \n",
            "\t   MIR17HG, GPC5, GPC5-AS2, AL356737.2, GPC5-AS1, MIR181A2HG, NR6A1, RPL35, ARPC5L, PSMB7 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:59:57] define_signif_tumor_subclusters(), tumor: komórki B pamięci\n",
            "INFO [2025-06-10 07:59:57] Setting auto leiden resolution for komórki B pamięci to 0.105818\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AL050403.2, JAG1, BTBD3, MKKS, SLX4IP, SNAP25-AS1, PAK5, AL035448.1, LAMP5, SPTLC3 \n",
            "\t   PLCB4, ISM1, PLCB1, TASP1, TMX4, BMP2, ESF1, FERMT1, NDUFAF5, CRLS1 \n",
            "\t   MACROD2, AL035461.2, AL079338.1, MCM8, KIF16B, TRMT6, SNRPB2, SHLD1, BFSP1, GPCPD1 \n",
            "Negative:  FANCE, RPL10A, TEAD3, FKBP5, AL033519.5, LINC01588, VCPKMT, ARF6, SOS2, AL627171.1 \n",
            "\t   LHFPL5, L2HGDH, AL627171.2, NEMF, DMAC2L, KLHDC2, KLHDC1, CDKL1, POLE2, DNAAF2 \n",
            "\t   MGAT2, MAP4K5, SRPK1, AL118556.1, ATL1, SLC26A8, SAV1, AL606834.1, NIN, MAPK14 \n",
            "PC_ 2 \n",
            "Positive:  UBXN11, CD52, SH3BGRL3, CRYBG2, ZNF683, CEP85, DHDDS, ZNF593, AL513365.2, FAM110D \n",
            "\t   HMGN2, PDIK1L, RPS6KA1, PAFAH2, ARID1A, STMN1, PIGV, ZDHHC18, PAQR7, GPN2 \n",
            "\t   AL033528.2, GPATCH3, NUDC, AUNIP, SLC9A1, MTFR1L, WDTC1, SELENON, TMEM222, SYTL1 \n",
            "Negative:  AC019197.1, COBLL1, SLC38A11, GRB14, SCN3A, GCA, GALNT3, IFIH1, AC009495.3, DPP4 \n",
            "\t   SLC4A10, AC009495.1, PSMD14, LINC01806, TTC21B, AC009299.3, TTC21B-AS1, SCN1A-AS1, SCN9A, SCN7A \n",
            "\t   XIRP2, AC073050.1, LINC01023, FER, NUDT12, AC008462.1, AC008871.1, EFNA5, C5orf30, AC016723.1 \n",
            "PC_ 3 \n",
            "Positive:  LINC01375, AL139340.1, LINC01374, HTR7, LINC00865, RPP30, KIF20B, PCGF5, PANK1, HECTD2 \n",
            "\t   IFIT5, TNKS2, IFIT1, IFIT3, FGFBP3, IFIT2, LIPA, BTAF1, CH25H, FAS \n",
            "\t   CPEB3, ACTA2, STAMBPL1, MARCH5, ANKRD22, LIPN, IDE, RNLS, KIF11, PTEN \n",
            "Negative:  XAB2, PET100, CAMSAP3, PNPLA6, MCOLN1, STXBP2, ZNF358, RETN, TEX45, PEX11G \n",
            "\t   MCEMP1, ARHGEF18, TRAPPC5, AC119396.2, AC119396.1, FCER2, INSR, EVI5L, ZNF557, ADGRE1 \n",
            "\t   MAP2K7, VAV1, SNAPC2, SH2D3A, TIMM44, TRIP10, GPR108, AC010336.8, TNFSF14, ELAVL1 \n",
            "PC_ 4 \n",
            "Positive:  ATG14, AL158801.3, KTN1-AS1, FBXO34, AL158801.2, DLGAP5, KTN1, AL355073.1, PELI2, AL355073.2 \n",
            "\t   LGALS3, LINC02284, MAPK1IP1L, TMEM260, AL161757.5, SOCS4, EXOC5, WDHD1, AP5M1, GCH1 \n",
            "\t   SAMD4A, NAA30, CGRRF1, GMFB, SLC35F4, CNIH1, ARMH4, CDKN3, ACTR10, AL365295.1 \n",
            "Negative:  PROS1, C3orf38, ZNF654, CGGBP1, HTR1F, AC108749.1, CHMP2B, AC107204.1, CADM2, LINC02008 \n",
            "\t   GBE1, ROBO1, ROBO2, ZNF717, LINC00960, COA1, BLVRA, MRPS24, URGCP, STK17A \n",
            "\t   UBE2D4, SPDYE1, LINC02018, AC004951.1, HECW1, LINC00957, CNTN3, MRPL32, DBNL, POLM \n",
            "PC_ 5 \n",
            "Positive:  DBNL, LINC00957, POLM, AC004951.1, AEBP1, SPDYE1, POLD2, UBE2D4, YKT6, URGCP \n",
            "\t   NUDCD3, MRPS24, DDX56, BLVRA, COA1, TMED4, STK17A, OGDH, HECW1, MRPL32 \n",
            "\t   ZMIZ2, PSMA2, PPIA, C7orf25, H2AFV, GLI3, SUGCT, LINC01952, MPLKIP, PURB \n",
            "Negative:  HINT3, TRMT11, CENPW, NCOA7-AS1, RNF146, ECHDC1, NCOA7, C6orf58, HDDC2, THEMIS \n",
            "\t   RNF217, PTPRK, RNF217-AS1, AL590006.1, NKAIN2, LAMA2, SMPDL3A, AL356124.1, PKIB, ARHGAP18 \n",
            "\t   SERINC1, HSF2, L3MBTL3, TBC1D32, AL355581.1, AL078600.1, SAMD3, MAN1A1, TMEM200A, FAM184A \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 07:59:58] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 1 (preB)\n",
            "INFO [2025-06-10 07:59:58] Less cells in group Komórki blastyczne 1 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 07:59:58] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 3 (preB)\n",
            "INFO [2025-06-10 07:59:58] Setting auto leiden resolution for Komórki blastyczne 3 (preB) to 0.0908055\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AC011498.3, CHAF1A, UBXN6, SH3GL1, AC011498.1, AC007292.3, HDGFL2, MPND, PLIN4, PLIN5 \n",
            "\t   LRG1, STAP2, SEMA6B, TMIGD2, TNFAIP8L1, MYDGF, SHD, DPP9, DPP9-AS1, YJU2 \n",
            "\t   FEM1A, SIRT6, AC005523.2, TICAM1, MAP2K2, PLIN3, ARRDC5, ZBTB7A, UHRF1, KDM4B \n",
            "Negative:  DHRS7, PCNX4, AL157911.1, RTN1, NEDD4, CCDC175, JKAMP, L3HYPDH, DNAAF4, GPR135 \n",
            "\t   DAAM1, C15orf65, LINC01500, KIAA0586, CCPG1, TIMM9, AL139021.1, PIGB, TOMM20L, AL139021.2 \n",
            "\t   ARID4A, PIGBOS1, PSMA3-AS1, AC018926.2, PSMA3, RAB27A, ACTR10, ARMH4, RSL24D1, SLC35F4 \n",
            "PC_ 2 \n",
            "Positive:  ARHGAP20, LINC02550, FDX1, AASDHPPT, AP000766.1, CWF19L2, POU2AF1, ZC3H12C, KBTBD3, ALKBH8 \n",
            "\t   MSANTD4, SLC35F2, RAB39A, LAYN, CARD16, RDX, AP003307.1, AP002433.2, NPAT, CASP1 \n",
            "\t   DDX10, ACAT1, CUL5, AP002433.1, SIK2, ATM, EXPH5, C11orf65, POGLUT3, PPP2R1B \n",
            "Negative:  ATXN10, Z95331.1, RIBC2, LINC00899, SMC1B, FAM118A, PRR34-AS1, UPK3A, MIRLET7BHG, AL121672.2 \n",
            "\t   KIAA0930, CDPF1, PPARA, TTC38, GTSE1, TRMU, NUP50, CELSR1, GRAMD4, NUP50-DT \n",
            "\t   CERK, AL118516.1, ARHGAP8, TBC1D22A, TAFA5, PRR5, AC207130.1, RTL6, C22orf34, AL031595.2 \n",
            "PC_ 3 \n",
            "Positive:  AC073343.2, ZNF316, ZNF12, ZNF853, CCZ1B, C7orf26, AC079804.3, ZDHHC4, GRID2IP, C1GALT1 \n",
            "\t   KDELR2, AC005532.1, DAGLB, AC004982.2, RAC1, CYTH3, MIOS, USP42, EIF2AK1, RPA3 \n",
            "\t   AIMP2, UMAD1, PMS2, CCZ1, AC006042.2, OCM, GLCCI1, RNF216, ICA1, FSCN1 \n",
            "Negative:  NEDD4L, ATP8B1, ALPK2, MALT1, AC027097.2, AC104365.1, AC027097.1, ZNF532, NARS, SEC11C \n",
            "\t   FECH, LMAN1, WDR7, TXNL1, CCBE1, AC009271.1, PMAIP1, RNF152, WNK1, RAD52 \n",
            "\t   AC004803.1, AC021054.1, PIGN, ERC1, NINJ2, WNT5B, CCDC77, FBXL14, RELCH, KDM5A \n",
            "PC_ 4 \n",
            "Positive:  AC006206.2, GALNT8, ANO2, NDUFA9, VWF, DYRK4, CD9, PLEKHG6, RAD51AP1, TNFRSF1A \n",
            "\t   C12orf4, LTBR, TIGAR, CD27-AS1, CD27, CCND2, TAPBPL, CCND2-AS1, VAMP1, MRPL51 \n",
            "\t   AC005842.1, NCAPD2, PARP11, GAPDH, CRACR2A, AC006064.4, TSPAN9, TULP3, IFFO1, RHNO1 \n",
            "Negative:  ELOVL6, GAR1, AC004704.1, PLA2G12A, CASP6, FAM241A, MCUB, AP1AR, SEC24B, SEC24B-AS1 \n",
            "\t   TIFA, COL25A1, OSTC, ALPK1, RPL34, ZGRF1, RPL34-AS1, LEF1-AS1, LARP7, LEF1 \n",
            "\t   ANK2, HADH, CAMK2D, CYP2U1, UGT8, AC096564.1, SGMS2, AC093765.2, PAPSS1, NDST3 \n",
            "PC_ 5 \n",
            "Positive:  METTL14, SNHG8, NDST3, SEC24D, AC093765.2, SYNPO2, UGT8, CAMK2D, USP53, ANK2 \n",
            "\t   C4orf3, LARP7, PDE5A, ZGRF1, MAD2L1, ALPK1, UVRAG, THAP12, GVQW3, DGAT2 \n",
            "\t   AP002360.1, SERPINH1, AC073475.1, EMSY, GDPD5, ACER3, TIFA, AP002498.1, KLHL35, CAPN5 \n",
            "Negative:  PTK2, DENND3, AC040970.1, SLC45A4, PTP4A3, TSNARE1, JRK, THEM6, LY6E-DT, LY6E \n",
            "\t   ZFP41, GLI4, MINCR, ZNF696, NDUFB9, TOP1MT, RHPN1, ZC3H3, AC067930.8, GSDMD \n",
            "\t   MROH6, NAPRT, EEF1D, TIGD5, PYCR3, TSTA3, ZNF623, ZNF707, AC105219.1, SCRIB \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:00:00] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 4 (preB)\n",
            "INFO [2025-06-10 08:00:00] Setting auto leiden resolution for Komórki blastyczne 4 (preB) to 0.172198\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  PKN1, DDX39A, GIPC1, AC008569.1, ADGRE5, DNAJB1, LINC01841, TECR, NDUFB7, CLEC17A \n",
            "\t   AC011509.2, ADGRE3, ADGRL1, ZNF333, AC022098.1, ADGRE2, ASF1B, PRKACA, SLC1A6, SAMD1 \n",
            "\t   IL27RA, ILVBL, RFX1, DCAF15, CC2D1A, BRD4, C19orf57, NANOS3, AKAP8, AC020916.1 \n",
            "Negative:  LAMTOR3, DNAJB14, H2AFZ, AC097460.1, EMCN, PPP3CA, AP001816.1, BANK1, AP002075.1, SLC39A8 \n",
            "\t   AC098487.1, AF213884.3, NFKB1, MANBA, UBE2D3, AC018797.3, UBE2D3-AS1, CISD2, SLC9B1, SLC9B2 \n",
            "\t   BDH2, CENPE, LINC02428, LINC01500, DAAM1, KIAA0586, CXXC4-AS1, AL139021.1, TIMM9, TOMM20L \n",
            "PC_ 2 \n",
            "Positive:  C11orf71, NNMT, RBM7, AP002518.2, REXO2, CADM1, BUD13, ZBTB16, ZPR1, USP28 \n",
            "\t   APOA1, ZW10, TAGLN, APOA1-AS, PCSK7, RNF214, TTC12, SIDT2, TMPRSS5, BACE1 \n",
            "\t   NCAM1, PAFAH1B2, LINC02762, BACE1-AS, SIK3, PLET1, CEP164, NKAPD1, DLAT, PTS \n",
            "Negative:  PLA2G12A, CASP6, MCUB, GAR1, ELOVL6, AC004704.1, SEC24B, FAM241A, SEC24B-AS1, AP1AR \n",
            "\t   COL25A1, TIFA, OSTC, RPL34, DDX11-AS1, TSPAN11, RPL34-AS1, AC010198.2, ALPK1, CAPRIN2 \n",
            "\t   LEF1-AS1, IPO8, TMTC1, ZGRF1, LEF1, OVCH1, OVCH1-AS1, LARP7, HADH, ERGIC2 \n",
            "PC_ 3 \n",
            "Positive:  GRAP2, FAM83F, CACNA1I, RPS19BP1, TNRC6B, ATF4, ADSL, MIEF1, SGSM3, MGAT3 \n",
            "\t   AL022238.2, MRTFA, TAB1, MRTFA-AS1, SYNGR1, SLC25A17, RPL3, ST13, XPNPEP3, AL031846.2 \n",
            "\t   DNAJB7, CBX7, RBX1, EP300, APOBEC3G, AL035658.1, EP300-AS1, RANGAP1, AL035681.1, L3MBTL2 \n",
            "Negative:  SLC45A4, AC040970.1, PTP4A3, DENND3, TSNARE1, JRK, PTK2, THEM6, LY6E-DT, LY6E \n",
            "\t   ZFP41, GLI4, MINCR, ZNF696, TOP1MT, RHPN1, ZC3H3, AC067930.8, GSDMD, MROH6 \n",
            "\t   NAPRT, EEF1D, TIGD5, PYCR3, TSTA3, ZNF623, ZNF707, AC105219.1, SCRIB, PUF60 \n",
            "PC_ 4 \n",
            "Positive:  LINC02666, MGMT, C10orf143, LINC01163, MKI67, GLRX3, PTPRE, INSYN2A, LINC02646, DOCK1 \n",
            "\t   ADAM12, FANK1, PPP2R2D, DHX32, BNIP3, BCCIP, AL162274.2, UROS, MMP21, JAKMIP3 \n",
            "\t   EDRF1-AS1, STK32C, EDRF1, LRRC27, PWWP2B, EDRF1-DT, CTBP2, ZRANB1, AL731577.2, ABRAXAS2 \n",
            "Negative:  ZWILCH, RPL4, SMAD6, AC087482.1, SMAD3, AC012568.1, AAGAB, IQCH, IQCH-AS1, SNAPC5 \n",
            "\t   C15orf61, MAP2K5, AC009292.2, MAP2K1, PIAS1, GLCE, SPESP1, CALML4, KIF23, RPLP1 \n",
            "\t   ANP32A, CLN6, CORO2B, FEM1B, TIPIN, AC027237.2, AC055855.1, DRAIC, DIS3L, TLE3 \n",
            "PC_ 5 \n",
            "Positive:  SPTY2D1OS, UEVLD, SPTY2D1, ZDHHC13, TSG101, LDHAL6A, CSRP3-AS1, LDHC, E2F8, LDHA \n",
            "\t   NAV2, GTF2H1, HTATIP2, HPS5, PRMT3, SAAL1, ANO5, TPH1, FANCF, SERGEF \n",
            "\t   GAS2, KCNC1, SVIP, NCR3LG1, AC006299.1, NUCB2, LINC02718, PIK3C2A, LUZP2, RPS13 \n",
            "Negative:  AC011287.1, CRPPA, CRPPA-AS1, ANKMY2, AC005014.4, BZW2, AC073333.1, TSPAN13, AHR, AC073332.1 \n",
            "\t   SNX13, HDAC9, AC007091.1, TWISTNB, AC005062.1, MACC1, MACC1-AS1, AC004130.2, ITGB8, SP4 \n",
            "\t   DNAH11, FOXK1, CDCA7L, SDK1, CHCHD3, PLXNA4, AC009365.4, EXOC4, LRGUK, SLC35B4 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:00:01] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 5 (preB)\n",
            "INFO [2025-06-10 08:00:01] Too few cells in group Komórki blastyczne 5 (preB) for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 08:00:01] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 6 (preB)\n",
            "INFO [2025-06-10 08:00:01] Less cells in group Komórki blastyczne 6 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 08:00:01] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 7 (preB)\n",
            "INFO [2025-06-10 08:00:01] Setting auto leiden resolution for Komórki blastyczne 7 (preB) to 0.57128\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  LMNB2, GADD45B, GNG7, AC092068.2, AC092068.3, DIRAS1, SLC39A3, SGTA, THOP1, FAAP100 \n",
            "\t   AC139149.1, NPLOC4, ACTG1, BAHCC1, PDE6G, ZNF554, AC110285.1, OXLD1, AC027601.6, CCDC137 \n",
            "\t   LINC00482, ARL16, HGS, SLC38A10, ZNF555, NDUFAF8, AC139530.1, TEPSIN, MRPL12, ZNF57 \n",
            "Negative:  ATF7IP, EMP1, PLBD1, GSG1, AC008114.1, FAM234B, PLBD1-AS1, AC023790.2, HEBP1, AC010168.1 \n",
            "\t   GPRC5D-AS1, DDX47, AC010168.2, APOLD1, HIST4H4, AC008115.3, CDKN1B, H2AFJ, CREBL2, C12orf60 \n",
            "\t   WBP11, AC007619.1, ART4, DUSP16, ERP27, ARHGDIB, BORCS5, MANSC1, PTPRO, LRP6 \n",
            "PC_ 2 \n",
            "Positive:  TMEM106B, THSD7A, AC060834.2, NDUFA4, AC004160.1, NXPH1, VWDE, PHF14, AC007009.1, ARL4A \n",
            "\t   AC011287.1, ICA1, CRPPA, GLCCI1, CRPPA-AS1, ANKMY2, AC005014.4, AC006042.2, BZW2, UMAD1 \n",
            "\t   RPA3, SNX13, MIOS, HDAC9, AC004982.2, AC007091.1, AC005532.1, TWISTNB, C1GALT1, AC005062.1 \n",
            "Negative:  RAPH1, CD28, ICOS, PARD3B, INO80D, AC007383.2, NDUFS1, EEF1B2, ZDBF2, ADAM23 \n",
            "\t   DYTN, FASTKD2, AC008269.1, KLF7, KLF7-IT1, MYOSLID, AC007879.3, AC009226.1, CREB1, METTL21A \n",
            "\t   LINC01857, CCNYL1, FZD5, PLEKHM3, CRYGD, IDH1, PIKFYVE, PTH2R, RPE, KANSL1L \n",
            "PC_ 3 \n",
            "Positive:  MXI1, SMNDC1, ADD3, AL355512.1, ADD3-AS1, DUSP5, XPNPEP1, SMC3, PDCD4-AS1, LINC01435 \n",
            "\t   PDCD4, BBIP1, SORCS1, SHOC2, GPAM, ITPRIP, ACSL5, GSTO2, AL157786.1, ZDHHC6 \n",
            "\t   GSTO1, VTI1A, AL139120.1, CFAP43, AL158212.3, AL158212.1, SFR1, TCF7L2, AL158212.2, COL17A1 \n",
            "Negative:  UBE2T, PPP1R12B, SYT2, LGR6, KDM5B, RABIF, KLHL12, PTPN7, ARL8A, ADIPOR1 \n",
            "\t   CYB5R1, P2RY13, ELF3, TMEM183A, P2RY12, ELF3-AS1, PPFIA4, IGSF10, LINC01136, RNPEP \n",
            "\t   BTG2, AADACL2-AS1, ATP2B4, TIMM17A, LAX1, SUCNR1, IPO9, ZC3H11A, AC108718.1, AC096992.2 \n",
            "PC_ 4 \n",
            "Positive:  C12orf42, HELLPAR, PARPBP, NUP37, WASHC3, DRAM1, SLC41A2, GNPTAB, C12orf45, SYCP3 \n",
            "\t   ALDH1L2, WASHC4, APPL2, C12orf75, CKAP4, TCP11L2, POLR3B, AC079385.1, RIC8B, TMEM263 \n",
            "\t   MTERF2, CRY1, AC078929.1, BTBD11, PWP1, PRDM4, AC007622.2, CMKLR1, FICD, SART3 \n",
            "Negative:  ABCC1, ABCC6, FOPNL, NOMO3, P4HA2, SLC22A4, AC138969.1, MIR3936HG, ACSL6, XYLT1 \n",
            "\t   SLC22A5, AF001548.2, MEIKIN, AC109446.3, C5orf56, MYH11, FNIP1, AC009152.3, AC116366.2, RAPGEF6 \n",
            "\t   AC026401.3, AC116366.1, NOMO2, NDE1, CDC42SE2, RPS15A, IRF1, LYRM7, MARF1, ARL6IP1 \n",
            "PC_ 5 \n",
            "Positive:  DUXAP8, AP000547.3, ANKRD62P1-PARP4P3, IL17RA, HDHD5, HDHD5-AS1, ADA2, CECR2, ATP6V1E1, BCL2L13 \n",
            "\t   BID, LINC00528, MICAL3, PEX26, TUBA8, AC008079.1, USP18, DGCR9, DGCR2, DGCR11 \n",
            "\t   ESS2, SLC25A1, CLTCL1, AC000085.1, HIRA, C22orf39, MRPL40, UFD1, CDC45, GP1BB \n",
            "Negative:  NORAD, CNBD2, SCAND1, AAR2, PHF20, DLGAP4, DLGAP4-AS1, TGIF2, RBM39, RAB5IF \n",
            "\t   SLA2, NDRG3, DSN1, SOGA1, ROMO1, TLDC2, NFS1, SAMHD1, RBM12, RBL1 \n",
            "\t   CPNE1, MROH8, ERGIC3, RPN2, MANBAL, FO393401.1, SRC, BLCAP, NNAT, CTNNBL1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:00:02] define_signif_tumor_subclusters(), tumor: Komórki blastyczne proliferujące\n",
            "INFO [2025-06-10 08:00:02] Setting auto leiden resolution for Komórki blastyczne proliferujące to 0.617678\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  IER2, CACNA1A, CCDC130, MRI1, AC008686.1, C19orf53, ZSWIM4, AC020916.1, NANOS3, C19orf57 \n",
            "\t   CC2D1A, ITGB4, SAP30BP, GALK1, SMIM6, H3F3B, UNK, SMIM5, DCAF15, UNC13D \n",
            "\t   WBP2, TRIM65, RECQL5, MRPL38, RFX1, FBF1, MYO15B, ACOX1, IL27RA, TEN1 \n",
            "Negative:  UCHL3, LMO7-AS1, LMO7, AL136441.1, LDLRAD4, KCTD12, LDLRAD4-AS1, AP005131.5, CLN5, AP005131.6 \n",
            "\t   FBXL3, ST8SIA4, FAM174A, AP005131.1, SLCO4C1, MYCBP2-AS1, AC021086.1, PAM, AC113385.1, AP005131.7 \n",
            "\t   MYCBP2, LINC02062, RGMB, LINC01846, CHD1, GIN1, AP005131.3, LINC02234, SLAIN1, LINC01340 \n",
            "PC_ 2 \n",
            "Positive:  HSF2BP, RRP1B, U2AF1, PKNOX1, PDXK, NDUFV3, WDR4, AP001628.1, CSTB, RRP1 \n",
            "\t   PDE9A, AATBC, AGPAT3, TRAPPC10, LINC01678, ICOSLG, AP001059.3, AP001059.2, PFKL, CFAP410 \n",
            "\t   AP001062.1, TRPM2, UBE2G2, SUMO3, PTTG1IP, ITGB2, ITGB2-AS1, FAM207A, AP001505.1, ADARB1 \n",
            "Negative:  COA4, MRPL48, PAAF1, RAB6A, DNAJB13, PLEKHB1, AP000763.3, AP003717.1, FAM168A, UCP2 \n",
            "\t   RELT, AP003717.4, C2CD3, ARHGEF17, PPME1, P2RY6, P4HA3, P2RY2, PGM2L1, KCNE3 \n",
            "\t   FCHSD2, LIPT2, RPS3, SPCS2, ARRB1, NEU3, POLD3, XRRA1, RNF169, ATG16L2 \n",
            "PC_ 3 \n",
            "Positive:  TPCN2, CCND1, IGHMBP2, LTO1, AP003555.2, MRPL21, ANO1, CPT1A, FADD, TESMIN \n",
            "\t   PPFIA1, GAL, AP002336.2, PPP6R3, AP000487.1, LRP5, CTTN, SHANK2, AP002813.1, DHCR7 \n",
            "\t   C11orf24, AP002387.2, KMT5B, NADSYN1, CHKA, ALG1L9P, AP002807.1, FAM86C1, TCIRG1, AP002495.1 \n",
            "Negative:  PPP1R14B, FKBP2, DNAJC4, VEGFB, AP001453.1, NUDT22, PPP1R14B-AS1, PLCB3, TRPT1, BAD \n",
            "\t   FERMT3, GPR137, STIP1, ESRRA, TRMT112, PRDX5, FLRT1, AP003774.2, CCDC88B, AP000721.2 \n",
            "\t   RPS6KA4, NRXN2, MACROD1, RASGRP2, OTUB1, PYGM, COX8A, SF1, NAA40, MARK2 \n",
            "PC_ 4 \n",
            "Positive:  SLC35A4, WDR55, APBB3, CD14, HARS, EIF4EBP3, IK, NDUFA2, SRA1, TMCO6 \n",
            "\t   HARS2, AC244517.11, TAF7, ZMAT2, ANKHD1, AC008438.1, PCDHGA6, HBEGF, DIAPH1, PFDN1 \n",
            "\t   CYSTM1, AC008781.2, PURA, AC008781.1, MALINC1, HDAC3, AC008667.1, CXXC5, RELL2, UBE2D2 \n",
            "Negative:  THAP1, SMIM19, SLC20A2, VDAC3, POLB, IKBKB, AC083973.1, PLAT, AP3M2, AC103724.4 \n",
            "\t   KAT6A, AC113133.1, ANK1, AC009630.3, GPAT4, GINS4, AC009630.1, GOLGA7, ZMAT4, ADAM18 \n",
            "\t   ADAM32, ADAM9, TM2D2, HTRA4, AC108863.2, PLEKHA2, AC067817.2, AC067817.1, TACC1, FGFR1 \n",
            "PC_ 5 \n",
            "Positive:  LDLRAD3, TRIM44, LRRC37B, SLC1A2, AL133330.1, CD44-AS1, SUZ12, CD44, AC004253.1, MGAT2 \n",
            "\t   PDHX, UTP6, APIP, DNAAF2, COPRS, CAT, POLE2, RAB11FIP4, ABTB2, KLHDC1 \n",
            "\t   EVI2A, NAT10, KLHDC2, EVI2B, CAPRIN1, LMO2, OMG, NEMF, NF1, FBXO3 \n",
            "Negative:  VPREB1, IGLV10-54, TOP3B, AC245452.1, PPM1F, MAPK1, AP000553.8, YPEL1, PPIL2, SDF2L1 \n",
            "\t   SMG6, YDJC, AC090617.4, SRR, TSR1, UBE2L3, SGSM2, MNT, METTL16, PAFAH1B1 \n",
            "\t   SHPK, CLUH, CTNS, RAP1GAP2, OR3A2, TRPV1, HIC2, THAP7-AS1, THAP7, LZTR1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:00:03] define_signif_tumor_subclusters(), tumor: komórki NK T\n",
            "INFO [2025-06-10 08:00:03] Setting auto leiden resolution for komórki NK T to 0.514215\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  ZNF554, ZNF555, THOP1, ZNF57, SGTA, AC119403.1, SLC39A3, ZNF77, DIRAS1, AC005944.1 \n",
            "\t   AC092068.3, TLE5, AC092068.2, GNA11, GNG7, GADD45B, GNA15, LMNB2, AC005264.1, S1PR4 \n",
            "\t   NCLN, NFIC, SMIM24, DOHH, FZR1, MFSD12, C19orf71, AC005786.3, HMG20B, LRRC45 \n",
            "Negative:  CAMK4, WDR36, SLC25A46, MAN2A1, TMEM232, AC010275.1, PJA2, STARD4, AC008467.1, STARD4-AS1 \n",
            "\t   AC116428.1, NREP, AC008871.1, EPB41L4A, FER, EPB41L4A-AS1, LINC01023, AC008462.1, APC, FBXL17 \n",
            "\t   SRP19, REEP5, EFNA5, DCP2, NUDT12, MCC, C5orf30, YTHDC2, PPIP5K2, PGGT1B \n",
            "PC_ 2 \n",
            "Positive:  CNTNAP3, ANKRD18A, GLIDR, ALDH1B1, CBWD6, SHB, FAM27C, SLC25A51, DCAF10, LINC01410 \n",
            "\t   AL512625.3, LERFS, AL512625.1, ANKRD20A4, CBWD5, AC104623.1, AC010880.1, FAM49A, ZBTB5, AC010745.5 \n",
            "\t   SMC6, GEN1, AC010745.2, RDH14, MYCN, LINC01376, DDX1, ZNF658, GRHPR, LINC00954 \n",
            "Negative:  RALGPS2, CLEC20A, RASAL2, AL136114.1, AL591043.2, COP1, LINC02803, KIAA0040, MRPS14, CACYBP \n",
            "\t   Z99127.4, Z99127.1, RABGAP1L, RABGAP1L-DT, RC3H1-IT1, RC3H1, ZBTB37, GAS5, DARS2, CENPL \n",
            "\t   KLHL20, PRDX6, AL645568.1, TNFSF4, AL031599.1, SUCO, PIGC, DNM3, AC124947.1, NUDT4 \n",
            "PC_ 3 \n",
            "Positive:  FAM160A1, AC095055.1, SH3D19, RPS3A, LRBA, DCLK2, LINC02355, AC108156.1, AC002460.2, AC069272.1 \n",
            "\t   NR3C2, ARHGAP10, PRMT9, MGAT2, DNAAF2, POLE2, KLHDC1, KLHDC2, NEMF, AL627171.2 \n",
            "\t   AL627171.1, ARF6, LINC01588, VCPKMT, SOS2, L2HGDH, DMAC2L, CDKL1, MAP4K5, AL118556.1 \n",
            "Negative:  DYNLL1, NRAV, SRSF9, COQ5, RNF10, POP5, MLEC, UNC119B, ACADS, AC069234.1 \n",
            "\t   SPPL3, C12orf43, OASL, P2RX7, P2RX4, CAMKK2, RNF34, ANAPC5, KDM2B, KDM2B-DT \n",
            "\t   ORAI1, MORN3, TMEM120B, RHOF, LINC01089, AC084018.2, SETD1B, PSMD9, WDR66, AC069503.1 \n",
            "PC_ 4 \n",
            "Positive:  BBS7, CCNA2, KIAA1109, EXOSC9, NUDT6, ANXA5, SPATA5, SPRY1, AC073475.1, LINC01091 \n",
            "\t   MAD2L1, ANKRD50, PDE5A, FAT4, C4orf3, AC104664.1, INTU, USP53, HSPA4L, PLK4 \n",
            "\t   SYNPO2, MFSD8, LARGE1, TIMP3, ABHD18, SYN3, FBXO7, LARGE-AS1, RTCB, HMGXB4 \n",
            "Negative:  PWP1, PRDM4, BTBD11, AC007622.2, CMKLR1, AC078929.1, FICD, SART3, CRY1, AC008119.1 \n",
            "\t   MTERF2, ISCU, TMEM263, SELPLG, AC007569.1, RIC8B, CORO1C, SSH1, AC079385.1, USP30 \n",
            "\t   POLR3B, USP30-AS1, TCP11L2, ALKBH2, CKAP4, UNG, C12orf75, ACACB, APPL2, WASHC4 \n",
            "PC_ 5 \n",
            "Positive:  CTR9, MRVI1, EIF4G2, AC116535.1, RNF141, ZBED5, MTRNR2L8, AMPD3, ZBED5-AS1, ADM \n",
            "\t   USP47, AC100763.1, AC011092.3, DKK3, AC011092.2, SBF2, MICAL2, SBF2-AS1, SWAP70, WEE1 \n",
            "\t   TEAD1, ZNF143, AC132192.2, AC132192.1, IPO7, LINC00958, TMEM41B, ARNTL, DENND5A, SETD5 \n",
            "Negative:  ZNF354C, ADAMTS2, RUFY1, AC136604.2, RMRP, CCDC107, SIT1, SIGMAR1, ARHGEF39, CD72 \n",
            "\t   TESK1, RUSC2, DCTN3, GALT, RPP25L, UNC13B, TPM2, FAM219A, C9orf24, FAM214B \n",
            "\t   NUDT2, KIF24, STOML2, VCP, TLN1, IL11RA, DNAJB5-DT, AL162231.1, FANCG, AL162231.2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:00:04] define_signif_tumor_subclusters(), tumor: MPP\n",
            "INFO [2025-06-10 08:00:04] Less cells in group MPP than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 08:00:04] define_signif_tumor_subclusters(), tumor: późne komórki erytroidalne\n",
            "INFO [2025-06-10 08:00:04] Less cells in group późne komórki erytroidalne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 08:00:04] define_signif_tumor_subclusters(), tumor: prekursory erytroidalne i megakariocytarne\n",
            "INFO [2025-06-10 08:00:04] Too few cells in group prekursory erytroidalne i megakariocytarne for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 08:00:04] define_signif_tumor_subclusters(), tumor: wczesne komórki erytroidalne\n",
            "INFO [2025-06-10 08:00:04] Too few cells in group wczesne komórki erytroidalne for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 08:00:04] define_signif_tumor_subclusters(), tumor: control\n",
            "INFO [2025-06-10 08:00:05] Setting auto leiden resolution for control to 0.0057241\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  IGSF10, P2RY12, AADACL2-AS1, P2RY13, MBNL1, SUCNR1, AC108718.1, MBNL1-AS1, AC091607.2, AC026347.1 \n",
            "\t   KCNAB1, P2RY1, GMPS, RAP2B, SLC33A1, ARHGEF26, C3orf33, DHX36, PLCH1, MME-AS1 \n",
            "\t   MME, ATP2A2, IFT81, C12orf76, ARPC3, ANAPC7, ANKRD13A, GIT2, GPN3, FAM216A \n",
            "Negative:  PARD6A, ENKD1, C16orf86, GFOD2, RANBP10, CENPT, THAP11, NUTF2, EDC4, PSKH1 \n",
            "\t   CTRL, PSMB10, LCAT, SLC12A4, DPEP2, DUS2, DDX28, NFATC3, AC020978.3, AC020978.5 \n",
            "\t   PLA2G15, SLC7A6, SLC7A6OS, PRMT7, PAGR1, MVP, CDIPT, ASPHD1, AC009133.1, KCTD13 \n",
            "PC_ 2 \n",
            "Positive:  TXNDC11, LITAF, AC007613.1, SNN, ZC3H7A, RSL1D1, AC007216.4, AC099489.1, GSPT1, AC007216.3 \n",
            "\t   SOCS1, AC007216.2, NPIPB2, RMI2, SNX29, AC010333.2, AC007220.1, CPPED1, AC007014.1, AC109597.1 \n",
            "\t   CLEC16A, ERCC4, DEXI, CIITA, MRTFB, AC133065.1, TVP23A, AC040173.1, NUBP1, PARN \n",
            "Negative:  AL392172.2, FAM177B, BROX, AL392172.1, AIDA, DISP1, AL592148.3, TLR5, MIA3, SUSD4 \n",
            "\t   CAPN2, TAF1A, TP53BP2, FBXO28, LINC02257, DEGS1, NVL, DUSP10, CNIH4, WDR26 \n",
            "\t   HLX, CNIH3, DNAH14, HLX-AS1, LBR, MARC1, ENAH, RAB3GAP2, AC099066.2, IARS2 \n",
            "PC_ 3 \n",
            "Positive:  PPP1R16A, KIFC2, CYHR1, TONSL, VPS28, CD8B, CD8A, RMND5A, RGPD1, PLGLB1 \n",
            "\t   RNF103, AC092651.3, AC015971.1, LINC01943, CYTOR, SLC39A4, CHMP3, PLGLB2, RGPD2, KDM3A \n",
            "\t   KRCC1, AC104134.1, MRPL35, AC009309.1, EIF2AK3, EIF2AK3-DT, RPIA, IMMT, IGKC, MAL \n",
            "Negative:  TRMT44, AC105345.1, AC108519.1, ACOX3, AC105916.1, SH3TC1, SLC2A9, ABLIM2, AC005674.2, AFAP1 \n",
            "\t   WDR1, SORCS2, ZNF518B, GRPEL1, AC005599.1, CLNK, TADA2B, HS3ST1, AC097382.2, AC006230.1 \n",
            "\t   AC097382.3, AC005699.1, RAB28, TBC1D14, BOD1L1, KIAA0232, LINC00504, BLOC1S4, CPEB2-DT, CPEB2 \n",
            "PC_ 4 \n",
            "Positive:  BCR, RSPH14, IGLC3, IGLL1, IGLC2, IGLC1, AP000345.2, IGLV2-8, RGL4, IGLV3-21 \n",
            "\t   ZNF70, ZNF280B, VPREB3, AC245060.5, CHCHD10, AC245060.6, SMARCB1, VPREB1, DERL3, IGLV10-54 \n",
            "\t   SLC2A11, TOP3B, MIF, AC245452.1, MIF-AS1, PPM1F, DDTL, DDT, MAPK1, CABIN1 \n",
            "Negative:  AIM2, ACKR1, IFI16, FCER1A, PYHIN1, MNDA, DUSP23, FCRL6, CFAP45, TAGLN2 \n",
            "\t   SPTA1, LINC01133, PIGM, IGSF8, CD1C, AL121987.2, PEA15, CD1D, DCAF8, PEX19 \n",
            "\t   COPA, KIRREL1, NCSTN, FCRL1, VANGL2, SLAMF6, FCRL2, AL138930.1, CD84, FCRL3 \n",
            "PC_ 5 \n",
            "Positive:  CASP1, CARD16, MSANTD4, KBTBD3, AASDHPPT, AP000766.1, CWF19L2, ALKBH8, SLC35F2, RAB39A \n",
            "\t   AP003307.1, CUL5, AP002433.1, ACAT1, TNFAIP8, DMXL1, HSD17B4, AP002433.2, AC008629.1, DTWD2 \n",
            "\t   AC008574.1, NPAT, AC093206.1, AC113418.1, LINC02208, LINC02147, ATM, PRR16, SEMA6A-AS1, AC008568.1 \n",
            "Negative:  AL590226.1, EGFL7, AGPAT2, DIPK1B, SNHG7, LCN10, LCN6, TMEM141, CCDC183, RABL6 \n",
            "\t   TADA2B, AC097382.2, GRPEL1, PHPT1, AC097382.3, SORCS2, EDF1, AFAP1, TBC1D14, TRAF2 \n",
            "\t   ABLIM2, KIAA0232, FBXW5, BLOC1S4, SH3TC1, MRFAP1L1, SIRT7, PTGDS, MAFG, PYCR1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:01:54] ::plot_cnv:Start\n",
            "INFO [2025-06-10 08:01:54] ::plot_cnv:Current data dimensions (r,c)=16167,5593 Total=90489980.2896351 Min=0.819203053674516 Max=8.66511299389459.\n",
            "INFO [2025-06-10 08:01:54] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 08:01:57] plot_cnv(): auto thresholding at: (0.897791 , 1.103712)\n",
            "INFO [2025-06-10 08:02:13] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 08:02:13] Observation data size: Cells= 687 Genes= 16167\n",
            "INFO [2025-06-10 08:02:13] clustering observations via method: ward.D\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(1) is 27\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  27,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(2) is 48\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  48,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(3) is 36\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  36,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(4) is 4\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  4,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(5) is 2\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  2,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(6) is 23\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  23,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(7) is 68\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  68,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(8) is 47\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  47,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(9) is 25\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  25,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(10) is 19\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  19,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(11) is 4\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  4,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(12) is 1\n",
            "INFO [2025-06-10 08:02:13] Skipping group: 12, since less than 2 entries\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(13) is 3\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  3,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(14) is 47\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  47,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(15) is 44\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  44,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(16) is 36\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  36,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(17) is 35\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  35,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(18) is 16\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  16,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(19) is 12\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  12,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(20) is 5\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  5,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(21) is 1\n",
            "INFO [2025-06-10 08:02:13] Skipping group: 21, since less than 2 entries\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(22) is 71\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  71,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(23) is 13\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  13,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(24) is 6\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  6,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(25) is 3\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  3,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(26) is 2\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  2,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(27) is 9\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  9,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(28) is 23\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  23,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(29) is 21\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  21,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(30) is 26\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  26,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(31) is 4\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  4,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(32) is 3\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  3,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(33) is 2\n",
            "INFO [2025-06-10 08:02:13] group size being clustered:  2,16167\n",
            "INFO [2025-06-10 08:02:13] Number of cells in group(34) is 1\n",
            "INFO [2025-06-10 08:02:13] Skipping group: 34, since less than 2 entries\n",
            "INFO [2025-06-10 08:02:13] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 08:02:13] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 08:02:14] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:02:14] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:02:16] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:02:16] Quantiles of plotted data range: 0.897791088431412,0.972242568437836,0.998361160873722,1.02713555106064,1.10371184816233\n",
            "INFO [2025-06-10 08:02:17] plot_cnv_references:Start\n",
            "INFO [2025-06-10 08:02:17] Reference data size: Cells= 4906 Genes= 16167\n",
            "INFO [2025-06-10 08:02:33] plot_cnv_references:Number reference groups= 50\n",
            "INFO [2025-06-10 08:02:35] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 08:02:57] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:02:57] Quantiles of plotted data range: 0.897791088431412,0.976010692131528,0.998359755820898,1.02268181396629,1.10371184816233\n",
            "INFO [2025-06-10 08:04:16] ::plot_cnv:Start\n",
            "INFO [2025-06-10 08:04:16] ::plot_cnv:Current data dimensions (r,c)=16167,5593 Total=90489980.2896351 Min=0.819203053674516 Max=8.66511299389459.\n",
            "INFO [2025-06-10 08:04:16] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 08:04:19] plot_cnv(): auto thresholding at: (0.897791 , 1.103712)\n",
            "INFO [2025-06-10 08:04:21] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 08:04:21] Observation data size: Cells= 687 Genes= 16167\n",
            "INFO [2025-06-10 08:04:21] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 08:04:21] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 08:04:21] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:04:21] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:04:24] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:04:24] Quantiles of plotted data range: 0.897791088431412,0.972242568437836,0.998361160873722,1.02713555106064,1.10371184816233\n",
            "INFO [2025-06-10 08:04:25] plot_cnv_references:Start\n",
            "INFO [2025-06-10 08:04:25] Reference data size: Cells= 4906 Genes= 16167\n",
            "INFO [2025-06-10 08:25:29] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 08:25:31] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 08:25:52] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:25:52] Quantiles of plotted data range: 0.897791088431412,0.976010692131528,0.998359755820898,1.02268181396629,1.10371184816233\n",
            "INFO [2025-06-10 08:25:58] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-06-10 08:25:58] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-06-10 08:25:58] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 08:26:01] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.9457355142319 - 1.05561274462978\n",
            "INFO [2025-06-10 08:26:41] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-06-10 08:26:42] ::plot_cnv:Start\n",
            "INFO [2025-06-10 08:26:42] ::plot_cnv:Current data dimensions (r,c)=16167,5593 Total=90666520.8922422 Min=0.819203053674516 Max=8.66511299389459.\n",
            "INFO [2025-06-10 08:26:42] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 08:26:44] plot_cnv(): auto thresholding at: (0.896288 , 1.103712)\n",
            "INFO [2025-06-10 08:26:46] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 08:26:46] Observation data size: Cells= 687 Genes= 16167\n",
            "INFO [2025-06-10 08:26:46] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 08:26:46] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 08:26:46] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:26:46] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 08:26:48] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:26:48] Quantiles of plotted data range: 0.896288151837673,1.00067412943084,1.00067412943084,1.00067412943084,1.10371184816233\n",
            "INFO [2025-06-10 08:26:49] plot_cnv_references:Start\n",
            "INFO [2025-06-10 08:26:49] Reference data size: Cells= 4906 Genes= 16167\n",
            "INFO [2025-06-10 08:47:52] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 08:47:54] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 08:48:15] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 08:48:15] Quantiles of plotted data range: 0.896288151837673,1.00067412943084,1.00067412943084,1.00067412943084,1.10371184816233\n"
          ]
        }
      ],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"1_vs_3\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "14J6s5qPSlHt",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "71ebee86-c2b9-49b2-c718-1929c429d534"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:57:45] Parsing matrix: /content/1_and_4.matrix\n",
            "INFO [2025-06-10 08:58:52] Parsing gene order file: gene_pos_file.txt\n",
            "INFO [2025-06-10 08:58:52] Parsing cell annotations file: /content/1_vs_4.annotations.tsv\n",
            "INFO [2025-06-10 08:58:52] ::order_reduce:Start.\n",
            "INFO [2025-06-10 08:58:53] .order_reduce(): expr and order match.\n",
            "INFO [2025-06-10 08:58:54] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 36601,7783 Total=91895129 Min=0 Max=9376.\n",
            "INFO [2025-06-10 08:58:55] num genes removed taking into account provided gene ordering list: 1280 = 3.49717220840961% removed.\n",
            "INFO [2025-06-10 08:58:55] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
            "WARN [2025-06-10 08:58:56] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-06-10 08:59:02] validating infercnv_obj\n"
          ]
        }
      ],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/1_and_4.matrix\",\n",
        "                                    annotations_file=\"/content/1_vs_4.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "N2t0ZjiuSoRM",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c2dd399f-4883-48e3-9861-9a8a1bb8bb84",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 08:59:02] ::process_data:Start\n",
            "INFO [2025-06-10 08:59:02] Creating output path 1_vs_4\n",
            "INFO [2025-06-10 08:59:02] Checking for saved results.\n",
            "INFO [2025-06-10 08:59:02] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-06-10 08:59:41] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-06-10 08:59:41] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-06-10 08:59:42] Removing 18557 genes from matrix as below mean expr threshold: 0.01\n",
            "INFO [2025-06-10 08:59:42] validating infercnv_obj\n",
            "INFO [2025-06-10 08:59:42] There are 16764 genes and 7783 cells remaining in the expr matrix.\n",
            "INFO [2025-06-10 08:59:48] no genes removed due to min cells/gene filter\n",
            "INFO [2025-06-10 09:00:18] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-06-10 09:00:18] normalizing counts matrix by depth\n",
            "INFO [2025-06-10 09:00:22] Computed total sum normalization factor as median libsize: 8174.000000\n",
            "INFO [2025-06-10 09:00:56] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-06-10 09:00:56] transforming log2xplus1()\n",
            "INFO [2025-06-10 09:01:31] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-06-10 09:01:31] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 09:01:31] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 09:01:43] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 09:03:02] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-06-10 09:03:02] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-06-10 09:04:12] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-06-10 09:04:12] smooth_by_chromosome: chr: chr1\n",
            "INFO [2025-06-10 09:04:29] smooth_by_chromosome: chr: chr2\n",
            "INFO [2025-06-10 09:04:43] smooth_by_chromosome: chr: chr3\n",
            "INFO [2025-06-10 09:04:56] smooth_by_chromosome: chr: chr4\n",
            "INFO [2025-06-10 09:05:09] smooth_by_chromosome: chr: chr5\n",
            "INFO [2025-06-10 09:05:21] smooth_by_chromosome: chr: chr6\n",
            "INFO [2025-06-10 09:05:35] smooth_by_chromosome: chr: chr7\n",
            "INFO [2025-06-10 09:05:48] smooth_by_chromosome: chr: chr8\n",
            "INFO [2025-06-10 09:06:00] smooth_by_chromosome: chr: chr9\n",
            "INFO [2025-06-10 09:06:12] smooth_by_chromosome: chr: chr10\n",
            "INFO [2025-06-10 09:06:25] smooth_by_chromosome: chr: chr11\n",
            "INFO [2025-06-10 09:06:38] smooth_by_chromosome: chr: chr12\n",
            "INFO [2025-06-10 09:06:52] smooth_by_chromosome: chr: chr13\n",
            "INFO [2025-06-10 09:07:02] smooth_by_chromosome: chr: chr14\n",
            "INFO [2025-06-10 09:07:14] smooth_by_chromosome: chr: chr15\n",
            "INFO [2025-06-10 09:07:26] smooth_by_chromosome: chr: chr16\n",
            "INFO [2025-06-10 09:07:39] smooth_by_chromosome: chr: chr17\n",
            "INFO [2025-06-10 09:07:52] smooth_by_chromosome: chr: chr18\n",
            "INFO [2025-06-10 09:08:03] smooth_by_chromosome: chr: chr19\n",
            "INFO [2025-06-10 09:08:17] smooth_by_chromosome: chr: chr20\n",
            "INFO [2025-06-10 09:08:28] smooth_by_chromosome: chr: chr21\n",
            "INFO [2025-06-10 09:08:38] smooth_by_chromosome: chr: chr22\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: GL000194.1\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: GL000195.1\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: GL000218.1\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: GL000219.1\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: KI270711.1\n",
            "INFO [2025-06-10 09:08:50] smooth_by_chromosome: chr: KI270721.1\n",
            "INFO [2025-06-10 09:09:59] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-06-10 09:09:59] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-06-10 09:11:32] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-06-10 09:11:32] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 09:11:32] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 09:11:44] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 09:13:10] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-06-10 09:13:10] invert_log2(), computing 2^x\n",
            "INFO [2025-06-10 09:14:55] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-06-10 09:14:55] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-06-10 09:15:00] define_signif_tumor_subclusters(), tumor: aktywowane HSC\n",
            "INFO [2025-06-10 09:15:00] Less cells in group aktywowane HSC than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:00] define_signif_tumor_subclusters(), tumor: CD4 T-cells\n",
            "INFO [2025-06-10 09:15:00] Too few cells in group CD4 T-cells for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 09:15:00] define_signif_tumor_subclusters(), tumor: CD8 T-cells\n",
            "INFO [2025-06-10 09:15:00] Less cells in group CD8 T-cells than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:00] define_signif_tumor_subclusters(), tumor: GMP\n",
            "INFO [2025-06-10 09:15:00] Setting auto leiden resolution for GMP to 0.166085\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  TMEM223, TAF6L, AP001160.3, TMEM179B, POLR2G, NXF1, ZBTB3, STX5, AP001160.1, TTC9C \n",
            "\t   WDR74, SNHG1, HNRNPUL2, SLC3A2, BSCL2, SLC22A10, UBXN1, LGALS12, UQCC3, PLAAT4 \n",
            "\t   CSKMT, PLAAT3, LBHD1, ATL3, C11orf98, AP000753.2, RTN3, INTS5, C11orf95, GANAB \n",
            "Negative:  AASS, FAM3C, CADPS2, NDUFA5, CPED1, LMOD2, ING3, WASL, AC006333.2, PCLO \n",
            "\t   SEMA3E, SEMA3A, POT1, CACNA2D1, AC074183.1, GRM3, HGF, KIAA1324L, SEMA3C, POT1-AS1 \n",
            "\t   CD36, AC005076.1, ZNF800, DMTF1, GNAI1, GCC1, TMEM243, MAGI2-AS3, TP53TG1, ARF5 \n",
            "PC_ 2 \n",
            "Positive:  FYN, TUBE1, FAM229B, Z97989.1, AL360015.1, LINC02541, TRAF3IP2, MARCKS, AL008730.1, HDAC2 \n",
            "\t   TRAF3IP2-AS1, AL080317.2, HDAC2-AS2, REV3L, AL080317.1, NT5DC1, MFSD4B, COL10A1, AL360227.1, SLC16A10 \n",
            "\t   TSPYL4, RPF2, DSE, GTF3C6, TSPYL1, AMD1, CDK19, CALHM6, AC002464.1, RWDD1 \n",
            "Negative:  AC073349.4, ZNF273, AC073349.1, AC073210.3, ZNF117, ZPBP, ERV3-1, ABCA13, ZNF92, UPP1 \n",
            "\t   VKORC1L1, HUS1, GUSB, TNS3, AC004870.3, ASL, AC004870.4, TBRG4, CRCP, CCM2 \n",
            "\t   SNHG15, MYO1G, AC068533.3, AC004847.1, AC004854.2, PURB, TPST1, LINC01952, H2AFV, LINC00174 \n",
            "PC_ 3 \n",
            "Positive:  SPART-AS1, CCNA1, RFXAP, ALG5, EXOSC8, SUPT20H, LINC02334, LINC00571, UFM1, PROSER1 \n",
            "\t   NHLRC3, AL354809.1, LHFPL6, COG6, LINC00598, FOXO1, MRPS31, POLE2, DNAAF2, MGAT2 \n",
            "\t   LRR1, KLHDC1, RPL36AL, SLC25A15, RPS29, AL591767.1, MDGA2, LINC00871, AL590064.1, MIS18BP1 \n",
            "Negative:  AL627309.5, LINC01409, LINC01128, LINC00115, FAM41C, AL645608.6, NOC2L, ISG15, AGRN, C1orf159 \n",
            "\t   AL390719.3, TNFRSF18, SDF4, B3GALT6, UBE2J2, LINC01786, SCNN1D, ACAP3, PUSL1, INTS11 \n",
            "\t   CPTP, DVL1, AURKAIP1, CCNL2, MRPL20-AS1, MRPL20, AL391244.2, ATAD3B, ATAD3A, SSU72 \n",
            "PC_ 4 \n",
            "Positive:  ELMO1, AOAH, ELMO1-AS1, AC006960.2, GPR141, EPDR1, ANLN, NME8, KIAA0895, EEPD1 \n",
            "\t   SFRP4, SEPTIN7, STARD3NL, SEPTIN7-AS1, TRGC2, AC018647.2, TRGC1, HERPUD2, TRG-AS1, AC005400.1 \n",
            "\t   AMPH, DPY19L1, VPS41, BBS9, POU6F2, RP9, YAE1, NT5C3A, RALA, FKBP9 \n",
            "Negative:  HOOK1, FGGY, AC093425.1, LINC01358, LINC01135, JUN, MYSM1, OMA1, DAB1, PLPP3 \n",
            "\t   AC119674.1, MIR4422HG, USP24, DHCR24, TTC4, LRGUK, EXOC4, SLC35B4, AC009365.4, AKR1B1 \n",
            "\t   CHCHD3, PLXNA4, BPGM, PODXL, CALD1, AC008264.2, MKLN1-AS, AC016831.5, AC016831.1, AC058791.1 \n",
            "PC_ 5 \n",
            "Positive:  RALA, LINC00265, YAE1, AC072061.1, CDK13, POU6F2, MPLKIP, VPS41, SUGCT, AMPH \n",
            "\t   GLI3, TRG-AS1, C7orf25, TRGC1, PSMA2, MRPL32, TRGC2, STK17A, COA1, STARD3NL \n",
            "\t   BLVRA, SFRP4, MRPS24, FBXW7, LINC02273, AC023424.3, MIR4453HG, TMEM154, ARFIP1, AC097375.5 \n",
            "Negative:  RTTN, CD226, DOK6, CCDC102B, SOCS6, TMX3, LINC01909, NETO1, AC114689.3, FBXO15 \n",
            "\t   DSEL, TIMM21, CDH7, LINC01924, SERPINB8, SERPINB10, SERPINB2, VPS4B, AC036176.1, KDSR \n",
            "\t   BCL2, PHLPP1, ZCCHC2, TNFRSF11A, RELCH, PIGN, RNF152, PMAIP1, LMAN1, SEC11C \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:01] define_signif_tumor_subclusters(), tumor: HSC\n",
            "INFO [2025-06-10 09:15:01] Setting auto leiden resolution for HSC to 0.0427056\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  VEGFB, DNAJC4, FKBP2, PPP1R14B, AP001453.1, PPP1R14B-AS1, NUDT22, TRPT1, PLCB3, FERMT3 \n",
            "\t   BAD, STIP1, GPR137, FLRT1, ESRRA, AP000721.2, MACROD1, TRMT112, OTUB1, PRDX5 \n",
            "\t   COX8A, NAA40, MARK2, CCDC88B, SPINDOC, C11orf95, RTN3, RPS6KA4, AP000753.2, ATL3 \n",
            "Negative:  DHRS7, PCNX4, AL157911.1, LRRC9, RTN1, CCDC175, JKAMP, L3HYPDH, GPR135, DAAM1 \n",
            "\t   LINC01500, KIAA0586, TIMM9, AL139021.1, TOMM20L, AL139021.2, ARID4A, AL132989.2, PSMA3-AS1, AL132989.1 \n",
            "\t   AL135978.2, ATP6V1D, PSMA3, EIF2S1, PLEK2, ACTR10, TMEM229B, PIGH, ARMH4, ARG2 \n",
            "PC_ 2 \n",
            "Positive:  TNRC18, SLC29A4, WIPI2, RBAK, AC092171.3, RADIL, AP5Z1, AC092171.5, AC092428.1, FBXL18 \n",
            "\t   ACTB, FOXK1, FSCN1, SDK1, RNF216, AC024028.1, OCM, CARD11, CCZ1, PMS2 \n",
            "\t   GNA12, AMZ1, AIMP2, TTYH3, EIF2AK1, IQCE, ANKRD61, BRAT1, USP42, LFNG \n",
            "Negative:  BCAS2, DENND2C, NRAS, CSDE1, LINC01765, TRIM33, SIKE1, TSPAN2, VANGL1, SLC22A15 \n",
            "\t   HIPK1, AL365318.1, HIPK1-AS1, ATP1A1, DCLRE1B, AP4B1, ATP1A1-AS1, AP4B1-AS1, LINC01762, PTPN22 \n",
            "\t   AL137856.1, AL390066.1, RSBN1, CD58, PHTF1, IGSF3, MAGI3, LRIG2, CD2, LRIG2-DT \n",
            "PC_ 3 \n",
            "Positive:  RPA3, UMAD1, MIOS, AC004982.2, AC006042.2, AC005532.1, GLCCI1, C1GALT1, ICA1, AC079804.3 \n",
            "\t   AC007009.1, CCZ1B, AC060834.2, ZNF12, NDUFA4, PHF14, AC073343.2, AC004160.1, THSD7A, ZNF316 \n",
            "\t   TMEM106B, ZNF853, VWDE, SCIN, C7orf26, ARL4A, ZDHHC4, AC011287.1, GRID2IP, CRPPA \n",
            "Negative:  RTRAF, GNG2, FRMD6-AS1, NID2, FRMD6, TMX1, PTGDR, TRIM9, PTGER2, ABHD12B \n",
            "\t   TXNDC16, PYGL, GPR137C, AL606834.2, ERO1A, NIN, AL133453.1, AL606834.1, PSMC6, SAV1 \n",
            "\t   STYX, GNPNAT1, ATL1, AL139317.5, AL118556.1, FERMT2, MAP4K5, DDHD1, CDKL1, AL356020.1 \n",
            "PC_ 4 \n",
            "Positive:  RTRAF, NID2, GNG2, FRMD6-AS1, PTGDR, PTGER2, FRMD6, TXNDC16, GPR137C, TMX1 \n",
            "\t   ERO1A, TRIM9, AL133453.1, ABHD12B, PSMC6, PYGL, STYX, AL606834.2, GNPNAT1, NIN \n",
            "\t   AL139317.5, AL606834.1, FERMT2, SAV1, DDHD1, ATL1, AL356020.1, AL118556.1, AL365295.1, MAP4K5 \n",
            "Negative:  PPP6R2, SBF1, DENND6B, LMF2, PLXNB2, NCAPH2, MAPK11, SCO2, MAPK12, HDAC10 \n",
            "\t   TYMP, TUBGCP6, SELENOO, ODF3B, AL022328.4, U62317.4, AL022328.3, TRABD, CPT1B, CHKB \n",
            "\t   PANX2, CHKB-DT, MLC1, ARSA, PIM3, AC000050.1, SHANK3, CRELD2, RABL2B, ALG12 \n",
            "PC_ 5 \n",
            "Positive:  TSPAN11, AC010198.2, DDX11-AS1, CAPRIN2, DDX11, IPO8, SINHCAF, TMTC1, DENND5B, OVCH1 \n",
            "\t   DENND5B-AS1, OVCH1-AS1, ETFBKMT, ERGIC2, AMN1, AC009318.3, AC023157.3, AC009318.1, LINC02422, FAR2 \n",
            "\t   RESF1, AC012150.1, AC022364.1, CCDC91, KLHL42, AC016957.2, MRPS35, AC009509.1, BICD1, PPFIBP1 \n",
            "Negative:  ADGRE5, LINC01841, AC008569.1, AC011509.2, ADGRL1, DDX39A, AC022098.1, ASF1B, PKN1, PRKACA \n",
            "\t   SAMD1, AC008569.2, IL27RA, GIPC1, RFX1, DNAJB1, DCAF15, TECR, CC2D1A, NDUFB7 \n",
            "\t   C19orf57, ADGRE3, ZNF333, NANOS3, ADGRE2, AC020916.1, SLC1A6, ILVBL, ZSWIM4, BRD4 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:04] define_signif_tumor_subclusters(), tumor: komórki B pamięci\n",
            "INFO [2025-06-10 09:15:04] Setting auto leiden resolution for komórki B pamięci to 0.149089\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  PALB2, NDUFAB1, AC008870.2, AC008870.4, UBFD1, EARS2, DCTN5, GGA2, PLK1, AC008915.2 \n",
            "\t   PRKCB, COG7, USP31, NPIPB5, RBBP6, AC092338.2, CDR2, AC092338.3, POLR3E, TNRC6A \n",
            "\t   EEF2K, SDR42E2, MOSMO, AC008731.1, UQCRC2, AC092119.2, NPIPB4, SLC5A11, OTOA, IGSF6 \n",
            "Negative:  DIO2-AS1, CEP128, DIO2, TSHR, AC008056.1, AC007262.2, AC022469.2, AL136040.2, NRXN3, GTF2A1 \n",
            "\t   MARCH6, ADCK1, AC012640.2, STON2, CMBL, ROPN1L, SEL1L, C14orf178, CCT5, ATPSCKMT \n",
            "\t   ANKRD33B, SEMA5A, SNW1, LINC02301, MIR4458HG, DAP, SLIRP, FASTKD3, ALKBH1, CTNND2 \n",
            "PC_ 2 \n",
            "Positive:  MME-AS1, PLCH1, MME, C3orf33, SLC33A1, GMPS, DHX36, KCNAB1, ARHGEF26, ARHGEF26-AS1 \n",
            "\t   AC091607.2, AC084036.1, RAP2B, SSR3, P2RY1, TIPARP-AS1, AC026347.1, TIPARP, MBNL1-AS1, LINC00886 \n",
            "\t   LEKR1, MBNL1, CCNL1, AC108718.1, AC092944.1, VEPH1, SUCNR1, PTX3, RSRC1, AADACL2-AS1 \n",
            "Negative:  DENND3, METAP1, AC019131.2, ADH5, AP002026.1, ADH4, TRMT10A, DAPP1, LAMTOR3, DNAJB14 \n",
            "\t   H2AFZ, AC097460.1, DDIT4L, EMCN, AC097459.1, PPP3CA, AP001816.1, BANK1, AP002075.1, SLC39A8 \n",
            "\t   AC098487.1, AF213884.3, NFKB1, MANBA, UBE2D3, DUSP4, SARAF, AC044849.1, LEPROTL1, MBOAT4 \n",
            "PC_ 3 \n",
            "Positive:  FKBP8, AC005387.2, KXD1, AC005253.1, UBA52, CRLF1, REX1BD, ESRRA, GPR137, BAD \n",
            "\t   PLCB3, KLHL26, PPP1R14B-AS1, PPP1R14B, FKBP2, CRTC1, VEGFB, DNAJC4, AP001453.1, NUDT22 \n",
            "\t   UPF1, TRPT1, FERMT3, STIP1, COPE, FLRT1, AP000721.2, MACROD1, DDX49, OTUB1 \n",
            "Negative:  RYBP, LINC00870, SHQ1, LINC00877, PROK2, GXYLT2, GPR27, PPP4R2, EIF4E3, EBLN2 \n",
            "\t   PDZRN3, AC097634.1, LINC02018, MFSD14A, SLC35A3, AC118553.1, AC093019.2, LINC00960, AGL, SASS6 \n",
            "\t   AC097634.3, ZNF717, FRRS1, TRMT13, ROBO2, DPYD-AS1, FOXP1-IT1, LRRC39, ALDH8A1, DBT \n",
            "PC_ 4 \n",
            "Positive:  EIF2B1, GTF2H3, DDX55, TMED2, TCTN2, ATP6V0A2, TMED2-DT, RILPL1, AC117503.5, SNRNP35 \n",
            "\t   DNAH10, RILPL2, CCDC92, KMT5A, ZNF664, SBNO1-AS1, SBNO1, RFLNA, CDK2AP1, C12orf65 \n",
            "\t   NCOR2, MPHOSPH9, AC073592.1, PITPNM2, ARL6IP4, SCARB1, OGFOD2, AC073593.2, ABCB9, VPS37B \n",
            "Negative:  CALHM6, RWDD1, DSE, TSPYL1, DCBLD1, GOPC, TSPYL4, RSPH4A, NUS1, COL10A1 \n",
            "\t   KPNA5, ZUP1, SLC35F1, CEP85L, NT5DC1, PLN, HDAC2-AS2, HDAC2, Z99496.2, MARCKS \n",
            "\t   MCM9, LINC02541, ASF1A, AL137009.1, AL360015.1, FAM184A, FAM229B, MAN1A1, AL078600.1, TUBE1 \n",
            "PC_ 5 \n",
            "Positive:  SEC24D, SYNPO2, METTL14, SNHG8, USP53, C4orf3, PDE5A, NDST3, MAD2L1, AC073475.1 \n",
            "\t   CAMK2D, ANK2, PRDM5, LARP7, ANXA5, ZGRF1, EXOSC9, ALPK1, CCNA2, TIFA \n",
            "\t   BBS7, AP1AR, KIAA1109, FAM241A, NUDT6, AC004704.1, SPATA5, AC139718.2, SPRY1, LINC01091 \n",
            "Negative:  E2F8, NAV2, HTATIP2, CSRP3-AS1, ZDHHC13, SPTY2D1, PRMT3, SPTY2D1OS, ANO5, FANCF \n",
            "\t   UEVLD, LINC02718, CCDC34, GAS2, AC006299.1, SVIP, LGR4, LIN7C, TSG101, BDNF-AS \n",
            "\t   LDHAL6A, KIF18A, METTL15, LINC02755, ARL14EP, LDHC, MPPED2, DNAJC24, LDHA, CCDC73 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:06] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 1 (preB)\n",
            "INFO [2025-06-10 09:15:06] Less cells in group Komórki blastyczne 1 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:06] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 2 (preB)\n",
            "INFO [2025-06-10 09:15:06] Less cells in group Komórki blastyczne 2 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:06] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 3 (preB)\n",
            "INFO [2025-06-10 09:15:06] Setting auto leiden resolution for Komórki blastyczne 3 (preB) to 0.0475954\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  UHRF1, RPL36, KDM4B, SAFB, MICOS13, SAFB2, TINCR, PTPRS, HSD11B1L, LONP1 \n",
            "\t   CATSPERD, PRR22, DUS3L, NDUFA11, VMAC, AC104532.2, CAPS, RANBP3, AC011444.1, RFX2 \n",
            "\t   AC011444.2, AC011444.3, ACSBG2, MLLT1, CLPP, ALKBH7, PSPN, GTF2F1, KHSRP, SLC25A23 \n",
            "Negative:  DUSP15, TTLL9, PDRG1, HCK, TM9SF4, PLAGL2, POFUT1, KIF3B, ASXL1, ZNF337-AS1 \n",
            "\t   NOL4L, NANP, AL034550.2, NFS1, RBM12, ROMO1, RBM39, CPNE1, PHF20, ERGIC3 \n",
            "\t   SCAND1, CNBD2, FO393401.1, C20orf203, NORAD, EPB41L1, AAR2, DLGAP4, CEP250, NINL \n",
            "PC_ 2 \n",
            "Positive:  HDDC2, NCOA7, NCOA7-AS1, RNF217, HINT3, RNF217-AS1, TRMT11, NKAIN2, CENPW, SMPDL3A \n",
            "\t   RNF146, PKIB, SERINC1, ECHDC1, HSF2, TBC1D32, SOGA3, AL078600.1, MAN1A1, FAM184A \n",
            "\t   AL137009.1, ASF1A, C6orf58, MCM9, Z99496.2, PLN, THEMIS, CEP85L, SLC35F1, PTPRK \n",
            "Negative:  AL133230.1, RIPOR3, CEBPB, SMIM25, PARD6B, PTPN1, CEBPB-AS1, TMEM189, BCAS4, ADNP \n",
            "\t   ADNP-AS1, SALL4, ATP9A, DPM1, MOCS3, ZFP64, NFATC2, AL109984.1, LINC01524, TSHZ2 \n",
            "\t   AL109930.1, AL157838.1, ZNF217, PFDN4, LINC01440, FAM210B, AURKA, CSTF1, CASS4, RTF2 \n",
            "PC_ 3 \n",
            "Positive:  AHDC1, FGR, IFI6, FO393419.2, FAM76A, WASF2, STX12, MAP3K6, PPP1R8, SYTL1 \n",
            "\t   THEMIS2, TMEM222, RPA2, WDTC1, SLC9A1, SMPDL3B, NUDC, AL512288.1, GPATCH3, XKR8 \n",
            "\t   GPN2, EYA3, ZDHHC18, PTAFR, PIGV, ARID1A, RPS6KA1, HMGN2, AL513365.2, DHDDS \n",
            "Negative:  SLC25A3, TMPO, IKBIP, TMPO-AS1, APAF1, LINC02453, MALT1, NEDD4L, AC104365.1, ATP8B1 \n",
            "\t   ANKS1B, AC027097.2, AC007656.2, ZNF532, AC027097.1, NARS, NEDD1, SEC11C, FECH, CFAP54 \n",
            "\t   LMAN1, UHRF1BP1L, WDR7, CDK17, TXNL1, PMAIP1, TCF4-AS2, TCF4-AS1, TCF4, AC008149.2 \n",
            "PC_ 4 \n",
            "Positive:  CNKSR1, ZNF593, CEP85, FAM110D, SH3BGRL3, PDIK1L, UBXN11, PAFAH2, CD52, STMN1 \n",
            "\t   CRYBG2, PAQR7, AL033528.2, DHDDS, AUNIP, MTFR1L, AL513365.2, SELENON, HMGN2, RPS6KA1 \n",
            "\t   ARID1A, PIGV, ZDHHC18, GPN2, GPATCH3, NUDC, SLC9A1, WDTC1, TMEM222, SYTL1 \n",
            "Negative:  TCF4, TCF4-AS1, RAB27B, TCF4-AS2, C18orf54, TXNL1, POLI, WDR7, MBD2, FECH \n",
            "\t   DCC, NARS, AC027097.1, AC027097.2, ATP8B1, NEDD4L, MALT1, AC104365.1, ZNF532, SEC11C \n",
            "\t   LMAN1, PMAIP1, RNF152, PIGN, RELCH, AC100778.3, TNFRSF11A, ZCCHC2, C18orf32, PHLPP1 \n",
            "PC_ 5 \n",
            "Positive:  EIF3B, SNX8, CHST12, LFNG, NUDT1, BRAT1, MRM2, IQCE, SDK1, FOXK1 \n",
            "\t   AC024028.1, AC092428.1, CARD11, TTYH3, GNA12, AMZ1, MAD1L1, AP5Z1, RBAK, WIPI2 \n",
            "\t   RADIL, SLC29A4, ELFN1-AS1, PSMG3-AS1, TNRC18, PSMG3, AC092171.3, TMEM184A, INTS1, MAFK \n",
            "Negative:  HPS5, SAAL1, GTF2H1, TPH1, LDHA, SERGEF, LDHC, KCNC1, NCR3LG1, LDHAL6A \n",
            "\t   NUCB2, TSG101, PIK3C2A, RPS13, UEVLD, PLEKHA7, SPTY2D1OS, C11orf58, SPTY2D1, AC103794.1 \n",
            "\t   ZDHHC13, SOX6, CSRP3-AS1, CALCB, E2F8, NAV2, CYP2R1, HTATIP2, PDE3B, PRMT3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:08] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 4 (preB)\n",
            "INFO [2025-06-10 09:15:08] Setting auto leiden resolution for Komórki blastyczne 4 (preB) to 0.0631938\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  TRAPPC12, EIPR1, ADI1, LINC01250, AC019118.2, AC093390.2, AC231981.1, AC108488.1, MYT1L, RNASEH1 \n",
            "\t   RNASEH1-AS1, PXDN, AC108488.3, RPS7, NRIR, CMPK2, RSAD2, RNF144A, AC068481.1, LINC01871 \n",
            "\t   LINC00299, LINC01814, SPPL2B, TMPRSS9, TIMM13, AC092068.2, LMNB2, GNG7, GADD45B, AC092068.3 \n",
            "Negative:  NORAD, CNBD2, SCAND1, EPB41L1, AAR2, PHF20, DLGAP4, RBM39, ROMO1, DLGAP4-AS1 \n",
            "\t   NFS1, RBM12, MYL9, CPNE1, TGIF2, ERGIC3, FO393401.1, RAB5IF, CEP250, UQCC1 \n",
            "\t   SLA2, EIF6, NDRG3, MMP24, MMP24OS, DSN1, PROCR, SOGA1, EDEM2, TLDC2 \n",
            "PC_ 2 \n",
            "Positive:  PSMD6-AS1, PSMD6, MAGI1, SLC25A26, PSMD6-AS2, LRIG1, ATXN7, KBTBD8, AC104162.2, SUCLG2 \n",
            "\t   THOC7, SUCLG2-AS1, TAFA1, C3orf49, EOGT, AC109587.1, LINC00698, TMF1, CADPS, UBA3 \n",
            "\t   C3orf14, ARL6IP5, PTPRG-AS1, LMOD3, PTPRG, FRMD4B, FHIT, MITF, AC126121.3, FOXP1 \n",
            "Negative:  CARD14, AC087741.1, SGSH, EIF4A3, GAA, SLC26A11, CCDC40, RNF213, TBC1D16, RNF213-AS1 \n",
            "\t   CBX4, ENDOV, BAIAP2-DT, BAIAP2, AC127496.5, CHMP6, RPTOR, AATK, CEP131, TEPSIN \n",
            "\t   NDUFAF8, SLC38A10, AC027601.6, AC110285.1, BAHCC1, ACTG1, AC139149.1, FAAP100, NPLOC4, TSPAN10 \n",
            "PC_ 3 \n",
            "Positive:  SEMA6A-AS1, SEMA6A, LINC02147, LINC02208, AC018752.1, AC093206.1, DTWD2, COMMD10, AC008629.1, AC034236.2 \n",
            "\t   DMXL1, TNFAIP8, AP3S1, HSD17B4, ATG12, AC008574.1, TMED7, AC113418.1, AC010226.1, PRR16 \n",
            "\t   TICAM2, FEM1C, AC008568.1, CCDC112, SRFBP1, PGGT1B, YTHDC2, LOX, MCC, DCP2 \n",
            "Negative:  FANCE, RPL10A, TEAD3, FKBP5, AL033519.5, LHFPL5, SRPK1, SLC26A8, MAPK14, MAPK13 \n",
            "\t   Z84485.1, BRPF3, PXT1, KCTD20, STK38, SRSF3, CDKN1A, RAB44, CPNE5, PPIL1 \n",
            "\t   C6orf89, MTCH1, FGD2, PIM1, TMEM217, TBC1D22B, AL096712.2, RNF8, CMTR1, CCDC167 \n",
            "PC_ 4 \n",
            "Positive:  PDIK1L, PAFAH2, FAM110D, STMN1, ZNF593, CNKSR1, PAQR7, CEP85, SH3BGRL3, UBXN11 \n",
            "\t   AL033528.2, CD52, CRYBG2, DHDDS, AL513365.2, HMGN2, AUNIP, RPS6KA1, ARID1A, PIGV \n",
            "\t   MTFR1L, ZDHHC18, GPN2, SELENON, GPATCH3, NUDC, SLC9A1, WDTC1, TMEM222, SYTL1 \n",
            "Negative:  LDHAL6A, LDHC, TSG101, LDHA, UEVLD, GTF2H1, SPTY2D1OS, HPS5, SPTY2D1, SAAL1 \n",
            "\t   ZDHHC13, TPH1, CSRP3-AS1, SERGEF, E2F8, KCNC1, NAV2, NCR3LG1, HTATIP2, NUCB2 \n",
            "\t   PRMT3, PIK3C2A, RPS13, ANO5, PLEKHA7, C11orf58, FANCF, AC103794.1, SOX6, CALCB \n",
            "PC_ 5 \n",
            "Positive:  PCCA-DT, AL137139.2, CLYBL, LINC01232, TM9SF2, GPR183, GPR18, GAA, CCDC40, UBAC2 \n",
            "\t   TBC1D16, FBXL4, EIF4A3, AL589740.1, MMS22L, COQ3, CBX4, PREP, PNISR, CARD14 \n",
            "\t   AL591518.1, AL096794.1, BVES, PRDM1, HACE1, KLHL32, AL513550.1, GRIK2, ATG5, AC087741.1 \n",
            "Negative:  ZBED5-AS1, USP47, GALNT18, ZBED5, DKK3, AC116535.1, MICAL2, TEAD1, EIF4G2, LINC00958 \n",
            "\t   CTR9, ARNTL, MRVI1, BTBD10, FAR1, MRVI1-AS1, FAR1-IT1, SPON1, RNF141, SPON1-AS1 \n",
            "\t   MTRNR2L8, RRAS2, COPB1, AMPD3, PSMA1, ADM, PDE3B, CYP2R1, CALCB, SOX6 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:11] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 5 (preB)\n",
            "INFO [2025-06-10 09:15:11] Setting auto leiden resolution for Komórki blastyczne 5 (preB) to 0.482523\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AL117379.1, GID8, DIDO1, SLC17A9, TCFL5, OGFR, MRGBP, NTSR1, SLCO4A1, BX640514.2 \n",
            "\t   CABLES2, RPS21, AL121832.2, LAMA5, ADRM1, OSBPL2, PNKP, AKT1S1, EIF5A, YBX2 \n",
            "\t   GPS2, TBC1D17, SLC2A4, NEURL4, CLDN7, ACAP1, PLSCR3, ELP5, CTDNEP1, TMEM256 \n",
            "Negative:  AC069542.1, STAM, STAM-AS1, HACD1, ST8SIA6-AS1, ST8SIA6, AL133415.1, VIM-AS1, VIM, TRDMT1 \n",
            "\t   CUBN, RSU1, AL353576.1, PTER, MINDY3, ITGA8, FAM171A1, NMT2, RPP38, RPP38-DT \n",
            "\t   ACBD7, OLAH, KLHDC10, AC073320.1, ZC3HC1, UBE2H, TMEM209, MEIG1, DCLRE1C, AC084864.2 \n",
            "PC_ 2 \n",
            "Positive:  Z97192.1, TAFA5, TBC1D22A, Z97192.2, AL118516.1, C22orf34, GRAMD4, CERK, Z97192.3, CELSR1 \n",
            "\t   BRD1, Z98885.3, ZBED4, TRMU, ALG12, GTSE1, CRELD2, TTC38, PIM3, CDPF1 \n",
            "\t   MLC1, PPARA, PANX2, AL121672.2, TRABD, MIRLET7BHG, AL022328.3, AL022328.4, PRR34-AS1, SELENOO \n",
            "Negative:  SH3BGRL3, CEP85, CNKSR1, ZNF593, FAM110D, UBXN11, PDIK1L, CD52, WASF2, FO393419.2 \n",
            "\t   PAFAH2, MAP3K6, SYTL1, CRYBG2, AHDC1, TMEM222, STMN1, FGR, BSCL2, HNRNPUL2 \n",
            "\t   WDTC1, UQCC3, DHDDS, UBXN1, AL033528.2, PAQR7, CSKMT, AUNIP, MTFR1L, IFI6 \n",
            "PC_ 3 \n",
            "Positive:  COA1, BLVRA, MRPS24, URGCP, UBE2D4, SPDYE1, AC004951.1, LINC00957, ZPBP, YKT6 \n",
            "\t   AEBP1, POLD2, POLM, NUDCD3, ABCA13, DBNL, UPP1, DDX56, HUS1, TMED4 \n",
            "\t   OGDH, TNS3, AC004870.3, ZMIZ2, AC004870.4, TBRG4, PPIA, CCM2, H2AFV, LINC01952 \n",
            "Negative:  UVRAG, THAP12, GVQW3, DGAT2, AP002360.1, EMSY, SERPINH1, ACER3, AP002498.1, CAPN5 \n",
            "\t   GDPD5, PAK1, KLHL35, RPS3, CLNS1A, AQP11, USP35, GAB2, KCTD21, LINC02728 \n",
            "\t   AP003086.1, KCTD21-AS1, RSF1, ALG8, AAMDC, AP002812.2, NDUFC2, INTS4, ARRB1, NEU3 \n",
            "PC_ 4 \n",
            "Positive:  TMEM43, AC093495.1, XPC, LSM3, SLC6A6, CCDC174, C3orf20, FGD5, FGD5-AS1, NR2C2 \n",
            "\t   MRPS25, PLK2, AC016644.1, GPBP1, AC008937.3, GAPT, MIER3, RBSN, RAB3C, SETD9 \n",
            "\t   AC008937.2, PDE4D, MAP3K1, ANKRD55, LINC01948, AC092343.1, PART1, DEPDC1B, ELOVL7, ERCC8 \n",
            "Negative:  AL731769.2, CYP2E1, SCART1, INPP5A, AL451069.2, PWWP2B, LRRC27, STK32C, JAKMIP3, ADAP1 \n",
            "\t   GPER1, GPR146, ZFAND2A, COX19, AC073957.1, AC091729.3, AC073957.3, INTS1, MICALL2, C7orf50 \n",
            "\t   AL162274.2, MAFK, GET4, AC093627.5, AC093627.4, AC215522.2, FAM20C, SUN1, PDGFA, TMEM184A \n",
            "PC_ 5 \n",
            "Positive:  KLHDC1, POLE2, AL591767.1, MGAT2, DNAAF2, RPL36AL, LRR1, KLHDC2, MDGA2, RPS29 \n",
            "\t   LINC00871, MIS18BP1, NEMF, AL627171.2, AL627171.1, ARF6, LINC01588, VCPKMT, SOS2, L2HGDH \n",
            "\t   DMAC2L, CDKL1, MAP4K5, AL118556.1, AL109628.2, ATL1, SAV1, TRAPPC6B, AL606834.1, GEMIN2 \n",
            "Negative:  SIDT2, TAGLN, PAFAH1B2, PCSK7, SIK3, APOA1-AS, APOA1, ZPR1, BUD13, REXO2 \n",
            "\t   CADM1, RBM7, C11orf71, AC007656.2, LINC02453, NEDD1, TMPO-AS1, CFAP54, TMPO, CDK17 \n",
            "\t   SLC25A3, AC008149.2, IKBIP, ELK3, APAF1, LINC02452, ANKS1B, LTA4H, HAL, UHRF1BP1L \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:12] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 6 (preB)\n",
            "INFO [2025-06-10 09:15:12] Setting auto leiden resolution for Komórki blastyczne 6 (preB) to 0.237326\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  Z97989.1, FYN, AL008730.1, TRAF3IP2, TRAF3IP2-AS1, TUBE1, AL080317.2, FAM229B, REV3L, RPF2 \n",
            "\t   GTF3C6, AL080317.1, SLC16A10, AL360227.1, AL360015.1, MFSD4B, AMD1, CDK19, LINC02541, AC002464.1 \n",
            "\t   SLC22A16, DDO, METTL24, CDC40, MARCKS, WASF1, FIG4, AK9, HDAC2, ZBTB24 \n",
            "Negative:  AC104083.1, GUCY1A1, GUCY1B1, MAP9, AC097467.3, PDGFC, CTSO, GLRB, AC093817.2, PLRG1 \n",
            "\t   GASK1B, FAM198B-AS1, RNF175, TMEM144, RXFP1, C4orf46, TLR2, ETFDH, AC106865.2, PPID \n",
            "\t   TMEM131L, FNIP2, MND1, C4orf45, TRIM2, FHDC1, RAPGEF2, ARFIP1, TMEM154, AC022272.1 \n",
            "PC_ 2 \n",
            "Positive:  PM20D2, AL353135.2, GABRR2, SRSF12, UBE2J1, RRAGD, ANKRD6, LYRM2, MDN1, CASP8AP2 \n",
            "\t   BACH2, AL158136.1, AL132996.1, MAP3K7, AL359987.1, EPHA7, AL591519.1, MANEA-DT, MANEA, UFL1-AS1 \n",
            "\t   UFL1, GPR63, NDUFAF4, KLHL32, MMS22L, AL589740.1, FBXL4, COQ3, PNISR, AL513550.1 \n",
            "Negative:  C19orf44, CHERP, AC020917.2, SLC35E1, EPS15L1, AC008764.8, MED26, KLF2, AC008764.7, AC008764.2 \n",
            "\t   AC020911.2, AP1M1, AC024075.3, FAM32A, SMIM7, HSH2D, AC008894.2, AC024075.2, RAB8A, TPM4 \n",
            "\t   AC024075.1, UCA1, LINC01764, TMEM38A, SIN3B, PGLYRP2, F2RL3, RASAL3, CPAMD8, HAUS8 \n",
            "PC_ 3 \n",
            "Positive:  AC100778.3, RPL17, C18orf32, AC044840.1, ACAA2, AC016866.2, SNHG22, DYM, CFAP53, MBD1 \n",
            "\t   AC016866.1, CXXC1, SMAD7, SKA1, AC048380.1, ME2, CTIF, ELAC1, SMAD2, SMAD4 \n",
            "\t   MEX3C, IER3IP1, DCC, MBD2, HDHD2, POLI, AC012254.1, C18orf54, KATNAL2, RAB27B \n",
            "Negative:  APMAP, CST7, CST3, DUSP15, NAPB, TTLL9, GZF1, PDRG1, LINC01431, HCK \n",
            "\t   NXT1, TM9SF4, PLAGL2, AL096677.1, POFUT1, KIF3B, CD93, ASXL1, NOL4L, THBD \n",
            "\t   AL034550.2, C20orf203, XRN2, COMMD7, DNMT3B, MAPRE1, AL117332.1, AL035071.2, AL035071.1, EFCAB8 \n",
            "PC_ 4 \n",
            "Positive:  MAST3, IL12RB1, PIK3R2, ARRDC2, KCNN1, FCHO1, CCDC124, JAK3, AC008761.1, RPL18A \n",
            "\t   IFI30, PGLS, MAP1S, AC010618.3, NIBAN3, SLC27A1, COLGALT1, AC010319.4, MPV17L2, MVB12A \n",
            "\t   BISPR, PDE4C, BST2, PLVAP, AC008397.1, GTPBP3, IQCN, ANO8, JUND, DDA1 \n",
            "Negative:  PMAIP1, LMAN1, RNF152, PIGN, SEC11C, RELCH, TNFRSF11A, ZNF532, ZCCHC2, AC104365.1 \n",
            "\t   PHLPP1, BCL2, MALT1, KDSR, NEDD4L, AC036176.1, ATP8B1, VPS4B, AC027097.2, SERPINB2 \n",
            "\t   AC027097.1, SERPINB10, NARS, SERPINB8, FECH, LINC01924, CDH7, WDR7, SDC4, SYS1 \n",
            "PC_ 5 \n",
            "Positive:  NFIX, GADD45GIP1, LYL1, TRMT1, NACC1, STX10, IER2, CACNA1A, CCDC130, MRI1 \n",
            "\t   AC008686.1, C19orf53, ZSWIM4, AC020916.1, NANOS3, C19orf57, CC2D1A, DCAF15, RFX1, IL27RA \n",
            "\t   SAMD1, PRKACA, ASF1B, AC022098.1, ADGRL1, AC011509.2, LINC01841, ADGRE5, AC008569.1, DDX39A \n",
            "Negative:  EMC1, MRTO4, EMC1-AS1, AKR7A2, UBR4, SLC66A1, IFFO2, CAPZB, MICOS10, AL031727.2 \n",
            "\t   NBL1, TMCO4, OTUD3, UBXN10-AS1, UBXN10, MUL1, CDA, PINK1, PINK1-AS, DDOST \n",
            "\t   KIF17, BX284668.6, HP1BP3, BX284668.5, EIF4G3, ECE1, NBPF3, RAP1GAP, USP48, LDLRAD2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:13] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 7 (preB)\n",
            "INFO [2025-06-10 09:15:13] Too few cells in group Komórki blastyczne 7 (preB) for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 09:15:13] define_signif_tumor_subclusters(), tumor: Komórki blastyczne proliferujące\n",
            "INFO [2025-06-10 09:15:13] Setting auto leiden resolution for Komórki blastyczne proliferujące to 0.0822339\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  RBM26, OBI1, POU4F1, UBAC2, AL445209.1, GPR18, OBI1-AS1, EDNRB-AS1, GPR183, SLAIN1 \n",
            "\t   MYCBP2, TM9SF2, MYCBP2-AS1, LINC01232, FBXL3, CLYBL, CLN5, AL137139.2, DHRS7, KCTD12 \n",
            "\t   PCCA-DT, PCNX4, AL157911.1, AL136441.1, LRRC9, RTN1, SETD9, AC008937.2, MIER3, MAP3K1 \n",
            "Negative:  DNAJC4, VEGFB, FKBP2, AP001453.1, PPP1R14B, NUDT22, TRPT1, TMEM179B, NXF1, AP001160.1 \n",
            "\t   STX5, WDR74, PPP1R14B-AS1, AP001160.3, TMEM223, FERMT3, SNHG1, STIP1, TAF6L, SLC3A2 \n",
            "\t   PLCB3, SLC22A10, OTUB1, FLRT1, COX8A, POLR2G, MACROD1, LGALS12, AP000721.2, BAD \n",
            "PC_ 2 \n",
            "Positive:  SDC4, STK4, SYS1, TP53TG5, TOMM34, PABPC1L, DBNDD2, PIGT, YWHAB, AL031663.3 \n",
            "\t   DNTTIP1, KCNK15-AS1, UBE2C, SNX21, ACOT8, TNNC2, ZSWIM3, AL133227.1, ZSWIM1, ELMO2 \n",
            "\t   SLC35C2, NEURL2, PCIF1, ZNF334, CTSA, SLC12A5, ZNF335, ADA, NCOA5, CD40 \n",
            "Negative:  DPYD-AS1, FRRS1, AGL, AC118553.1, SLC35A3, MFSD14A, AC093019.2, SASS6, TRMT13, LRRC39 \n",
            "\t   DBT, AL445928.2, RTCA-AS1, RTCA, CDC14A, AL589990.1, EXTL2, SLC30A7, DPH5, AC093157.1 \n",
            "\t   S1PR1, OLFM3, RNPC3, AMY2B, PRMT6, VAV3-AS1, LINC02785, SLC25A24, VAV3, NTNG1 \n",
            "PC_ 3 \n",
            "Positive:  RTN4IP1, QRSL1, CRYBG1, LINC02532, ATG5, CD24, PRDM1, AL591518.1, MTRES1, AL096794.1 \n",
            "\t   PREP, BEND3, BVES, PDSS2, HACE1, SOBP, GRIK2, SCML4, AL133338.1, ASCC3 \n",
            "\t   SEC63, CCNC, AL024507.2, OSTM1, TSTD3, SNX3, USP45, AFG1L, AL513550.1, AL139106.1 \n",
            "Negative:  NCOA6, PIGU, TP53INP2, MAP1LC3A, AL109824.1, DYNLRB1, GGT7, ITCH-AS1, ACSS2, ITCH \n",
            "\t   GSS, AHCY, MYH7B, AL035458.2, TRPC4AP, ASIP, EIF2S2, EDEM2, AL031668.1, RALY \n",
            "\t   PROCR, RALY-AS1, MMP24OS, CHMP4B, ZNF341, MMP24, PXMP4, EIF6, E2F1, UQCC1 \n",
            "PC_ 4 \n",
            "Positive:  ESS2, DGCR11, DGCR2, DGCR9, TMEM191B, USP18, AC008079.1, TUBA8, PEX26, MICAL3 \n",
            "\t   LINC00528, BID, BCL2L13, ATP6V1E1, CECR2, ADA2, HDHD5-AS1, HDHD5, TMEM121B, IL17RA \n",
            "\t   ANKRD62P1-PARP4P3, AP000547.3, ZC2HC1A, PKIA, PKIA-AS1, PEX2, AC011029.1, CRISPLD1, GDAP1, AC087627.1 \n",
            "Negative:  SMC4, C3orf80, IL12A, IL12A-AS1, TRIM59, SCHIP1, IFT80, IQCJ-SCHIP1, KPNA4, MFSD1 \n",
            "\t   RARRES1, PPM1L, LXN, B3GALNT1, GFM1, NMD3, LINC02067, MLF1, LINC01322, AC106707.1 \n",
            "\t   SERPINI2, RSRC1, PTX3, VEPH1, AC092944.1, CCNL1, LEKR1, LINC00886, TIPARP, TIPARP-AS1 \n",
            "PC_ 5 \n",
            "Positive:  UEVLD, TSG101, LDHAL6A, SPTY2D1OS, LDHC, SPTY2D1, LDHA, ZDHHC13, CSRP3-AS1, GTF2H1 \n",
            "\t   E2F8, HPS5, SAAL1, NAV2, TPH1, HTATIP2, SERGEF, PRMT3, KCNC1, ANO5 \n",
            "\t   FANCF, NCR3LG1, GAS2, SVIP, NUCB2, PIK3C2A, AC006299.1, RPS13, LINC02718, PLEKHA7 \n",
            "Negative:  MFSD1, IQCJ-SCHIP1, RARRES1, SCHIP1, LXN, GFM1, IL12A-AS1, MLF1, IL12A, AC106707.1 \n",
            "\t   C3orf80, RSRC1, IFT80, PTX3, VEPH1, SMC4, AC092944.1, TRIM59, CCNL1, KPNA4 \n",
            "\t   LEKR1, LINC00886, PPM1L, TIPARP, TIPARP-AS1, SSR3, B3GALNT1, AC084036.1, NMD3, AC091607.2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:15] define_signif_tumor_subclusters(), tumor: komórki dendrytyczne mieloidalne\n",
            "INFO [2025-06-10 09:15:15] Less cells in group komórki dendrytyczne mieloidalne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:15] define_signif_tumor_subclusters(), tumor: komórki dendrytyczne plazmacytoidalne\n",
            "INFO [2025-06-10 09:15:15] Setting auto leiden resolution for komórki dendrytyczne plazmacytoidalne to 0.32736\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AF131215.4, AF131215.6, XKR6, PINX1, AC105001.1, PRSS51, AC104964.1, AC104964.3, AC023385.1, MSRA \n",
            "\t   TNKS, AC022784.1, DDX19B, PPP1R3B, ERI1, MFHAS1, PRAG1, FAM85B, FAM66B, AF233439.1 \n",
            "\t   BCAS3, AC011921.1, PPM1D, XKR5, APPBP2, AC025048.4, C17orf64, AC025048.2, USP32, AGPAT5 \n",
            "Negative:  AHNAK, AP001363.1, EEF1G, ASRGL1, INCENP, TUT1, FTH1, MTA2, BEST1, EML3 \n",
            "\t   B3GAT3, RAB3IL1, GANAB, INTS5, C11orf98, FADS3, LBHD1, CSKMT, UQCC3, FADS1 \n",
            "\t   UBXN1, BSCL2, FADS2, HNRNPUL2, FEN1, TTC9C, TMEM258, ZBTB3, PPP1R32, POLR2G \n",
            "PC_ 2 \n",
            "Positive:  ATG2B, BDKRB2, C14orf132, TUNAR, GSKIP, AL139020.1, TCL1A, AK7, TCL1B, AL163051.1 \n",
            "\t   TCL6, PAPOLA, GLRX5, SNHG10, VRK1, SYNE3, LINC00618, CLMN, LINC02325, DICER1-AS1 \n",
            "\t   LINC01550, BCL11B, SETD3, CCNK, CCDC85C, AL160313.1, CYP46A1, EML1, EVL, AL157912.1 \n",
            "Negative:  ATP9A, SALL4, NFATC2, MOCS3, ZFP64, DPM1, AL109984.1, ADNP-AS1, LINC01524, ADNP \n",
            "\t   TSHZ2, BCAS4, AL109930.1, AL157838.1, PARD6B, ZNF217, RIPOR3, PFDN4, AL133230.1, LINC01440 \n",
            "\t   PTPN1, FAM210B, SMIM25, AURKA, CEBPB, CSTF1, CEBPB-AS1, CASS4, TMEM189, RTF2 \n",
            "PC_ 3 \n",
            "Positive:  ANKS3, NUDT16L1, AC023830.1, C16orf71, MGRN1, UBALD1, ZNF500, ROGDI, GLYR1, AC020663.3 \n",
            "\t   UBN1, PPL, NAGPA, ALG1, EEF2KMT, RBFOX1, METTL22, ABAT, TMEM186, FUT7 \n",
            "\t   C9orf139, ABCA2, PMM2, NPDC1, CLIC3, PAXX, SAPCD2, LCNL1, AC022167.2, UAP1L1 \n",
            "Negative:  SNX16, LINC02235, AC060765.2, LINC02839, AC105031.2, LRRCC1, E2F5, RBIS, CA13, CA1 \n",
            "\t   CA3, CA3-AS1, CA2, AC100801.1, WWP1, RMDN1, THOC7, C3orf49, LINC00698, AC104162.2 \n",
            "\t   CADPS, ATXN7, PSMD6-AS2, C3orf14, CPNE3, PSMD6, PSMD6-AS1, PTPRG-AS1, MAGI1, PTPRG \n",
            "PC_ 4 \n",
            "Positive:  IL10RB, IFNAR1, IL10RB-DT, IFNGR2, IFNAR2, AP000282.1, C21orf62, C21orf62-AS1, PAXBP1, CFAP298 \n",
            "\t   PAXBP1-AS1, SYNJ1, EVA1C, URB1-AS1, URB1, MRAP, MIS18A, LINC00159, SCAF4, AP000254.2 \n",
            "\t   SOD1, AP000251.1, TIAM1, GRIK1, CYYR1, BACH1-AS1, CYYR1-AS1, ADAMTS1, AP001439.1, AF165147.1 \n",
            "Negative:  ENTPD3-AS1, ENTPD3, EIF1B, ZNF619, RPL14, ZNF620, ZNF621, EIF1B-AS1, AC099560.1, MYRIP \n",
            "\t   CTNNB1, ULK4, TRAK1, AC099332.2, VIPR1, RPSA, SEC22C, SLC25A38, AC104850.2, SS18L2 \n",
            "\t   CX3CR1, NKTR, AC092053.3, AC006059.1, CSRNP1, ZBTB47, TTC21A, CCDC13, HIGD1A, GORASP1 \n",
            "PC_ 5 \n",
            "Positive:  C1orf112, SELP, F5, Z99572.1, SLC19A2, CCDC181, BLZF1, Z99758.1, NME7, ATP1B1 \n",
            "\t   LINC00970, XCL1, XCL2, TBX19, SFT2D2, TIPRL, GPR161, DCAF6, MPC2, ADCY10 \n",
            "\t   MPZL1, RCSD1, AL031733.2, CREG1, CD247, AL136984.1, POU2F1, MAEL, ILDR2, TADA1 \n",
            "Negative:  TERF2IP, AC025287.4, KARS, AC106729.1, MON1B, ADAT1, SYCE1L, AC009139.2, GABARAPL2, NUDT7 \n",
            "\t   VAT1L, WWOX, TMEM231, AC079414.3, WWOX-AS1, AC046158.1, AC027279.1, TMEM170A, MAF, LINC01229 \n",
            "\t   CFDP1, MAFTRR, AC105411.1, ZFP1, DYNLRB2, CDYL2, AC099508.2, CMC2, ZNRF1, CENPN \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:16] define_signif_tumor_subclusters(), tumor: komórki NK T\n",
            "INFO [2025-06-10 09:15:16] Less cells in group komórki NK T than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 09:15:16] define_signif_tumor_subclusters(), tumor: monocyty\n",
            "INFO [2025-06-10 09:15:16] Setting auto leiden resolution for monocyty to 0.514215\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  CCDC124, RPL18A, KCNN1, JAK3, ARRDC2, FCHO1, IL12RB1, MAST3, AC008761.1, PIK3R2 \n",
            "\t   MAP1S, IFI30, MPV17L2, COLGALT1, PDE4C, AC008397.1, NIBAN3, IQCN, JUND, PGLS \n",
            "\t   LSM4, AC010618.3, PGPEP1, SLC27A1, LRRC25, AC010319.4, SSBP4, MVB12A, ISYNA1, BISPR \n",
            "Negative:  ETFBKMT, AMN1, DENND5B-AS1, DENND5B, AC023157.3, SINHCAF, LINC02422, DDX11, RESF1, DDX11-AS1 \n",
            "\t   AC016957.2, TSPAN11, AC010198.2, BICD1, CAPRIN2, AC026356.2, IPO8, FGD4, DNM1L, TMTC1 \n",
            "\t   YARS2, OVCH1, PKP2, AC046130.1, OVCH1-AS1, ALG10, ERGIC2, AC140847.2, AC009318.3, ALG10B \n",
            "PC_ 2 \n",
            "Positive:  BICD2, ZNF484, FGD3, SUSD3, CARD19, NINJ1, FAM120AOS, PHF2, PTPDC1, FAM120A \n",
            "\t   AL158152.1, LINC02603, ZNF169, AL691447.2, FBP1, MFSD14B, AOPEP, AL353768.2, FANCC, PTCH1 \n",
            "\t   LINC00476, ERCC6L2, SLC35D2, ZNF367, AL133477.2, HABP4, CDC14B, PRXL2C, AL589843.1, ZNF510 \n",
            "Negative:  SLC12A7, TERT, TRIP13, CLPTM1L, AC026748.3, BRD9, ZDHHC11, CEP72, LPCAT1, SLC9A3-AS1 \n",
            "\t   EXOC3, EXOC3-AS1, MRPL36, AHRR, PDCD6, SDHA, NDUFS6, AC124852.1, IRX1, AC091978.1 \n",
            "\t   RPRD1B, ACTR5, PPP1R16B, TTI1, RALGAPB, TGM2, BLCAP, SNHG11, SRC, CTNNBL1 \n",
            "PC_ 3 \n",
            "Positive:  ZPBP, AC073349.4, ZNF273, ABCA13, AC073349.1, AC073210.3, UPP1, ZNF117, HUS1, ERV3-1 \n",
            "\t   TNS3, ZNF92, AC004870.3, VKORC1L1, AC004870.4, GUSB, TBRG4, ASL, CCM2, CRCP \n",
            "\t   SNHG15, AC068533.3, TPST1, MYO1G, AC004847.1, LINC00174, AC008267.5, AC004854.2, PURB, KCTD7 \n",
            "Negative:  TMEM189, CEBPB-AS1, CEBPB, SMIM25, PTPN1, AL133230.1, RIPOR3, PARD6B, BCAS4, ADNP \n",
            "\t   ADNP-AS1, DPM1, MOCS3, NFATC2, ATP9A, SALL4, ZFP64, AL109984.1, LINC01524, EYA2 \n",
            "\t   TSHZ2, SLC2A10, TP53RK, AL109930.1, SLC13A3, ZNF334, AL157838.1, ELMO2, AL133227.1, ZNF217 \n",
            "PC_ 4 \n",
            "Positive:  ELMO1, AOAH, ELMO1-AS1, GPR141, EPDR1, AC006960.2, NME8, ANLN, SFRP4, KIAA0895 \n",
            "\t   STARD3NL, EEPD1, TRGC2, SEPTIN7, SEPTIN7-AS1, AC018647.2, TRGC1, HERPUD2, AC005400.1, TRG-AS1 \n",
            "\t   DPY19L1, AMPH, BBS9, VPS41, RP9, NT5C3A, POU6F2, FKBP9, KBTBD2, YAE1 \n",
            "Negative:  WNK1, AC021054.1, NINJ2, CCDC77, KDM5A, PTPRA, MRPS26, VPS16, OXT, AC007406.5 \n",
            "\t   AVP, PCED1A, UBOX5-AS1, AC026369.3, CPXM1, UBOX5, PLCB4, LAMP5, EBF4, PAK5 \n",
            "\t   FASTKD5, IQSEC3, ANKEF1, IDH3B, LZTS3, SNAP25-AS1, SMIM15, NDUFAF2, ERCC8, SMIM15-AS1 \n",
            "PC_ 5 \n",
            "Positive:  UQCRC2, NPIPB4, OTOA, AC092119.2, IGSF6, MOSMO, SDR42E2, POLR3E, CDR2, AC092338.3 \n",
            "\t   EEF2K, AC005632.3, AC092338.2, NPIPB5, USP31, COG7, AC008915.2, METTL9, GGA2, EARS2 \n",
            "\t   AC005632.6, UBFD1, AC008870.2, NPIPB3, NDUFAB1, CRYM, PALB2, AC008870.4, TMEM159, DCTN5 \n",
            "Negative:  ANKRD61, USP42, CYTH3, EIF2AK1, AIMP2, RAC1, PMS2, DAGLB, KDELR2, CCZ1 \n",
            "\t   GRID2IP, OCM, ZDHHC4, RNF216, C7orf26, ZNF853, FSCN1, ZNF316, AC073343.2, ACTB \n",
            "\t   ZNF12, FBXL18, AC092171.5, AC092171.3, TNRC18, SLC29A4, WIPI2, RBAK, RADIL, AP5Z1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:17] define_signif_tumor_subclusters(), tumor: MPP\n",
            "INFO [2025-06-10 09:15:17] Setting auto leiden resolution for MPP to 0.0400268\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  ETFBKMT, AMN1, DENND5B-AS1, AC023157.3, DENND5B, LINC02422, SINHCAF, RESF1, DDX11, DDX11-AS1 \n",
            "\t   AC016957.2, TSPAN11, BICD1, AC010198.2, CAPRIN2, AC026356.2, IPO8, FGD4, TMTC1, DNM1L \n",
            "\t   OVCH1, YARS2, OVCH1-AS1, PKP2, ERGIC2, AC046130.1, AC009318.3, ALG10, AC009318.1, AC140847.2 \n",
            "Negative:  PPP1R14B, SPPL2B, PPP1R14B-AS1, FKBP2, PLCB3, VEGFB, BAD, DNAJC4, TMPRSS9, GPR137 \n",
            "\t   AP001453.1, TIMM13, ESRRA, NUDT22, TRMT112, LMNB2, PRDX5, TRPT1, GADD45B, FERMT3 \n",
            "\t   CCDC88B, STIP1, GNG7, RPS6KA4, TAF6L, TMEM223, FLRT1, AP001160.3, TMEM179B, POLR2G \n",
            "PC_ 2 \n",
            "Positive:  PIM3, CRELD2, MLC1, ALG12, PANX2, ZBED4, TRABD, Z98885.3, AL022328.3, AL022328.4 \n",
            "\t   BRD1, SELENOO, Z97192.3, Z97192.2, TUBGCP6, Z97192.1, HDAC10, C22orf34, MAPK12, MAPK11 \n",
            "\t   PLXNB2, DENND6B, PPP6R2, SBF1, LMF2, NCAPH2, SCO2, TYMP, ODF3B, U62317.4 \n",
            "Negative:  ABHD12B, TRIM9, NID2, RTRAF, PYGL, PTGDR, TMX1, AL606834.2, FRMD6-AS1, PTGER2 \n",
            "\t   GNG2, FRMD6, NIN, TXNDC16, AL606834.1, GPR137C, SAV1, ERO1A, ATL1, AL133453.1 \n",
            "\t   AL118556.1, PSMC6, MAP4K5, STYX, CDKL1, GNPNAT1, DMAC2L, AL139317.5, L2HGDH, FERMT2 \n",
            "PC_ 3 \n",
            "Positive:  GNG2, FRMD6-AS1, FRMD6, RTRAF, NID2, PTGDR, TMX1, PTGER2, TXNDC16, TRIM9 \n",
            "\t   GPR137C, ERO1A, AL133453.1, PSMC6, ABHD12B, STYX, GNPNAT1, AL139317.5, PYGL, FERMT2 \n",
            "\t   AL606834.2, DDHD1, AL356020.1, NIN, AL365295.1, AL606834.1, AL365295.2, SAV1, CDKN3, ATL1 \n",
            "Negative:  AC023590.1, SAMD12-AS1, CCN3, ENPP2, TAF2, SELENON, AP005717.2, MTFR1L, STMN1, PAFAH2 \n",
            "\t   PAQR7, AL033528.2, PDIK1L, AUNIP, FAM110D, ZNF593, CNKSR1, CEP85, SH3BGRL3, UBXN11 \n",
            "\t   CD52, CRYBG2, DHDDS, DSCC1, RCAN3AS, AL513365.2, NIPAL3, HMGN2, RPS6KA1, STPG1 \n",
            "PC_ 4 \n",
            "Positive:  ECHS1, FUOM, AL360181.1, PAOX, ZNF511, TUBGCP2, ADAM8, MTG1, VENTX, KNDC1 \n",
            "\t   INPP5A, AL451069.2, PWWP2B, LRRC27, SCART1, STK32C, JAKMIP3, AL162274.2, CYP2E1, BNIP3 \n",
            "\t   PPP2R2D, AL731769.2, LINC02646, GLRX3, C10orf143, EBF3, LINC02666, AL355531.1, MGMT, LINC01163 \n",
            "Negative:  TTC5, CCNB1IP1, AL355075.4, PARP2, AL355075.1, TM7SF3, AC024896.1, MED21, TEP1, FGFR1OP2 \n",
            "\t   AC092747.4, INTS13, AC092747.1, AC023051.1, KLHL33, STK38L, ITPR2, OSGEP, ARNTL2, ARNTL2-AS1 \n",
            "\t   AC024145.1, SMCO2, APEX1, PPFIBP1, AC022509.2, AC009509.1, PIP4P1, MRPS35, SSPN, PNP \n",
            "PC_ 5 \n",
            "Positive:  TUBE1, FAM229B, FYN, Z97989.1, AL360015.1, TRAF3IP2, LINC02541, AL008730.1, TRAF3IP2-AS1, MARCKS \n",
            "\t   AL080317.2, REV3L, AL080317.1, MFSD4B, HDAC2, AL360227.1, SLC16A10, HDAC2-AS2, RPF2, GTF3C6 \n",
            "\t   AMD1, NT5DC1, CDK19, AC002464.1, COL10A1, SLC22A16, DDO, TSPYL4, METTL24, CDC40 \n",
            "Negative:  ARNTL2-AS1, SMCO2, AC009509.1, PPFIBP1, MRPS35, ARNTL2, KLHL42, STK38L, CCDC91, AC022364.1 \n",
            "\t   AC092747.1, AC012150.1, FAR2, AC092747.4, AC009318.1, AC009318.3, MED21, ERGIC2, OVCH1-AS1, AC024896.1 \n",
            "\t   OVCH1, TMTC1, TM7SF3, IPO8, FGFR1OP2, CAPRIN2, INTS13, AC010198.2, AC023051.1, TSPAN11 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:20] define_signif_tumor_subclusters(), tumor: późne komórki erytroidalne\n",
            "INFO [2025-06-10 09:15:20] Setting auto leiden resolution for późne komórki erytroidalne to 0.266148\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AC073349.4, ZNF273, AC073349.1, AC073210.3, ZNF117, ERV3-1, ZNF92, VKORC1L1, GUSB, ASL \n",
            "\t   CRCP, AC068533.3, TPST1, LINC00174, AC008267.5, KCTD7, AC006001.2, RABGEF1, AC027644.3, LINC02604 \n",
            "\t   TMEM248, SBDS, TYW1, SPDYE21P, AC006480.2, AC092100.1, AUTS2, GALNT17, CALN1, TYW1B \n",
            "Negative:  SYT2, PPP1R12B, KDM5B, UBE2T, LGR6, PCAT6, PTPN7, ARL8A, RABIF, ELF3-AS1 \n",
            "\t   RNPEP, KLHL12, TIMM17A, IPO9, ADIPOR1, IPO9-AS1, NAV1, AC096677.1, CSRP1, CYB5R1 \n",
            "\t   TMEM9, KIF21B, CAMSAP2, TMEM183A, DDX59-AS1, PPFIA4, DDX59, KIF14, BTG2, ATP2B4 \n",
            "PC_ 2 \n",
            "Positive:  HACE1, BVES, PREP, GRIK2, AL096794.1, PRDM1, AL591518.1, ATG5, AL133338.1, CRYBG1 \n",
            "\t   RTN4IP1, QRSL1, LINC02532, ASCC3, CCNC, CD24, MTRES1, BEND3, TSTD3, PDSS2 \n",
            "\t   USP45, AL513550.1, SOBP, PNISR, COQ3, SCML4, FBXL4, SEC63, AL589740.1, AL024507.2 \n",
            "Negative:  ARRDC2, IL12RB1, MAST3, PIK3R2, IFI30, MPV17L2, PDE4C, AC008397.1, IQCN, JUND \n",
            "\t   LSM4, PGPEP1, LRRC25, IGHD, IGHG3, SSBP4, IGHG1, ISYNA1, IGHA1, ELL \n",
            "\t   IGHGP, FKBP8, IGHG2, AC005387.2, IGHG4, IGHA2, KXD1, TMEM121, AC005253.1, TEDC1 \n",
            "PC_ 3 \n",
            "Positive:  MNT, SGSM2, METTL16, TSR1, SRR, PAFAH1B1, AC090617.4, SMG6, CLUH, HIC1 \n",
            "\t   AC005696.4, FANCC, PTCH1, AL353768.2, AOPEP, DPH1, LINC00476, ERCC6L2, RAP1GAP2, FBP1 \n",
            "\t   SLC35D2, AL133477.2, ZNF367, HABP4, RPA1, MFSD14B, CDC14B, OR3A2, PRXL2C, AL691447.2 \n",
            "Negative:  AC002460.2, AC108156.1, LINC02355, RPS3A, AC069272.1, SH3D19, AC110813.1, DCLK2, AC095055.1, LRBA \n",
            "\t   NR3C2, FAM160A1, ARHGAP10, GATB, PRMT9, TMEM184C, AC092611.1, SLC10A7, LSM6, ZNF827 \n",
            "\t   AC097375.1, AC097375.5, LINC02273, FBXW7, AC023424.3, PAFAH1B2, SIDT2, TAGLN, SIK3, APOA1-AS \n",
            "PC_ 4 \n",
            "Positive:  RMDN1, WWP1, CPNE3, CNGB3, AC100801.1, CNBD1, MMP16, AC090578.1, RIPK2, AF117829.1 \n",
            "\t   OSGIN2, NBN, DECR1, AC004083.1, LINC00534, TMEM64, NECAB1, AC103770.1, C8orf88, PIP4P2 \n",
            "\t   OTUD6B-AS1, OTUD6B, LRRC69, AC104211.1, AC117834.1, TRIQK, LINC00535, FAM92A, AC120053.1, RBM12B \n",
            "Negative:  NEURL2, CTSA, PLTP, AL162458.1, PCIF1, ZNF335, SLC12A5, NCOA5, CD40, SLC35C2 \n",
            "\t   AL133227.1, ELMO2, ZNF334, SLC13A3, TP53RK, SLC2A10, EYA2, TMEM189, CEBPB-AS1, PARD6B \n",
            "\t   CEBPB, BCAS4, RIPOR3, SMIM25, OSER1, ADNP, AL133230.1, PTPN1, ADNP-AS1, DPM1 \n",
            "PC_ 5 \n",
            "Positive:  TRAF3, RCOR1, AMN, ANKRD9, CDC42BPB, TECPR2, AL117209.1, CINP, TNFAIP2, ZNF839 \n",
            "\t   EIF5, MARK3, MOK, CKB, WDR20, HSP90AA1, TRMT61A, DYNC1H1, AL118558.3, AL139300.2 \n",
            "\t   AL137779.1, BAG5, AL137779.2, KLC1, PPP2R5C, LINC02320, COA8, AL132709.8, AL049840.2, AL049840.6 \n",
            "Negative:  ACO2, PHF5A, TOB2, TEF, ZC3H7B, RANGAP1, CHADL, AL035681.1, L3MBTL2, TBC1D22B \n",
            "\t   EP300-AS1, AL096712.2, RNF8, CMTR1, CCDC167, AL035658.1, MDGA1, AL121574.1, EP300, ZFAND3 \n",
            "\t   RBX1, BTBD9, DNAJB7, GLO1, XPNPEP3, DNAH8, ST13, SAYSD1, SLC25A17, MRTFA-AS1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:21] define_signif_tumor_subclusters(), tumor: prekursory erytroidalne i megakariocytarne\n",
            "INFO [2025-06-10 09:15:21] Setting auto leiden resolution for prekursory erytroidalne i megakariocytarne to 0.0605936\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  ESRRA, GPR137, TRMT112, BAD, PRDX5, PLCB3, CCDC88B, PPP1R14B-AS1, RPS6KA4, PPP1R14B \n",
            "\t   AP005273.2, FKBP2, NRXN2, VEGFB, RASGRP2, PYGM, DNAJC4, SF1, AP001453.1, AP001462.1 \n",
            "\t   MAP4K2, NUDT22, MEN1, CDC42BPG, TRPT1, EHD1, ATG2A, FERMT3, PPP2R5B, ARL2 \n",
            "Negative:  AC097634.1, EIF4E3, AC097634.3, GPR27, FOXP1-IT1, PROK2, LINC00877, AC097634.2, LINC00870, FOXP1-AS1 \n",
            "\t   RYBP, SHQ1, FOXP1, GXYLT2, PPP4R2, MITF, EBLN2, PDZRN3, FRMD4B, LINC02018 \n",
            "\t   LMOD3, LINC00960, ARL6IP5, ZNF717, UBA3, ZNF485, ROBO2, ZNF32, TMF1, NETO1 \n",
            "PC_ 2 \n",
            "Positive:  AIMP1, TBCK, AC092445.1, AC109361.1, PAPSS1, GSTCD, SGMS2, INTS12, AC096564.1, ARHGEF38 \n",
            "\t   CYP2U1, PPA2, HADH, LEF1, TET2-AS1, LEF1-AS1, TET2, RPL34-AS1, AC004069.1, AC096577.1 \n",
            "\t   RPL34, AC004053.1, CXXC4-AS1, OSTC, LINC02428, CENPE, BDH2, SLC9B2, SLC9B1, CISD2 \n",
            "Negative:  SLC43A1, RTN4RL2, SLC43A3, TIMM10, SMTNL1, PRG2, UBE2L6, SERPING1, AP000662.1, SSRP1 \n",
            "\t   YPEL4, CLP1, ZDHHC5, TNKS1BP1, MED19, TMX2, PTPRJ, SELENOH, BTBD18, CTNND1 \n",
            "\t   LPXN, ZFP91, GLYATL2, GLYATL1, NUP160, AP001636.3, FAM111B, FAM111A-DT, FAM111A, DTX4 \n",
            "PC_ 3 \n",
            "Positive:  CDC45, UFD1, CLDN5, AC000068.1, MRPL40, SEPTIN5, C22orf39, GP1BB, HIRA, AC000085.1 \n",
            "\t   CLTCL1, GNB1L, SLC25A1, LINC01311, ESS2, RTL10, DGCR11, DGCR2, TXNRD2, DGCR9 \n",
            "\t   TMEM191B, USP18, COMT, AC008079.1, ARVCF, TUBA8, PEX26, MICAL3, LINC00528, BID \n",
            "Negative:  TMC5, AC130456.3, SYT17, GDE1, ITPRIPL2, COQ7, CCP110, AC092287.1, VPS35L, SMG1 \n",
            "\t   AC138811.1, KNOP1, ARL6IP1, IQCK, ACSM3, RPS15A, ACSM1, THUMPD1, NOMO2, AC004381.1 \n",
            "\t   AC009152.3, ERI2, REXO5, AC009152.1, DCUN1D3, AC009152.6, LYRM1, DNAH3, AC109446.3, AC008551.1 \n",
            "PC_ 4 \n",
            "Positive:  ACSM3, ACSM1, THUMPD1, IQCK, AC004381.1, KNOP1, ERI2, VPS35L, CCP110, REXO5 \n",
            "\t   GDE1, DCUN1D3, AC130456.3, TMC5, LYRM1, SYT17, ITPRIPL2, COQ7, DNAH3, AC092287.1 \n",
            "\t   SMG1, AC008551.1, AC138811.1, TMEM159, ARL6IP1, RPS15A, CRYM, NOMO2, NPIPB3, AC009152.3 \n",
            "Negative:  RSBN1, PHTF1, AL137856.1, MAGI3, PTPN22, LRIG2, AP4B1-AS1, LRIG2-DT, AP4B1, DCLRE1B \n",
            "\t   AL390242.1, HIPK1-AS1, SLC16A1-AS1, HIPK1, SLC16A1, PPM1J, TRIM33, RHOC, BCAS2, DENND2C \n",
            "\t   NRAS, CSDE1, SIKE1, TSPAN2, LINC01765, VANGL1, SLC22A15, MINDY3, ITGA8, PTER \n",
            "PC_ 5 \n",
            "Positive:  CDC40, WASF1, METTL24, DDO, SLC22A16, FIG4, AC002464.1, AK9, CDK19, ZBTB24 \n",
            "\t   AMD1, MICAL1, GTF3C6, SMPD2, RPF2, PPIL6, SLC16A10, AL359711.2, AL360227.1, CD164 \n",
            "\t   MFSD4B, CEP57L1, AL080317.1, SESN1, REV3L, ARMC2, AL080317.2, TRAF3IP2-AS1, FOXO3, AL008730.1 \n",
            "Negative:  IGHV3-49, LINC00221, IGHV1-24, IGHV3-23, DNAJC3-DT, CLDN10, ABCC4, DNAJC3, GPR180, IGHV3-11 \n",
            "\t   TGDS, AL138955.1, GPC6, IGHV3-7, UGGT2, GPC5-AS1, IGHV1-3, MBNL2, AL356737.2, AL161430.1 \n",
            "\t   IGHV1-2, GPC5-AS2, AL442067.3, GPC5, IGHV6-1, RAP2A, MIR17HG, FAM30A, IPO5, SLITRK5 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:23] define_signif_tumor_subclusters(), tumor: wczesne komórki erytroidalne\n",
            "INFO [2025-06-10 09:15:23] Setting auto leiden resolution for wczesne komórki erytroidalne to 0.11046\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  PPP1R14B, PPP1R14B-AS1, PLCB3, FKBP2, BAD, VEGFB, GPR137, DNAJC4, ESRRA, AP001453.1 \n",
            "\t   TRMT112, NUDT22, PRDX5, CCDC88B, TRPT1, RPS6KA4, FERMT3, AP005273.2, NRXN2, STIP1 \n",
            "\t   RASGRP2, FLRT1, PYGM, SF1, AP001462.1, AP000721.2, MAP4K2, MEN1, MACROD1, OTUB1 \n",
            "Negative:  ASCC3, CCNC, TSTD3, AL133338.1, USP45, GRIK2, AL513550.1, HACE1, BVES, PNISR \n",
            "\t   PREP, AL096794.1, COQ3, AL591518.1, FBXL4, PRDM1, ATG5, AL589740.1, CRYBG1, MMS22L \n",
            "\t   RTN4IP1, QRSL1, KLHL32, LINC02532, CD24, MTRES1, BEND3, NDUFAF4, PDSS2, SEC63 \n",
            "PC_ 2 \n",
            "Positive:  AC215522.2, AC093627.5, AC093627.4, PRKAR1B, FAM20C, AC147651.4, PDGFA, DNAAF5, SUN1, GET4 \n",
            "\t   AC073957.3, ADAP1, COX19, C7orf50, AC073957.1, GPR146, GPER1, ZFAND2A, AC091729.3, MICALL2 \n",
            "\t   INTS1, MAFK, TMEM184A, PSMG3, PSMG3-AS1, ELFN1-AS1, AOPEP, FBP1, AL353768.2, MFSD14B \n",
            "Negative:  STRAP, EPS8, PTPRO, DERA, MGST1, PLEKHA5, ARHGDIB, AC092828.1, AEBP2, PDE3A \n",
            "\t   ERP27, PYROXD1, RECQL, ART4, GOLT1B, LDHB, C12orf60, ABCC9, WBP11, CMAS \n",
            "\t   H2AFJ, ST8SIA1, HIST4H4, AC053513.1, AC010168.2, C2CD5, ETNK1, AC010168.1, AC084816.1, PLBD1-AS1 \n",
            "PC_ 3 \n",
            "Positive:  HSD11B1-AS1, G0S2, LAMB3, HSD11B1, PLXNA2, LINC02767, TRAF3IP3, CD34, UTP25, MIR29B2CHG \n",
            "\t   HHAT, CD46, KCNH1, LINC00467, RCOR3, SLC30A1, CR1L, TRAF5, NEK2, AL137789.1 \n",
            "\t   LPGAT1, CR1, CR2, INTS7, PPP2R5A, AL606468.1, DTL, CD55, TATDN3, FLVCR1-DT \n",
            "Negative:  AC009831.3, RNF125, RNF138, AC022960.1, AC011825.2, TRAPPC8, AC017100.1, B4GALT6, GAREM1, DSG2 \n",
            "\t   CDH2, PCAT18, AC025887.2, AQP4-AS1, KCTD1, TAF4B, CABLES1, RBBP8, TMEM241, AC090912.1 \n",
            "\t   KLHL14, RIOK3, SS18, RMC1, NPC1, LAMA3, TTC39C, AC090772.2, MIR133A1HG, AC105114.2 \n",
            "PC_ 4 \n",
            "Positive:  CCDC92, DNAH10, ZNF664, AC117503.5, RFLNA, ATP6V0A2, NCOR2, TCTN2, AC073592.1, GTF2H3 \n",
            "\t   SCARB1, EIF2B1, DDX55, AC073593.2, TMED2, TMED2-DT, UBC, RILPL1, DHX37, SNRNP35 \n",
            "\t   BRI3BP, RILPL2, THRIL, KMT5A, AACS, SBNO1-AS1, LINC00944, SBNO1, SLC15A4, CDK2AP1 \n",
            "Negative:  TPP1, AC091564.5, DCHS1, TAF10, MRPL17, ILK, RRP8, AC087280.2, DNHD1, TIMM10B \n",
            "\t   ZNF215, ARFIP2, TRIM3, SYT9, APBB1, SMPD1, CNGA4, AC107884.1, FAM160A2, OLFML1 \n",
            "\t   TRIM22, TRIM5, TRIM34, TRIM6, AC104389.5, HBG2, HBG1, HBD, HBB, MMP26 \n",
            "PC_ 5 \n",
            "Positive:  RTN4IP1, QRSL1, CRYBG1, ATG5, LINC02532, PRDM1, CD24, AL591518.1, AL096794.1, MTRES1 \n",
            "\t   PREP, BEND3, BVES, PDSS2, HACE1, SOBP, GRIK2, AL133338.1, SCML4, ASCC3 \n",
            "\t   CCNC, TSTD3, SEC63, USP45, AL024507.2, AL513550.1, PNISR, OSTM1, COQ3, FBXL4 \n",
            "Negative:  TP53BP1, MAP1A, PPIP5K1, TUBGCP4, ZSCAN29, CATSPER2, ADAL, AC011330.2, LCMT2, PDIA3 \n",
            "\t   EPB42, CCNDBP1, TMEM62, AC068724.3, ELL3, UBR1, AC090510.1, TTBK2, SERF2, CDAN1 \n",
            "\t   STARD9, HAUS2, HYPK, LRRC57, AC018362.1, MFAP1, SNAP23, WDR76, ZNF106, CAPN3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:15:25] define_signif_tumor_subclusters(), tumor: control\n",
            "INFO [2025-06-10 09:15:25] Setting auto leiden resolution for control to 0.0057241\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  C21orf91-OT1, C21orf91, AL109761.1, CHODL, AF130417.1, TMPRSS15, MIR548XHG, LINC02573, NCAM2, LINC01684 \n",
            "\t   AP001341.1, LINC00158, AADACL2-AS1, IGSF10, SUCNR1, AC108718.1, MBNL1, SLC33A1, MBNL1-AS1, AC026347.1 \n",
            "\t   P2RY1, C3orf33, RAP2B, PLCH1, ARHGEF26-AS1, MME-AS1, ARHGEF26, MME, DHX36, ARPC3 \n",
            "Negative:  TLCD3B, AC093512.2, PAGR1, MVP, CDIPT, ALDOA, DOC2A, SEZ6L2, AC009133.1, PPP4C \n",
            "\t   INO80E, TBX6, ASPHD1, YPEL3, HIRIP3, KCTD13, MAZ, C16orf86, GDPD3, TMEM219 \n",
            "\t   TAOK2, KIF22, MAPK3, GFOD2, CORO1A, AC009133.3, AC012645.3, BOLA2B, SULT1A3, RANBP10 \n",
            "PC_ 2 \n",
            "Positive:  BROX, FAM177B, AIDA, AL392172.2, AL592148.3, MIA3, AL392172.1, TAF1A, DISP1, AL356108.1 \n",
            "\t   DUSP10, TLR5, HLX, SUSD4, CAPN2, HLX-AS1, TP53BP2, MARC1, FBXO28, DEGS1 \n",
            "\t   RAB3GAP2, NVL, IARS2, CNIH4, BPNT1, EPRS, WDR26, LYPLAL1, CNIH3, LYPLAL1-DT \n",
            "Negative:  SNN, LITAF, TXNDC11, AC099489.1, AC007613.1, SOCS1, ZC3H7A, RSL1D1, RMI2, AC007216.4 \n",
            "\t   STYK1, GSPT1, AC007220.1, AC007216.3, AC007216.2, AC007220.2, NPIPB2, AC007014.1, SNX29, AC010333.2 \n",
            "\t   YBX3, CLEC16A, CPPED1, DEXI, CIITA, AC109597.1, PRH1, AC133065.1, ERCC4, TVP23A \n",
            "PC_ 3 \n",
            "Positive:  PPP1R16A, KIFC2, CYHR1, TONSL, VPS28, MRPS5, MAL, ZNF514, ZNF2, IGKV1D-13 \n",
            "\t   AC092835.1, IGKV3-20, FAHD2A, IGKV3-11, LINC00342, IGKV1-6, IGKV1-5, ANKRD36C, KRCC1, IGKC \n",
            "\t   RPIA, RGPD2, EIF2AK3-DT, AC104134.1, PLGLB2, GPAT2, EIF2AK3, CYTOR, ASTL, DUSP2 \n",
            "Negative:  AC104825.1, ACOX3, SH3TC1, TRMT44, AC105345.1, ABLIM2, AC108519.1, AC073648.7, AFAP1, AC105916.1 \n",
            "\t   SORCS2, SLC2A9, GRPEL1, AC005674.2, TADA2B, WDR1, AC097382.2, ZNF518B, AC097382.3, AC005599.1 \n",
            "\t   TBC1D14, CLNK, KIAA0232, HS3ST1, BLOC1S4, AC006230.1, MRFAP1L1, AC025539.1, S100P, AC005699.1 \n",
            "PC_ 4 \n",
            "Positive:  PSMD11, ZNF207, C17orf75, CDK5R1, RHBDL3, MYO1D, RHOT1, AC116407.2, AC025211.1, LRRC37B \n",
            "\t   TMEM98, SUZ12, CCT6B, AC004253.1, ZNF830, LIG3, UTP6, RFFL, COPRS, RAD51D \n",
            "\t   RAB11FIP4, NLE1, EVI2A, AC022916.1, EVI2B, OMG, SLFN5, AC079915.1, AC022706.1, NF1 \n",
            "Negative:  VPREB1, IGLV6-57, IGLV4-60, AC245060.6, IGLV10-54, AC245060.5, TOP3B, IGLV1-47, IGLV1-40, AC245452.1 \n",
            "\t   ZNF280B, IGLV3-21, PPM1F, IGLV3-19, IGLV2-14, MAPK1, IGLV3-1, IGLC1, YPEL1, IGLC2 \n",
            "\t   IGLC3, RSPH14, PPIL2, BCR, AP000553.2, IGLL1, SDF2L1, AP000345.2, YDJC, DRICH1 \n",
            "PC_ 5 \n",
            "Positive:  PRR16, AC113418.1, AC008574.1, AC008568.1, HSD17B4, ASCC3, SRFBP1, TNFAIP8, DMXL1, LOX \n",
            "\t   AL133338.1, AC008629.1, DTWD2, SNX2, AC093206.1, GRIK2, LINC02208, HACE1, LINC02147, SEMA6A-AS1 \n",
            "\t   BVES, SEMA6A, PREP, AC018752.1, AL096794.1, COMMD10, AL591518.1, ANKRD49, PRDM1, AC034236.2 \n",
            "Negative:  MNDA, PYHIN1, IFI16, AIM2, SPTA1, ACKR1, FCER1A, CD1C, DUSP23, FCRL6 \n",
            "\t   CD1D, CFAP45, KIRREL1, TAGLN2, LINC01133, FCRL1, PIGM, FCRL2, IGSF8, FCRL3 \n",
            "\t   AL121987.2, FCRL5, PEA15, LINC02772, DCAF8, ETV3, AL139011.1, ARHGEF11, PEX19, PEAR1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 09:17:42] ::plot_cnv:Start\n",
            "INFO [2025-06-10 09:17:42] ::plot_cnv:Current data dimensions (r,c)=16764,7783 Total=130676727.26151 Min=0.766442817427447 Max=8.73681684296526.\n",
            "INFO [2025-06-10 09:17:43] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 09:17:46] plot_cnv(): auto thresholding at: (0.880550 , 1.122554)\n",
            "INFO [2025-06-10 09:18:01] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 09:18:01] Observation data size: Cells= 2877 Genes= 16764\n",
            "INFO [2025-06-10 09:18:01] clustering observations via method: ward.D\n",
            "INFO [2025-06-10 09:18:01] Number of cells in group(1) is 11\n",
            "INFO [2025-06-10 09:18:01] group size being clustered:  11,16764\n",
            "INFO [2025-06-10 09:18:01] Number of cells in group(2) is 2\n",
            "INFO [2025-06-10 09:18:01] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:01] Number of cells in group(3) is 19\n",
            "INFO [2025-06-10 09:18:01] group size being clustered:  19,16764\n",
            "INFO [2025-06-10 09:18:01] Number of cells in group(4) is 42\n",
            "INFO [2025-06-10 09:18:01] group size being clustered:  42,16764\n",
            "INFO [2025-06-10 09:18:01] Number of cells in group(5) is 27\n",
            "INFO [2025-06-10 09:18:01] group size being clustered:  27,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(6) is 26\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  26,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(7) is 1\n",
            "INFO [2025-06-10 09:18:02] Skipping group: 7, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(8) is 1\n",
            "INFO [2025-06-10 09:18:02] Skipping group: 8, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(9) is 51\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  51,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(10) is 47\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  47,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(11) is 45\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  45,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(12) is 43\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  43,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(13) is 41\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  41,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(14) is 37\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  37,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(15) is 36\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  36,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(16) is 34\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  34,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(17) is 31\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  31,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(18) is 28\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  28,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(19) is 27\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  27,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(20) is 26\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  26,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(21) is 23\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  23,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(22) is 3\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  3,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(23) is 47\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  47,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(24) is 44\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  44,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(25) is 16\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  16,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(26) is 3\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  3,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(27) is 3\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  3,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(28) is 6\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  6,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(29) is 99\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  99,16764\n",
            "INFO [2025-06-10 09:18:02] Number of cells in group(30) is 97\n",
            "INFO [2025-06-10 09:18:02] group size being clustered:  97,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(31) is 69\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  69,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(32) is 49\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  49,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(33) is 35\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  35,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(34) is 21\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  21,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(35) is 18\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  18,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(36) is 10\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  10,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(37) is 7\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  7,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(38) is 7\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  7,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(39) is 2\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(40) is 1\n",
            "INFO [2025-06-10 09:18:03] Skipping group: 40, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(41) is 1\n",
            "INFO [2025-06-10 09:18:03] Skipping group: 41, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(42) is 100\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  100,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(43) is 83\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  83,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(44) is 58\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  58,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(45) is 40\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  40,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(46) is 15\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  15,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(47) is 2\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(48) is 1\n",
            "INFO [2025-06-10 09:18:03] Skipping group: 48, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(49) is 28\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  28,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(50) is 44\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  44,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(51) is 20\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  20,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(52) is 2\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:03] Number of cells in group(53) is 71\n",
            "INFO [2025-06-10 09:18:03] group size being clustered:  71,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(54) is 52\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  52,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(55) is 48\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  48,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(56) is 47\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  47,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(57) is 2\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(58) is 16\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  16,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(59) is 34\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  34,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(60) is 10\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  10,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(61) is 7\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  7,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(62) is 26\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  26,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(63) is 92\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  92,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(64) is 90\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  90,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(65) is 83\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  83,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(66) is 73\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  73,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(67) is 56\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  56,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(68) is 49\n",
            "INFO [2025-06-10 09:18:04] group size being clustered:  49,16764\n",
            "INFO [2025-06-10 09:18:04] Number of cells in group(69) is 30\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  30,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(70) is 15\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  15,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(71) is 15\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  15,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(72) is 4\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  4,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(73) is 1\n",
            "INFO [2025-06-10 09:18:05] Skipping group: 73, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(74) is 1\n",
            "INFO [2025-06-10 09:18:05] Skipping group: 74, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(75) is 36\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  36,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(76) is 20\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  20,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(77) is 94\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  94,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(78) is 56\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  56,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(79) is 47\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  47,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(80) is 45\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  45,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(81) is 34\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  34,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(82) is 20\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  20,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(83) is 13\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  13,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(84) is 5\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  5,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(85) is 60\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  60,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(86) is 54\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  54,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(87) is 27\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  27,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(88) is 7\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  7,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(89) is 5\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  5,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(90) is 2\n",
            "INFO [2025-06-10 09:18:05] group size being clustered:  2,16764\n",
            "INFO [2025-06-10 09:18:05] Number of cells in group(91) is 1\n",
            "INFO [2025-06-10 09:18:05] Skipping group: 91, since less than 2 entries\n",
            "INFO [2025-06-10 09:18:06] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 09:18:06] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 09:18:06] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:18:06] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:18:18] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 09:18:18] Quantiles of plotted data range: 0.880549880691281,0.972308546540498,0.998288297493428,1.02794867855461,1.12255441491482\n",
            "INFO [2025-06-10 09:18:23] plot_cnv_references:Start\n",
            "INFO [2025-06-10 09:18:23] Reference data size: Cells= 4906 Genes= 16764\n",
            "INFO [2025-06-10 09:18:41] plot_cnv_references:Number reference groups= 53\n",
            "INFO [2025-06-10 09:18:42] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 09:19:04] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 09:19:04] Quantiles of plotted data range: 0.880549880691281,0.974514578011378,0.998296202856621,1.02418718081683,1.12255441491482\n",
            "INFO [2025-06-10 09:20:52] ::plot_cnv:Start\n",
            "INFO [2025-06-10 09:20:52] ::plot_cnv:Current data dimensions (r,c)=16764,7783 Total=130676727.26151 Min=0.766442817427447 Max=8.73681684296526.\n",
            "INFO [2025-06-10 09:20:53] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 09:20:56] plot_cnv(): auto thresholding at: (0.880550 , 1.122554)\n",
            "INFO [2025-06-10 09:20:59] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 09:20:59] Observation data size: Cells= 2877 Genes= 16764\n",
            "INFO [2025-06-10 09:21:00] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 09:21:00] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 09:21:00] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:21:00] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:21:12] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 09:21:12] Quantiles of plotted data range: 0.880549880691281,0.972308546540498,0.998288297493428,1.02794867855461,1.12255441491482\n",
            "INFO [2025-06-10 09:21:17] plot_cnv_references:Start\n",
            "INFO [2025-06-10 09:21:17] Reference data size: Cells= 4906 Genes= 16764\n",
            "INFO [2025-06-10 09:42:46] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 09:42:47] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 09:43:08] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 09:43:08] Quantiles of plotted data range: 0.880549880691281,0.974514578011378,0.998296202856621,1.02418718081683,1.12255441491482\n",
            "INFO [2025-06-10 09:43:13] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-06-10 09:43:13] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-06-10 09:43:13] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 09:43:16] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.942381800161571 - 1.05913404577161\n",
            "INFO [2025-06-10 09:44:17] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-06-10 09:44:18] ::plot_cnv:Start\n",
            "INFO [2025-06-10 09:44:18] ::plot_cnv:Current data dimensions (r,c)=16764,7783 Total=130970824.881686 Min=0.766442817427447 Max=8.73681684296526.\n",
            "INFO [2025-06-10 09:44:19] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 09:44:21] plot_cnv(): auto thresholding at: (0.877446 , 1.122554)\n",
            "INFO [2025-06-10 09:44:24] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 09:44:24] Observation data size: Cells= 2877 Genes= 16764\n",
            "INFO [2025-06-10 09:44:25] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 09:44:25] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 09:44:25] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:44:25] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 09:44:34] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 09:44:34] Quantiles of plotted data range: 0.877445585085183,1.00075792296659,1.00075792296659,1.00075792296659,1.12255441491482\n",
            "INFO [2025-06-10 09:44:39] plot_cnv_references:Start\n",
            "INFO [2025-06-10 09:44:39] Reference data size: Cells= 4906 Genes= 16764\n",
            "INFO [2025-06-10 10:06:01] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 10:06:03] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 10:06:21] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:06:21] Quantiles of plotted data range: 0.877445585085183,1.00075792296659,1.00075792296659,1.00075792296659,1.12255441491482\n"
          ]
        }
      ],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"1_vs_4\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_tnU-Oj5Sp6J",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "cbcdfd7d-df54-43dd-9502-5daa91de138a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:06:26] Parsing matrix: /content/1_and_95.matrix\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, :\n",
            "“number of items read is not a multiple of the number of columns”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:08:07] Parsing gene order file: gene_pos_file.txt\n",
            "INFO [2025-06-10 10:08:07] Parsing cell annotations file: /content/1_vs_95.annotations.tsv\n",
            "INFO [2025-06-10 10:08:07] ::order_reduce:Start.\n",
            "INFO [2025-06-10 10:08:09] .order_reduce(): expr and order match.\n",
            "INFO [2025-06-10 10:08:11] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 31052,14818 Total=NA Min=NA Max=NA.\n",
            "INFO [2025-06-10 10:08:12] num genes removed taking into account provided gene ordering list: 7 = 0.0225428313796213% removed.\n",
            "INFO [2025-06-10 10:08:13] -filtering out cells < 100 or > Inf, removing 2.76016 % of cells\n",
            "WARN [2025-06-10 10:08:16] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-06-10 10:08:26] validating infercnv_obj\n"
          ]
        }
      ],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/1_and_95.matrix\",\n",
        "                                    annotations_file=\"/content/1_vs_95.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QAmqcgWgSu_t",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "85318c96-9852-4475-b7b0-b3c9833bc9a3",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:08:26] ::process_data:Start\n",
            "INFO [2025-06-10 10:08:26] Creating output path 1_vs_95\n",
            "INFO [2025-06-10 10:08:26] Checking for saved results.\n",
            "INFO [2025-06-10 10:08:26] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-06-10 10:09:12] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-06-10 10:09:12] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-06-10 10:09:13] Removing 18828 genes from matrix as below mean expr threshold: 0.01\n",
            "INFO [2025-06-10 10:09:15] validating infercnv_obj\n",
            "INFO [2025-06-10 10:09:15] There are 12217 genes and 14407 cells remaining in the expr matrix.\n",
            "INFO [2025-06-10 10:09:22] no genes removed due to min cells/gene filter\n",
            "INFO [2025-06-10 10:09:53] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-06-10 10:09:53] normalizing counts matrix by depth\n",
            "INFO [2025-06-10 10:09:59] Computed total sum normalization factor as median libsize: 2297.000000\n",
            "INFO [2025-06-10 10:10:34] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-06-10 10:10:34] transforming log2xplus1()\n",
            "INFO [2025-06-10 10:11:12] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-06-10 10:11:12] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 10:11:12] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 10:11:25] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 10:13:09] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-06-10 10:13:09] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-06-10 10:14:40] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-06-10 10:14:40] smooth_by_chromosome: chr: chr1\n",
            "INFO [2025-06-10 10:15:08] smooth_by_chromosome: chr: chr2\n",
            "INFO [2025-06-10 10:15:33] smooth_by_chromosome: chr: chr3\n",
            "INFO [2025-06-10 10:15:57] smooth_by_chromosome: chr: chr4\n",
            "INFO [2025-06-10 10:16:19] smooth_by_chromosome: chr: chr5\n",
            "INFO [2025-06-10 10:16:42] smooth_by_chromosome: chr: chr6\n",
            "INFO [2025-06-10 10:17:05] smooth_by_chromosome: chr: chr7\n",
            "INFO [2025-06-10 10:17:28] smooth_by_chromosome: chr: chr8\n",
            "INFO [2025-06-10 10:17:50] smooth_by_chromosome: chr: chr9\n",
            "INFO [2025-06-10 10:18:11] smooth_by_chromosome: chr: chr10\n",
            "INFO [2025-06-10 10:18:34] smooth_by_chromosome: chr: chr11\n",
            "INFO [2025-06-10 10:18:57] smooth_by_chromosome: chr: chr12\n",
            "INFO [2025-06-10 10:19:21] smooth_by_chromosome: chr: chr13\n",
            "INFO [2025-06-10 10:19:41] smooth_by_chromosome: chr: chr14\n",
            "INFO [2025-06-10 10:20:02] smooth_by_chromosome: chr: chr15\n",
            "INFO [2025-06-10 10:20:24] smooth_by_chromosome: chr: chr16\n",
            "INFO [2025-06-10 10:20:47] smooth_by_chromosome: chr: chr17\n",
            "INFO [2025-06-10 10:21:11] smooth_by_chromosome: chr: chr18\n",
            "INFO [2025-06-10 10:21:30] smooth_by_chromosome: chr: chr19\n",
            "INFO [2025-06-10 10:23:11] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-06-10 10:23:11] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-06-10 10:25:13] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-06-10 10:25:13] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 10:25:13] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 10:25:25] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 10:27:12] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-06-10 10:27:12] invert_log2(), computing 2^x\n",
            "INFO [2025-06-10 10:29:42] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-06-10 10:29:42] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-06-10 10:29:46] define_signif_tumor_subclusters(), tumor: aktywowane HSC\n",
            "INFO [2025-06-10 10:29:46] Setting auto leiden resolution for aktywowane HSC to 0.408451\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  MYOSLID, METTL21A, CREB1, CCNYL1, AC009226.1, KLF7-IT1, KLF7, FZD5, AC008269.1, FASTKD2 \n",
            "\t   DYTN, PLEKHM3, ADAM23, ZDBF2, CRYGD, EEF1B2, IDH1, NDUFS1, PIKFYVE, AC007383.2 \n",
            "\t   PTH2R, PDCL, AL359636.2, RPE, PTGS1, RC3H2, INO80D, MRRF, ZBTB6, RBM18 \n",
            "Negative:  GM2A, SLC36A1, CCDC69, SPARC, ANXA6, TNIP1, ATOX1, IRGM, SMIM3, G3BP1 \n",
            "\t   DCTN4, FAM114A2, MFAP3, RBM22, GALNT10, SAP30L-AS1, NDST1, SAP30L, RPS14, CSNK1A1 \n",
            "\t   LARP1, CD74, SLC26A2, ARHGEF37, PPARGC1B, HMGXB3, CSF1R, ARSI, PCYOX1L, TCOF1 \n",
            "PC_ 2 \n",
            "Positive:  CFLAR-AS1, CASP10, CASP8, CFLAR, FLACC1, TRAK2, NDUFB3, STRADB, FAM117B, ICA1L \n",
            "\t   BMPR2, TMEM237, WDR12, CARF, AC064836.4, NIF3L1, ALS2, SUMO1, NBEAL1, PPIL3 \n",
            "\t   NOP58, CYP20A1, ABI2, RAPH1, CD28, ICOS, PARD3B, NRP2, EIF2S1, ATP6V1D \n",
            "Negative:  CCDC183, RABL6, TMEM141, PHPT1, LCN6, EDF1, TRAF2, FBXW5, LCN10, PTGDS \n",
            "\t   SNHG7, DIPK1B, AGPAT2, EGFL7, AL590226.1, NOTCH1, SEC16A, INPP5E, PMPCA, ENTR1 \n",
            "\t   SNAPC4, CARD9, GPSM1, QSOX2, LHX3, TMEM250, NACC2, UBAC1, AL355574.1, CAMSAP1 \n",
            "PC_ 3 \n",
            "Positive:  SOX6, C11orf58, CALCB, CYP2R1, PDE3B, PLEKHA7, RRAS2, PSMA1, COPB1, SPON1 \n",
            "\t   RPS13, FAR1-IT1, PIK3C2A, FAR1, NUCB2, BTBD10, NCR3LG1, ARNTL, KCNC1, LINC00958 \n",
            "\t   TEAD1, SERGEF, MICAL2, DKK3, SAAL1, USP47, ZBED5-AS1, HPS5, ZBED5, AC116535.1 \n",
            "Negative:  OSBPL3, AC003093.1, CCDC126, TRA2A, MPP6, FAM221A, NPY, TMPO-AS1, AC007656.2, TMPO \n",
            "\t   SLC25A3, NEDD1, IKBIP, CFAP54, APAF1, CDK17, ANKS1B, UHRF1BP1L, ACTR6, AC008149.2 \n",
            "\t   DEPDC4, ELK3, SCYL2, LTA4H, GAS2L3, HAL, UTP20, CCDC38, ARL1, SNRPF \n",
            "PC_ 4 \n",
            "Positive:  SLC48A1, HDAC7, VDR, RAPGEF3, TMEM106C, RPAP3, PCED1B-AS1, SENP1, PCED1B, PFKM \n",
            "\t   SLC38A4, ASB8, AC008014.1, AC024257.3, SLC38A2, ZNF641, KANSL2, SLC38A1, CCNT1, SCAF11 \n",
            "\t   ADCY6, ARID2, CACNB3, L2HGDH, DMAC2L, SOS2, CDKL1, VCPKMT, LINC01588, DDX23 \n",
            "Negative:  LRCH1, ESD, SUCLA2, RUBCNL, LRRC63, NUDT15, LCP1, CPB2-AS1, MED4, ZC3H13 \n",
            "\t   COG3, MED4-AS1, SLC25A30, ITM2B, TPT1-AS1, AL138963.4, RB1-DT, TPT1, GTF2F2, RB1 \n",
            "\t   GPALPP1, LPAR6, AL359706.1, RCBTB2, NUFIP1, TSC22D1, CYSLTR2, LACC1, AL161421.1, CCDC122 \n",
            "PC_ 5 \n",
            "Positive:  MIOS, F3, CNN3, ABCD3, AC005532.1, GCLM, ARHGAP29, DNTTIP2, AL109613.1, BCAR3 \n",
            "\t   ALG14, FNBP1L, C1GALT1, DR1, TLCD4, CCDC18-AS1, AC092802.1, AC079804.3, RWDD3, CCDC18 \n",
            "\t   CCZ1B, PTBP2, TMED5, DPYD, ZNF12, MTF2, AC073343.2, ZNF316, C7orf26, DPYD-AS1 \n",
            "Negative:  ZNF839, MOK, WDR20, CINP, TECPR2, ANKRD9, HSP90AA1, RCOR1, TNFAIP2, CDC42BPB \n",
            "\t   EIF5, TRAF3, MARK3, CKB, TRMT61A, AL139300.2, BAG5, DYNC1H1, KLC1, COA8 \n",
            "\t   AL049840.2, AL049840.3, AL049840.4, AL118558.3, AL049840.5, XRCC3, ZFYVE21, PPP1R13B, ATP5MPL, TDRD9 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:29:47] define_signif_tumor_subclusters(), tumor: CD4 T-cells\n",
            "INFO [2025-06-10 10:29:47] Too few cells in group CD4 T-cells for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 10:29:47] define_signif_tumor_subclusters(), tumor: CD8 T-cells\n",
            "INFO [2025-06-10 10:29:47] Less cells in group CD8 T-cells than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 10:29:47] define_signif_tumor_subclusters(), tumor: HSC\n",
            "INFO [2025-06-10 10:29:47] Setting auto leiden resolution for HSC to 0.514215\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  BDKRB2, C14orf132, AL355102.1, AL139020.1, TCL1A, BDKRB1, GLRX5, ATG2B, SNHG10, GSKIP \n",
            "\t   SYNE3, AK7, CLMN, PAPOLA, VRK1, DICER1-AS1, LINC00618, DICER1, LINC01550, SERPINA1 \n",
            "\t   BCL11B, IFI27L2, SETD3, ATP5MD, PCGF6, TAF5, NT5C2, PDCD11, CNNM2, NEURL1 \n",
            "Negative:  CCDC88B, PRDX5, TRMT112, ESRRA, GPR137, OTUB1, BAD, MARK2, NAA40, MACROD1 \n",
            "\t   AP000721.2, COX8A, STIP1, FERMT3, PPP1R14B, SPINDOC, TRPT1, FKBP2, NUDT22, VEGFB \n",
            "\t   RTN3, DNAJC4, AP000753.2, EMC2, NUDCD1, ENY2, EBAG9, TRPS1, EIF3H, UTP23 \n",
            "PC_ 2 \n",
            "Positive:  PANK1, IFIT5, KIF20B, IFIT3, LINC01374, LINC01375, IFIT2, AL139340.1, RPP30, PCGF5 \n",
            "\t   LIPA, HECTD2, CH25H, FAS, TNKS2, SLC35G1, FRA10AC1, PLCE1, CEP55, FGFBP3 \n",
            "\t   NOC3L, MYOF, ACTA2, EXOC6, HHEX, TBC1D12, KIF11, IDE, BTAF1, MARCH5 \n",
            "Negative:  CDC123, AL512770.1, CAMK1D, CCDC3, OPTN, MCM10, PHYH, SEPHS1, BEND7, PRPF18 \n",
            "\t   AL157392.3, FRMD4A, OTUD4, ABCE1, ANAPC10, SMAD1, HHIP, HHIP-AS1, AC098588.3, GYPA \n",
            "\t   GYPB, GYPE, AC107223.1, SMARCA5, AC097658.3, GAB1, USP38, AC104596.1, AC139720.1, INPP4B \n",
            "PC_ 3 \n",
            "Positive:  RNF139, RNF139-AS1, TRMT12, TMEM65, NDRG1, ST3GAL1, AC103706.1, FBXO32, ZFAT, KHDRBS3 \n",
            "\t   LINC02055, WDYHV1, COL22A1, TRAPPC9, ATAD2, AC107375.1, ZHX1, CHRAC1, C8orf76, AGO2 \n",
            "\t   PTK2, TBC1D31, DENND3, DERL1, AC040970.1, AC104316.1, SLC45A4, AC016405.1, TSNARE1, ZHX2 \n",
            "Negative:  TNS1, DIRC3, ARPC2, IGFBP2, AC021016.2, RPL37A, GPBAR1, AAMP, SMARCAL1, PNKD \n",
            "\t   TMBIM1, XRCC5, SLC11A1, PECR, CTDSP1, MREG, VIL1, FN1, USP37, ATIC \n",
            "\t   CNOT9, SNHG31, ZNF142, BARD1, BCS1L, AC068051.1, RNF25, SPAG16, IKZF2, AC093865.1 \n",
            "PC_ 4 \n",
            "Positive:  KAT6A, AC103724.4, ANK1, AP3M2, IKBKB, GPAT4, POLB, VDAC3, THAP1, SMIM19 \n",
            "\t   GINS4, RNF170, SLC20A2, HOOK3, FNTA, AC009630.1, POMK, GOLGA7, ADAM32, HGSNAT \n",
            "\t   ADAM9, SPIDR, TM2D2, HTRA4, CEBPD, AC108863.2, PRKDC, PLEKHA2, MCM4, AC067817.2 \n",
            "Negative:  DAGLB, KDELR2, RAC1, GRID2IP, ZDHHC4, CYTH3, C7orf26, ZNF316, USP42, AC073343.2 \n",
            "\t   ZNF12, EIF2AK1, CCZ1B, AC079804.3, AIMP2, C1GALT1, AC005532.1, PMS2, MIOS, RPA3 \n",
            "\t   CCZ1, RNF216, UMAD1, FSCN1, GLCCI1, ACTB, ICA1, FBXL18, TNRC18, NDUFA4 \n",
            "PC_ 5 \n",
            "Positive:  ZNF891, ZNF140, ZNF10, ZNF268, ZNF84, ZNF26, ZNF605, CHFR, GOLGA3, ANKLE2 \n",
            "\t   PGAM5, PXMP2, POLE, LRCOL1, FBRSL1, LINC02361, NOC4L, DDX51, EP400, PUS1 \n",
            "\t   AC131009.1, ULK1, MMP17, SFSWAP, AC078925.1, ADGRD1, RAN, STX2, GLT1D1, SLC15A4 \n",
            "Negative:  FAM13B, PKD2L2, AC106791.1, BRD8, HNRNPA0, KIF20A, KLHL3, CDC23, SPOCK1, CDC25C \n",
            "\t   SMAD5, FAM53C, TGFBI, SLC25A48, KDM3B, H2AFY, EGR1, ETF1, C5orf66, HSPA9 \n",
            "\t   PCBD2, CTNNA1, TXNDC15, SIL1, C5orf24, DDX46, AC011405.1, CAMLG, SNHG4, SEC24A \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:29:49] define_signif_tumor_subclusters(), tumor: komórki B pamięci\n",
            "INFO [2025-06-10 10:29:49] Too few cells in group komórki B pamięci for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 10:29:49] define_signif_tumor_subclusters(), tumor: Komórki blastyczne (preB/CLP)\n",
            "INFO [2025-06-10 10:29:49] Setting auto leiden resolution for Komórki blastyczne (preB/CLP) to 0.00781197\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  ZWILCH, RPL4, SMAD3, SNAPC5, MAP2K1, TIPIN, AC055855.1, AC012568.1, DIS3L, AAGAB \n",
            "\t   MEGF11, IQCH, AC011939.2, RAB11A, IQCH-AS1, DENND4A, C15orf61, SLC24A1, MAP2K5, INTS14 \n",
            "\t   PIAS1, HACD3, CALML4, DPP8, CLN6, PARP16, FEM1B, CLPX, ITGA11, PDCD7 \n",
            "Negative:  USP53, SEC24D, C4orf3, PDE5A, METTL14, MAD2L1, SNHG8, CAMK2D, AC073475.1, LARP7 \n",
            "\t   ANXA5, EXOSC9, ZGRF1, CCNA2, BBS7, ALPK1, KIAA1109, TIFA, NUDT6, AP1AR \n",
            "\t   SPATA5, SPRY1, FAM241A, LINC01091, AC004704.1, ANKRD50, ELOVL6, FAT4, GAR1, INTU \n",
            "PC_ 2 \n",
            "Positive:  AC106865.2, TMEM131L, TLR2, RNF175, MND1, PLRG1, MAP9, TRIM2, GUCY1A1, FHDC1 \n",
            "\t   GUCY1B1, ARFIP1, CTSO, PDGFC, TMEM154, GASK1B, MIR4453HG, FAM198B-AS1, AC023424.3, TMEM144 \n",
            "\t   FBXW7, GATB, C4orf46, FAM160A1, NAIP, SMN1, GTF2H2, SMN2, ETFDH, BDP1 \n",
            "Negative:  RNF213, SLC26A11, SGSH, EIF4A3, GAA, TBC1D16, CBX4, CBX2, ENGASE, CANT1 \n",
            "\t   LGALS3BP, CEP295NL, TIMP2, AC022966.2, USP36, CYTH1, SCAT1, DNAH17, PGS1, SOCS3 \n",
            "\t   BIRC5, AFMID, TK1, SYNGR2, C17orf99, TMC8, TMC6, TNRC6C-AS1, TNRC6C, AC021683.2 \n",
            "PC_ 3 \n",
            "Positive:  CIAO2A, SNX1, SNX22, PPIB, CSNK1G1, PCLAF, TRIP4, ZNF609, OAZ2, AC100830.1 \n",
            "\t   RBPMS2, PLEKHO2, ANKDD1A, AC103691.1, SPG21, MTFMT, UBAP1L, PDCD7, CLPX, PARP16 \n",
            "\t   DPP8, AQP9, HACD3, ALDH1A2, INTS14, POLR2M, SLC24A1, LINC00926, DENND4A, RAB11A \n",
            "Negative:  PIK3R1, SLC30A5, AC010273.2, CCNB1, MAST4, CENPH, MRPS36, SREK1, CDK7, CCDC125 \n",
            "\t   AC025442.2, AK6, AC010359.3, TAF9, RAD17, ERBIN, MARVELD2, NLN, OCLN, SGTB \n",
            "\t   GTF2H2C, TRAPPC13, SHLD3, SMN2, TRIM23, SMN1, PPWD1, NAIP, CENPK, GTF2H2 \n",
            "PC_ 4 \n",
            "Positive:  AC002460.2, AC108156.1, NR3C2, LINC02355, ARHGAP10, DCLK2, LRBA, PRMT9, RPS3A, TMEM184C \n",
            "\t   SH3D19, SLC10A7, AC095055.1, LSM6, FAM160A1, ZNF827, GATB, MMAA, FBXW7, AC023424.3 \n",
            "\t   MIR4453HG, SMAD1-AS2, TMEM154, ARFIP1, SMAD1, FHDC1, TRIM2, OTUD4, MND1, ABCE1 \n",
            "Negative:  SLC30A5, AC010273.2, CCNB1, PIK3R1, MAST4, CENPH, SREK1, MRPS36, AC025442.2, CDK7 \n",
            "\t   AC010359.3, CCDC125, ERBIN, AK6, NLN, TAF9, SGTB, RAD17, TRAPPC13, MARVELD2 \n",
            "\t   SHLD3, OCLN, TRIM23, GTF2H2C, PPWD1, SMN2, CENPK, SMN1, ADAMTS6, NAIP \n",
            "PC_ 5 \n",
            "Positive:  ADM, AC100763.1, AMPD3, MTRNR2L8, AC011092.3, RNF141, AC011092.2, MRVI1-AS1, SBF2, SBF2-AS1 \n",
            "\t   MRVI1, SWAP70, CTR9, WEE1, EIF4G2, ZNF143, AC116535.1, AC132192.2, ZBED5, AC132192.1 \n",
            "\t   ZBED5-AS1, IPO7, USP47, TMEM41B, DKK3, DENND5A, MICAL2, SCUBE2, NRIP3, TEAD1 \n",
            "Negative:  FOXP1-AS1, FOXP1, AC097634.2, MITF, FOXP1-IT1, FRMD4B, AC097634.3, LMOD3, AC097634.1, ARL6IP5 \n",
            "\t   UBA3, EIF4E3, TMF1, AC109587.1, PROK2, EOGT, TAFA1, LINC00877, SUCLG2-AS1, RYBP \n",
            "\t   SUCLG2, KBTBD8, SHQ1, LRIG1, PPP4R2, SLC25A26, EBLN2, MAGI1, PSMD6-AS1, PDZRN3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:07] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 1 (preB)\n",
            "INFO [2025-06-10 10:30:07] Setting auto leiden resolution for Komórki blastyczne 1 (preB) to 0.00792161\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  KLHL6, KLHL6-AS1, MAGEF1, KLHL24, B3GNT5, YEATS2, VPS8, PARL, MCF2L2, EHHADH \n",
            "\t   MCCC1, MAP3K13, DCUN1D1, ATP11B, TMEM41A, SOX2-OT, DNAJC19, SENP2, IGF2BP2, FXR1 \n",
            "\t   TRA2B, ETV5, CCDC39, DGKG, TTC14, CRYGS, TBCCD1, USP13, DNAJB11, NDUFB5 \n",
            "Negative:  MAD2L1, PDE5A, AC073475.1, C4orf3, ANXA5, USP53, EXOSC9, CCNA2, SEC24D, BBS7 \n",
            "\t   KIAA1109, METTL14, NUDT6, SNHG8, SPATA5, CAMK2D, SPRY1, LINC01091, LARP7, ANKRD50 \n",
            "\t   ZGRF1, FAT4, ALPK1, INTU, TIFA, PLK4, AP1AR, MFSD8, FAM241A, ABHD18 \n",
            "PC_ 2 \n",
            "Positive:  POLE2, DNAAF2, KLHDC1, MGAT2, KLHDC2, RPL36AL, NEMF, LRR1, AL627171.2, RPS29 \n",
            "\t   AL627171.1, MDGA2, ARF6, MIS18BP1, LINC01588, FANCM, VCPKMT, FKBP3, SOS2, PRPF39 \n",
            "\t   L2HGDH, TOGARAM1, DMAC2L, KLHL28, CDKL1, C14orf28, AL049828.1, MAP4K5, FBXO33, AL118556.1 \n",
            "Negative:  STXBP1, PTRH1, NIBAN2, TOR2A, LRSAM1, SH2D3C, CDK9, RPL12, FPGS, ZNF79 \n",
            "\t   ENG, AL162586.1, GARNL3, RALGPS1, ZBTB34, ZBTB43, MVB12B, PBX3, MAPKAP1, GAPVD1 \n",
            "\t   HSPA5, RABEPK, PPP6C, SCAI, GOLGA1, ARPC5L, RPL35, MIR181A2HG, NR6A1, PSMB7 \n",
            "PC_ 3 \n",
            "Positive:  EHHADH, MAP3K13, VPS8, TMEM41A, MAGEF1, SENP2, IGF2BP2, TRA2B, ETV5, DGKG \n",
            "\t   CRYGS, TBCCD1, DNAJB11, AC068631.1, EIF4A2, RFC4, ST6GAL1, PARL, RPL39L, YEATS2 \n",
            "\t   BCL6, KLHL24, KLHL6-AS1, LPP, KLHL6, TPRG1, B3GNT5, MCF2L2, P3H2, MCCC1 \n",
            "Negative:  AC090971.2, LEO1, MAPK6, TMOD3, AC023906.5, TMOD2, GNB5, AC023906.4, LYSMD2, CERNA1 \n",
            "\t   DMXL2, MYO5C, MYO5A, AC066613.2, ARPP19, AC025917.1, AP4E1, FAM214A, AC021752.1, AC011912.1 \n",
            "\t   RSL24D1, SPPL2A, RAB27A, TRPM7, AC018926.2, USP50, PIGBOS1, PIGB, USP8, CCPG1 \n",
            "PC_ 4 \n",
            "Positive:  DNAAF2, POLE2, MGAT2, KLHDC1, RPL36AL, KLHDC2, LRR1, NEMF, RPS29, AL627171.2 \n",
            "\t   MDGA2, AL627171.1, MIS18BP1, ARF6, FANCM, LINC01588, FKBP3, VCPKMT, PRPF39, SOS2 \n",
            "\t   TOGARAM1, L2HGDH, KLHL28, DMAC2L, C14orf28, AL049828.1, CDKL1, FBXO33, MAP4K5, MIA2 \n",
            "Negative:  AC025917.1, FAM214A, AC011912.1, ARPP19, MYO5A, MYO5C, RSL24D1, CERNA1, AC023906.4, RAB27A \n",
            "\t   GNB5, AC018926.2, AC023906.5, PIGBOS1, MAPK6, PIGB, LEO1, CCPG1, AC090971.2, DNAAF4 \n",
            "\t   TMOD3, NEDD4, TMOD2, RFX7, LYSMD2, DMXL2, TEX9, MNS1, AC066613.2, ZNF280D \n",
            "PC_ 5 \n",
            "Positive:  TOGARAM1, PRPF39, KLHL28, FKBP3, C14orf28, FANCM, AL049828.1, MIS18BP1, FBXO33, MDGA2 \n",
            "\t   MIA2, RPS29, PNN, LRR1, AL109628.2, RPL36AL, TRAPPC6B, MGAT2, GEMIN2, DNAAF2 \n",
            "\t   SEC23A, POLE2, AL355835.1, KLHDC1, CLEC14A, KLHDC2, MIPOL1, NEMF, AL162464.2, AL627171.2 \n",
            "Negative:  MRPS36, CDK7, CENPH, CCDC125, CCNB1, AK6, AC010273.2, TAF9, SLC30A5, RAD17 \n",
            "\t   PIK3R1, MAST4, MARVELD2, SREK1, AC025442.2, AC010359.3, OCLN, ERBIN, NLN, GTF2H2C \n",
            "\t   SGTB, TRAPPC13, SMN2, SHLD3, SMN1, TRIM23, PPWD1, NAIP, CENPK, GTF2H2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:24] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 2 (preB)\n",
            "INFO [2025-06-10 10:30:24] Setting auto leiden resolution for Komórki blastyczne 2 (preB) to 0.0138945\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  RPS6KA1, ARID1A, PIGV, HMGN2, AL513365.2, ZDHHC18, DHDDS, GPN2, CD52, GPATCH3 \n",
            "\t   UBXN11, NUDC, SH3BGRL3, SLC9A1, CEP85, WDTC1, ZNF593, TMEM222, PDIK1L, SYTL1 \n",
            "\t   PAFAH2, WASF2, FO393419.2, STMN1, AHDC1, PAQR7, FGR, AUNIP, IFI6, MTFR1L \n",
            "Negative:  MCUB, CASP6, HADH, SEC24B-AS1, CYP2U1, PLA2G12A, LEF1, SEC24B, COL25A1, AC096564.1 \n",
            "\t   LEF1-AS1, GAR1, OSTC, SGMS2, RPL34-AS1, RPL34, ELOVL6, PAPSS1, AC004704.1, AC092445.1 \n",
            "\t   FAM241A, AIMP1, AP1AR, TBCK, TIFA, GSTCD, ALPK1, INTS12, ZGRF1, ARHGEF38 \n",
            "PC_ 2 \n",
            "Positive:  AC010273.2, SLC30A5, CCNB1, PIK3R1, CENPH, MRPS36, MAST4, CDK7, SREK1, AC025442.2 \n",
            "\t   CCDC125, AC010359.3, AK6, ERBIN, TAF9, NLN, RAD17, SGTB, MARVELD2, TRAPPC13 \n",
            "\t   OCLN, SHLD3, GTF2H2C, TRIM23, SMN2, SMN1, PPWD1, NAIP, CENPK, GTF2H2 \n",
            "Negative:  COL25A1, OSTC, SEC24B-AS1, RPL34, SEC24B, RPL34-AS1, MCUB, CASP6, PLA2G12A, LEF1-AS1 \n",
            "\t   GAR1, ELOVL6, LEF1, HADH, AC004704.1, CYP2U1, FAM241A, AC096564.1, AP1AR, TIFA \n",
            "\t   SGMS2, ALPK1, PAPSS1, ZGRF1, LARP7, AC092445.1, CAMK2D, AIMP1, SNHG8, TBCK \n",
            "PC_ 3 \n",
            "Positive:  DMXL2, AC066613.2, LYSMD2, AP4E1, TMOD2, AC021752.1, TMOD3, SPPL2A, AC090971.2, LEO1 \n",
            "\t   TRPM7, MAPK6, USP50, AC023906.5, USP8, GNB5, AC022087.1, AC023906.4, GABPB1-AS1, CERNA1 \n",
            "\t   GABPB1-IT1, MYO5C, GABPB1, MYO5A, HDC, ARPP19, SLC27A2, AC025917.1, ATP8B4, FAM214A \n",
            "Negative:  INTS1, MICALL2, MAFK, ZFAND2A, GPR146, TMEM184A, C7orf50, COX19, ADAP1, PSMG3 \n",
            "\t   AC073957.3, PSMG3-AS1, PRKAR1B, DNAAF5, AC147651.1, SUN1, PDGFA, AC147651.3, FAM20C, MAD1L1 \n",
            "\t   MRM2, NUDT1, SNX8, EIF3B, CHST12, LFNG, BRAT1, IQCE, TTYH3, AMZ1 \n",
            "PC_ 4 \n",
            "Positive:  CDK7, MRPS36, CCDC125, CENPH, AK6, CCNB1, TAF9, AC010273.2, RAD17, SLC30A5 \n",
            "\t   MARVELD2, PIK3R1, MAST4, OCLN, SREK1, AC025442.2, GTF2H2C, AC010359.3, ERBIN, SMN2 \n",
            "\t   NLN, SGTB, SMN1, TRAPPC13, NAIP, SHLD3, TRIM23, GTF2H2, PPWD1, BDP1 \n",
            "Negative:  TMOD3, TMOD2, AC090971.2, LYSMD2, LEO1, DMXL2, MAPK6, AC066613.2, AC023906.5, AP4E1 \n",
            "\t   GNB5, AC021752.1, AC023906.4, SPPL2A, CERNA1, TRPM7, MYO5C, USP50, MYO5A, USP8 \n",
            "\t   ARPP19, AC022087.1, AC025917.1, FAM214A, GABPB1-AS1, AC011912.1, GABPB1-IT1, RSL24D1, GABPB1, RAB27A \n",
            "PC_ 5 \n",
            "Positive:  PTRH2, VMP1, CLTC, TUBD1, RPS6KB1, AC091271.1, DHX40, RNFT1, YPEL2, RNFT1-DT \n",
            "\t   HEATR6, AC025048.2, USP32, APPBP2, AC011921.1, PPM1D, BCAS3, AC005884.1, AC005856.1, AC005746.2 \n",
            "\t   BRIP1, INTS2, MED13, AC018628.2, METTL2A, TLK2, MRC2, TANC2, AC015923.1, CYB561 \n",
            "Negative:  KIF3A, AC004039.1, SEPTIN8, TH2LCRR, GDF9, RAD50, UQCRQ, IRF1, LEAP2, AC116366.1 \n",
            "\t   AFF4, AC116366.2, ZCCHC10, HSPA4, C5orf56, C5orf15, SLC22A5, VDAC1, MIR3936HG, TCF7 \n",
            "\t   SLC22A4, SKP1, P4HA2, PPP2CA, ACSL6, CDKL3, FNIP1, UBE2B, RAPGEF6, AC109454.2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:32] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 5 (preB)\n",
            "INFO [2025-06-10 10:30:32] Setting auto leiden resolution for Komórki blastyczne 5 (preB) to 0.0892444\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AP2M1, ABCF3, DVL3, ALG3, EIF2B5, EEF1AKMT4, PSMD2, ABCC5, EIF4G1, FAM131A \n",
            "\t   PARL, POLR2H, YEATS2, MAGEF1, VPS8, KLHL24, EHHADH, KLHL6-AS1, MAP3K13, KLHL6 \n",
            "\t   TMEM41A, B3GNT5, SENP2, MCF2L2, IGF2BP2, IGF2BP2-AS1, MCCC1, TRA2B, DCUN1D1, ETV5 \n",
            "Negative:  MALSU1, IGF2BP3, NUP42, TRA2A, KLHL7, CCDC126, FAM221A, FAM126A, NPY, MPP6 \n",
            "\t   AC003093.1, SNHG26, OSBPL3, CYCS, TOMM7, C7orf31, STEAP1B, NFE2L3, RAPGEF5, CBX3 \n",
            "\t   SNX10, HNRNPA2B1, CDCA7L, AC004540.2, DNAH11, AC004540.1, SKAP2, SP4, HOTAIRM1, HOXA3 \n",
            "PC_ 2 \n",
            "Positive:  CDKL3, PPP2CA, SKP1, AC109454.2, UBE2B, TCF7, CDKN2AIPNL, JADE2, VDAC1, SAR1B \n",
            "\t   C5orf15, SEC24A, CAMLG, HSPA4, DDX46, ZCCHC10, AFF4, C5orf24, LEAP2, TXNDC15 \n",
            "\t   UQCRQ, PCBD2, GDF9, C5orf66, SEPTIN8, H2AFY, SLC25A48, KIF3A, TGFBI, AC004039.1 \n",
            "Negative:  FAM49A, AC104623.1, AC010880.1, SMC6, AC010745.2, GEN1, MYCN, RDH14, DDX1, LINC01376 \n",
            "\t   NBAS, TTC32, TRIB2, WDR35, MIR3681HG, LAPTM4A, LPIN1, PUM2, RHOB, HS1BP3 \n",
            "\t   E2F6, LDAH, AC096570.2, ROCK2, KLHL29, ATAD2B, UBXN2A, SLC66A3, MFSD2B, SF3B6 \n",
            "PC_ 3 \n",
            "Positive:  PAQR7, STMN1, PAFAH2, PDIK1L, AUNIP, ZNF593, MTFR1L, CEP85, SELENON, SH3BGRL3 \n",
            "\t   AL031280.1, UBXN11, MAN1C1, CD52, LDLRAP1, DHDDS, MACO1, AL513365.2, RHCE, HMGN2 \n",
            "\t   RPS6KA1, ARID1A, PIGV, ZDHHC18, GPN2, GPATCH3, NUDC, NCMAP, SLC9A1, NCMAP-DT \n",
            "Negative:  AC068631.1, EIF4A2, DNAJB11, TBCCD1, RFC4, CRYGS, ST6GAL1, DGKG, RPL39L, ETV5 \n",
            "\t   BCL6, TRA2B, IGF2BP2-AS1, LPP, IGF2BP2, TPRG1, SENP2, P3H2, TMEM41A, MAP3K13 \n",
            "\t   IL1RAP, EHHADH, CCDC50, VPS8, MB21D2, MAGEF1, ATP13A4, OPA1, POLR2H, OPA1-AS1 \n",
            "PC_ 4 \n",
            "Positive:  LEF1, LEF1-AS1, HADH, RPL34-AS1, CYP2U1, RPL34, AC096564.1, OSTC, SGMS2, PAPSS1 \n",
            "\t   COL25A1, AC092445.1, SEC24B-AS1, AIMP1, SEC24B, TBCK, GSTCD, MCUB, INTS12, CASP6 \n",
            "\t   ARHGEF38, PLA2G12A, PPA2, GAR1, TET2-AS1, ELOVL6, TET2, AC004704.1, AC096577.1, AC004053.1 \n",
            "Negative:  DAGLB, RAC1, CYTH3, USP42, EIF2AK1, AIMP2, PMS2, CCZ1, RNF216, FSCN1 \n",
            "\t   ACTB, FBXL18, TNRC18, WIPI2, RBAK, RADIL, AP5Z1, AC092428.1, AC073333.1, FOXK1 \n",
            "\t   SDK1, TSPAN13, CARD11, AHR, GNA12, AMZ1, AC073332.1, TTYH3, SNX13, IQCE \n",
            "PC_ 5 \n",
            "Positive:  ERO1A, STYX, AL133453.1, PSMC6, GNPNAT1, GPR137C, AL139317.5, TXNDC16, FERMT2, PTGER2 \n",
            "\t   DDHD1, PTGDR, AL365295.1, CDKN3, RTRAF, CNIH1, GMFB, GNG2, CGRRF1, FRMD6 \n",
            "\t   SAMD4A, TMX1, GCH1, WDHD1, PYGL, SOCS4, AL606834.2, MAPK1IP1L, NIN, SAV1 \n",
            "Negative:  AC083870.1, CALD1, AGBL3, BPGM, CYREN, AKR1B1, TMEM140, SLC35B4, AC083862.3, LRGUK \n",
            "\t   WDR91, EXOC4, AC009365.4, AC009542.1, CHCHD3, CNOT4, PLXNA4, NUP205, PODXL, STMP1 \n",
            "\t   MKLN1, AC015987.1, LINC-PINT, MTPN, AC058791.1, DGKI, AC016831.5, CREB3L2, AC016831.1, AKR1D1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:34] define_signif_tumor_subclusters(), tumor: Komórki blastyczne proliferujące\n",
            "INFO [2025-06-10 10:30:34] Setting auto leiden resolution for Komórki blastyczne proliferujące to 0.0415009\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AATK, CEP131, BAIAP2, TEPSIN, BAIAP2-DT, NDUFAF8, CHMP6, SLC38A10, RPTOR, AC027601.6 \n",
            "\t   ENDOV, BAHCC1, RNF213-AS1, ACTG1, RNF213, SLC26A11, FAAP100, SGSH, NPLOC4, EIF4A3 \n",
            "\t   PDE6G, GAA, OXLD1, TBC1D16, CCDC137, CBX4, ARL16, CBX2, HGS, ENGASE \n",
            "Negative:  DR1, FNBP1L, BCAR3, CCDC18-AS1, AL109613.1, CCDC18, DNTTIP2, TMED5, MTF2, GCLM \n",
            "\t   TGFBR3, BTBD8, GLMN, CDC7, RPAP2, DIPK1A, HFM1, GFI1, ARHGAP29, EVI5 \n",
            "\t   ZNF644, RPL5, ZNF326, MRPS9, ABCD3, LRRC8D, TGFBRAP1, TMEM182, AC012360.1, MFSD9 \n",
            "PC_ 2 \n",
            "Positive:  CORO2A, TBC1D2, TRIM14, ANKS6, GALNT12, NANS, TGFBR1, ANP32B, ALG2, HEMGN \n",
            "\t   SEC61B, AL499604.1, STX17-AS1, ADH5, TRMO, NR4A3, METAP1, AP002026.1, STX17, INVS \n",
            "\t   ERP44, TEX10, TRMT10A, AC019131.1, MSANTD3, EIF4E, DAPP1, TMEFF1, XPA, TSPAN5 \n",
            "Negative:  CSRNP1, AC092053.3, TTC21A, AC104850.2, GORASP1, SLC25A38, WDR48, SCN11A, EXOG, RPSA \n",
            "\t   ACVR2B, MYD88, ACAA1, DLEC1, OXSR1, XYLB, PLCD1, VILL, CTDSPL, AC099332.2 \n",
            "\t   ITGA9-AS1, ITGA9, EPM2AIP1, MLH1, C3orf35, LRRFIP2, GOLGA4, MYRIP, EIF1B-AS1, EIF1B \n",
            "PC_ 3 \n",
            "Positive:  RNF217-AS1, NKAIN2, SAMD3, SMPDL3A, EPB41L2, PKIB, AKAP7, SERINC1, ARG1, HSF2 \n",
            "\t   MED23, TBC1D32, CCN2, AL078600.1, AL133346.1, MAN1A1, LINC01013, FAM184A, STX7, VNN1 \n",
            "\t   AL137009.1, VNN3, ASF1A, VNN2, MCM9, SLC18B1, Z99496.2, RPS12, PLN, TBPL1 \n",
            "Negative:  PPM1D, BCAS3, AC005884.1, AC011921.1, AC005856.1, APPBP2, AC005746.2, BRIP1, USP32, AC025048.2 \n",
            "\t   INTS2, HEATR6, MED13, AC018628.2, RNFT1-DT, METTL2A, RNFT1, TLK2, RPS6KB1, MRC2 \n",
            "\t   TUBD1, TANC2, VMP1, AC015923.1, PTRH2, CYB561, CLTC, DCAF7, AC091271.1, TACO1 \n",
            "PC_ 4 \n",
            "Positive:  CALD1, BPGM, AKR1B1, AC083870.1, AGBL3, SLC35B4, LRGUK, CYREN, EXOC4, PLXNA4 \n",
            "\t   AC009365.4, CHCHD3, TMEM140, AC083862.3, WDR91, AC009542.1, CNOT4, NUP205, STMP1, AC015987.1 \n",
            "\t   MTPN, DGKI, CREB3L2, AKR1D1, TRIM24, SVOPL, ATP6V0A4, KIAA1549, ZC3HAV1L, ZC3HAV1 \n",
            "Negative:  AIMP2, EIF2AK1, PMS2, USP42, CYTH3, CCZ1, RNF216, RAC1, FSCN1, DAGLB \n",
            "\t   ACTB, KDELR2, FBXL18, GRID2IP, TNRC18, ZDHHC4, WIPI2, C7orf26, RBAK, ZNF316 \n",
            "\t   RADIL, AC073343.2, AP5Z1, ZNF12, AC092428.1, CCZ1B, FOXK1, AC079804.3, SDK1, C1GALT1 \n",
            "PC_ 5 \n",
            "Positive:  DNAJB14, H2AFZ, LAMTOR3, AC097460.1, DAPP1, DDIT4L, TRMT10A, EMCN, AP002026.1, PPP3CA \n",
            "\t   AP001816.1, ADH5, BANK1, AP002075.1, SLC39A8, METAP1, NFKB1, AC019131.1, MANBA, EIF4E \n",
            "\t   UBE2D3, MARCKS, HDAC2, HDAC2-AS2, FAM229B, NT5DC1, TRAF3IP2-AS1, TRAF3IP2, TUBE1, TSPAN5 \n",
            "Negative:  FAM160A1, GATB, FBXW7, AC095055.1, AC023424.3, SH3D19, MIR4453HG, RPS3A, LRBA, DCLK2 \n",
            "\t   TMEM154, LINC02355, AC108156.1, ARFIP1, AC002460.2, FHDC1, NR3C2, TRIM2, ARHGAP10, MND1 \n",
            "\t   PRMT9, TMEM131L, TMEM184C, AC106865.2, SLC10A7, TLR2, RNF175, LSM6, PLRG1, ZNF827 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:36] define_signif_tumor_subclusters(), tumor: komórki NK T\n",
            "INFO [2025-06-10 10:30:36] Less cells in group komórki NK T than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 10:30:36] define_signif_tumor_subclusters(), tumor: MPP\n",
            "INFO [2025-06-10 10:30:36] Too few cells in group MPP for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 10:30:36] define_signif_tumor_subclusters(), tumor: późne komórki erytroidalne\n",
            "INFO [2025-06-10 10:30:36] Setting auto leiden resolution for późne komórki erytroidalne to 0.106375\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  EEF1B2, NDUFS1, ZDBF2, ADAM23, AC007383.2, DYTN, INO80D, NRP2, FASTKD2, PARD3B \n",
            "\t   AC008269.1, ICOS, CD28, KLF7, RAPH1, KLF7-IT1, ABI2, MYOSLID, CYP20A1, AC009226.1 \n",
            "\t   NBEAL1, CARF, CREB1, WDR12, METTL21A, ICA1L, CCNYL1, FAM117B, BMPR2, FZD5 \n",
            "Negative:  MYCBP2, SLAIN1, POU4F1, FBXL3, OBI1, MYCBP2-AS1, CLN5, RBM26, KCTD12, LMO7 \n",
            "\t   LMO7-AS1, RBM26-AS1, UCHL3, NDFIP2, SPRY2, COMMD6, AL590807.1, TBC1D4, AL162493.1, KLF12 \n",
            "\t   SLITRK5, KLF5, MIR17HG, PIBF1, GPC5, TGDS, DIS3, GPR180, BORA, ABCC4 \n",
            "PC_ 2 \n",
            "Positive:  TBCK, GSTCD, AIMP1, INTS12, AC092445.1, PAPSS1, ARHGEF38, SGMS2, PPA2, TET2-AS1 \n",
            "\t   AC096564.1, TET2, AC096577.1, CYP2U1, AC004053.1, HADH, LINC02428, CENPE, LEF1, BDH2 \n",
            "\t   SLC9B2, SLC9B1, LEF1-AS1, CISD2, UBE2D3-AS1, AC018797.3, RPL34-AS1, UBE2D3, MANBA, NFKB1 \n",
            "Negative:  FOXK1, SDK1, AC092428.1, CARD11, AMZ1, GNA12, AP5Z1, TTYH3, RADIL, IQCE \n",
            "\t   RBAK, BRAT1, WIPI2, TNRC18, LFNG, FBXL18, CHST12, ACTB, FSCN1, EIF3B \n",
            "\t   RNF216, CCZ1, SNX8, PMS2, AIMP2, NUDT1, CYTH3, RAC1, EIF2AK1, MRM2 \n",
            "PC_ 3 \n",
            "Positive:  C1GALT1, AC005532.1, AC079804.3, MIOS, CCZ1B, RPA3, UMAD1, ZNF12, GLCCI1, ICA1 \n",
            "\t   AC073343.2, NDUFA4, ZNF316, PHF14, C7orf26, THSD7A, TMEM106B, ZDHHC4, VWDE, GRID2IP \n",
            "\t   ARL4A, KDELR2, AC011287.1, DAGLB, CRPPA, RAC1, CRPPA-AS1, CYTH3, ANKMY2, USP42 \n",
            "Negative:  ALDH1L2, C12orf45, SLC41A2, TXNRD1, EID3, CHST11, NFYB, HCFC2, GLT8D2, TDG \n",
            "\t   C12orf73, HSP90B1, NT5DC3, C12orf42, HELLPAR, PARPBP, NUP37, WASHC3, DRAM1, GNPTAB \n",
            "\t   SYCP3, CHPT1, ARL1, UTP20, GAS2L3, SCYL2, DEPDC4, AC008149.2, ELK3, LTA4H \n",
            "PC_ 4 \n",
            "Positive:  INTS6, INTS6-AS1, RNASEH2B, WDFY2, DHRS12, RNASEH2B-AS1, AL162377.1, DLEU7, ATP7B, DLEU1 \n",
            "\t   ALG11, TRIM13, DLEU2, NEK5, SPRYD7, NEK3, KPNA3, EBPL, THSD1, ARL11 \n",
            "\t   VPS36, RCBTB1, PHF11, AL359513.1, TMEM222, SETDB2, WDTC1, SYTL1, SLC9A1, WASF2 \n",
            "Negative:  ZNF532, AC104365.1, SEC11C, LMAN1, MALT1, PMAIP1, NEDD4L, RNF152, PIGN, ATP8B1 \n",
            "\t   RELCH, AC027097.2, TNFRSF11A, AC027097.1, ZCCHC2, PHLPP1, NARS, BCL2, FECH, KDSR \n",
            "\t   AC036176.1, WDR7, VPS4B, TXNL1, SERPINB2, TCF4-AS2, SERPINB10, SERPINB8, TCF4-AS1, AC090348.1 \n",
            "PC_ 5 \n",
            "Positive:  MED18, SESN2, PHACTR4, ATP5IF1, RCC1, DNAJC8, SNHG3, PTAFR, TRNAU1AP, EYA3 \n",
            "\t   SNHG12, XKR8, TAF12, AL512288.1, AL360012.1, GMEB1, SMPDL3B, YTHDF2, RPA2, EPB41 \n",
            "\t   THEMIS2, SRSF4, MECR, PPP1R8, MATN1, MATN1-AS1, STX12, AL137857.1, LAPTM5, FAM76A \n",
            "Negative:  LARP1B, PGRMC2, ABHD18, LINC02615, JADE1, MFSD8, SCLT1, PLK4, INTU, C4orf33 \n",
            "\t   FAT4, ANKRD50, AC096711.2, LINC01091, AC093916.1, SPRY1, SPATA5, PCDH18, NUDT6, SLC7A11 \n",
            "\t   KIAA1109, BBS7, CCNA2, AC109927.1, EXOSC9, ANXA5, NOCT, AC073475.1, MAD2L1, ELF2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:37] define_signif_tumor_subclusters(), tumor: prekursory erytroidalne i megakariocytarne\n",
            "INFO [2025-06-10 10:30:37] Less cells in group prekursory erytroidalne i megakariocytarne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 10:30:37] define_signif_tumor_subclusters(), tumor: wczesne komórki erytroidalne\n",
            "INFO [2025-06-10 10:30:37] Setting auto leiden resolution for wczesne komórki erytroidalne to 0.274587\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  KLHL5, TMEM156, FAM114A1, TLR6, TLR1, KLF3, KLF3-AS1, TBC1D1, PGM2, RELL1 \n",
            "\t   C4orf19, ARAP2, PRELID2, AC091887.1, AC137770.1, LARS, YIPF5, HMHB1, RBM27, LINC02506 \n",
            "\t   AC008696.2, TCERG1, NR3C1, WBP4, ELF1, PPP2R2B, AL354696.2, KBTBD7, STIM2, MTRF1 \n",
            "Negative:  EML3, B3GAT3, MTA2, GANAB, INTS5, TUT1, EEF1G, AHNAK, ASRGL1, INCENP \n",
            "\t   UBAC1, FTH1, NACC2, AL355574.1, TMEM250, CAMSAP1, LHX3, MRPS2, BEST1, FCN1 \n",
            "\t   QSOX2, GPSM1, COL5A1, CARD9, RXRA, SARDH, VAV2, WDR5, SLC2A6, FUT7 \n",
            "PC_ 2 \n",
            "Positive:  C3orf20, FGD5, CCDC174, SLC6A6, FGD5-AS1, LSM3, NR2C2, XPC, MRPS25, AC093495.1 \n",
            "\t   RBSN, TMEM43, CHCHD4, CAPN7, HDAC11, SH3BP5-AS1, RNF213-AS1, BAIAP2-DT, CHMP6, ENDOV \n",
            "\t   BAIAP2, RNF213, RPTOR, AATK, SLC26A11, CEP131, TEPSIN, SGSH, ACTG1, BAHCC1 \n",
            "Negative:  PCBD2, C5orf66, TXNDC15, H2AFY, C5orf24, SLC25A48, DDX46, TGFBI, CAMLG, SMAD5 \n",
            "\t   SPOCK1, SEC24A, KLHL3, SAR1B, HNRNPA0, JADE2, AC106791.1, CDKN2AIPNL, PKD2L2, AC109454.2 \n",
            "\t   UBE2B, FAM13B, CDKL3, BRD8, PPP2CA, KIF20A, SKP1, CDC23, TCF7, CDC25C \n",
            "PC_ 3 \n",
            "Positive:  ATG12, TMED7, AC010226.1, AP3S1, TICAM2, COMMD10, FEM1C, SEMA6A, CCDC112, SEMA6A-AS1 \n",
            "\t   PGGT1B, LINC02147, DTWD2, YTHDC2, DMXL1, MCC, TNFAIP8, DCP2, HSD17B4, REEP5 \n",
            "\t   AC008574.1, PRR16, SRP19, APC, SRFBP1, EPB41L4A-AS1, EPB41L4A, LOX, NREP, STARD4-AS1 \n",
            "Negative:  LINC00467, SLC30A1, RCOR3, TRAF5, LPGAT1, NEK2, KCNH1, INTS7, DTL, PPP2R5A \n",
            "\t   HHAT, AL360091.3, UTP25, PACC1, NENF, TRAF3IP3, ATF3, G0S2, NSL1, TATDN3 \n",
            "\t   PLXNA2, FLVCR1-DT, FLVCR1, LINC02767, VASH2, ANGEL2, CD34, RPS6KC1, MIR29B2CHG, SMYD2 \n",
            "PC_ 4 \n",
            "Positive:  H3F3B, MYO15B, RECQL5, LLGL2, GALK1, SMIM5, UNK, SAP30BP, UNC13D, TSEN54 \n",
            "\t   WBP2, TMEM94, TRIM65, MRPL38, FBF1, AC011933.2, ACOX1, UBE2O, TEN1, SPHK1 \n",
            "\t   SRP68, RHBDF2, PRPSAP1, SNHG16, EXOC7, UBALD2, RNF157, QRICH2, AC015802.6, AC011933.4 \n",
            "Negative:  FAM210A, AP005131.3, AP005131.1, TCEA1, LYPLA1, RGS20, AP005131.6, MRPL15, ATP6V1H, AP005131.5 \n",
            "\t   AC022034.3, LDLRAD4, RB1CC1, CEP192, ST18, AP002449.1, AC064807.1, AC103769.1, PCMTD1, PXDNL \n",
            "\t   UBE2V2, MCM4, PRKDC, CEBPD, SPIDR, HGSNAT, POMK, FNTA, HOOK3, RNF170 \n",
            "PC_ 5 \n",
            "Positive:  AQP9, ALDH1A2, POLR2M, LINC00926, TCF12, AC090517.5, ZNF280D, MNS1, TEX9, RFX7 \n",
            "\t   NEDD4, DNAAF4, CCPG1, PIGB, PIGBOS1, CIAO2A, AC018926.2, SNX1, RAB27A, SNX22 \n",
            "\t   RSL24D1, PPIB, AC011912.1, CSNK1G1, PCLAF, FAM214A, TRIP4, AC025917.1, ZNF609, ARPP19 \n",
            "Negative:  CNR2, FUCA1, HMGCL, NCMAP-DT, PNRC2, GALE, RCAN3, NCMAP, NIPAL3, STPG1 \n",
            "\t   IFNLR1, RCAN3AS, LYPLA2, SRSF10, MDS2, RPL11, ELOA-AS1, PITHD1, ELOA, ID3 \n",
            "\t   E2F2, ASAP3, CDC42, CDC42-AS1, ZNF436-AS1, CDC42-IT1, HNRNPR, ZBTB40, LUZP1, LINC00339 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:30:38] define_signif_tumor_subclusters(), tumor: control\n",
            "INFO [2025-06-10 10:30:39] Setting auto leiden resolution for control to 0.0057241\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  IFT81, C12orf76, ANKRD13A, GIT2, TCHP, GLTP, MMAB, UBE3B, KCTD10, MYO1H \n",
            "\t   ACACB, UNG, ALKBH2, USP30, SSH1, CORO1C, AC007569.1, SELPLG, ISCU, AC008119.1 \n",
            "\t   FAM3C, AASS, NDUFA5, LMOD2, WASL, AC006333.2, POT1, POT1-AS1, ZNF800, GCC1 \n",
            "Negative:  SLX4, DNASE1, NLRC3, TRAP1, CLUAP1, CREBBP, ADCY9, NAA60, TFAP4, ZNF174 \n",
            "\t   PAM16, ZSCAN32, CORO7, VASN, AC025283.2, DNAJA3, LRRC37A2, LRRC37A, ARL17A, ARL17B \n",
            "\t   ZNF75A, NMRAL1, FAM215B, KANSL1-AS1, NSF, KANSL1, MAPT, HMOX2, AC091132.2, GOSR2 \n",
            "PC_ 2 \n",
            "Positive:  MIA3, AIDA, TAF1A, BROX, DUSP10, FAM177B, HLX, AL392172.1, HLX-AS1, MARC1 \n",
            "\t   DISP1, RAB3GAP2, TLR5, IARS2, CAPN2, BPNT1, EPRS, TP53BP2, FBXO28, LYPLAL1 \n",
            "\t   DEGS1, C1orf143, NVL, TGFB2, RRP15, CNIH4, WDR26, GPATCH2, CNIH3, DNAH14 \n",
            "Negative:  TMEM123, BIRC2, DCUN1D5, DYNC2H1, BIRC3, AP002989.1, CEP126, PDGFD, ARHGAP42, CASP4 \n",
            "\t   JRKL, CCDC82, AP000779.1, CASP1, MAML2, CARD16, MTMR2, MSANTD4, CEP57, KBTBD3 \n",
            "\t   AASDHPPT, FAM76B, AP000766.1, SESN3, FOXP1, CWF19L2, FOXP1-AS1, MITF, AP000787.1, AC097634.2 \n",
            "PC_ 3 \n",
            "Positive:  RECQL4, TMSB10, TRABD2A, SUCLG1, AC022210.2, AC073091.3, MFSD3, KCMF1, TGOLN2, RETSAT \n",
            "\t   ELMOD3, CAPG, PPP1R16A, PARTICL, MAT2A, GGCX, VAMP8, VAMP5, CYHR1, RNF181 \n",
            "\t   USP39, C2orf68, ATOH8, GNLY, ST3GAL5, POLR1A, PTCD3, TONSL, IMMT, MRPL35 \n",
            "Negative:  UBE2H, AC073320.1, NRF1, ZC3HC1, KLHDC10, AC078846.1, TMEM209, STRIP2, AC087071.2, AHCYL2 \n",
            "\t   CEP41, PODXL, PLXNA4, MEST, CHCHD3, MKLN1, SMO, AC009365.4, LINC-PINT, COPG2 \n",
            "\t   AC058791.1, EXOC4, AUXG01000058.1, AC016831.5, TSPAN33, LRGUK, AC016831.1, AC016831.7, SLC35B4, LINC00513 \n",
            "PC_ 4 \n",
            "Positive:  EFHD2, FHAD1, CASP9, KAZN, DNAJC16, PRDM2, AGMAT, DHRS3, DDI2, PLEKHM2 \n",
            "\t   AL450998.3, SLC25A34, SLC25A34-AS1, TMEM82, UQCRHL, AL450998.2, SPEN, ZBTB17, SIRT7, AC145207.5 \n",
            "\t   PCYT2, ANAPC11, MAFG, ALYREF, PYCR1, ASPSCR1, CENPX, LRRC45, DCXR, GPS1 \n",
            "Negative:  PARP1, LIN9, ACBD3-AS1, ITPKB, ACBD3, LINC01703, AL512343.2, SDE2, PYCR2, COQ8A \n",
            "\t   H3F3A, TMEM63A, AL591895.1, EPHX1, SRP9, CDC42BPA, ENAH, ZNF678, LBR, DNAH14 \n",
            "\t   SNAP47, MROH7, TTC4, ACOT11, SSBP3, TCEANC2, JMJD4, CYB5RL, CNIH3, MRPL37 \n",
            "PC_ 5 \n",
            "Positive:  REV1, AFF3, EIF5B, TXNDC9, MRPL30, MITD1, LIPT1, CHST10, AC079447.1, TSGA10 \n",
            "\t   LINC02611, PDCL3, MGAT4A, NPAS2, UNC50, COA5, RPL31, INPP4A, TBC1D8, TMEM131 \n",
            "\t   CNOT11, ZAP70, C2orf92, RNF149, ACTR1B, CREG2, RFX8, COX5B, MAP4K4, ANKRD36B \n",
            "Negative:  LINC00937, CLEC4D, AC092745.5, CLEC4E, CLEC4A, NECAP1, FAM66C, C3AR1, FOXJ2, AC092490.2 \n",
            "\t   SLC2A3, RIMKLB, SLC2A14, A2ML1-AS1, CLEC4C, PHC1, CD163, M6PR, KLRG1, CD163L1 \n",
            "\t   LINC00612, PEX5, A2M-AS1, A2M, PZP, AC092821.3, KLRB1, AC010186.3, CLEC2D, CLECL1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 10:33:35] ::plot_cnv:Start\n",
            "INFO [2025-06-10 10:33:35] ::plot_cnv:Current data dimensions (r,c)=12217,14407 Total=176297617.949414 Min=0.873826021269021 Max=1.46428206966171.\n",
            "INFO [2025-06-10 10:33:36] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 10:33:41] plot_cnv(): auto thresholding at: (0.912757 , 1.090508)\n",
            "INFO [2025-06-10 10:33:56] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 10:33:56] Observation data size: Cells= 9501 Genes= 12217\n",
            "INFO [2025-06-10 10:33:56] clustering observations via method: ward.D\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(1) is 34\n",
            "INFO [2025-06-10 10:33:56] group size being clustered:  34,12217\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(2) is 2\n",
            "INFO [2025-06-10 10:33:56] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(3) is 3\n",
            "INFO [2025-06-10 10:33:56] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(4) is 26\n",
            "INFO [2025-06-10 10:33:56] group size being clustered:  26,12217\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(5) is 1\n",
            "INFO [2025-06-10 10:33:56] Skipping group: 5, since less than 2 entries\n",
            "INFO [2025-06-10 10:33:56] Number of cells in group(6) is 269\n",
            "INFO [2025-06-10 10:33:56] group size being clustered:  269,12217\n",
            "INFO [2025-06-10 10:33:57] Number of cells in group(7) is 259\n",
            "INFO [2025-06-10 10:33:57] group size being clustered:  259,12217\n",
            "INFO [2025-06-10 10:33:58] Number of cells in group(8) is 210\n",
            "INFO [2025-06-10 10:33:58] group size being clustered:  210,12217\n",
            "INFO [2025-06-10 10:33:59] Number of cells in group(9) is 175\n",
            "INFO [2025-06-10 10:33:59] group size being clustered:  175,12217\n",
            "INFO [2025-06-10 10:33:59] Number of cells in group(10) is 143\n",
            "INFO [2025-06-10 10:33:59] group size being clustered:  143,12217\n",
            "INFO [2025-06-10 10:33:59] Number of cells in group(11) is 141\n",
            "INFO [2025-06-10 10:33:59] group size being clustered:  141,12217\n",
            "INFO [2025-06-10 10:34:00] Number of cells in group(12) is 132\n",
            "INFO [2025-06-10 10:34:00] group size being clustered:  132,12217\n",
            "INFO [2025-06-10 10:34:00] Number of cells in group(13) is 128\n",
            "INFO [2025-06-10 10:34:00] group size being clustered:  128,12217\n",
            "INFO [2025-06-10 10:34:00] Number of cells in group(14) is 117\n",
            "INFO [2025-06-10 10:34:00] group size being clustered:  117,12217\n",
            "INFO [2025-06-10 10:34:00] Number of cells in group(15) is 116\n",
            "INFO [2025-06-10 10:34:00] group size being clustered:  116,12217\n",
            "INFO [2025-06-10 10:34:00] Number of cells in group(16) is 103\n",
            "INFO [2025-06-10 10:34:00] group size being clustered:  103,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(17) is 99\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  99,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(18) is 96\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  96,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(19) is 92\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  92,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(20) is 90\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  90,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(21) is 84\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  84,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(22) is 83\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  83,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(23) is 74\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  74,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(24) is 71\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  71,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(25) is 67\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  67,12217\n",
            "INFO [2025-06-10 10:34:01] Number of cells in group(26) is 66\n",
            "INFO [2025-06-10 10:34:01] group size being clustered:  66,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(27) is 63\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  63,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(28) is 60\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  60,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(29) is 57\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  57,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(30) is 48\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  48,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(31) is 45\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(32) is 45\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(33) is 45\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(34) is 45\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(35) is 40\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  40,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(36) is 36\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  36,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(37) is 32\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  32,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(38) is 32\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  32,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(39) is 32\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  32,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(40) is 31\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  31,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(41) is 29\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  29,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(42) is 27\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  27,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(43) is 18\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  18,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(44) is 18\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  18,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(45) is 18\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  18,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(46) is 17\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  17,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(47) is 15\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  15,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(48) is 14\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  14,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(49) is 14\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  14,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(50) is 4\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  4,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(51) is 4\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  4,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(52) is 3\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(53) is 3\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(54) is 2\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(55) is 1\n",
            "INFO [2025-06-10 10:34:02] Skipping group: 55, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(56) is 1\n",
            "INFO [2025-06-10 10:34:02] Skipping group: 56, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(57) is 1\n",
            "INFO [2025-06-10 10:34:02] Skipping group: 57, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(58) is 116\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  116,12217\n",
            "INFO [2025-06-10 10:34:02] Number of cells in group(59) is 108\n",
            "INFO [2025-06-10 10:34:02] group size being clustered:  108,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(60) is 107\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  107,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(61) is 106\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  106,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(62) is 104\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  104,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(63) is 104\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  104,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(64) is 97\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  97,12217\n",
            "INFO [2025-06-10 10:34:03] Number of cells in group(65) is 97\n",
            "INFO [2025-06-10 10:34:03] group size being clustered:  97,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(66) is 95\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  95,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(67) is 92\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  92,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(68) is 90\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  90,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(69) is 88\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  88,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(70) is 88\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  88,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(71) is 87\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  87,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(72) is 87\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  87,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(73) is 83\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  83,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(74) is 81\n",
            "INFO [2025-06-10 10:34:04] group size being clustered:  81,12217\n",
            "INFO [2025-06-10 10:34:04] Number of cells in group(75) is 79\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  79,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(76) is 75\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  75,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(77) is 73\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  73,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(78) is 70\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  70,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(79) is 65\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  65,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(80) is 61\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  61,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(81) is 60\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  60,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(82) is 58\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  58,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(83) is 57\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  57,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(84) is 56\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  56,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(85) is 56\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  56,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(86) is 55\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  55,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(87) is 55\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  55,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(88) is 52\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  52,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(89) is 51\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  51,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(90) is 51\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  51,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(91) is 51\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  51,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(92) is 50\n",
            "INFO [2025-06-10 10:34:05] group size being clustered:  50,12217\n",
            "INFO [2025-06-10 10:34:05] Number of cells in group(93) is 50\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  50,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(94) is 48\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  48,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(95) is 45\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(96) is 44\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  44,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(97) is 42\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  42,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(98) is 39\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  39,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(99) is 37\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  37,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(100) is 35\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  35,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(101) is 32\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  32,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(102) is 32\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  32,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(103) is 30\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  30,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(104) is 28\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  28,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(105) is 26\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  26,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(106) is 25\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  25,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(107) is 24\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  24,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(108) is 23\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  23,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(109) is 23\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  23,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(110) is 23\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  23,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(111) is 22\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  22,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(112) is 21\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  21,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(113) is 4\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  4,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(114) is 2\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(115) is 146\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  146,12217\n",
            "INFO [2025-06-10 10:34:06] Number of cells in group(116) is 137\n",
            "INFO [2025-06-10 10:34:06] group size being clustered:  137,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(117) is 98\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  98,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(118) is 77\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  77,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(119) is 76\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  76,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(120) is 71\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  71,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(121) is 71\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  71,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(122) is 64\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  64,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(123) is 62\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  62,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(124) is 56\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  56,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(125) is 53\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  53,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(126) is 49\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  49,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(127) is 49\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  49,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(128) is 47\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  47,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(129) is 45\n",
            "INFO [2025-06-10 10:34:07] group size being clustered:  45,12217\n",
            "INFO [2025-06-10 10:34:07] Number of cells in group(130) is 44\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  44,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(131) is 43\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  43,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(132) is 42\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  42,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(133) is 41\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  41,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(134) is 40\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  40,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(135) is 39\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  39,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(136) is 37\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  37,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(137) is 37\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  37,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(138) is 34\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  34,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(139) is 33\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  33,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(140) is 29\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  29,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(141) is 28\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  28,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(142) is 27\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  27,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(143) is 26\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  26,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(144) is 25\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  25,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(145) is 23\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  23,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(146) is 19\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  19,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(147) is 19\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  19,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(148) is 17\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  17,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(149) is 17\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  17,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(150) is 11\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  11,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(151) is 10\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  10,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(152) is 2\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(153) is 1\n",
            "INFO [2025-06-10 10:34:08] Skipping group: 153, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(154) is 1\n",
            "INFO [2025-06-10 10:34:08] Skipping group: 154, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(155) is 54\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  54,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(156) is 36\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  36,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(157) is 34\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  34,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(158) is 26\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  26,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(159) is 23\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  23,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(160) is 20\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  20,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(161) is 3\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(162) is 2\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(163) is 1\n",
            "INFO [2025-06-10 10:34:08] Skipping group: 163, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(164) is 1\n",
            "INFO [2025-06-10 10:34:08] Skipping group: 164, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:08] Number of cells in group(165) is 138\n",
            "INFO [2025-06-10 10:34:08] group size being clustered:  138,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(166) is 69\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  69,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(167) is 62\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  62,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(168) is 36\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  36,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(169) is 35\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  35,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(170) is 28\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  28,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(171) is 28\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  28,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(172) is 27\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  27,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(173) is 17\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  17,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(174) is 13\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  13,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(175) is 13\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  13,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(176) is 8\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  8,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(177) is 7\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  7,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(178) is 4\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  4,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(179) is 2\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(180) is 1\n",
            "INFO [2025-06-10 10:34:09] Skipping group: 180, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(181) is 3\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(182) is 2\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(183) is 62\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  62,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(184) is 58\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  58,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(185) is 34\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  34,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(186) is 3\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  3,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(187) is 2\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(188) is 2\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(189) is 2\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  2,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(190) is 4\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  4,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(191) is 53\n",
            "INFO [2025-06-10 10:34:09] group size being clustered:  53,12217\n",
            "INFO [2025-06-10 10:34:09] Number of cells in group(192) is 1\n",
            "INFO [2025-06-10 10:34:09] Skipping group: 192, since less than 2 entries\n",
            "INFO [2025-06-10 10:34:11] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 10:34:11] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 10:34:11] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:34:11] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:34:44] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:34:44] Quantiles of plotted data range: 0.912756716170172,0.980100756875301,0.998798888504086,1.02059320989278,1.0905078532147\n",
            "INFO [2025-06-10 10:34:58] plot_cnv_references:Start\n",
            "INFO [2025-06-10 10:34:58] Reference data size: Cells= 4906 Genes= 12217\n",
            "INFO [2025-06-10 10:35:13] plot_cnv_references:Number reference groups= 48\n",
            "INFO [2025-06-10 10:35:13] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 10:35:30] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:35:30] Quantiles of plotted data range: 0.912756716170172,0.985276121496026,0.998795555814984,1.01352556078786,1.0905078532147\n",
            "INFO [2025-06-10 10:38:01] ::plot_cnv:Start\n",
            "INFO [2025-06-10 10:38:01] ::plot_cnv:Current data dimensions (r,c)=12217,14407 Total=176297617.949414 Min=0.873826021269021 Max=1.46428206966171.\n",
            "INFO [2025-06-10 10:38:01] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 10:38:07] plot_cnv(): auto thresholding at: (0.912757 , 1.090508)\n",
            "INFO [2025-06-10 10:38:13] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 10:38:13] Observation data size: Cells= 9501 Genes= 12217\n",
            "INFO [2025-06-10 10:38:15] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 10:38:15] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 10:38:16] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:38:16] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:38:50] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:38:50] Quantiles of plotted data range: 0.912756716170172,0.980100756875301,0.998798888504086,1.02059320989278,1.0905078532147\n",
            "INFO [2025-06-10 10:39:05] plot_cnv_references:Start\n",
            "INFO [2025-06-10 10:39:05] Reference data size: Cells= 4906 Genes= 12217\n",
            "INFO [2025-06-10 10:55:00] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 10:55:01] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 10:55:17] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:55:17] Quantiles of plotted data range: 0.912756716170172,0.985276121496026,0.998795555814984,1.01352556078786,1.0905078532147\n",
            "INFO [2025-06-10 10:55:22] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-06-10 10:55:22] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-06-10 10:55:22] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 10:55:24] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.966252909180011 - 1.03427542393777\n",
            "INFO [2025-06-10 10:56:46] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-06-10 10:56:47] ::plot_cnv:Start\n",
            "INFO [2025-06-10 10:56:47] ::plot_cnv:Current data dimensions (r,c)=12217,14407 Total=176542493.54218 Min=0.873826021269021 Max=1.46428206966171.\n",
            "INFO [2025-06-10 10:56:48] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 10:56:52] plot_cnv(): auto thresholding at: (0.909492 , 1.090508)\n",
            "INFO [2025-06-10 10:56:58] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 10:56:58] Observation data size: Cells= 9501 Genes= 12217\n",
            "INFO [2025-06-10 10:57:01] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 10:57:01] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 10:57:01] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:57:01] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 10:57:33] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 10:57:33] Quantiles of plotted data range: 0.9094921467853,1.00026416655889,1.00026416655889,1.00026416655889,1.0905078532147\n",
            "INFO [2025-06-10 10:57:46] plot_cnv_references:Start\n",
            "INFO [2025-06-10 10:57:46] Reference data size: Cells= 4906 Genes= 12217\n",
            "INFO [2025-06-10 11:13:38] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 11:13:39] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 11:13:54] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 11:13:54] Quantiles of plotted data range: 0.9094921467853,1.00026416655889,1.00026416655889,1.00026416655889,1.0905078532147\n"
          ]
        }
      ],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"1_vs_95\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Oo2JOxSKSzzj",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "26323ca9-06cf-4afb-b921-428c17b88e54"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:24:40] Parsing matrix: /content/1_and_99.matrix\n",
            "INFO [2025-06-10 11:25:43] Parsing gene order file: gene_pos_file.txt\n",
            "INFO [2025-06-10 11:25:43] Parsing cell annotations file: /content/1_vs_99.annotations.tsv\n",
            "INFO [2025-06-10 11:25:43] ::order_reduce:Start.\n",
            "INFO [2025-06-10 11:25:44] .order_reduce(): expr and order match.\n",
            "INFO [2025-06-10 11:25:45] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 36601,7764 Total=80363403 Min=0 Max=5566.\n",
            "INFO [2025-06-10 11:25:46] num genes removed taking into account provided gene ordering list: 1280 = 3.49717220840961% removed.\n",
            "INFO [2025-06-10 11:25:46] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
            "WARN [2025-06-10 11:25:49] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-06-10 11:25:54] validating infercnv_obj\n"
          ]
        }
      ],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/1_and_99.matrix\",\n",
        "                                    annotations_file=\"/content/1_vs_99.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kioSZzBzS2fS",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "42e36196-c230-4b4c-9733-1f4cc7b989c8",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:25:54] ::process_data:Start\n",
            "INFO [2025-06-10 11:25:54] Creating output path 1_vs_99\n",
            "INFO [2025-06-10 11:25:54] Checking for saved results.\n",
            "INFO [2025-06-10 11:25:54] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-06-10 11:26:30] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-06-10 11:26:30] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-06-10 11:26:31] Removing 18720 genes from matrix as below mean expr threshold: 0.01\n",
            "INFO [2025-06-10 11:26:32] validating infercnv_obj\n",
            "INFO [2025-06-10 11:26:32] There are 16601 genes and 7764 cells remaining in the expr matrix.\n",
            "INFO [2025-06-10 11:26:37] no genes removed due to min cells/gene filter\n",
            "INFO [2025-06-10 11:27:05] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-06-10 11:27:05] normalizing counts matrix by depth\n",
            "INFO [2025-06-10 11:27:09] Computed total sum normalization factor as median libsize: 7496.500000\n",
            "INFO [2025-06-10 11:27:40] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-06-10 11:27:40] transforming log2xplus1()\n",
            "INFO [2025-06-10 11:28:13] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-06-10 11:28:13] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 11:28:13] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 11:28:25] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 11:29:43] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-06-10 11:29:43] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-06-10 11:30:52] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-06-10 11:30:52] smooth_by_chromosome: chr: chr1\n",
            "INFO [2025-06-10 11:31:08] smooth_by_chromosome: chr: chr2\n",
            "INFO [2025-06-10 11:31:22] smooth_by_chromosome: chr: chr3\n",
            "INFO [2025-06-10 11:31:36] smooth_by_chromosome: chr: chr4\n",
            "INFO [2025-06-10 11:31:48] smooth_by_chromosome: chr: chr5\n",
            "INFO [2025-06-10 11:32:01] smooth_by_chromosome: chr: chr6\n",
            "INFO [2025-06-10 11:32:16] smooth_by_chromosome: chr: chr7\n",
            "INFO [2025-06-10 11:32:28] smooth_by_chromosome: chr: chr8\n",
            "INFO [2025-06-10 11:32:41] smooth_by_chromosome: chr: chr9\n",
            "INFO [2025-06-10 11:32:53] smooth_by_chromosome: chr: chr10\n",
            "INFO [2025-06-10 11:33:06] smooth_by_chromosome: chr: chr11\n",
            "INFO [2025-06-10 11:33:19] smooth_by_chromosome: chr: chr12\n",
            "INFO [2025-06-10 11:33:34] smooth_by_chromosome: chr: chr13\n",
            "INFO [2025-06-10 11:33:44] smooth_by_chromosome: chr: chr14\n",
            "INFO [2025-06-10 11:33:56] smooth_by_chromosome: chr: chr15\n",
            "INFO [2025-06-10 11:34:09] smooth_by_chromosome: chr: chr16\n",
            "INFO [2025-06-10 11:34:22] smooth_by_chromosome: chr: chr17\n",
            "INFO [2025-06-10 11:34:37] smooth_by_chromosome: chr: chr18\n",
            "INFO [2025-06-10 11:34:47] smooth_by_chromosome: chr: chr19\n",
            "INFO [2025-06-10 11:35:02] smooth_by_chromosome: chr: chr20\n",
            "INFO [2025-06-10 11:35:14] smooth_by_chromosome: chr: chr21\n",
            "INFO [2025-06-10 11:35:24] smooth_by_chromosome: chr: chr22\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: GL000194.1\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: GL000195.1\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: GL000218.1\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: GL000219.1\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: KI270711.1\n",
            "INFO [2025-06-10 11:35:35] smooth_by_chromosome: chr: KI270721.1\n",
            "INFO [2025-06-10 11:36:43] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-06-10 11:36:43] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-06-10 11:38:15] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-06-10 11:38:15] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-06-10 11:38:15] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 11:38:27] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-06-10 11:39:47] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-06-10 11:39:47] invert_log2(), computing 2^x\n",
            "INFO [2025-06-10 11:41:31] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-06-10 11:41:31] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-06-10 11:41:37] define_signif_tumor_subclusters(), tumor: aktywowane HSC\n",
            "INFO [2025-06-10 11:41:37] Less cells in group aktywowane HSC than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:37] define_signif_tumor_subclusters(), tumor: CD8 T-cells\n",
            "INFO [2025-06-10 11:41:37] Less cells in group CD8 T-cells than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:37] define_signif_tumor_subclusters(), tumor: GMP\n",
            "INFO [2025-06-10 11:41:37] Less cells in group GMP than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:37] define_signif_tumor_subclusters(), tumor: HSC\n",
            "INFO [2025-06-10 11:41:37] Setting auto leiden resolution for HSC to 0.0394946\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  RPL18A, JAK3, FCHO1, AC008761.1, CCDC124, MAP1S, COLGALT1, KCNN1, NIBAN3, ARRDC2 \n",
            "\t   PGLS, AC010618.3, IL12RB1, SLC27A1, MAST3, AC010319.4, PIK3R2, MVB12A, BISPR, IFI30 \n",
            "\t   BST2, MPV17L2, GTPBP3, PDE4C, ANO8, AC008397.1, IQCN, DDA1, JUND, MRPL34 \n",
            "Negative:  TMEM232, SLC25A46, WDR36, CAMK4, AC010275.1, STARD4, STARD4-AS1, NREP, EPB41L4A, ANKRD49 \n",
            "\t   MRE11, EPB41L4A-AS1, AP002784.1, AL442003.1, APC, PANX1, WASHC2A, SRP19, MED17, SGMS1 \n",
            "\t   REEP5, AC069547.2, C11orf54, SGMS1-AS1, DCP2, PRKG1, MCC, LINC01806, CSTF2T, PSMA3 \n",
            "PC_ 2 \n",
            "Positive:  AC008569.2, GIPC1, PKN1, DNAJB1, TECR, DDX39A, NDUFB7, AC008569.1, CLEC17A, ADGRE3 \n",
            "\t   ZNF333, ADGRE5, ADGRE2, OR7C1, SLC1A6, ILVBL, LINC01841, BRD4, AKAP8, AKAP8L \n",
            "\t   WIZ, RASAL3, AC011509.2, PGLYRP2, CYP4F22, UCA1, ADGRL1, TPM4, RAB8A, AC008894.2 \n",
            "Negative:  FRMD6-AS1, FRMD6, GNG2, TMX1, RTRAF, NID2, TRIM9, PTGDR, ABHD12B, PTGER2 \n",
            "\t   TXNDC16, PYGL, GPR137C, AL606834.2, ERO1A, NIN, AL133453.1, AL606834.1, PSMC6, SAV1 \n",
            "\t   STYX, ATL1, GNPNAT1, AL118556.1, AL139317.5, MAP4K5, FERMT2, CDKL1, DDHD1, DMAC2L \n",
            "PC_ 3 \n",
            "Positive:  CCDC115, MZT2B, SMPD4, HS6ST1, UGGT1, BCL2, KDSR, AC036176.1, VPS4B, PHLPP1 \n",
            "\t   SERPINB2, ZCCHC2, SERPINB10, TNFRSF11A, SERPINB8, AC027514.1, LINC01924, RELCH, PIGN, CDH7 \n",
            "\t   RNF152, PMAIP1, LMAN1, SEC11C, DSEL, ZNF532, AC104365.1, AC114689.3, MALT1, NEDD4L \n",
            "Negative:  LINC01588, ARF6, VCPKMT, AL627171.1, SOS2, L2HGDH, AL627171.2, DMAC2L, NEMF, CDKL1 \n",
            "\t   KLHDC2, MAP4K5, AL118556.1, KLHDC1, ATL1, POLE2, SAV1, AL606834.1, DNAAF2, NIN \n",
            "\t   MGAT2, AL606834.2, RPL36AL, PYGL, LRR1, ABHD12B, RPS29, TRIM9, MDGA2, TMX1 \n",
            "PC_ 4 \n",
            "Positive:  KCNK10, LINC01146, SPATA7, GPR65, PTPN21, GALC, ZC3H14, LINC02296, AL162171.3, LINC02316 \n",
            "\t   EML5, LINC02328, TTC8, SEL1L, FOXN3, STON2, AL357093.2, GTF2A1, AL356805.1, AC007262.2 \n",
            "\t   FOXN3-AS1, TSHR, FOXN3-AS2, CEP128, EFCAB11, DIO2-AS1, TDP1, NRXN3, KCNK13, ADCK1 \n",
            "Negative:  ARL6IP5, LMOD3, UBA3, FRMD4B, MITF, TMF1, FOXP1, FOXP1-AS1, AC109587.1, AC097634.2 \n",
            "\t   EOGT, FOXP1-IT1, TAFA1, AC097634.3, SUCLG2-AS1, SUCLG2, AC097634.1, KBTBD8, EIF4E3, LRIG1 \n",
            "\t   SLC25A26, GPR27, MAGI1, PROK2, PSMD6-AS1, LINC00877, PSMD6, PSMD6-AS2, LINC00870, ATXN7 \n",
            "PC_ 5 \n",
            "Positive:  ARL6IP5, LMOD3, UBA3, FRMD4B, TMF1, MITF, TRIP11, FOXP1, ATXN3, NDUFB1 \n",
            "\t   FBLN5, AC109587.1, FOXP1-AS1, CPSF2, TC2N, SLC24A4, AL121839.2, AC097634.2, EOGT, RIN3 \n",
            "\t   CATSPERB, LGMN, FOXP1-IT1, GOLGA5, PPP4R3A, AC097634.3, TAFA1, ITPK1, CCDC88C, MOAP1 \n",
            "Negative:  PPP2R2D, LINC02646, BNIP3, GLRX3, AL162274.2, C10orf143, JAKMIP3, LINC02666, STK32C, AL355531.1 \n",
            "\t   LRRC27, MGMT, PWWP2B, SLC35E1, AC008764.8, AC008764.6, AP1M1, CHERP, LINC01163, MED26 \n",
            "\t   FAM32A, C19orf44, AC020911.2, EPS15L1, HSH2D, AC008764.7, KLF2, AL451069.2, AC008894.2, MKI67 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:40] define_signif_tumor_subclusters(), tumor: komórki B pamięci\n",
            "INFO [2025-06-10 11:41:40] Setting auto leiden resolution for komórki B pamięci to 0.100576\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  GPR27, EIF4E3, PROK2, AC097634.1, LINC00877, AC097634.3, FOXP1-IT1, LINC00870, AC097634.2, RYBP \n",
            "\t   FOXP1-AS1, FOXP1, MITF, SHQ1, FRMD4B, LMOD3, GXYLT2, ARL6IP5, UBA3, TMF1 \n",
            "\t   PPP4R2, AC109587.1, EOGT, EBLN2, TAFA1, SUCLG2-AS1, PDZRN3, CNTN3, SUCLG2, LINC02018 \n",
            "Negative:  AKT1S1, TBC1D17, IL4I1, NUP62, ATF5, PLCB3, PPP1R14B-AS1, PPP1R14B, BAD, GPR137 \n",
            "\t   FKBP2, ESRRA, TRMT112, VEGFB, DNAJC4, PRDX5, AP001453.1, CCDC88B, NUDT22, RPS6KA4 \n",
            "\t   TRPT1, AP005273.2, VRK3, FERMT3, NRXN2, STIP1, RASGRP2, FLRT1, CCDC124, KCNN1 \n",
            "PC_ 2 \n",
            "Positive:  LRMP, BCAT1, SOX5, AC023510.1, CASC1, AC084816.1, ETFRF1, KRAS, ETNK1, AC092794.1 \n",
            "\t   C2CD5, AC087239.1, AC053513.1, ST8SIA1, CMAS, RASSF8-AS1, ABCC9, LDHB, RASSF8, GOLT1B \n",
            "\t   RECQL, SSPN, PYROXD1, AC022509.2, PDE3A, AEBP2, AC024145.1, AC092828.1, PLEKHA5, ITPR2 \n",
            "Negative:  POLR3B, TCP11L2, AC079385.1, RFX4, CKAP4, RIC8B, AC007541.1, C12orf75, TMEM263, APPL2 \n",
            "\t   MTERF2, WASHC4, CRY1, ALDH1L2, AC078929.1, C12orf45, BTBD11, SLC41A2, PWP1, CHST11 \n",
            "\t   PRDM4, EID3, AC007622.2, TXNRD1, NFYB, CMKLR1, HCFC2, FICD, GLT8D2, SART3 \n",
            "PC_ 3 \n",
            "Positive:  CKAP4, TCP11L2, C12orf75, POLR3B, APPL2, AC079385.1, WASHC4, RFX4, RIC8B, ALDH1L2 \n",
            "\t   C12orf45, AC007541.1, SLC41A2, TMEM263, CHST11, MTERF2, CRY1, EID3, AC078929.1, BTBD11 \n",
            "\t   PWP1, TXNRD1, PRDM4, NFYB, AC007622.2, HCFC2, CMKLR1, FICD, GLT8D2, SART3 \n",
            "Negative:  RSPH4A, RWDD1, ZUP1, KPNA5, CALHM6, DCBLD1, TSPYL1, GOPC, DSE, NUS1 \n",
            "\t   TSPYL4, SLC35F1, COL10A1, CEP85L, NT5DC1, PLN, HDAC2-AS2, HDAC2, Z99496.2, LINC01268 \n",
            "\t   MCM9, MARCKS, ASF1A, LINC02541, AL137009.1, AL360015.1, FAM184A, FAM229B, MAN1A1, TUBE1 \n",
            "PC_ 4 \n",
            "Positive:  DDHD1, FERMT2, AL356020.1, AL139317.5, AL365295.1, AL365295.2, GNPNAT1, CDKN3, STYX, CNIH1 \n",
            "\t   PSMC6, AL133453.1, GMFB, ERO1A, GPR137C, CGRRF1, TXNDC16, SAMD4A, PTGER2, PTGDR \n",
            "\t   GCH1, NID2, WDHD1, RTRAF, SOCS4, GNG2, MAPK1IP1L, FRMD6-AS1, LGALS3, FRMD6 \n",
            "Negative:  MPPED2, ARL14EP, DNAJC24, METTL15, IMMP1L, ELP4, KIF18A, AC131571.1, BDNF-AS, AL035078.1 \n",
            "\t   LIN7C, LGR4, RCN1, CCDC34, AL078612.1, LUZP2, EIF3M, LINC02726, CCDC73, LINC02718 \n",
            "\t   PRRG4, QSER1, AC006299.1, DEPDC7, SVIP, TCP11L1, GAS2, LINC00294, FANCF, CSTF3 \n",
            "PC_ 5 \n",
            "Positive:  ITGB1BP1, CPSF3, ASAP2, IAH1, MBOAT2, KIDINS220, ADAM17, ID2, AC080162.1, ID2-AS1 \n",
            "\t   LINC01814, AC073195.1, LINC00299, YWHAQ, AC007463.1, TAF1B, LINC01871, AC010969.2, AC068481.1, GRHL1 \n",
            "\t   RNF144A, AC104794.3, RSAD2, KLF11, CMPK2, RRM2, NRIR, HPCAL1, LINC00487, ODC1 \n",
            "Negative:  LIN7C, BDNF-AS, LGR4, KIF18A, CCDC34, METTL15, LUZP2, ARL14EP, LINC02726, LINC02718 \n",
            "\t   MPPED2, AC006299.1, DNAJC24, SVIP, GAS2, IMMP1L, FANCF, ANO5, ELP4, NELL1 \n",
            "\t   AC131571.1, PRMT3, AL035078.1, HTATIP2, RCN1, NAV2, AL078612.1, E2F8, EIF3M, CSRP3-AS1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:42] define_signif_tumor_subclusters(), tumor: Komórki blastyczne (preB/CLP)\n",
            "INFO [2025-06-10 11:41:42] Less cells in group Komórki blastyczne (preB/CLP) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:42] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 1 (preB)\n",
            "INFO [2025-06-10 11:41:42] Setting auto leiden resolution for Komórki blastyczne 1 (preB) to 0.550787\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  WARS, WDR25, AL157871.4, MEG3, SLC25A29, MEG8, AL157871.6, AL132709.8, YY1, LINC02320 \n",
            "\t   AL133523.1, DEGS2, PPP2R5C, AL137779.2, AL137779.1, AL118558.3, DYNC1H1, AL157912.1, HSP90AA1, WDR20 \n",
            "\t   VRK1, PAPOLA, MOK, AL163051.1, AK7, ZNF839, GSKIP, ATG2B, CINP, C14orf132 \n",
            "Negative:  TAFA1, EOGT, SUCLG2-AS1, AC109587.1, FRMD4B, LMOD3, ARL6IP5, SUCLG2, MITF, TMF1 \n",
            "\t   UBA3, FOXP1, KBTBD8, FOXP1-AS1, LRIG1, SLC25A26, AC097634.2, MAGI1, FOXP1-IT1, PSMD6-AS1 \n",
            "\t   AC097634.3, PSMD6, AC097634.1, PSMD6-AS2, EIF4E3, GPR27, ATXN7, PROK2, AC104162.2, LINC00877 \n",
            "PC_ 2 \n",
            "Positive:  ERP29, MAPKAPK5, TMEM116, NAA25, MAPKAPK5-AS1, ALDH2, AC073575.1, ACAD10, TRAFD1, BRAP \n",
            "\t   ATXN2, HECTD4, RPL6, PTPN11, RPH3A, OAS1, AC004551.1, OAS3, OAS2, DTX1 \n",
            "\t   RASAL1, MROH9, FMO4, CFAP73, BX284613.2, C2orf42, TIA1, GORAB, PRRC2C, PCYOX1 \n",
            "Negative:  LAT2, RFC2, CLIP2, EIF4H, GTF2IRD1, CASTOR2, NCF1, GTF2IRD2, AC211433.1, GTF2I \n",
            "\t   RCC1L, LIMK1, GTF2IRD2B, ELN, AC211486.5, AC211486.2, ABHD11, TRIM73, BUD23, POM121C \n",
            "\t   DNAJC30, SPDYE5, TBL2, HIP1, BCL7B, RHBDD2, BAZ1B, POR, NSUN5, AC211476.6 \n",
            "PC_ 3 \n",
            "Positive:  AC008764.6, CHERP, C19orf44, EPS15L1, KLF2, AC020911.2, AP1M1, RNF145, FAM32A, ZNRF1 \n",
            "\t   AC099508.2, WDR59, ZFP1, CFDP1, LINC02202, TMEM170A, TMEM231, FA2H, HSH2D, GABARAPL2 \n",
            "\t   ADAT1, MSL2, AC136424.2, MLKL, KARS, AC008894.2, TERF2IP, PPP2R3A, RFWD3, AC025287.4 \n",
            "Negative:  ARHGAP1, ZNF408, ATG13, CKAP5, HARBI1, AC127035.1, LRP4-AS1, AMBRA1, MDK, LRP4 \n",
            "\t   DGKZ, PHF21A, C11orf49, LARGE2, PEX16, AC090589.2, MAPK8IP1, CRY2, ARFGAP2, SLC35C1 \n",
            "\t   PRDM11, DDB2, TP53I11, TSPAN18, ACP2, CD82, EXT2, ACCS, ALKBH3, AC087521.2 \n",
            "PC_ 4 \n",
            "Positive:  DENR, CCDC62, HIP1R, HCAR3, VPS37B, KNTC1, ABCB9, RSRC2, OGFOD2, ARL6IP4 \n",
            "\t   ZCCHC8, PITPNM2, MPHOSPH9, CLIP1-AS1, C12orf65, CDK2AP1, SBNO1, SBNO1-AS1, CLIP1, KMT5A \n",
            "\t   RILPL2, SNRNP35, VPS33A, OSER1, TOX2, OSER1-DT, MYBL2, RILPL1, IFT52, GDAP1L1 \n",
            "Negative:  SIAH3, COG3, ZC3H13, SLC25A30, CPB2-AS1, TPT1-AS1, AL138963.4, TPT1, GTF2F2, GPALPP1 \n",
            "\t   AL359706.1, NUFIP1, AL356515.1, LINC00407, TSC22D1-AS1, TSC22D1, TUSC8, SERP2, SMIM2-AS1, LACC1 \n",
            "\t   CCDC122, ENOX1, DNAJC15, EPSTI1, AKAP11, DGKH, VWA8, RGCC, NAA16, MTRF1 \n",
            "PC_ 5 \n",
            "Positive:  FEM1A, AC005523.2, ARRDC5, PLIN3, TICAM1, DPP9-AS1, GTF2F1, UHRF1, ALKBH7, KDM4B \n",
            "\t   DPP9, CLPP, MLLT1, PTPRS, ACSBG2, MYDGF, TNFAIP8L1, AC011444.3, TINCR, AC011444.2 \n",
            "\t   SEMA6B, SAFB2, RFX2, LRG1, SAFB, AC011444.1, PLIN5, RANBP3, RPL36, CAPS \n",
            "Negative:  SEC63, AL024507.2, SCML4, OSTM1, SOBP, SNX3, PDSS2, AFG1L, BEND3, AL139106.1 \n",
            "\t   MTRES1, CD24, LINC02532, QRSL1, RTN4IP1, CRYBG1, ATG5, PRDM1, AL591518.1, AL096794.1 \n",
            "\t   PREP, BVES-AS1, BVES, HACE1, PNISR, AL513550.1, GRIK2, USP45, TSTD3, AL133338.1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:43] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 2 (preB)\n",
            "INFO [2025-06-10 11:41:43] Less cells in group Komórki blastyczne 2 (preB) than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:43] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 3 (preB)\n",
            "INFO [2025-06-10 11:41:43] Setting auto leiden resolution for Komórki blastyczne 3 (preB) to 0.02847\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  CATSPERD, VMAC, DUS3L, NDUFA11, LONP1, AC104532.2, CAPS, HSD11B1L, RANBP3, MICOS13 \n",
            "\t   AC011444.1, RPL36, RFX2, SAFB, SAFB2, AC011444.2, TINCR, PTPRS, AC011444.3, KDM4B \n",
            "\t   ARRDC5, UHRF1, ACSBG2, MLLT1, CLPP, ALKBH7, GTF2F1, KHSRP, SLC25A23, DENND1C \n",
            "Negative:  ZNF608, LINC02240, CSNK1G3, AC116362.1, GRAMD2B, CEP120, SNX24, SNX2, AC010255.1, AC093535.1 \n",
            "\t   LOX, ALDH7A1, PHAX, SRFBP1, LMNB1-DT, AC008568.1, LMNB1, PRR16, MARCH3, AC113418.1 \n",
            "\t   C5orf63, AC011416.3, PRRC1, AC008574.1, CCDC192, LINC01184, SLC12A2, HSD17B4, FBN2, SLC27A6 \n",
            "PC_ 2 \n",
            "Positive:  CALCB, CYP2R1, INSC, PDE3B, PSMA1, SOX6, COPB1, AC103794.1, RRAS2, C11orf58 \n",
            "\t   SPON1-AS1, PLEKHA7, SPON1, RPS13, PIK3C2A, NUCB2, NCR3LG1, FAR1-IT1, KCNC1, SERGEF \n",
            "\t   FAR1, TPH1, BTBD10, SAAL1, ARNTL, HPS5, LINC00958, GTF2H1, TEAD1, MICAL2 \n",
            "Negative:  SEC24A, SAR1B, CAMLG, DDX46, JADE2, C5orf24, CDKN2AIPNL, TXNDC15, PCBD2, AC109454.2 \n",
            "\t   C5orf66, UBE2B, H2AFY, CDKL3, CXCL14, PPP2CA, SLC25A48, SKP1, TGFBI, TCF7 \n",
            "\t   SMAD5, AC008608.2, SPOCK1, VDAC1, KLHL3, C5orf15, HNRNPA0, FSTL4, AC106791.1, HSPA4 \n",
            "PC_ 3 \n",
            "Positive:  ZDHHC4, GRID2IP, C7orf26, KDELR2, ZNF853, DAGLB, ZNF316, RAC1, CYTH3, AC073343.2 \n",
            "\t   USP42, ZNF12, EIF2AK1, RSPH10B2, AIMP2, CCZ1B, PMS2, CCZ1, AC079804.3, OCM \n",
            "\t   C1GALT1, RNF216, AC005532.1, FSCN1, AC004982.2, ACTB, FBXL18, MIOS, AC092171.5, RPA3 \n",
            "Negative:  MAP3K6, WASF2, SYTL1, TMEM222, FO393419.2, WDTC1, SLC9A1, BX293535.1, NUDC, GPATCH3 \n",
            "\t   AHDC1, FGR, GPN2, IFI6, ZDHHC18, FAM76A, PIGV, STX12, ARID1A, PPP1R8 \n",
            "\t   RPS6KA1, THEMIS2, HMGN2, RPA2, AL513365.2, SMPDL3B, DHDDS, AL512288.1, CRYBG2, CD52 \n",
            "PC_ 4 \n",
            "Positive:  AL513365.2, HMGN2, DHDDS, RPS6KA1, CRYBG2, GPATCH3, PIGV, ZDHHC18, GPN2, ARID1A \n",
            "\t   NUDC, CD52, SLC9A1, WDTC1, UBXN11, TMEM222, SH3BGRL3, SYTL1, MAP3K6, CEP85 \n",
            "\t   WASF2, CNKSR1, FO393419.2, BX293535.1, ZNF593, AHDC1, FAM110D, FGR, IFI6, PDIK1L \n",
            "Negative:  HCFC2, GLT8D2, C12orf73, TDG, HSP90B1, NT5DC3, C12orf42, HELLPAR, PARPBP, NUP37 \n",
            "\t   AC079907.1, WASHC3, DRAM1, GNPTAB, SYCP3, CHPT1, ARL1, UTP20, GAS2L3, SCYL2 \n",
            "\t   DEPDC4, ACTR6, UHRF1BP1L, ANKS1B, APAF1, IKBIP, SLC25A3, TMPO, TMPO-AS1, LINC02453 \n",
            "PC_ 5 \n",
            "Positive:  INSC, SOX6, CALCB, AC103794.1, CYP2R1, C11orf58, PDE3B, PLEKHA7, PSMA1, RPS13 \n",
            "\t   COPB1, PIK3C2A, RRAS2, SPON1-AS1, NUCB2, SPON1, NCR3LG1, FAR1-IT1, KCNC1, FAR1 \n",
            "\t   SERGEF, BTBD10, TPH1, ARNTL, SAAL1, LINC00958, HPS5, TEAD1, GTF2H1, MICAL2 \n",
            "Negative:  RSPH10B2, CCZ1B, AC079804.3, ZNF12, C1GALT1, AC005532.1, AC004982.2, AC073343.2, MIOS, ZNF316 \n",
            "\t   RPA3, ZNF853, UMAD1, AC006042.2, RPL36AL, LRR1, C7orf26, GLCCI1, RPS29, MGAT2 \n",
            "\t   ICA1, MDGA2, DNAAF2, ZDHHC4, AC007009.1, LINC00871, POLE2, AC060834.2, GRID2IP, KLHDC1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:47] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 4 (preB)\n",
            "INFO [2025-06-10 11:41:47] Setting auto leiden resolution for Komórki blastyczne 4 (preB) to 0.0712485\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AC092802.1, TLCD4, RWDD3, PTBP2, ALG14, DPYD, AC105942.1, DPYD-AS1, CNN3, FRRS1 \n",
            "\t   SLC44A3, AGL, F3, AC118553.1, ABCD3, SLC35A3, ARHGAP29, MFSD14A, ABCA4, AC093019.2 \n",
            "\t   GCLM, DNTTIP2, AL109613.1, BCAR3, FNBP1L, DR1, CCDC18-AS1, AC114485.1, AC095032.1, RNPC3 \n",
            "Negative:  UCA1, TPM4, RAB8A, CYP4F22, AC008894.2, HSH2D, PGLYRP2, FAM32A, RASAL3, AP1M1 \n",
            "\t   AC020911.2, WIZ, KLF2, AKAP8L, EPS15L1, C19orf44, AKAP8, CHERP, AC008764.6, SLC35E1 \n",
            "\t   BRD4, AC008764.8, MED26, ILVBL, AC008764.7, SLC1A6, AC008764.2, AC024075.3, OR7C1, SMIM7 \n",
            "PC_ 2 \n",
            "Positive:  MTRNR2L8, MRVI1, CTR9, RNF141, MRVI1-AS1, AMPD3, EIF4G2, ADM, AC116535.1, AC100763.1 \n",
            "\t   ZBED5, AC011092.3, ZBED5-AS1, USP47, AC011092.2, DKK3, SBF2, MICAL2, SBF2-AS1, TEAD1 \n",
            "\t   SWAP70, LINC00958, WEE1, ARNTL, ZNF143, BTBD10, AC132192.2, FAR1, AC132192.1, FAR1-IT1 \n",
            "Negative:  NCBP1, TSTD2, XPA, PTCSC2, TMOD1, TRMO, TDRD7, AL499604.1, CCDC180, HEMGN \n",
            "\t   AL590705.1, ANP32B, NANS, NUTM2G, TRIM14, CORO2A, MFSD14C, TBC1D2, ANKS6, ZNF782 \n",
            "\t   GALNT12, ZNF510, TGFBR1, ALG2, AL589843.1, SEC61B, PRXL2C, STX17-AS1, NR4A3, CDC14B \n",
            "PC_ 3 \n",
            "Positive:  AGBL3, CYREN, AC083870.1, TMEM140, CALD1, AC083862.3, BPGM, WDR91, AKR1B1, SLC35B4 \n",
            "\t   AC009542.1, LRGUK, CNOT4, EXOC4, AC009365.4, CHCHD3, PLXNA4, NUP205, PODXL, AC008264.2 \n",
            "\t   MKLN1-AS, STMP1, MKLN1, SLC13A4, LINC-PINT, AC058791.1, AC015987.1, AC016831.5, MTPN, AC016831.1 \n",
            "Negative:  HMGN2, AL513365.2, RPS6KA1, DHDDS, CRYBG2, ARID1A, PIGV, CD52, ZDHHC18, GPN2 \n",
            "\t   GPATCH3, UBXN11, NUDC, SH3BGRL3, CEP85, CNKSR1, ZNF593, FAM110D, PDIK1L, PAFAH2 \n",
            "\t   STMN1, PAQR7, AL033528.2, AUNIP, MTFR1L, SELENON, TXN, PALM2-AKAP2, TMEM245, LPAR1 \n",
            "PC_ 4 \n",
            "Positive:  AC018647.3, HERPUD2, AC005400.1, AC018647.2, DPY19L1, BMPER, SEPTIN7-AS1, BBS9, SEPTIN7, EEPD1 \n",
            "\t   RP9, KIAA0895, ANLN, NT5C3A, AC006960.2, FKBP9, AOAH, KBTBD2, ELMO1, LINC00997 \n",
            "\t   ELMO1-AS1, GPR141, AC018645.3, EPDR1, NME8, AVL9, SFRP4, LSM5, STARD3NL, PDE1C \n",
            "Negative:  CTNNBL1, BLCAP, SRC, MANBAL, TTI1, RPRD1B, RPN2, TGM2, BPI, MROH8 \n",
            "\t   AL391095.1, SNHG17, SAMHD1, TLDC2, RBL1, SOGA1, SNHG11, DSN1, NDRG3, SLA2 \n",
            "\t   RALGAPB, ACTR5, RAB5IF, TGIF2, PPP1R16B, MYL9, FAM83D, DLGAP4-AS1, DHX35, MAFB \n",
            "PC_ 5 \n",
            "Positive:  AL109628.2, TRAPPC6B, PNN, GEMIN2, MIA2, SEC23A, AL355835.1, FBXO33, CLEC14A, AL121790.1 \n",
            "\t   AL049828.1, MIPOL1, AL049828.2, SLC25A21-AS1, AL390800.1, AL162464.2, LINC02315, SLC25A21, MBIP, LINC02307 \n",
            "\t   BRMS1L, AL049870.2, RALGAPA1, NFKBIA, C14orf28, PSMA6, KLHL28, PRORP, TOGARAM1, PPP2R3C \n",
            "Negative:  CHD6, PTPRT, ZHX3, PLCG1-AS1, PLCG1, TOP1, SRSF6, L3MBTL1, MAFB, SGK2 \n",
            "\t   DHX35, IFT52, FAM83D, MYBL2, PPP1R16B, ACTR5, RALGAPB, SNHG11, TOX2, SNHG17 \n",
            "\t   OSER1, AL391095.1, OSER1-DT, BPI, TGM2, GDAP1L1, RPRD1B, FAM221A, CCDC126, FITM2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:49] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 5 (preB)\n",
            "INFO [2025-06-10 11:41:49] Setting auto leiden resolution for Komórki blastyczne 5 (preB) to 0.333884\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AC211433.1, GTF2I, GTF2IRD1, CLIP2, RFC2, LAT2, EIF4H, LIMK1, ELN, ABHD11 \n",
            "\t   BUD23, DNAJC30, TBL2, BCL7B, BAZ1B, BRMS1L, SLC25A21, RALGAPA1, MBIP, NFKBIA \n",
            "\t   PRORP, AL162464.2, PPP2R3C, PSMA6, NSUN5, PLPBP, ERLIN2, ADAM9, ADGRA2, FAM177A1 \n",
            "Negative:  ACSBG2, MLLT1, AC011444.3, CLPP, AC011444.2, ALKBH7, RFX2, GTF2F1, AC011444.1, RANBP3 \n",
            "\t   KHSRP, CAPS, SLC25A23, AC104532.2, DENND1C, VMAC, GPR108, TRIP10, NDUFA11, SH2D3A \n",
            "\t   DUS3L, VAV1, ADGRE1, CATSPERD, ZNF557, LONP1, INSR, AC119396.1, HSD11B1L, AC119396.2 \n",
            "PC_ 2 \n",
            "Positive:  ARL6IP4, OGFOD2, ABCB9, VPS37B, PITPNM2, HIP1R, MPHOSPH9, CCDC62, C12orf65, CDK2AP1 \n",
            "\t   DENR, HCAR3, KNTC1, SBNO1, RSRC2, ZCCHC8, SBNO1-AS1, CLIP1-AS1, KMT5A, RILPL2 \n",
            "\t   CLIP1, SNRNP35, RILPL1, TMED2-DT, VPS33A, TMED2, DDX55, DIABLO, EIF2B1, LRRC43 \n",
            "Negative:  CCDC127, SDHA, PDCD6, AHRR, EXOC3-AS1, EXOC3, SLC9A3-AS1, CEP72, ZDHHC11, BRD9 \n",
            "\t   TRIP13, SLC12A7, TERT, CLPTM1L, AC026748.3, LPCAT1, MRPL36, NDUFS6, AC124852.1, IRX1 \n",
            "\t   CLN8, DLGAP2-AS1, AC091978.1, ICE1, GP1BB, MED10, GNB1L, ZNF195, RTL10, AC127526.1 \n",
            "PC_ 3 \n",
            "Positive:  MFSD14C, ZNF782, NUTM2G, ZNF510, AL590705.1, CCDC180, AL589843.1, TDRD7, TMOD1, PRXL2C \n",
            "\t   TSTD2, CDC14B, HABP4, NCBP1, AL133477.2, XPA, ZNF367, AGO1, PTCSC2, CLSPN \n",
            "\t   AGO4, AL139286.3, C1orf216, PSMB2, SLC35D2, TRMO, AGO3, TFAP2E, NCDN, KIAA0319L \n",
            "Negative:  ELK3, AC008149.2, CDK17, AC007513.1, LTA4H, CFAP54, HAL, NEDD1, CCDC38, AC007656.2 \n",
            "\t   SNRPF, LINC02453, USP44, TMPO-AS1, TMPO, METAP2, SLC25A3, VEZT, CCNG1, IKBIP \n",
            "\t   NUDCD2, FGD6, APAF1, HMMR, ANKS1B, NR2C1, UHRF1BP1L, MAT2B, ACTR6, AC109466.1 \n",
            "PC_ 4 \n",
            "Positive:  BHLHE40, ARL8B, AC026202.2, EDEM1, AC069277.1, GRM7-AS3, GRM7, LMCD1-AS1, CAV3, RAD18 \n",
            "\t   SRGAP3, THUMPD3-AS1, THUMPD3, SETD5, TTLL3, ARPC4, TADA3, JAGN1, RPUSD3, LHFPL4 \n",
            "\t   CAMK1, IL17RE, OGG1, BRPF1, AC022382.2, CPNE9, MTMR14, IL17RC, CRELD1, PRRT3 \n",
            "Negative:  PXT1, PNPLA1, KCTD20, BRPF3, STK38, SRSF3, Z84485.1, CDKN1A, MAPK13, RAB44 \n",
            "\t   CPNE5, MAPK14, PPIL1, SLC26A8, C6orf89, MTCH1, SRPK1, FGD2, PIM1, TMEM217 \n",
            "\t   LHFPL5, TBC1D22B, RNF8, AL033519.5, CMTR1, CCDC167, FKBP5, MDGA1, AL121574.1, ZFAND3 \n",
            "PC_ 5 \n",
            "Positive:  AC079296.1, NRIP3, SCUBE2, TMEM9B-AS1, TMEM9B, DENND5A, AKIP1, TMEM41B, IPO7, ST5 \n",
            "\t   AC132192.1, AC132192.2, RPL27A, TRIM66, ZNF143, STK33, RIC3, WEE1, TUB, EIF3F \n",
            "\t   SWAP70, AC044810.2, SBF2-AS1, SBF2, AC104237.3, AC011092.2, AC011092.3, AC100763.1, ADM, AMPD3 \n",
            "Negative:  PDLIM1, ACSM6, SORBS1, AL157834.1, ALDH18A1, HELLS, TCTN3, TBC1D12, NOC3L, ENTPD1 \n",
            "\t   PLCE1, ENTPD1-AS1, SLC35G1, FRA10AC1, CC2D2B, PDE6C, CEP55, CCNJ, MYOF, EXOC6 \n",
            "\t   ZNF518A, HHEX, KIF11, IDE, MARCH5, CPEB3, BLNK, BTAF1, AL136181.1, FGFBP3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:51] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 6 (preB)\n",
            "INFO [2025-06-10 11:41:51] Setting auto leiden resolution for Komórki blastyczne 6 (preB) to 0.11169\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  MCUB, SEC24B, CASP6, SEC24B-AS1, PLA2G12A, COL25A1, GAR1, AC107071.1, RRH, OSTC \n",
            "\t   EGF, RPL34, ELOVL6, RPL34-AS1, ENPEP, LEF1-AS1, LEF1, AC083795.2, HADH, CYP2U1 \n",
            "\t   AC139718.2, AC096564.1, AC139718.1, SGMS2, AC004704.1, PAPSS1, FAM241A, AC092445.1, AP1AR, AIMP1 \n",
            "Negative:  Z97192.3, Z97192.2, Z97192.4, TAFA5, Z97192.1, TBC1D22A, C22orf34, BRD1, AL118516.1, Z98885.3 \n",
            "\t   CERK, ZBED4, GRAMD4, ALG12, CELSR1, CRELD2, TRMU, GTSE1, PIM3, TTC38 \n",
            "\t   MLC1, CDPF1, PANX2, PPARA, MIRLET7BHG, TRABD, PRR34-AS1, AL022328.3, LINC00899, AL022328.4 \n",
            "PC_ 2 \n",
            "Positive:  SEC24A, SAR1B, CAMLG, JADE2, DDX46, CDKN2AIPNL, C5orf24, TXNDC15, TGFBI, SLC25A48 \n",
            "\t   PCBD2, SPOCK1, KLHL3, SMAD5, AC109454.2, C5orf66, CXCL14, HNRNPA0, H2AFY, AC106791.1 \n",
            "\t   MYOT, UBE2B, PKD2L2, CDKL3, FAM13B, PPP2CA, AC113382.1, SKP1, BRD8, TCF7 \n",
            "Negative:  KDELR2, DAGLB, GRID2IP, ZDHHC4, RAC1, C7orf26, ZNF853, CYTH3, ZNF316, USP42 \n",
            "\t   AC073343.2, ZNF12, EIF2AK1, RSPH10B2, AIMP2, CCZ1B, AC079804.3, PMS2, C1GALT1, CCZ1 \n",
            "\t   AC005532.1, AC004982.2, OCM, MIOS, RNF216, RPA3, UMAD1, FSCN1, AC006042.2, ACTB \n",
            "PC_ 3 \n",
            "Positive:  NAV2, HTATIP2, PRMT3, E2F8, NELL1, CSRP3-AS1, ANO5, ZDHHC13, FANCF, SPTY2D1 \n",
            "\t   SPTY2D1OS, GAS2, UEVLD, SVIP, TSG101, LDHAL6A, AC006299.1, LDHC, LINC02718, LDHA \n",
            "\t   LINC02726, GTF2H1, LUZP2, HPS5, CCDC34, SAAL1, LGR4, LIN7C, TPH1, BDNF-AS \n",
            "Negative:  AC011287.1, CRPPA, ARL4A, CRPPA-AS1, VWDE, ANKMY2, TMEM106B, AC005014.4, THSD7A, BZW2 \n",
            "\t   AC004160.1, AC073333.1, PHF14, NDUFA4, TSPAN13, AC060834.2, AHR, AC007009.1, AC073332.1, ICA1 \n",
            "\t   SNX13, GLCCI1, HDAC9, AC006042.2, AC007091.1, UMAD1, TWISTNB, RPA3, AC005062.1, MIOS \n",
            "PC_ 4 \n",
            "Positive:  RAB44, CDKN1A, CPNE5, PPIL1, SRSF3, C6orf89, STK38, MTCH1, FGD2, PIM1 \n",
            "\t   TMEM217, KCTD20, TBC1D22B, PXT1, RNF8, PNPLA1, CMTR1, BRPF3, CCDC167, MDGA1 \n",
            "\t   AL121574.1, Z84485.1, ZFAND3, MAPK13, SAYSD1, DNAH8, GLO1, BTBD9, KCNK5, MAPK14 \n",
            "Negative:  AC008438.1, ANKHD1, EIF4EBP3, SRA1, HBEGF, APBB3, SLC35A4, PFDN1, CD14, CYSTM1 \n",
            "\t   NDUFA2, TMCO6, IGIP, IK, WDR55, HARS, PURA, HARS2, ZMAT2, MALINC1 \n",
            "\t   TAF7, AC008667.1, PCDHGA6, CXXC5, PCDHGA10, TAF1B, DIAPH1, AC010969.2, GRHL1, UBE2D2 \n",
            "PC_ 5 \n",
            "Positive:  Z84485.1, MAPK13, BRPF3, SLC26A8, MAPK14, PNPLA1, SRPK1, LHFPL5, PXT1, AL033519.5 \n",
            "\t   FKBP5, KCTD20, TEAD3, STK38, RPL10A, FANCE, SRSF3, CDKN1A, RAB44, STMN1 \n",
            "\t   PAQR7, AL033528.2, AUNIP, MTFR1L, SELENON, PAFAH2, PDIK1L, CPNE5, FAM110D, ZNF593 \n",
            "Negative:  NUDT7, AC009139.2, VAT1L, SYCE1L, WWOX, MON1B, AC079414.3, AC106729.1, AC046158.1, AC025287.4 \n",
            "\t   AC009141.1, TERF2IP, AC027279.1, KARS, MAF, ADAT1, LINC01229, GABARAPL2, MAFTRR, TMEM231 \n",
            "\t   AC105411.1, DYNLRB2, TMEM170A, CDYL2, CFDP1, ZFP1, CMC2, AC099508.2, CENPN, ZNRF1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:52] define_signif_tumor_subclusters(), tumor: Komórki blastyczne 7 (preB)\n",
            "INFO [2025-06-10 11:41:52] Setting auto leiden resolution for Komórki blastyczne 7 (preB) to 0.130032\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  PGLS, NIBAN3, AC010618.3, SLC27A1, COLGALT1, MAP1S, AC008761.1, AC010319.4, FCHO1, MVB12A \n",
            "\t   JAK3, BISPR, RPL18A, BST2, CCDC124, KCNN1, ARRDC2, IL12RB1, MAST3, PIK3R2 \n",
            "\t   IFI30, MPV17L2, PDE4C, AC008397.1, IQCN, JUND, LSM4, PGPEP1, LRRC25, SSBP4 \n",
            "Negative:  SESN1, ARMC2, FOXO3, CEP57L1, AL139106.1, CD164, AFG1L, AL359711.2, PPIL6, SNX3 \n",
            "\t   SMPD2, OSTM1, MICAL1, AL024507.2, ZBTB24, SEC63, AK9, SCML4, FIG4, SOBP \n",
            "\t   WASF1, CDC40, PDSS2, METTL24, BEND3, DDO, MTRES1, SLC22A16, AC002464.1, CD24 \n",
            "PC_ 2 \n",
            "Positive:  MRPS36, CENPH, CDK7, CCNB1, AC010273.2, CCDC125, SLC30A5, AK6, TAF9, PIK3R1 \n",
            "\t   AC112206.2, RAD17, CD180, MARVELD2, MAST4, OCLN, SREK1, GTF2H2C, AC025442.2, AC010359.3 \n",
            "\t   SERF1B, ERBIN, NLN, SMN2, SGTB, TRAPPC13, AC146944.4, SHLD3, SERF1A, TRIM23 \n",
            "Negative:  AC102953.2, INTS1, MAFK, MICALL2, TMEM184A, AC091729.3, PSMG3, ZFAND2A, PSMG3-AS1, ELFN1 \n",
            "\t   GPER1, MAD1L1, GPR146, MRM2, AC073957.1, C7orf50, NUDT1, COX19, ADAP1, SNX8 \n",
            "\t   AC073957.3, EIF3B, GET4, SUN1, CHST12, DNAAF5, LFNG, AC147651.4, PRKAR1B, BRAT1 \n",
            "PC_ 3 \n",
            "Positive:  AL157912.1, DEGS2, AL133523.1, YY1, AL157871.6, SLC25A29, AL157871.4, WARS, VRK1, PAPOLA \n",
            "\t   AL163051.1, AK7, GSKIP, ATG2B, C14orf132, AL139020.1, TCL1A, TCL1B, TCL6, GLRX5 \n",
            "\t   SNHG10, SYNE3, CLMN, DICER1-AS1, DICER1, SERPINA1, IFI27L2, IFI27, IFI27L1, DDX24 \n",
            "Negative:  SEC24A, CAMLG, SAR1B, JADE2, DDX46, CDKN2AIPNL, C5orf24, TXNDC15, PCBD2, AC109454.2 \n",
            "\t   C5orf66, UBE2B, H2AFY, CXCL14, CDKL3, SLC25A48, TGFBI, PPP2CA, SMAD5, SKP1 \n",
            "\t   SPOCK1, TCF7, KLHL3, AC008608.2, HNRNPA0, VDAC1, C5orf15, AC106791.1, FSTL4, MYOT \n",
            "PC_ 4 \n",
            "Positive:  LINC02315, LINC02307, AL390800.1, AL049870.2, AL049828.2, C14orf28, AL049828.1, KLHL28, FBXO33, TOGARAM1 \n",
            "\t   MIA2, PRPF39, PNN, AL109628.2, TRAPPC6B, FKBP3, GEMIN2, SEC23A, FANCM, AL355835.1 \n",
            "\t   CLEC14A, MIS18BP1, AL121790.1, MIPOL1, LINC00871, SLC25A21-AS1, MDGA2, AL162464.2, SLC25A21, RPS29 \n",
            "Negative:  ORAI1, MORN3, KDM2B-DT, KDM2B, P2RX7, TMEM120B, P2RX4, RHOF, RNF34, CAMKK2 \n",
            "\t   LINC01089, AC084018.2, ANAPC5, SETD1B, PSMD9, WDR66, AC069503.1, BCL7A, AC156455.1, SDHB \n",
            "\t   ATP13A2, PADI2, AL049569.1, PADI4, MFAP2, MLXIP, PADI6, BX284668.6, RCC2, LRRC43 \n",
            "PC_ 5 \n",
            "Positive:  FAM160A1, AC095055.1, GATB, SH3D19, RPS3A, AC092611.1, AC097375.1, LRBA, DCLK2, AC097375.5 \n",
            "\t   LINC02355, AC108156.1, LINC02273, AC002460.2, FBXW7, AC069272.1, AC023424.3, MIR4453HG, NR3C2, TMEM154 \n",
            "\t   ARFIP1, ARHGAP10, FHDC1, TRIM2, PRMT9, MND1, TMEM131L, AC106865.2, TMEM184C, TLR2 \n",
            "Negative:  MIR3142HG, PTTG1, ATP10B, AC113414.1, SLU7, CCNG1, ZBED8, NUDCD2, HMMR, CCNJL \n",
            "\t   MAT2B, AC109466.1, PWWP2A, LINC02143, TTC1, AC008415.1, TENM2, AC008691.1, RARS, UBLCP1 \n",
            "\t   PANK3, RNF145, SPDL1, DOCK2, LINC02202, AC008680.1, AC136424.2, INSYN2B, EBF1, C5orf58 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:53] define_signif_tumor_subclusters(), tumor: Komórki blastyczne proliferujące\n",
            "INFO [2025-06-10 11:41:54] Setting auto leiden resolution for Komórki blastyczne proliferujące to 0.0681121\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  EXOC5, AL161757.5, AP5M1, TMEM260, NAA30, ARMH4, LINC02284, ACTR10, SLC35F4, AL355073.2 \n",
            "\t   PSMA3, AL355073.1, PSMA3-AS1, ARID4A, AL139021.2, PELI2, TOMM20L, AL139021.1, KTN1, TIMM9 \n",
            "\t   KIAA0586, KTN1-AS1, DACT1, ATG14, LINC01500, AL158801.3, DAAM1, FBXO34, GPR135, AL158801.2 \n",
            "Negative:  MED26, AC008764.8, SLC35E1, AC008764.6, CHERP, C19orf44, EPS15L1, KLF2, ILVBL, SLC1A6 \n",
            "\t   BRD4, OR7C1, ADGRE2, ZNF333, ADGRE3, AKAP8, CLEC17A, AKAP8L, NDUFB7, WIZ \n",
            "\t   AC020911.2, RASAL3, TECR, PGLYRP2, AP1M1, CYP4F22, DNAJB1, FAM32A, HSH2D, UCA1 \n",
            "PC_ 2 \n",
            "Positive:  CD52, UBXN11, SH3BGRL3, CRYBG2, CEP85, DHDDS, CNKSR1, AL513365.2, HMGN2, ZNF593 \n",
            "\t   FAM110D, RPS6KA1, PDIK1L, PAFAH2, ARID1A, STMN1, PIGV, PAQR7, ZDHHC18, GPN2 \n",
            "\t   AL033528.2, GPATCH3, AUNIP, NUDC, MTFR1L, SLC9A1, SELENON, WDTC1, TMEM222, SYTL1 \n",
            "Negative:  AC215522.2, AC093627.5, AC093627.4, FAM20C, PRKAR1B, AC147651.4, DNAAF5, SUN1, GET4, AC073957.3 \n",
            "\t   ADAP1, COX19, C7orf50, AC073957.1, GPR146, GPER1, ZFAND2A, AC091729.3, MICALL2, CPT1B \n",
            "\t   CHKB, U62317.4, CHKB-DT, ODF3B, ARSA, TYMP, SHANK3, SCO2, NCAPH2, AC102953.2 \n",
            "PC_ 3 \n",
            "Positive:  RPS12, SLC18B1, VNN2, EYA4, VNN3, TBPL1, VNN1, STX7, SLC2A12, MOXD1 \n",
            "\t   LINC01013, SGK1, AL021408.1, AL133346.1, LINC01010, CCN2, AL078590.2, ENPP3, MED23, ALDH8A1 \n",
            "\t   ARG1, HBS1L, AKAP7, MYB, EPB41L2, AHI1, TMEM200A, AL049552.1, SAMD3, LINC00271 \n",
            "Negative:  APAF1, IKBIP, SLC25A3, TMPO, TMPO-AS1, ANKS1B, LINC02453, UHRF1BP1L, AC007656.2, ACTR6 \n",
            "\t   NEDD1, CFAP54, DEPDC4, AC007513.1, SCYL2, CDK17, GAS2L3, AC008149.2, UTP20, ELK3 \n",
            "\t   ARL1, LTA4H, CHPT1, HAL, SYCP3, CCDC38, GNPTAB, SNRPF, DRAM1, USP44 \n",
            "PC_ 4 \n",
            "Positive:  TSPYL4, COL10A1, DSE, TSPYL1, CALHM6, NT5DC1, RWDD1, HDAC2-AS2, RSPH4A, ZUP1 \n",
            "\t   HDAC2, KPNA5, LINC01268, DCBLD1, MARCKS, GOPC, LINC02541, NUS1, AL360015.1, SLC35F1 \n",
            "\t   FAM229B, CEP85L, TUBE1, PLN, Z99496.2, FYN, MCM9, ASF1A, TRAF3IP2, AL137009.1 \n",
            "Negative:  LINC01588, ARF6, VCPKMT, AL627171.1, SOS2, AL627171.2, L2HGDH, NEMF, DMAC2L, KLHDC2 \n",
            "\t   CDKL1, KLHDC1, MAP4K5, POLE2, AL118556.1, ATL1, DNAAF2, SAV1, AL606834.1, MGAT2 \n",
            "\t   NIN, RPL36AL, AL606834.2, LRR1, PYGL, ABHD12B, RPS29, TRIM9, MDGA2, LINC00871 \n",
            "PC_ 5 \n",
            "Positive:  CYTH3, USP42, EIF2AK1, AIMP2, PMS2, CCZ1, OCM, RNF216, FSCN1, ACTB \n",
            "\t   FBXL18, AC092171.5, AC092171.3, TNRC18, SLC29A4, WIPI2, RPA3, RBAK, UMAD1, RADIL \n",
            "\t   AC006042.2, AP5Z1, AC092428.1, GLCCI1, FOXK1, ICA1, SDK1, AC007009.1, AC024028.1, AC060834.2 \n",
            "Negative:  SSR3, TIPARP-AS1, TIPARP, LINC00886, LEKR1, CCNL1, AC092944.1, VEPH1, PTX3, RSRC1 \n",
            "\t   AC106707.1, MLF1, GFM1, LXN, RARRES1, MFSD1, IQCJ-SCHIP1, SCHIP1, IL12A-AS1, IL12A \n",
            "\t   C3orf80, IFT80, SMC4, TRIM59, KPNA4, PPM1L, B3GALNT1, TOX2, MYBL2, NMD3 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:56] define_signif_tumor_subclusters(), tumor: komórki dendrytyczne mieloidalne\n",
            "INFO [2025-06-10 11:41:56] Too few cells in group komórki dendrytyczne mieloidalne for any (sub)clustering. Keeping as is.\n",
            "INFO [2025-06-10 11:41:56] define_signif_tumor_subclusters(), tumor: komórki dendrytyczne plazmacytoidalne\n",
            "INFO [2025-06-10 11:41:56] Less cells in group komórki dendrytyczne plazmacytoidalne than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:56] define_signif_tumor_subclusters(), tumor: komórki NK T\n",
            "INFO [2025-06-10 11:41:56] Less cells in group komórki NK T than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-06-10 11:41:56] define_signif_tumor_subclusters(), tumor: monocyty\n",
            "INFO [2025-06-10 11:41:56] Setting auto leiden resolution for monocyty to 0.419053\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  CD52, CRYBG2, UBXN11, DHDDS, AL513365.2, RPS6KA1, SH3BGRL3, ARID1A, HMGN2, PIGV \n",
            "\t   ZDHHC18, GPN2, GPATCH3, CEP85, NUDC, SLC9A1, WDTC1, TMEM222, SYTL1, CNKSR1 \n",
            "\t   MAP3K6, WASF2, ZNF593, FO393419.2, BX293535.1, FAM110D, AHDC1, PDIK1L, FGR, PAFAH2 \n",
            "Negative:  ZC3HC1, KLHDC10, AC073320.1, TMEM209, UBE2H, AC087071.2, AC084864.2, CEP41, NRF1, AC084816.1 \n",
            "\t   MEST, AC078846.1, AC007938.2, STRIP2, ETNK1, COPG2, AHCYL2, SMO, AC007938.3, C2CD5 \n",
            "\t   TSPAN33, AASS, AUXG01000058.1, TNPO3, AC015983.2, NDUFA5, IRF5, AC053513.1, AC016831.7, LMOD2 \n",
            "PC_ 2 \n",
            "Positive:  PTPRK, AL590006.1, LAMA2, THEMIS, AL356124.1, C6orf58, ARHGAP18, L3MBTL3, ECHDC1, RNF146 \n",
            "\t   AL355581.1, CENPW, SAMD3, TRMT11, HINT3, NCOA7-AS1, NCOA7, HEY2, RCC1, PHACTR4 \n",
            "\t   MED18, SESN2, HDDC2, SNHG3, AL353622.1, ATP5IF1, DNAJC8, TRNAU1AP, PTAFR, SNHG12 \n",
            "Negative:  TCAIM, ABHD5, ANO10, SNRK-AS1, SNRK, POMGNT2, ZNF445, GASK1A, ZNF852, ZKSCAN7 \n",
            "\t   ZNF662, ZKSCAN7-AS1, KRBOX1, ZNF660, AC099329.1, ZNF197, HIGD1A, ZNF35, CCDC13, ZNF502 \n",
            "\t   ZBTB47, ZNF501, AC006059.1, KIAA1143, KIF15, NKTR, SS18L2, TMEM42, SEC22C, ZDHHC3 \n",
            "PC_ 3 \n",
            "Positive:  NBL1, AL031727.2, TMCO4, OTUD3, MICOS10, UBXN10, CAPZB, SLC66A1, AKR7A2, MRTO4 \n",
            "\t   EMC1, EMC1-AS1, UBR4, ARHGEF10L, RCC2, PADI6, PADI4, PADI2, SDHB, ATP13A2 \n",
            "\t   AL049569.1, MFAP2, BX284668.6, BX284668.5, CROCC, GUCY1B1, AC104083.1, CTSO, PDGFC, GASK1B \n",
            "Negative:  TLE3, UACA, DRAIC, LRRC49, THSD4, AC027237.2, THSD4-AS1, MYO9A, RPLP1, KIF23 \n",
            "\t   AC022872.1, GLCE, SENP8, NOX5, SPESP1, PKM, ANP32A, PARP6, CORO2B, FEM1B \n",
            "\t   AC009690.2, CLN6, CELF6, CALML4, PIAS1, HEXA, MAP2K5, C15orf61, TMEM202-AS1, IQCH-AS1 \n",
            "PC_ 4 \n",
            "Positive:  NETO1, LINC01909, SOCS6, RTTN, CD226, DOK6, CCDC102B, AL139317.5, FERMT2, DDHD1 \n",
            "\t   AL356020.1, GNPNAT1, TMX3, AL365295.1, STYX, AL365295.2, PSMC6, AC114689.3, CDKN3, AL133453.1 \n",
            "\t   CNIH1, DSEL, ERO1A, GMFB, CDH7, BRD9, CGRRF1, ZDHHC11, TRIP13, LINC01924 \n",
            "Negative:  AL591519.1, AL359987.1, AL356094.2, MAP3K7, EPHA7, MANEA-DT, MANEA, AL132996.1, UFL1-AS1, AL158136.1 \n",
            "\t   BACH2, UFL1, CASP8AP2, GPR63, NDUFAF4, MDN1, KLHL32, LYRM2, MMS22L, ANKRD6 \n",
            "\t   AL589740.1, FBXL4, RRAGD, UBE2J1, COQ3, GABRR2, PNISR, AL513550.1, PM20D2, USP45 \n",
            "PC_ 5 \n",
            "Positive:  USP50, AC012170.3, USP8, TRPM7, AC022087.1, SPPL2A, GABPB1-AS1, AC021752.1, GABPB1-IT1, GABPB1 \n",
            "\t   HDC, AP4E1, SLC27A2, AC025040.1, ATP8B4, AC066613.2, COPS2, AC022306.2, GALK2, FGF7 \n",
            "\t   DTWD1, FAM227B, DMXL2, LYSMD2, TMOD2, GFM1, TMOD3, MLF1, AC090971.2, AC106707.1 \n",
            "Negative:  FAF1, AL049637.1, CDKN2C, RNF11, AL162430.2, TTC39A, EPS15, AC104170.2, AC104170.1, OSBPL9 \n",
            "\t   NRDC, AL050343.2, TXNDC12, KTI12, AL445685.1, BTF3L4, ZFYVE9, CC2D1B, ORC1, PRPF38A \n",
            "\t   TUT4, GPX7, SHISAL2A, COA7, ZYG11B, ECHDC2, SCP2, CPT2, AL606760.2, CZIB \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:57] define_signif_tumor_subclusters(), tumor: MPP\n",
            "INFO [2025-06-10 11:41:57] Setting auto leiden resolution for MPP to 0.118316\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  AC092068.3, AC092068.1, DIRAS1, AC092068.2, ZNF554, SLC39A3, THOP1, SGTA, ZNF555, GNG7 \n",
            "\t   ZNF57, GADD45B, AC119403.1, LMNB2, AP001453.1, DNAJC4, NUDT22, VEGFB, ZNF77, FKBP2 \n",
            "\t   TRPT1, PPP1R14B, FERMT3, TLE2, STIP1, AC005944.1, PPP1R14B-AS1, MPND, TLE5, STAP2 \n",
            "Negative:  CDK19, AC002464.1, SLC22A16, AMD1, DDO, METTL24, GTF3C6, CDC40, WASF1, FIG4 \n",
            "\t   RPF2, AK9, SLC16A10, ZBTB24, MFSD4B, AL080317.1, MICAL1, REV3L, SMPD2, AL080317.2 \n",
            "\t   PPIL6, TRAF3IP2-AS1, AL359711.2, AL008730.1, CD164, TRAF3IP2, CEP57L1, FYN, SESN1, ARMC2 \n",
            "PC_ 2 \n",
            "Positive:  AC109361.1, TBCK, GSTCD, AIMP1, AC092445.1, PAPSS1, INTS12, ARHGEF38, PPA2, TET2-AS1 \n",
            "\t   TET2, AC004069.1, AC096577.1, AC004053.1, CXXC4-AS1, LINC02428, CENPE, BDH2, SLC9B2, SLC9B1 \n",
            "\t   CISD2, UBE2D3-AS1, AC018797.3, UBE2D3, MANBA, NFKB1, AF213884.3, AC098487.1, SLC39A8, AP002075.1 \n",
            "Negative:  AKAP8L, WIZ, AKAP8, RASAL3, BRD4, PGLYRP2, ILVBL, CYP4F22, SLC1A6, UCA1 \n",
            "\t   OR7C1, TPM4, ADGRE2, RAB8A, ZNF333, ADGRE3, AC008894.2, CLEC17A, NDUFB7, HSH2D \n",
            "\t   TECR, DNAJB1, FAM32A, GIPC1, AC008569.2, AP1M1, PKN1, DDX39A, AC020911.2, AC008569.1 \n",
            "PC_ 3 \n",
            "Positive:  RNASE4, RNASE6, RNASE1, RNASE3, AL355922.1, RNASE2, METTL17, NDRG2, ARHGEF40, ZNF219 \n",
            "\t   LINC00641, HNRNPC, RPGRIP1, SUPT16H, CHD8, RAB2B, TOX4, METTL3, AC244502.1, TRDC \n",
            "\t   TRAC, DAD1, KBTBD8, ABHD4, SUCLG2, SUCLG2-AS1, TAFA1, AL160314.2, EOGT, AC109587.1 \n",
            "Negative:  KDM1A, EPHB2, AL031428.1, LUZP1, ZBTB40, LINC01355, CDC42-IT1, AL109936.6, CDC42-AS1, HNRNPR \n",
            "\t   CDC42, ZNF436, LINC00339, ZNF436-AS1, HSPG2, TCEA3, LDLRAD2, CHD6, ZHX3, PLCG1-AS1 \n",
            "\t   TOP1, PLCG1, MAFB, PTPRT, DHX35, ASAP3, FAM83D, PPP1R16B, SRSF6, USP48 \n",
            "PC_ 4 \n",
            "Positive:  NCOA4, AGAP4, TIMM23, PTPN20, WASHC2C, GDF10, ZFAND4, FAM25G, MARCH8, AGAP9 \n",
            "\t   MAPK8, AL731567.1, ARHGAP22, ALOX5, WDFY4, AC060234.3, VSTM4, TMEM273, OR13A1, AL138760.1 \n",
            "\t   ERCC6, ZNF22, PARG, TIMM23B, TIMM23B-AGAP6, C10orf25, AL442003.1, WASHC2A, SGMS1, DEPP1 \n",
            "Negative:  TTI1, RPRD1B, CTNNBL1, TGM2, BLCAP, BPI, SRC, AL391095.1, MANBAL, RPN2 \n",
            "\t   SNHG17, SNHG11, MROH8, RALGAPB, RBL1, SAMHD1, TLDC2, ACTR5, SOGA1, DSN1 \n",
            "\t   NDRG3, SLA2, RAB5IF, PPP1R16B, TGIF2, MYL9, FAM83D, DLGAP4-AS1, DLGAP4, AAR2 \n",
            "PC_ 5 \n",
            "Positive:  C3orf33, PLCH1, MME-AS1, MME, DHX36, SLC33A1, ARHGEF26, GMPS, KCNAB1, RAP2B \n",
            "\t   AC091607.2, P2RY1, SSR3, AC026347.1, TIPARP-AS1, MBNL1-AS1, TIPARP, MBNL1, LINC00886, AC108718.1 \n",
            "\t   LEKR1, SUCNR1, CCNL1, AADACL2-AS1, AC092944.1, VEPH1, PTX3, RSRC1, STX10, IER2 \n",
            "Negative:  SELENOO, AL022328.4, TUBGCP6, AL022328.3, HDAC10, TRABD, MAPK12, PANX2, PLXNB2, MLC1 \n",
            "\t   DENND6B, PIM3, PPP6R2, SBF1, CRELD2, LMF2, ALG12, NCAPH2, SCO2, ZBED4 \n",
            "\t   TYMP, Z98885.3, BRD1, ODF3B, U62317.4, Z97192.4, CPT1B, Z97192.3, CHKB, CHKB-DT \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:41:58] define_signif_tumor_subclusters(), tumor: późne komórki erytroidalne\n",
            "INFO [2025-06-10 11:41:58] Setting auto leiden resolution for późne komórki erytroidalne to 0.21198\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  ZPBP, IKZF1, ABCA13, UPP1, HUS1, TNS3, AC004870.3, AC004870.4, TBRG4, CCM2 \n",
            "\t   SNHG15, MYO1G, AC004847.1, AC004854.2, PURB, LINC01952, H2AFV, PPIA, ZMIZ2, OGDH \n",
            "\t   TMED4, DDX56, NUDCD3, YKT6, POLD2, AEBP1, PPIG, POLM, FASTKD1, DBNL \n",
            "Negative:  DCAF8, PEA15, AL139011.1, AL121987.2, PEX19, VANGL2, SLAMF6, NCSTN, IGSF8, COPA \n",
            "\t   AL138930.1, CD84, SLAMF1, AL121985.1, CD48, SLAMF7, LY9, CD244, F11R, TSTD1 \n",
            "\t   USF1, ARHGAP30, PFDN2, NIT1, DEDD, UFC1, USP21, PPOX, B4GALT3, NDUFS2 \n",
            "PC_ 2 \n",
            "Positive:  SNX15, SAC3D1, ARL2, NAALADL1, PPP2R5B, ATG2A, CDCA5, EHD1, ZFPL1, CDC42BPG \n",
            "\t   MEN1, VPS51, MAP4K2, TM7SF2, AP001462.1, ZNHIT2, SF1, PYGM, RASGRP2, FAU \n",
            "\t   NRXN2, SYVN1, AP005273.2, MRPL49, AP003068.2, RPS6KA4, CAPN1, CCDC88B, POLA2, PRDX5 \n",
            "Negative:  AL080317.2, TRAF3IP2-AS1, REV3L, AL080317.1, AL008730.1, MFSD4B, TRAF3IP2, SLC16A10, RPF2, CDK19 \n",
            "\t   AMD1, GTF3C6, AC002464.1, FYN, SLC22A16, TUBE1, DDO, METTL24, CDC40, WASF1 \n",
            "\t   FIG4, FAM229B, AK9, ZBTB24, MICAL1, AL360015.1, SMPD2, PPIL6, AL359711.2, LINC02541 \n",
            "PC_ 3 \n",
            "Positive:  TMEM173, DNAJC18, UBE2D2, SPATA24, CXXC5, MZB1, AC008667.1, SLC23A1, MALINC1, AC135457.1 \n",
            "\t   PURA, PAIP2, IGIP, AC008438.1, HBEGF, CYSTM1, EIF4EBP3, SRA1, ANKHD1, APBB3 \n",
            "\t   PFDN1, SNHG4, SLC35A4, CD14, MATR3, NDUFA2, TMCO6, AC011405.1, IK, SIL1 \n",
            "Negative:  AC009271.1, AC006305.1, LINC01416, LINC01905, TCF4-AS2, TXNL1, WDR7, TCF4-AS1, FECH, NARS \n",
            "\t   TCF4, AC027097.1, AC027097.2, ATP8B1, CCDC68, NEDD4L, RAB27B, MALT1, AC104365.1, C18orf54 \n",
            "\t   ZNF532, POLI, SEC11C, MBD2, LMAN1, PMAIP1, MEX3C, RNF152, SMAD4, PIGN \n",
            "PC_ 4 \n",
            "Positive:  SLC3A2, WDR74, SNHG1, SLC22A10, LGALS12, AP001160.1, PLAAT4, PLAAT3, STX5, ATL3 \n",
            "\t   AP000753.2, RTN3, NXF1, C11orf95, SPINDOC, MARK2, TMEM179B, NAA40, COX8A, OTUB1 \n",
            "\t   AP001160.3, MACROD1, TMEM223, AP000721.2, FLRT1, TAF6L, STIP1, FERMT3, POLR2G, TRPT1 \n",
            "Negative:  GOLGA8N, AC123768.3, ARHGAP11A, AC123768.4, LINC02256, GOLGA8O, OTUD7A, LINC02352, MTMR10, AC091057.6 \n",
            "\t   AC123768.5, ARHGAP11B, FAN1, AC091057.2, KLF13, SCG5, FMN1, AC055874.1, RYR3, AC010809.2 \n",
            "\t   AC010809.1, AVEN, CHRM5, EMC7, AC007938.3, AC058791.1, COPG2, LINC-PINT, AUXG01000058.1, AC016831.5 \n",
            "PC_ 5 \n",
            "Positive:  TMEM246, RNF20, ZNF189, GRIN3A, MRPL50, SMC2, CAVIN4, TMEFF1, NIPSNAP3A, MSANTD3 \n",
            "\t   NIPSNAP3B, TEX10, ABCA1, INVS, AL591506.1, SLC44A1, ERP44, FSD1L, AL359711.2, PPIL6 \n",
            "\t   CD164, SMPD2, FKTN, AL358937.1, MICAL1, CEP57L1, ZBTB24, AK9, TMEM38B, FIG4 \n",
            "Negative:  AC211486.5, GTF2IRD2B, AC211486.2, RCC1L, CASTOR2, TRIM73, GTF2IRD2, POM121C, SPDYE5, NCF1 \n",
            "\t   HIP1, RHBDD2, AC211433.1, POR, GTF2I, TMEM120A, GTF2IRD1, CLIP2, STYXL1, RFC2 \n",
            "\t   MDH2, LAT2, HSPB1, EIF4H, LIMK1, YWHAG, ELN, ABHD11, SSC4D, BUD23 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:42:00] define_signif_tumor_subclusters(), tumor: prekursory erytroidalne i megakariocytarne\n",
            "INFO [2025-06-10 11:42:00] Setting auto leiden resolution for prekursory erytroidalne i megakariocytarne to 0.108668\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  LMNB2, GADD45B, GNG7, AC092068.2, ARHGDIA, P4HB, AC145207.5, ALYREF, MCRIP1, ANAPC11 \n",
            "\t   SLC25A10, PCYT2, AC092068.1, MRPL12, SIRT7, AC139530.1, MAFG, PYCR1, HGS, ASPSCR1 \n",
            "\t   ARL16, AC092068.3, CENPX, CCDC137, LRRC45, OXLD1, RAC3, PDE6G, DIRAS1, CC2D1A \n",
            "Negative:  AL589990.1, EXTL2, CDC14A, SLC30A7, DPH5, RTCA, AC093157.1, RTCA-AS1, S1PR1, AL445928.2 \n",
            "\t   DBT, AC114485.1, LRRC39, AC095032.1, TRMT13, SASS6, RNPC3, AC093019.2, AMY2B, AL591888.1 \n",
            "\t   MFSD14A, AL355306.2, SLC35A3, PRMT6, AC118553.1, NTNG1, AGL, VAV3, FRRS1, DPYD-AS1 \n",
            "PC_ 2 \n",
            "Positive:  PSTK, C10orf88, FAM24B, IKZF5, ACADSB, CUZD1, BUB3, PLEKHA1, AL731566.2, LINC02641 \n",
            "\t   NSMCE4A, CHST15, ATE1, OAT, WDR11, LHPP, WDR11-AS1, FAM53B, AC023282.1, FAM53B-AS1 \n",
            "\t   EEF1AKMT2, PLPP4, ABRAXAS2, SEC23IP, AL731577.2, MCMBP, ZRANB1, CTBP2, INPP5F, EDRF1-DT \n",
            "Negative:  SMIM7, AC024075.2, AC024075.3, AC008764.2, AC024075.1, AC008764.7, TMEM38A, MED26, SIN3B, AC008764.8 \n",
            "\t   CPAMD8, SLC35E1, HAUS8, AC008764.6, AC020908.1, CHERP, MYO9B, C19orf44, AC020908.3, EPS15L1 \n",
            "\t   KLF2, AC020913.1, AC020911.2, USE1, AP1M1, OCEL1, FAM32A, NR2F6, HSH2D, USHBP1 \n",
            "PC_ 3 \n",
            "Positive:  SLC9A1, WDTC1, NUDC, TMEM222, GPATCH3, GPN2, SYTL1, ZDHHC18, MAP3K6, AL513365.2 \n",
            "\t   DHDDS, HMGN2, PIGV, CRYBG2, CD52, WASF2, ARID1A, RPS6KA1, UBXN11, SH3BGRL3 \n",
            "\t   FO393419.2, CEP85, BX293535.1, CNKSR1, AHDC1, ZNF593, FGR, FAM110D, IFI6, PDIK1L \n",
            "Negative:  ACSM3, IQCK, ACSM1, KNOP1, THUMPD1, VPS35L, AC004381.1, CCP110, ERI2, GDE1 \n",
            "\t   AC130456.3, REXO5, TMC5, DCUN1D3, SYT17, LYRM1, ITPRIPL2, DNAH3, AC008551.1, TMEM159 \n",
            "\t   CRYM, NPIPB3, AC005632.6, METTL9, AC005632.3, PANX2, TRABD, MLC1, AL022328.3, MAPK12 \n",
            "PC_ 4 \n",
            "Positive:  GEMIN2, TRAPPC6B, AL109628.2, SEC23A, AL355835.1, PNN, CLEC14A, AL121790.1, MIA2, MIPOL1 \n",
            "\t   FBXO33, SLC25A21-AS1, AL162464.2, AL049828.1, SLC25A21, MBIP, AL049828.2, BRMS1L, AL390800.1, RALGAPA1 \n",
            "\t   NFKBIA, PSMA6, LINC02315, PRORP, PPP2R3C, LINC02307, AL049870.2, C14orf28, KLHL28, TOGARAM1 \n",
            "Negative:  ST14, APLP2, PRDM10, NFRKB, ARHGAP32, KCNJ1, SENCR, FLI1, AP001122.1, ETS1 \n",
            "\t   KIRREL3, ARHGEF10, ST3GAL4, DLGAP2-AS1, KBTBD11, AC100810.1, CLN8, MYOM2, AC245123.1, AC245187.2 \n",
            "\t   AC246817.1, AC246817.2, CSMD1, GSEC, AC016065.1, MCPH1, ANGPT2, SLC38A10, NDUFAF8, TEPSIN \n",
            "PC_ 5 \n",
            "Positive:  WDR77, ATP5PB, AL390195.1, OVGP1, CHI3L2, C1orf162, DENND2D, CEPT1, TMIGD3, RAP1A \n",
            "\t   INKA2, DDX20, CTTNBP2NL, WNT2B, ST7L, CAPZA1, MOV10, AL603832.1, LAMTOR5, RHOC \n",
            "\t   AL355488.1, SLC16A1, SLC16A1-AS1, SLC16A4, AL390242.1, LAMTOR5-AS1, LRIG2-DT, RBM15, LRIG2, RBM15-AS1 \n",
            "Negative:  MEF2B, TMEM161A, SLC25A42, ARMC6, SUGP2, HOMER3, DDX49, COPE, SMAD1-AS1, UPF1 \n",
            "\t   MMAA, C4orf51, CRTC1, ZNF827, LSM6, KLHL26, SLC10A7, REX1BD, TTC29, AC093916.1 \n",
            "\t   CRLF1, TMEM184C, AC096711.2, PRMT9, UBA52, LINC02479, ARHGAP10, AC005253.1, C4orf33, NR3C2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:42:01] define_signif_tumor_subclusters(), tumor: wczesne komórki erytroidalne\n",
            "INFO [2025-06-10 11:42:01] Setting auto leiden resolution for wczesne komórki erytroidalne to 0.321105\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  PTCD1, CPSF4, BUD31, ARPC1B, PDAP1, ARPC1A, AC004834.1, ATP5MF, SMURF1, TRRAP \n",
            "\t   BAIAP2L1, BRI3, ZNF789, TECPR1, ZNF394, LMTK2, AC079781.5, ZKSCAN5, SDHAF3, SEM1 \n",
            "\t   FAM200A, SLC25A13, ZNF655, DYNC1I1, ZSCAN25, PDK4, CYP3A5, CYP3A7, TRIM4, AC004522.5 \n",
            "Negative:  LCOR, SLIT1, ARHGAP19, PIK3AP1, FRAT1, TM9SF3, FRAT2, TLL2, OPALIN, AL355490.1 \n",
            "\t   DNTT, RRP12, AL355490.2, AL136181.1, PGAM1, BLNK, EXOSC1, ZNF518A, ZDHHC16, CCNJ \n",
            "\t   CC2D2B, MMS19, ENTPD1-AS1, UBTD1, ENTPD1, TCTN3, MORN4, ALDH18A1, SORBS1, PI4K2A \n",
            "PC_ 2 \n",
            "Positive:  DENND5B-AS1, DENND5B, SINHCAF, AC068774.1, ETFBKMT, DDX11, AMN1, AC023157.3, DDX11-AS1, LINC02422 \n",
            "\t   TSPAN11, AC010198.2, AC010198.1, CAPRIN2, IPO8, TMTC1, OVCH1, OVCH1-AS1, ERGIC2, LINC01806 \n",
            "\t   PSMD14, SLC4A10, DPP4, AC009318.3, IFIH1, GCA, SLC38A11, AC019197.1, GRB14, SCN3A \n",
            "Negative:  SMIM6, SAP30BP, ITGB4, GALK1, H3F3B, UNK, UNC13D, WBP2, AP2A2, MUC6 \n",
            "\t   TRIM65, TOLLIP, MRPL38, FTH1, BEST1, INCENP, RAB3IL1, TOLLIP-AS1, FADS3, ASRGL1 \n",
            "\t   FADS1, AHNAK, VRK3, FBF1, BRSK2, FADS2, FEN1, AP001363.1, TMEM258, EEF1G \n",
            "PC_ 3 \n",
            "Positive:  CAMK1D, AL512770.1, CDC123, AL731559.1, NUDT5, CCDC3, SEC61A2, OPTN, DHTKD1, MCM10 \n",
            "\t   PHYH, SEPHS1, UPF2, BEND7, PROSER2-AS1, PRPF18, LRGUK, EXOC4, AC009365.4, AL157392.3 \n",
            "\t   CHCHD3, PLXNA4, FRMD4A, PODXL, AC008264.2, FAM107B, MKLN1-AS, CDNF, MKLN1, LINC-PINT \n",
            "Negative:  BTBD7, UNC79, OTUB2, DDX24, IFI27L1, IFI27, IFI27L2, SERPINA1, DICER1, DICER1-AS1 \n",
            "\t   CLMN, SYNE3, SNHG10, GLRX5, WASHC3, HOOK1, FGGY, AC099792.1, DRAM1, NFIA \n",
            "\t   AC093425.1, TCL6, LINC01358, LINC01135, JUN, MYSM1, GNPTAB, OMA1, TCL1B, DAB1 \n",
            "PC_ 4 \n",
            "Positive:  TSPYL4, DSE, COL10A1, TSPYL1, NT5DC1, CALHM6, RWDD1, HDAC2-AS2, RSPH4A, HDAC2 \n",
            "\t   ZUP1, LINC01268, KPNA5, MARCKS, LINC02541, DCBLD1, AL360015.1, GOPC, FAM229B, NUS1 \n",
            "\t   TUBE1, SLC35F1, FYN, MRPL42, NUDT4, UBE2N, SOCS2-AS1, SOCS2, CEP85L, TRAF3IP2 \n",
            "Negative:  GAR1, PLA2G12A, AC132872.1, CASP6, LINC01970, CSNK1D, LINC01694, LSS, SPATC1L, COL6A2 \n",
            "\t   PCBP3, SLC16A3, MCUB, CCDC57, FASN, SEC24B, FAM71F2, DUS1L, CALU, GPS1 \n",
            "\t   SEC24B-AS1, CCDC136, RFNG, KCP, ATP6V1F, DCXR, COL25A1, IRF5, RAC3, LRRC45 \n",
            "PC_ 5 \n",
            "Positive:  MYO1D, CDK5R1, PSMD11, AC025211.1, AC005899.6, CCT6B, ZNF207, ZNF830, C17orf75, RHBDL3 \n",
            "\t   RAD51D, LIG3, RHOT1, RFFL, NLE1, AC116407.2, LRRC37B, AC022916.1, SUZ12, SLFN5 \n",
            "\t   AC004253.1, AC022706.1, UTP6, COPRS, RAB11FIP4, EVI2A, EVI2B, OMG, AC079915.1, NF1 \n",
            "Negative:  INTU, HSPA4L, PLK4, MFSD8, ABHD18, LARP1B, PGRMC2, LINC02615, JADE1, SCLT1 \n",
            "\t   C4orf33, LINC02479, AC096711.2, AC093916.1, AC096631.1, MARCH1, MIR181A1HG, TMA16, NAF1, AC022272.1 \n",
            "\t   RAPGEF2, C4orf45, FNIP2, PPID, ETFDH, C4orf46, RPS3A, RXFP1, SH3D19, LRBA \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:42:02] define_signif_tumor_subclusters(), tumor: control\n",
            "INFO [2025-06-10 11:42:02] Setting auto leiden resolution for control to 0.0057241\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  KCNAB1, GMPS, SLC33A1, MBNL1, MBNL1-AS1, C3orf33, AC026347.1, P2RY1, PLCH1, RAP2B \n",
            "\t   ARHGEF26, MME-AS1, DHX36, MME, ARPC3, GPN3, FAM216A, PHETA1, VPS29, ANAPC7 \n",
            "\t   RAD9B, PPTC7, ATP2A2, TCTN1, HVCN1, PPP1CC, CUX2, IFT81, C12orf76, ANKRD13A \n",
            "Negative:  C16orf86, GFOD2, RANBP10, CENPT, THAP11, NUTF2, EDC4, PSKH1, CTRL, PSMB10 \n",
            "\t   LCAT, SLC12A4, DPEP2, DUS2, DDX28, NFATC3, AC020978.3, KIF22, AC009133.3, MAZ \n",
            "\t   AC009133.1, AC020978.7, PAGR1, C16orf54, MVP, QPRT, CDIPT, INO80E, DOC2A, HIRIP3 \n",
            "PC_ 2 \n",
            "Positive:  ERCC4, AC130650.2, MRTFB, SHISA9, AC130650.1, AC109597.1, AC040173.1, CPPED1, PARN, AC010333.2 \n",
            "\t   BFAR, PLA2G10, SNX29, NPIPB2, NOMO1, AC007216.2, NPIPA1, AC007216.3, PDXDC1, NTAN1 \n",
            "\t   RRN3, GSPT1, MPV17L, BMERB1, AC007216.4, MARF1, NDE1, AC026401.3, MYH11, RSL1D1 \n",
            "Negative:  HLX-AS1, HLX, MARC1, DUSP10, RAB3GAP2, LINC02257, AC096642.1, BPNT1, EPRS, IARS2 \n",
            "\t   LYPLAL1, TAF1A, MIA3, LYPLAL1-DT, AL592148.3, C1orf143, AIDA, TGFB2, BROX, FAM177B \n",
            "\t   RRP15, AL392172.2, SPATA17, AL392172.1, GPATCH2, DISP1, ESRRG, TLR5, USH2A, SUSD4 \n",
            "PC_ 3 \n",
            "Positive:  CYHR1, TONSL, VPS28, SLC39A4, CPSF1, ADCK5, FBXL6, EDF1, PHPT1, TRAF2 \n",
            "\t   RABL6, FBXW5, CCDC183, PTGDS, LCNL1, TMEM141, PAXX, CLIC3, NPDC1, ABCA2 \n",
            "\t   C9orf139, SAPCD2, FUT7, LCN6, UAP1L1, SNHG7, MAN1B1-DT, MAN1B1, DIPK1B, DPP7 \n",
            "Negative:  AC005674.2, SLC2A9, AC105916.1, WDR1, AC073648.7, ZNF518B, AC108519.1, AC005599.1, AC105345.1, TRMT44 \n",
            "\t   CLNK, ACOX3, HS3ST1, AC104825.1, AC006230.1, AC025539.1, SH3TC1, AC005699.1, ABLIM2, RAB28 \n",
            "\t   BOD1L1, AFAP1, AC092546.1, LINC00504, SORCS2, CPEB2-DT, GRPEL1, CPEB2, TADA2B, AC098829.1 \n",
            "PC_ 4 \n",
            "Positive:  IFI16, AIM2, PYHIN1, MNDA, ACKR1, FCER1A, DUSP23, FCRL6, CFAP45, TAGLN2 \n",
            "\t   SPTA1, IGSF9, OR10Z1, LINC01133, CD1C, PIGM, IGSF8, CD1D, AL121987.2, KIRREL1 \n",
            "\t   PEA15, DCAF8, FCRL1, AL139011.1, PEX19, FCRL2, COPA, FCRL3, NCSTN, FCRL5 \n",
            "Negative:  MMS22L, AL589740.1, FBXL4, COQ3, PNISR, AL513550.1, USP45, TSTD3, CCNC, ASCC3 \n",
            "\t   AL133338.1, GRIK2, HACE1, BVES, SRFBP1, AC008568.1, LOX, AC010255.1, PRR16, SNX2 \n",
            "\t   BVES-AS1, SNX24, AC113418.1, CEP120, CSNK1G3, AC008574.1, PREP, HSD17B4, AL096794.1, TNFAIP8 \n",
            "PC_ 5 \n",
            "Positive:  PPIL2, AP000553.2, YPEL1, SDF2L1, YDJC, MAPK1, UBE2L3, PPM1F, HIC2, AC245452.1 \n",
            "\t   THAP7-AS1, THAP7, TOP3B, LZTR1, IGLV10-54, LINC01637, VPREB1, CRKL, AC245060.6, SNAP29 \n",
            "\t   AC245060.5, SERPIND1, ZNF280B, IGLV3-21, PI4KA, IGLC1, MED15, IGLC2, KLHL22, IGLC3 \n",
            "Negative:  C17orf75, AC005899.6, ZNF207, PSMD11, RHBDL3, CDK5R1, RHOT1, MYO1D, AC116407.2, AC025211.1 \n",
            "\t   LRRC37B, SUZ12, CCT6B, AC004253.1, ZNF830, UTP6, LIG3, COPRS, RFFL, RAB11FIP4 \n",
            "\t   RAD51D, EVI2A, NLE1, EVI2B, AC022916.1, OMG, SLFN5, AC079915.1, NF1, AC022706.1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-06-10 11:44:21] ::plot_cnv:Start\n",
            "INFO [2025-06-10 11:44:21] ::plot_cnv:Current data dimensions (r,c)=16601,7764 Total=129049398.672976 Min=0.798279710191454 Max=8.72924187359209.\n",
            "INFO [2025-06-10 11:44:22] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 11:44:25] plot_cnv(): auto thresholding at: (0.889253 , 1.113218)\n",
            "INFO [2025-06-10 11:44:44] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 11:44:44] Observation data size: Cells= 2858 Genes= 16601\n",
            "INFO [2025-06-10 11:44:44] clustering observations via method: ward.D\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(1) is 12\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  12,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(2) is 10\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  10,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(3) is 5\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  5,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(4) is 68\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  68,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(5) is 63\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  63,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(6) is 58\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  58,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(7) is 56\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  56,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(8) is 52\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  52,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(9) is 46\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  46,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(10) is 45\n",
            "INFO [2025-06-10 11:44:44] group size being clustered:  45,16601\n",
            "INFO [2025-06-10 11:44:44] Number of cells in group(11) is 36\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  36,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(12) is 31\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  31,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(13) is 19\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  19,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(14) is 14\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  14,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(15) is 10\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  10,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(16) is 8\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  8,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(17) is 4\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  4,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(18) is 3\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(19) is 2\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(20) is 1\n",
            "INFO [2025-06-10 11:44:45] Skipping group: 20, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(21) is 1\n",
            "INFO [2025-06-10 11:44:45] Skipping group: 21, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(22) is 64\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  64,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(23) is 44\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  44,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(24) is 37\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  37,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(25) is 25\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  25,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(26) is 3\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(27) is 1\n",
            "INFO [2025-06-10 11:44:45] Skipping group: 27, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(28) is 9\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  9,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(29) is 24\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  24,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(30) is 18\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  18,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(31) is 79\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  79,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(32) is 76\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  76,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(33) is 72\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  72,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(34) is 72\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  72,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(35) is 69\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  69,16601\n",
            "INFO [2025-06-10 11:44:45] Number of cells in group(36) is 65\n",
            "INFO [2025-06-10 11:44:45] group size being clustered:  65,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(37) is 46\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  46,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(38) is 42\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  42,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(39) is 42\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  42,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(40) is 39\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  39,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(41) is 31\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  31,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(42) is 29\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  29,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(43) is 23\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  23,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(44) is 21\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  21,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(45) is 16\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  16,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(46) is 10\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  10,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(47) is 10\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  10,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(48) is 8\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  8,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(49) is 4\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  4,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(50) is 2\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(51) is 1\n",
            "INFO [2025-06-10 11:44:46] Skipping group: 51, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(52) is 77\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  77,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(53) is 50\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  50,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(54) is 49\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  49,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(55) is 36\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  36,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(56) is 27\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  27,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(57) is 12\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  12,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(58) is 4\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  4,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(59) is 2\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(60) is 2\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(61) is 1\n",
            "INFO [2025-06-10 11:44:46] Skipping group: 61, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(62) is 43\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  43,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(63) is 56\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  56,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(64) is 53\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  53,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(65) is 18\n",
            "INFO [2025-06-10 11:44:46] group size being clustered:  18,16601\n",
            "INFO [2025-06-10 11:44:46] Number of cells in group(66) is 17\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  17,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(67) is 4\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  4,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(68) is 3\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(69) is 1\n",
            "INFO [2025-06-10 11:44:47] Skipping group: 69, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(70) is 1\n",
            "INFO [2025-06-10 11:44:47] Skipping group: 70, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(71) is 1\n",
            "INFO [2025-06-10 11:44:47] Skipping group: 71, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(72) is 46\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  46,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(73) is 29\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  29,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(74) is 22\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  22,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(75) is 12\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  12,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(76) is 11\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  11,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(77) is 9\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  9,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(78) is 82\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  82,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(79) is 66\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  66,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(80) is 50\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  50,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(81) is 36\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  36,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(82) is 23\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  23,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(83) is 6\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  6,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(84) is 6\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  6,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(85) is 3\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(86) is 2\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(87) is 2\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(88) is 12\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  12,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(89) is 4\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  4,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(90) is 33\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  33,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(91) is 69\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  69,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(92) is 59\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  59,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(93) is 13\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  13,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(94) is 2\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  2,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(95) is 1\n",
            "INFO [2025-06-10 11:44:47] Skipping group: 95, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(96) is 35\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  35,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(97) is 34\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  34,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(98) is 3\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(99) is 1\n",
            "INFO [2025-06-10 11:44:47] Skipping group: 99, since less than 2 entries\n",
            "INFO [2025-06-10 11:44:47] Number of cells in group(100) is 73\n",
            "INFO [2025-06-10 11:44:47] group size being clustered:  73,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(101) is 49\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  49,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(102) is 31\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  31,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(103) is 3\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(104) is 3\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  3,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(105) is 26\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  26,16601\n",
            "INFO [2025-06-10 11:44:48] Number of cells in group(106) is 19\n",
            "INFO [2025-06-10 11:44:48] group size being clustered:  19,16601\n",
            "INFO [2025-06-10 11:44:48] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 11:44:48] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 11:44:48] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 11:44:48] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 11:45:01] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 11:45:01] Quantiles of plotted data range: 0.889252643133043,0.9715935119125,0.998332371362393,1.02880198588484,1.11321821527902\n",
            "INFO [2025-06-10 11:45:07] plot_cnv_references:Start\n",
            "INFO [2025-06-10 11:45:07] Reference data size: Cells= 4906 Genes= 16601\n",
            "INFO [2025-06-10 11:45:25] plot_cnv_references:Number reference groups= 53\n",
            "INFO [2025-06-10 11:45:27] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 11:45:48] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 11:45:48] Quantiles of plotted data range: 0.889252643133043,0.975498019078326,0.998337630382622,1.02319444155191,1.11321821527902\n",
            "INFO [2025-06-10 11:47:35] ::plot_cnv:Start\n",
            "INFO [2025-06-10 11:47:35] ::plot_cnv:Current data dimensions (r,c)=16601,7764 Total=129049398.672976 Min=0.798279710191454 Max=8.72924187359209.\n",
            "INFO [2025-06-10 11:47:35] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 11:47:39] plot_cnv(): auto thresholding at: (0.889253 , 1.113218)\n",
            "INFO [2025-06-10 11:47:43] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 11:47:43] Observation data size: Cells= 2858 Genes= 16601\n",
            "INFO [2025-06-10 11:47:43] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 11:47:43] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 11:47:43] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 11:47:43] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 11:47:55] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 11:47:55] Quantiles of plotted data range: 0.889252643133043,0.9715935119125,0.998332371362393,1.02880198588484,1.11321821527902\n",
            "INFO [2025-06-10 11:47:59] plot_cnv_references:Start\n",
            "INFO [2025-06-10 11:47:59] Reference data size: Cells= 4906 Genes= 16601\n",
            "INFO [2025-06-10 12:09:13] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 12:09:14] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 12:09:36] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 12:09:36] Quantiles of plotted data range: 0.889252643133043,0.975498019078326,0.998337630382622,1.02319444155191,1.11321821527902\n",
            "INFO [2025-06-10 12:09:42] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-06-10 12:09:42] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-06-10 12:09:42] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-06-10 12:09:45] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.944580434314598 - 1.05682375801066\n",
            "INFO [2025-06-10 12:10:43] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-06-10 12:10:44] ::plot_cnv:Start\n",
            "INFO [2025-06-10 12:10:44] ::plot_cnv:Current data dimensions (r,c)=16601,7764 Total=129319326.997359 Min=0.798279710191454 Max=8.72924187359209.\n",
            "INFO [2025-06-10 12:10:45] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-06-10 12:10:48] plot_cnv(): auto thresholding at: (0.886782 , 1.113218)\n",
            "INFO [2025-06-10 12:10:52] plot_cnv_observation:Start\n",
            "INFO [2025-06-10 12:10:52] Observation data size: Cells= 2858 Genes= 16601\n",
            "INFO [2025-06-10 12:10:52] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-06-10 12:10:52] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-06-10 12:10:53] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 12:10:53] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-06-10 12:11:04] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 12:11:04] Quantiles of plotted data range: 0.886781784720977,1.00070209616263,1.00070209616263,1.00070209616263,1.11321821527902\n",
            "INFO [2025-06-10 12:11:08] plot_cnv_references:Start\n",
            "INFO [2025-06-10 12:11:08] Reference data size: Cells= 4906 Genes= 16601\n",
            "INFO [2025-06-10 12:32:25] plot_cnv_references:Number reference groups= 1\n",
            "INFO [2025-06-10 12:32:27] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-06-10 12:32:48] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-06-10 12:32:48] Quantiles of plotted data range: 0.886781784720977,1.00070209616263,1.00070209616263,1.00070209616263,1.11321821527902\n"
          ]
        }
      ],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"1_vs_99\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ysdkNz7Ljun6"
      },
      "outputs": [],
      "source": [
        "# create the infercnv object\n",
        "library(infercnv)\n",
        "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/content/all_samples.matrix\",\n",
        "                                    annotations_file=\"/content/all_samples.annotations.tsv\",\n",
        "                                    delim=\"\\t\",\n",
        "                                    gene_order_file=\"gene_pos_file.txt\",\n",
        "                                    ref_group_names=c(\"control\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qtY_f5ZJjv3N"
      },
      "outputs": [],
      "source": [
        "# perform infercnv operations to reveal cnv signal\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=0.01,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
        "                             out_dir=\"all_samples\",  # dir is auto-created for storing outputs\n",
        "                             cluster_by_groups=T,   # cluster\n",
        "                             denoise=T,\n",
        "                             plot_steps=F,\n",
        "                             resume_mode=TRUE,\n",
        "                             HMM=T,\n",
        "                             num_threads=5,\n",
        "                             window_length = 151\n",
        "                             )"
      ]
    }
  ],
  "metadata": {
    "accelerator": "TPU",
    "colab": {
      "gpuType": "V28",
      "machine_shape": "hm",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    },
    "language_info": {
      "name": "R"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}